{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "\n",
    "import os\n",
    "import torch \n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import math\n",
    "import mygene\n",
    "import sklearn.preprocessing as sk\n",
    "import seaborn as sns\n",
    "from sklearn import metrics\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "from torch.utils.data.sampler import WeightedRandomSampler\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import average_precision_score, precision_recall_curve\n",
    "import random\n",
    "import warnings\n",
    "from random import randint\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define directories\n",
    "\n",
    "cell_line_dir = \"/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC/\"\n",
    "models_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/EGFRi/Expression_Mutation_CNA/Models/GDSC_PDX_TCGA_Fourth_Strategy/\"\n",
    "finetuned_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/EGFRi/Expression_Mutation_CNA/Models/GDSC_PDX_TCGA_Fourth_Strategy/Fourth_Strategy_Finetuned_Models/\"\n",
    "DEGs_dir = \"/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/EGFRi/\"\n",
    "dataset_dir = \"/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/\"\n",
    "pdx_data_dir = \"/Volumes/Expansion/Thesis Work/Datasets/PDX/Expression/\"\n",
    "save_results_to = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/EGFRi/Expression_Mutation_CNA/Predictions/GDSC_PDX_TCGA_Fourth_Strategy/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set random seeds\n",
    "\n",
    "torch.manual_seed(42)\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(cell_line_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Whole Exome Sequencing (WES)</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>Drug\\nResponse</th>\n",
       "      <th>GDSC\\nTissue descriptor 1</th>\n",
       "      <th>GDSC\\nTissue\\ndescriptor 2</th>\n",
       "      <th>Cancer Type\\n(matching TCGA label)</th>\n",
       "      <th>Microsatellite \\ninstability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1299069</th>\n",
       "      <td>TMK-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>digestive_system</td>\n",
       "      <td>stomach</td>\n",
       "      <td>STAD</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1298218</th>\n",
       "      <td>KP-2</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>PAAD</td>\n",
       "      <td></td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330932</th>\n",
       "      <td>KO52</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>leukemia</td>\n",
       "      <td>acute_myeloid_leukaemia</td>\n",
       "      <td>LAML</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331030</th>\n",
       "      <td>SC-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>lymphoma</td>\n",
       "      <td>B_cell_lymphoma</td>\n",
       "      <td>DLBC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503373</th>\n",
       "      <td>U-CH2</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>bone</td>\n",
       "      <td>bone_other</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample Name Whole Exome Sequencing (WES)  \\\n",
       "COSMIC identifier                                            \n",
       "1299069                 TMK-1                            Y   \n",
       "1298218                  KP-2                            Y   \n",
       "1330932                  KO52                            Y   \n",
       "1331030                  SC-1                            Y   \n",
       "1503373                 U-CH2                            Y   \n",
       "\n",
       "                  Copy Number Alterations (CNA) Gene Expression Methylation  \\\n",
       "COSMIC identifier                                                             \n",
       "1299069                                       Y               N           N   \n",
       "1298218                                       N               Y           N   \n",
       "1330932                                       Y               N           N   \n",
       "1331030                                       Y               N           N   \n",
       "1503373                                       Y               N           N   \n",
       "\n",
       "                  Drug\\nResponse GDSC\\nTissue descriptor 1  \\\n",
       "COSMIC identifier                                            \n",
       "1299069                        Y          digestive_system   \n",
       "1298218                        Y                  pancreas   \n",
       "1330932                        N                  leukemia   \n",
       "1331030                        N                  lymphoma   \n",
       "1503373                        N                      bone   \n",
       "\n",
       "                  GDSC\\nTissue\\ndescriptor 2  \\\n",
       "COSMIC identifier                              \n",
       "1299069                              stomach   \n",
       "1298218                             pancreas   \n",
       "1330932              acute_myeloid_leukaemia   \n",
       "1331030                      B_cell_lymphoma   \n",
       "1503373                           bone_other   \n",
       "\n",
       "                  Cancer Type\\n(matching TCGA label)  \\\n",
       "COSMIC identifier                                      \n",
       "1299069                                         STAD   \n",
       "1298218                                         PAAD   \n",
       "1330932                                         LAML   \n",
       "1331030                                         DLBC   \n",
       "1503373                                                \n",
       "\n",
       "                  Microsatellite \\ninstability Status (MSI) Screen Medium  \\\n",
       "COSMIC identifier                                                           \n",
       "1299069                                                                 R   \n",
       "1298218                                                             D/F12   \n",
       "1330932                                           MSS/MSI-L         D/F12   \n",
       "1331030                                           MSS/MSI-L             R   \n",
       "1503373                                           MSS/MSI-L         D/F12   \n",
       "\n",
       "                  Growth Properties  \n",
       "COSMIC identifier                    \n",
       "1299069                    Adherent  \n",
       "1298218                    Adherent  \n",
       "1330932                          NA  \n",
       "1331030                          NA  \n",
       "1503373                    Adherent  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read cell line details table in GDSC\n",
    "\n",
    "GDSC_cell_line_details = pd.read_excel(\"GDSC_Cell_Lines_Details.xlsx\", keep_default_na = False)\n",
    "GDSC_cell_line_details.set_index(\"COSMIC identifier\", inplace = True)\n",
    "GDSC_cell_line_details = GDSC_cell_line_details.iloc[:-1,]\n",
    "GDSC_cell_line_details.index = GDSC_cell_line_details.index.astype(str)\n",
    "GDSC_cell_line_details.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC expression dataset (EGFRi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/exprs/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>107984199</th>\n",
       "      <th>107984208</th>\n",
       "      <th>107984325</th>\n",
       "      <th>107984411</th>\n",
       "      <th>107984648</th>\n",
       "      <th>107984889</th>\n",
       "      <th>107984989</th>\n",
       "      <th>107986313</th>\n",
       "      <th>107986782</th>\n",
       "      <th>107986810</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>-0.4485253908432412</td>\n",
       "      <td>-0.3964608177552915</td>\n",
       "      <td>0.6112242395986884</td>\n",
       "      <td>-0.4268956145117655</td>\n",
       "      <td>-0.4170391628564061</td>\n",
       "      <td>-0.4792877174580186</td>\n",
       "      <td>-0.8963781730111828</td>\n",
       "      <td>-0.5325490279167789</td>\n",
       "      <td>1.3556206755229008</td>\n",
       "      <td>-0.6777986169743725</td>\n",
       "      <td>...</td>\n",
       "      <td>0.157304424865776</td>\n",
       "      <td>1.3128169701288817</td>\n",
       "      <td>1.581148961036248</td>\n",
       "      <td>-1.0233928812648092</td>\n",
       "      <td>-0.5261275394433881</td>\n",
       "      <td>-1.2544249102538338</td>\n",
       "      <td>-0.1893406128651996</td>\n",
       "      <td>-0.6772425488930321</td>\n",
       "      <td>-1.178970090534446</td>\n",
       "      <td>-0.3515273698794716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753548</th>\n",
       "      <td>0.54434399328065</td>\n",
       "      <td>-0.2097132915252216</td>\n",
       "      <td>0.7428823880631317</td>\n",
       "      <td>-0.4654204339690438</td>\n",
       "      <td>-0.5008015014248242</td>\n",
       "      <td>-0.4957218058012111</td>\n",
       "      <td>0.1123933266524433</td>\n",
       "      <td>-0.1159950501276046</td>\n",
       "      <td>-0.2995795718953774</td>\n",
       "      <td>0.7723990325321687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5031855878328687</td>\n",
       "      <td>2.1399370118140006</td>\n",
       "      <td>-0.1876516557343364</td>\n",
       "      <td>0.7612631050871597</td>\n",
       "      <td>-1.1694375733774025</td>\n",
       "      <td>-1.059385500933128</td>\n",
       "      <td>0.0472500273184416</td>\n",
       "      <td>0.2408070809791057</td>\n",
       "      <td>0.811515939332738</td>\n",
       "      <td>-0.5952609010470329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906870</th>\n",
       "      <td>0.2131901385736997</td>\n",
       "      <td>0.0780957254627422</td>\n",
       "      <td>0.4962581009511371</td>\n",
       "      <td>0.3132923777155636</td>\n",
       "      <td>-0.4346530575506854</td>\n",
       "      <td>-0.484081268732043</td>\n",
       "      <td>-2.2608419901281893</td>\n",
       "      <td>-0.692542244004405</td>\n",
       "      <td>-2.4698679220343585</td>\n",
       "      <td>-0.646918188167355</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6053841334831193</td>\n",
       "      <td>-1.0523147187370203</td>\n",
       "      <td>1.1184767808124156</td>\n",
       "      <td>3.1961868353521354</td>\n",
       "      <td>1.4698980731759494</td>\n",
       "      <td>-1.178220046996365</td>\n",
       "      <td>1.0882357903061055</td>\n",
       "      <td>1.567400529589298</td>\n",
       "      <td>1.4265693390197831</td>\n",
       "      <td>-1.242892613324148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907053</th>\n",
       "      <td>-0.4583258871402069</td>\n",
       "      <td>-0.2649278330546487</td>\n",
       "      <td>0.6792607671752738</td>\n",
       "      <td>-0.698282021878342</td>\n",
       "      <td>-0.3437006707719533</td>\n",
       "      <td>-0.4292248293825993</td>\n",
       "      <td>-0.1483686444323214</td>\n",
       "      <td>0.5264302510638762</td>\n",
       "      <td>1.0351456808342203</td>\n",
       "      <td>-0.2891606802619315</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0410126737854252</td>\n",
       "      <td>-1.1291700058226537</td>\n",
       "      <td>1.720665295186618</td>\n",
       "      <td>0.0164675996610106</td>\n",
       "      <td>1.910374333839876</td>\n",
       "      <td>0.8096458123532942</td>\n",
       "      <td>-0.4944018709001623</td>\n",
       "      <td>0.1083432782752229</td>\n",
       "      <td>2.091402424813153</td>\n",
       "      <td>0.6440439348021278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905938</th>\n",
       "      <td>-0.8429588834713087</td>\n",
       "      <td>-0.2841139508211123</td>\n",
       "      <td>0.7684726437129762</td>\n",
       "      <td>0.0769543357163258</td>\n",
       "      <td>-0.3685331180491761</td>\n",
       "      <td>-0.1573291597843853</td>\n",
       "      <td>0.1413639780822689</td>\n",
       "      <td>2.239857376724212</td>\n",
       "      <td>0.454570902072972</td>\n",
       "      <td>-0.2655973272587233</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4917406673594041</td>\n",
       "      <td>-0.0906714733183549</td>\n",
       "      <td>-0.7432117949453445</td>\n",
       "      <td>0.0774994852637504</td>\n",
       "      <td>1.114524806379214</td>\n",
       "      <td>-0.7695705450315459</td>\n",
       "      <td>0.9716758554533608</td>\n",
       "      <td>-1.0177790402874862</td>\n",
       "      <td>1.0767256903064175</td>\n",
       "      <td>2.679071507799088</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18597 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  1                    2                   9          \\\n",
       "910688  -0.4485253908432412  -0.3964608177552915  0.6112242395986884   \n",
       "753548     0.54434399328065  -0.2097132915252216  0.7428823880631317   \n",
       "906870   0.2131901385736997   0.0780957254627422  0.4962581009511371   \n",
       "907053  -0.4583258871402069  -0.2649278330546487  0.6792607671752738   \n",
       "905938  -0.8429588834713087  -0.2841139508211123  0.7684726437129762   \n",
       "\n",
       "                  10                   12                   13         \\\n",
       "910688  -0.4268956145117655  -0.4170391628564061  -0.4792877174580186   \n",
       "753548  -0.4654204339690438  -0.5008015014248242  -0.4957218058012111   \n",
       "906870   0.3132923777155636  -0.4346530575506854   -0.484081268732043   \n",
       "907053   -0.698282021878342  -0.3437006707719533  -0.4292248293825993   \n",
       "905938   0.0769543357163258  -0.3685331180491761  -0.1573291597843853   \n",
       "\n",
       "                  14                   15                   16         \\\n",
       "910688  -0.8963781730111828  -0.5325490279167789   1.3556206755229008   \n",
       "753548   0.1123933266524433  -0.1159950501276046  -0.2995795718953774   \n",
       "906870  -2.2608419901281893   -0.692542244004405  -2.4698679220343585   \n",
       "907053  -0.1483686444323214   0.5264302510638762   1.0351456808342203   \n",
       "905938   0.1413639780822689    2.239857376724212    0.454570902072972   \n",
       "\n",
       "                  18         ...           107984199            107984208  \\\n",
       "910688  -0.6777986169743725  ...   0.157304424865776   1.3128169701288817   \n",
       "753548   0.7723990325321687  ...  0.5031855878328687   2.1399370118140006   \n",
       "906870   -0.646918188167355  ...  0.6053841334831193  -1.0523147187370203   \n",
       "907053  -0.2891606802619315  ...  0.0410126737854252  -1.1291700058226537   \n",
       "905938  -0.2655973272587233  ...  0.4917406673594041  -0.0906714733183549   \n",
       "\n",
       "                  107984325            107984411            107984648  \\\n",
       "910688    1.581148961036248  -1.0233928812648092  -0.5261275394433881   \n",
       "753548  -0.1876516557343364   0.7612631050871597  -1.1694375733774025   \n",
       "906870   1.1184767808124156   3.1961868353521354   1.4698980731759494   \n",
       "907053    1.720665295186618   0.0164675996610106    1.910374333839876   \n",
       "905938  -0.7432117949453445   0.0774994852637504    1.114524806379214   \n",
       "\n",
       "                  107984889            107984989            107986313  \\\n",
       "910688  -1.2544249102538338  -0.1893406128651996  -0.6772425488930321   \n",
       "753548   -1.059385500933128   0.0472500273184416   0.2408070809791057   \n",
       "906870   -1.178220046996365   1.0882357903061055    1.567400529589298   \n",
       "907053   0.8096458123532942  -0.4944018709001623   0.1083432782752229   \n",
       "905938  -0.7695705450315459   0.9716758554533608  -1.0177790402874862   \n",
       "\n",
       "                 107986782            107986810  \n",
       "910688  -1.178970090534446  -0.3515273698794716  \n",
       "753548   0.811515939332738  -0.5952609010470329  \n",
       "906870  1.4265693390197831   -1.242892613324148  \n",
       "907053   2.091402424813153   0.6440439348021278  \n",
       "905938  1.0767256903064175    2.679071507799088  \n",
       "\n",
       "[5 rows x 18597 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC expression dataset (EGFRi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCE = pd.read_csv(\"GDSC_exprs.z.EGFRi.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index_col=0,\n",
    "                    decimal=\",\")\n",
    "GDSCE = pd.DataFrame.transpose(GDSCE)\n",
    "GDSCE = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCE,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCE[\"GDSC\\nTissue descriptor 1\"] != \"lung_NSCLC\") & \\\n",
    "         (GDSCE[\"GDSC\\nTissue descriptor 1\"] != \"large_intestine\") & \\\n",
    "         (GDSCE[\"GDSC\\nTissue descriptor 1\"] != \"aero_dig_tract\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"COREAD\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"LUAD\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"HNSC\")\n",
    "GDSCE = GDSCE.loc[filter, ]\n",
    "GDSCE.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCE.index = GDSCE.index.astype(int)\n",
    "\n",
    "GDSCE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/mutations\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC mutation dataset (EGFRi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/mutations/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>143872</th>\n",
       "      <th>728577</th>\n",
       "      <th>2</th>\n",
       "      <th>2050</th>\n",
       "      <th>10243</th>\n",
       "      <th>79365</th>\n",
       "      <th>256006</th>\n",
       "      <th>4609</th>\n",
       "      <th>118788</th>\n",
       "      <th>113675</th>\n",
       "      <th>...</th>\n",
       "      <th>26517</th>\n",
       "      <th>28972</th>\n",
       "      <th>53917</th>\n",
       "      <th>283507</th>\n",
       "      <th>514</th>\n",
       "      <th>339210</th>\n",
       "      <th>64981</th>\n",
       "      <th>7335</th>\n",
       "      <th>5739</th>\n",
       "      <th>83442</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753548</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906870</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907053</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905938</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        143872  728577  2       2050    10243   79365   256006  4609    \\\n",
       "910688     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "753548     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "906870     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "907053     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "905938     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        118788  113675  ...  26517   28972   53917   283507  514     339210  \\\n",
       "910688     0.0    0.00  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "753548     0.0   -0.05  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "906870     0.0    0.00  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "907053     0.0    0.00  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "905938     0.0    0.00  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        64981   7335    5739    83442   \n",
       "910688     0.0     0.0     0.0     0.0  \n",
       "753548     0.0     0.0     0.0     0.0  \n",
       "906870     0.0     0.0     0.0     0.0  \n",
       "907053     0.0     0.0     0.0     0.0  \n",
       "905938     0.0     0.0     0.0     0.0  \n",
       "\n",
       "[5 rows x 18377 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC mutation dataset (EGFRi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCM = pd.read_csv(\"GDSC_mutations.EGFRi.tsv\",\n",
    "                    sep=\"\\t\", index_col=0, decimal=\".\")\n",
    "GDSCM = pd.DataFrame.transpose(GDSCM)\n",
    "GDSCM = GDSCM.loc[:, ~GDSCM.columns.duplicated()]\n",
    "GDSCM = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCM,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCM[\"GDSC\\nTissue descriptor 1\"] != \"lung_NSCLC\") & \\\n",
    "         (GDSCM[\"GDSC\\nTissue descriptor 1\"] != \"large_intestine\") & \\\n",
    "         (GDSCM[\"GDSC\\nTissue descriptor 1\"] != \"aero_dig_tract\") & \\\n",
    "         (GDSCM[\"Cancer Type\\n(matching TCGA label)\"] != \"COREAD\") & \\\n",
    "         (GDSCM[\"Cancer Type\\n(matching TCGA label)\"] != \"LUAD\") & \\\n",
    "         (GDSCM[\"Cancer Type\\n(matching TCGA label)\"] != \"HNSC\")\n",
    "GDSCM = GDSCM.loc[filter, ]\n",
    "GDSCM.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCM.index = GDSCM.index.astype(int)\n",
    "\n",
    "GDSCM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC CNA dataset (EGFRi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/CNA/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>107985535</th>\n",
       "      <th>107985759</th>\n",
       "      <th>107986588</th>\n",
       "      <th>107986809</th>\n",
       "      <th>107986898</th>\n",
       "      <th>107987337</th>\n",
       "      <th>107987341</th>\n",
       "      <th>109731405</th>\n",
       "      <th>112441434</th>\n",
       "      <th>121676927</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753548</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906870</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907053</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905938</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24402 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        1          2          9          10         12         13         \\\n",
       "910688        0.0        0.0   0.000000   0.000000   0.000000   0.584963   \n",
       "753548        0.0        0.0   0.000000   0.000000   0.000000   0.000000   \n",
       "906870        0.0        0.0   0.584963   0.584963   0.000000   0.000000   \n",
       "907053        0.0        0.0   0.415037   0.415037  -0.584963   0.000000   \n",
       "905938        0.0        0.0   0.000000   0.000000   0.000000   0.000000   \n",
       "\n",
       "        14         15         16         18         ...  107985535  107985759  \\\n",
       "910688        0.0        0.0   0.000000   0.000000  ...        0.0   0.000000   \n",
       "753548        0.0        0.0   0.000000   0.000000  ...        0.0   0.000000   \n",
       "906870        0.0        0.0  -1.000000   0.000000  ...        0.0   0.000000   \n",
       "907053        0.0        0.0  -0.584963  -0.584963  ...        0.0   0.415037   \n",
       "905938        0.0        0.0   0.000000   0.000000  ...        0.0   0.000000   \n",
       "\n",
       "        107986588  107986809  107986898  107987337  107987341  109731405  \\\n",
       "910688        0.0   0.000000   0.000000      -1.00   0.584963   0.000000   \n",
       "753548        0.0   0.000000   0.000000      -4.32  -4.320000   0.000000   \n",
       "906870        0.0   0.000000   0.584963      -4.32  -4.320000  -1.000000   \n",
       "907053        0.0  -0.584963   0.415037      -4.32  -4.320000  -0.584963   \n",
       "905938        0.0   0.000000   0.000000      -4.32  -4.320000   0.000000   \n",
       "\n",
       "        112441434  121676927  \n",
       "910688   0.000000   0.000000  \n",
       "753548   0.000000   0.000000  \n",
       "906870   0.000000   0.584963  \n",
       "907053  -0.584963   0.415037  \n",
       "905938  -1.000000   0.000000  \n",
       "\n",
       "[5 rows x 24402 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC CNA dataset (EGFRi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCC = pd.read_csv(\"GDSC_CNA.EGFRi.tsv\",\n",
    "                    sep=\"\\t\", index_col=0, decimal=\".\")\n",
    "GDSCC.drop_duplicates(keep='last')\n",
    "GDSCC = pd.DataFrame.transpose(GDSCC)\n",
    "GDSCC = GDSCC.loc[:, ~GDSCC.columns.duplicated()]\n",
    "GDSCC = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCC,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCC[\"GDSC\\nTissue descriptor 1\"] != \"lung_NSCLC\") & \\\n",
    "         (GDSCC[\"GDSC\\nTissue descriptor 1\"] != \"large_intestine\") & \\\n",
    "         (GDSCC[\"GDSC\\nTissue descriptor 1\"] != \"aero_dig_tract\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"COREAD\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"LUAD\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"HNSC\")\n",
    "GDSCC = GDSCC.loc[filter, ]\n",
    "GDSCC.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCC.index = GDSCC.index.astype(int)\n",
    "\n",
    "GDSCC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC response dataset (EGFRi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/response/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>1.24811569418395</td>\n",
       "      <td>Afatinib</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>3.82039284182925</td>\n",
       "      <td>Afatinib (rescreen)</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>-2.24580154812369</td>\n",
       "      <td>CUDC-101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>1</td>\n",
       "      <td>-0.939077266753545</td>\n",
       "      <td>EKB-569</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>2.75638048253033</td>\n",
       "      <td>Erlotinib</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   response             logIC50                 drug  exprs  \\\n",
       "COSMIC identifier                                                             \n",
       "910688                    0    1.24811569418395             Afatinib      1   \n",
       "910688                    0    3.82039284182925  Afatinib (rescreen)      1   \n",
       "910688                    0   -2.24580154812369             CUDC-101      1   \n",
       "910688                    1  -0.939077266753545              EKB-569      1   \n",
       "910688                    0    2.75638048253033            Erlotinib      1   \n",
       "\n",
       "                   CNA  mutations  \n",
       "COSMIC identifier                  \n",
       "910688               1          1  \n",
       "910688               1          1  \n",
       "910688               1          1  \n",
       "910688               1          1  \n",
       "910688               1          1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC response dataset (EGFRi) and binarize\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCR = pd.read_csv(\"GDSC_response.EGFRi.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index_col=0,\n",
    "                    decimal=\",\")\n",
    "GDSCR.dropna(inplace=True)\n",
    "GDSCR.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "GDSCR[\"response\"] = GDSCR.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "GDSCR = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCR,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCR[\"GDSC\\nTissue descriptor 1\"] != \"lung_NSCLC\") & \\\n",
    "         (GDSCR[\"GDSC\\nTissue descriptor 1\"] != \"large_intestine\") & \\\n",
    "         (GDSCR[\"GDSC\\nTissue descriptor 1\"] != \"aero_dig_tract\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"COREAD\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"LUAD\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"HNSC\")\n",
    "GDSCR = GDSCR.loc[filter, ]\n",
    "GDSCR.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCR.index = GDSCR.index.astype(int)\n",
    "GDSCR = GDSCR.loc[GDSCR[\"drug\"] != \"Cetuximab\", ]\n",
    "\n",
    "GDSCR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "drug\n",
       "EKB-569                586\n",
       "Afatinib (rescreen)    575\n",
       "CUDC-101               571\n",
       "Afatinib               563\n",
       "Gefitinib              559\n",
       "Lapatinib              324\n",
       "Erlotinib              316\n",
       "HG-5-88-01             309\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print count of cell lines for each drug (EGFRi)\n",
    "\n",
    "GDSCR[\"drug\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs_homogenized\n"
     ]
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA expression datasets homogenized with GDSC expression dataset (Cetuximab)\n",
    "\n",
    "os.chdir(dataset_dir + \"exprs_homogenized/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100507206</th>\n",
       "      <th>100507254</th>\n",
       "      <th>100507436</th>\n",
       "      <th>100507472</th>\n",
       "      <th>100526773</th>\n",
       "      <th>100527978</th>\n",
       "      <th>100532746</th>\n",
       "      <th>100820829</th>\n",
       "      <th>102724473</th>\n",
       "      <th>105375355</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>3.60061063639454</td>\n",
       "      <td>3.07817969365007</td>\n",
       "      <td>6.32765157689639</td>\n",
       "      <td>2.93005596217687</td>\n",
       "      <td>2.64304491832882</td>\n",
       "      <td>2.56618657438719</td>\n",
       "      <td>8.09168714940745</td>\n",
       "      <td>2.55483284192774</td>\n",
       "      <td>8.59989127217117</td>\n",
       "      <td>2.80664192172726</td>\n",
       "      <td>...</td>\n",
       "      <td>2.5997416042245</td>\n",
       "      <td>3.06591431154376</td>\n",
       "      <td>5.50994883395808</td>\n",
       "      <td>2.93037262394265</td>\n",
       "      <td>2.93460472043208</td>\n",
       "      <td>2.44987572619243</td>\n",
       "      <td>2.48989773232543</td>\n",
       "      <td>3.88458609037183</td>\n",
       "      <td>3.95227909263729</td>\n",
       "      <td>2.46055443755884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>3.54613916595946</td>\n",
       "      <td>3.07817969365007</td>\n",
       "      <td>7.46956729139981</td>\n",
       "      <td>4.20521626673769</td>\n",
       "      <td>2.929569339555</td>\n",
       "      <td>5.57576806896234</td>\n",
       "      <td>8.20319464923893</td>\n",
       "      <td>2.41443776874882</td>\n",
       "      <td>7.7754694329959</td>\n",
       "      <td>3.34200309260696</td>\n",
       "      <td>...</td>\n",
       "      <td>2.67274979347372</td>\n",
       "      <td>3.11395605082</td>\n",
       "      <td>8.98094314882696</td>\n",
       "      <td>2.93037262394265</td>\n",
       "      <td>2.93460472043208</td>\n",
       "      <td>2.44987572619243</td>\n",
       "      <td>2.48989773232543</td>\n",
       "      <td>4.81457724903364</td>\n",
       "      <td>3.95227909263729</td>\n",
       "      <td>2.84309989182372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1156</th>\n",
       "      <td>6.99909567844156</td>\n",
       "      <td>8.47130432034951</td>\n",
       "      <td>6.15135176832441</td>\n",
       "      <td>2.36952374925443</td>\n",
       "      <td>3.6713478754664</td>\n",
       "      <td>2.53437362934482</td>\n",
       "      <td>8.43809032488933</td>\n",
       "      <td>2.53376073162296</td>\n",
       "      <td>7.11948011162002</td>\n",
       "      <td>3.28816715734563</td>\n",
       "      <td>...</td>\n",
       "      <td>2.72150643954151</td>\n",
       "      <td>3.17396413482904</td>\n",
       "      <td>6.58328775150283</td>\n",
       "      <td>2.93037262394265</td>\n",
       "      <td>2.93460472043208</td>\n",
       "      <td>2.44987572619243</td>\n",
       "      <td>2.48989773232543</td>\n",
       "      <td>4.74727922278015</td>\n",
       "      <td>4.15062701412473</td>\n",
       "      <td>2.46987656241725</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18217 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID         1                 2                 9          \\\n",
       "X-1027    3.60061063639454  3.07817969365007  6.32765157689639   \n",
       "X-1119    3.54613916595946  3.07817969365007  7.46956729139981   \n",
       "X-1156    6.99909567844156  8.47130432034951  6.15135176832441   \n",
       "\n",
       "ENTREZID         10                12                13         \\\n",
       "X-1027    2.93005596217687  2.64304491832882  2.56618657438719   \n",
       "X-1119    4.20521626673769    2.929569339555  5.57576806896234   \n",
       "X-1156    2.36952374925443   3.6713478754664  2.53437362934482   \n",
       "\n",
       "ENTREZID         14                15                16         \\\n",
       "X-1027    8.09168714940745  2.55483284192774  8.59989127217117   \n",
       "X-1119    8.20319464923893  2.41443776874882   7.7754694329959   \n",
       "X-1156    8.43809032488933  2.53376073162296  7.11948011162002   \n",
       "\n",
       "ENTREZID         18         ...         100507206         100507254  \\\n",
       "X-1027    2.80664192172726  ...   2.5997416042245  3.06591431154376   \n",
       "X-1119    3.34200309260696  ...  2.67274979347372     3.11395605082   \n",
       "X-1156    3.28816715734563  ...  2.72150643954151  3.17396413482904   \n",
       "\n",
       "ENTREZID         100507436         100507472         100526773  \\\n",
       "X-1027    5.50994883395808  2.93037262394265  2.93460472043208   \n",
       "X-1119    8.98094314882696  2.93037262394265  2.93460472043208   \n",
       "X-1156    6.58328775150283  2.93037262394265  2.93460472043208   \n",
       "\n",
       "ENTREZID         100527978         100532746         100820829  \\\n",
       "X-1027    2.44987572619243  2.48989773232543  3.88458609037183   \n",
       "X-1119    2.44987572619243  2.48989773232543  4.81457724903364   \n",
       "X-1156    2.44987572619243  2.48989773232543  4.74727922278015   \n",
       "\n",
       "ENTREZID         102724473         105375355  \n",
       "X-1027    3.95227909263729  2.46055443755884  \n",
       "X-1119    3.95227909263729  2.84309989182372  \n",
       "X-1156    4.15062701412473  2.46987656241725  \n",
       "\n",
       "[3 rows x 18217 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX expression dataset homogenized with GDSC expression dataset (Cetuximab)\n",
    "\n",
    "PDXEcetuximab = pd.read_csv(\"PDX_exprs.Cetuximab.eb_with.GDSC_exprs.Cetuximab.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "PDXEcetuximab = pd.DataFrame.transpose(PDXEcetuximab)\n",
    "PDXEcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100188893</th>\n",
       "      <th>100190940</th>\n",
       "      <th>100190949</th>\n",
       "      <th>100191040</th>\n",
       "      <th>100192386</th>\n",
       "      <th>100268168</th>\n",
       "      <th>100271715</th>\n",
       "      <th>100287718</th>\n",
       "      <th>100288778</th>\n",
       "      <th>100289635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-A4IG-01A</th>\n",
       "      <td>3.53364972782538</td>\n",
       "      <td>2.63169744477874</td>\n",
       "      <td>6.46913538384915</td>\n",
       "      <td>3.11954321356217</td>\n",
       "      <td>1.78425710174861</td>\n",
       "      <td>4.78953909500701</td>\n",
       "      <td>8.06446753523746</td>\n",
       "      <td>3.04479803527151</td>\n",
       "      <td>9.18974247446462</td>\n",
       "      <td>2.91775104542515</td>\n",
       "      <td>...</td>\n",
       "      <td>9.88694622925868</td>\n",
       "      <td>3.1254895591217</td>\n",
       "      <td>0</td>\n",
       "      <td>3.5074580817852</td>\n",
       "      <td>3.01824927947277</td>\n",
       "      <td>2.84422970206547</td>\n",
       "      <td>3.5035589781396</td>\n",
       "      <td>3.06737618948643</td>\n",
       "      <td>4.01010719689955</td>\n",
       "      <td>4.66295551081748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-4731-01A</th>\n",
       "      <td>4.12585378817146</td>\n",
       "      <td>2.1506582678597</td>\n",
       "      <td>6.44489273685529</td>\n",
       "      <td>3.21409376420109</td>\n",
       "      <td>3.15286503399873</td>\n",
       "      <td>3.28492483066951</td>\n",
       "      <td>8.22062958425597</td>\n",
       "      <td>2.67285528315238</td>\n",
       "      <td>8.51873835886811</td>\n",
       "      <td>5.1735664410497</td>\n",
       "      <td>...</td>\n",
       "      <td>9.26858559890324</td>\n",
       "      <td>3.08733295866026</td>\n",
       "      <td>0</td>\n",
       "      <td>3.30585011523952</td>\n",
       "      <td>3.3487497141691</td>\n",
       "      <td>2.6930439585389</td>\n",
       "      <td>2.45792584512192</td>\n",
       "      <td>2.99299508958328</td>\n",
       "      <td>3.94714478968898</td>\n",
       "      <td>4.17054508896713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-A49C-01A</th>\n",
       "      <td>4.37925369978313</td>\n",
       "      <td>3.43283585192836</td>\n",
       "      <td>6.59346803138021</td>\n",
       "      <td>3.11954321356217</td>\n",
       "      <td>3.31289809486857</td>\n",
       "      <td>3.10283527665291</td>\n",
       "      <td>8.2759744669096</td>\n",
       "      <td>2.74770994492163</td>\n",
       "      <td>8.98377132955025</td>\n",
       "      <td>3.96878580867183</td>\n",
       "      <td>...</td>\n",
       "      <td>9.71555609970402</td>\n",
       "      <td>3.0546486734275</td>\n",
       "      <td>0</td>\n",
       "      <td>3.67110924364761</td>\n",
       "      <td>3.04644627367641</td>\n",
       "      <td>2.61249547417883</td>\n",
       "      <td>3.00945260505516</td>\n",
       "      <td>3.06200290033863</td>\n",
       "      <td>5.05997145230091</td>\n",
       "      <td>4.13533467313432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18197 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID                 1                 2                 9          \\\n",
       "TCGA-BA-A4IG-01A  3.53364972782538  2.63169744477874  6.46913538384915   \n",
       "TCGA-CN-4731-01A  4.12585378817146   2.1506582678597  6.44489273685529   \n",
       "TCGA-CN-A49C-01A  4.37925369978313  3.43283585192836  6.59346803138021   \n",
       "\n",
       "ENTREZID                 10                12                13         \\\n",
       "TCGA-BA-A4IG-01A  3.11954321356217  1.78425710174861  4.78953909500701   \n",
       "TCGA-CN-4731-01A  3.21409376420109  3.15286503399873  3.28492483066951   \n",
       "TCGA-CN-A49C-01A  3.11954321356217  3.31289809486857  3.10283527665291   \n",
       "\n",
       "ENTREZID                 14                15                16         \\\n",
       "TCGA-BA-A4IG-01A  8.06446753523746  3.04479803527151  9.18974247446462   \n",
       "TCGA-CN-4731-01A  8.22062958425597  2.67285528315238  8.51873835886811   \n",
       "TCGA-CN-A49C-01A   8.2759744669096  2.74770994492163  8.98377132955025   \n",
       "\n",
       "ENTREZID                 18         ...         100188893         100190940  \\\n",
       "TCGA-BA-A4IG-01A  2.91775104542515  ...  9.88694622925868   3.1254895591217   \n",
       "TCGA-CN-4731-01A   5.1735664410497  ...  9.26858559890324  3.08733295866026   \n",
       "TCGA-CN-A49C-01A  3.96878580867183  ...  9.71555609970402   3.0546486734275   \n",
       "\n",
       "ENTREZID         100190949         100191040         100192386  \\\n",
       "TCGA-BA-A4IG-01A         0   3.5074580817852  3.01824927947277   \n",
       "TCGA-CN-4731-01A         0  3.30585011523952   3.3487497141691   \n",
       "TCGA-CN-A49C-01A         0  3.67110924364761  3.04644627367641   \n",
       "\n",
       "ENTREZID                 100268168         100271715         100287718  \\\n",
       "TCGA-BA-A4IG-01A  2.84422970206547   3.5035589781396  3.06737618948643   \n",
       "TCGA-CN-4731-01A   2.6930439585389  2.45792584512192  2.99299508958328   \n",
       "TCGA-CN-A49C-01A  2.61249547417883  3.00945260505516  3.06200290033863   \n",
       "\n",
       "ENTREZID                 100288778         100289635  \n",
       "TCGA-BA-A4IG-01A  4.01010719689955  4.66295551081748  \n",
       "TCGA-CN-4731-01A  3.94714478968898  4.17054508896713  \n",
       "TCGA-CN-A49C-01A  5.05997145230091  4.13533467313432  \n",
       "\n",
       "[3 rows x 18197 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA expression dataset homogenized with GDSC expression dataset (Cetuximab)\n",
    "\n",
    "TCGAEcetuximab = pd.read_csv(\"TCGA_exprs.Cetuximab.eb_with.GDSC_exprs.Cetuximab.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "TCGAEcetuximab = pd.DataFrame.transpose(TCGAEcetuximab)\n",
    "TCGAEcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/mutations'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX mutation dataset (Cetuximab)\n",
    "\n",
    "os.chdir(dataset_dir + \"mutations\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>3075.0</th>\n",
       "      <th>171017.0</th>\n",
       "      <th>2058.0</th>\n",
       "      <th>79890.0</th>\n",
       "      <th>7701.0</th>\n",
       "      <th>3104.0</th>\n",
       "      <th>4642.0</th>\n",
       "      <th>9778.0</th>\n",
       "      <th>100132406.0</th>\n",
       "      <th>23607.0</th>\n",
       "      <th>...</th>\n",
       "      <th>253738.0</th>\n",
       "      <th>154807.0</th>\n",
       "      <th>79140.0</th>\n",
       "      <th>6307.0</th>\n",
       "      <th>6352.0</th>\n",
       "      <th>56005.0</th>\n",
       "      <th>115650.0</th>\n",
       "      <th>64946.0</th>\n",
       "      <th>5777.0</th>\n",
       "      <th>253639.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1156</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 15010 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID 3075.0      171017.0    2058.0      79890.0     7701.0       \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1119           0.0       -0.05         0.0       -0.05         0.0   \n",
       "X-1156           0.0       -0.05         0.0       -0.05         0.0   \n",
       "\n",
       "ENTREZID 3104.0      4642.0      9778.0      100132406.0 23607.0      ...  \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0  ...   \n",
       "X-1119           0.0         0.0         0.0       -0.05         0.0  ...   \n",
       "X-1156           0.0         0.0       -0.05       -0.05         0.0  ...   \n",
       "\n",
       "ENTREZID 253738.0    154807.0    79140.0     6307.0      6352.0       \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1119           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1156           0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "ENTREZID 56005.0     115650.0    64946.0     5777.0      253639.0     \n",
       "X-1027           0.0         0.0         0.0         0.0         0.0  \n",
       "X-1119           0.0         0.0         0.0         0.0         0.0  \n",
       "X-1156           0.0         0.0         0.0         0.0         0.0  \n",
       "\n",
       "[3 rows x 15010 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX mutation dataset (Cetuximab)\n",
    "\n",
    "PDXMcetuximab = pd.read_csv(\"PDX_mutations.Cetuximab.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "PDXMcetuximab.drop_duplicates(keep='last')\n",
    "PDXMcetuximab = pd.DataFrame.transpose(PDXMcetuximab)\n",
    "PDXMcetuximab = PDXMcetuximab.loc[:, ~PDXMcetuximab.columns.duplicated()]\n",
    "\n",
    "PDXMcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>27136</th>\n",
       "      <th>84993</th>\n",
       "      <th>56834</th>\n",
       "      <th>10256</th>\n",
       "      <th>6672</th>\n",
       "      <th>253461</th>\n",
       "      <th>90134</th>\n",
       "      <th>162333</th>\n",
       "      <th>345630</th>\n",
       "      <th>10278</th>\n",
       "      <th>...</th>\n",
       "      <th>374395</th>\n",
       "      <th>23173</th>\n",
       "      <th>646498</th>\n",
       "      <th>23016</th>\n",
       "      <th>10888</th>\n",
       "      <th>8347</th>\n",
       "      <th>29901</th>\n",
       "      <th>9445</th>\n",
       "      <th>766</th>\n",
       "      <th>2944</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-A4IG-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-4731-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-A49C-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 15848 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 27136  84993  56834  10256  6672   253461 90134  162333  \\\n",
       "TCGA-BA-A4IG-01A    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-CN-4731-01A    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-CN-A49C-01A    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "                 345630 10278   ... 374395 23173  646498 23016  10888  8347    \\\n",
       "TCGA-BA-A4IG-01A    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-CN-4731-01A    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-CN-A49C-01A    0.0    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "                 29901  9445   766    2944    \n",
       "TCGA-BA-A4IG-01A    0.0    0.0    0.0    0.0  \n",
       "TCGA-CN-4731-01A    0.0    0.0    0.0    0.0  \n",
       "TCGA-CN-A49C-01A    0.0    0.0    0.0    0.0  \n",
       "\n",
       "[3 rows x 15848 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA mutation dataset (Cetuximab)\n",
    "\n",
    "TCGAMcetuximab = pd.read_csv(\"TCGA_mutations.Cetuximab.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "TCGAMcetuximab.drop_duplicates(keep='last')\n",
    "TCGAMcetuximab = pd.DataFrame.transpose(TCGAMcetuximab)\n",
    "TCGAMcetuximab = TCGAMcetuximab.loc[:, ~TCGAMcetuximab.columns.duplicated()]\n",
    "\n",
    "TCGAMcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX CNA dataset (Cetuximab)\n",
    "\n",
    "os.chdir(dataset_dir + \"CNA\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>...</th>\n",
       "      <th>101340252</th>\n",
       "      <th>102723547</th>\n",
       "      <th>102724473</th>\n",
       "      <th>103091865</th>\n",
       "      <th>105375355</th>\n",
       "      <th>107126285</th>\n",
       "      <th>109623460</th>\n",
       "      <th>109731405</th>\n",
       "      <th>118126072</th>\n",
       "      <th>128385369</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.1844245711374275</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7311832415721999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4276061727818994</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1156</th>\n",
       "      <td>0.9781956296816516</td>\n",
       "      <td>0.7004397181410922</td>\n",
       "      <td>0.7004397181410922</td>\n",
       "      <td>-0.2429767534925404</td>\n",
       "      <td>-0.2429767534925404</td>\n",
       "      <td>-0.2863041851566411</td>\n",
       "      <td>0.5607149544744789</td>\n",
       "      <td>0.3895668117627256</td>\n",
       "      <td>-0.2688167584278</td>\n",
       "      <td>-0.2688167584278</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4594316186372972</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.4246876693125631</td>\n",
       "      <td>-0.7369655941662062</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.100977647724821</td>\n",
       "      <td>-0.3770696490798233</td>\n",
       "      <td>-0.3219280948873623</td>\n",
       "      <td>-0.6551715030025588</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID           1                   2                   3          \\\n",
       "X-1027                   0.0                 0.0                 0.0   \n",
       "X-1119                   0.0                 0.0                 0.0   \n",
       "X-1156    0.9781956296816516  0.7004397181410922  0.7004397181410922   \n",
       "\n",
       "ENTREZID            9                    10                   12         \\\n",
       "X-1027                    0.0                  0.0                  0.0   \n",
       "X-1119                    0.0                  0.0                  0.0   \n",
       "X-1156    -0.2429767534925404  -0.2429767534925404  -0.2863041851566411   \n",
       "\n",
       "ENTREZID           13                  14                15         \\\n",
       "X-1027                   0.0                 0.0               0.0   \n",
       "X-1119                   0.0                 0.0               0.0   \n",
       "X-1156    0.5607149544744789  0.3895668117627256  -0.2688167584278   \n",
       "\n",
       "ENTREZID         16         ...           101340252 102723547  \\\n",
       "X-1027                 0.0  ...                 0.0       0.0   \n",
       "X-1119                 0.0  ...                 0.0       0.0   \n",
       "X-1156    -0.2688167584278  ...  0.4594316186372972       0.0   \n",
       "\n",
       "ENTREZID            102724473            103091865           105375355  \\\n",
       "X-1027                    0.0                  0.0                 0.0   \n",
       "X-1119                    0.0                  0.0  0.7311832415721999   \n",
       "X-1156    -0.4246876693125631  -0.7369655941662062                 0.0   \n",
       "\n",
       "ENTREZID            107126285            109623460            109731405  \\\n",
       "X-1027    -1.1844245711374275                  0.0                  0.0   \n",
       "X-1119                    0.0                  0.0                  0.0   \n",
       "X-1156      1.100977647724821  -0.3770696490798233  -0.3219280948873623   \n",
       "\n",
       "ENTREZID            118126072 128385369  \n",
       "X-1027                    0.0       0.0  \n",
       "X-1119     0.4276061727818994       0.0  \n",
       "X-1156    -0.6551715030025588       0.0  \n",
       "\n",
       "[3 rows x 23205 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX CNA dataset (Cetuximab)\n",
    "\n",
    "PDXCcetuximab = pd.read_csv(\"PDX_CNA.Cetuximab.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "PDXCcetuximab.drop_duplicates(keep='last')\n",
    "PDXCcetuximab = pd.DataFrame.transpose(PDXCcetuximab)\n",
    "PDXCcetuximab = PDXCcetuximab.loc[:, ~PDXCcetuximab.columns.duplicated()]\n",
    "\n",
    "PDXCcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>24</th>\n",
       "      <th>30</th>\n",
       "      <th>34</th>\n",
       "      <th>...</th>\n",
       "      <th>100652795</th>\n",
       "      <th>100652800</th>\n",
       "      <th>2104</th>\n",
       "      <th>9534</th>\n",
       "      <th>359762</th>\n",
       "      <th>730087</th>\n",
       "      <th>100101266</th>\n",
       "      <th>100419711</th>\n",
       "      <th>100421702</th>\n",
       "      <th>100505851</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-BA-A4IG-01A</th>\n",
       "      <td>0.3326</td>\n",
       "      <td>0.3326</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.504</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-4731-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3089</td>\n",
       "      <td>0.3089</td>\n",
       "      <td>0.3089</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4264</td>\n",
       "      <td>-0.5391</td>\n",
       "      <td>-0.5456</td>\n",
       "      <td>-0.5391</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-CN-A49C-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 31922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 2         3         9         10        11        12         \\\n",
       "TCGA-BA-A4IG-01A    0.3326    0.3326       0.0       0.0       0.0       0.0   \n",
       "TCGA-CN-4731-01A       0.0       0.0    0.3089    0.3089    0.3089       0.0   \n",
       "TCGA-CN-A49C-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "                 13        24        30        34         ... 100652795  \\\n",
       "TCGA-BA-A4IG-01A     0.504       0.0       0.0       0.0  ...       0.0   \n",
       "TCGA-CN-4731-01A    0.4264   -0.5391   -0.5456   -0.5391  ...       0.0   \n",
       "TCGA-CN-A49C-01A     0.763       0.0       0.0       0.0  ...       0.0   \n",
       "\n",
       "                 100652800 2104      9534      359762    730087    100101266  \\\n",
       "TCGA-BA-A4IG-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "TCGA-CN-4731-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "TCGA-CN-A49C-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "                 100419711 100421702 100505851  \n",
       "TCGA-BA-A4IG-01A       0.0       0.0       0.0  \n",
       "TCGA-CN-4731-01A       0.0       0.0       0.0  \n",
       "TCGA-CN-A49C-01A       0.0       0.0       0.0  \n",
       "\n",
       "[3 rows x 31922 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA CNA dataset (Cetuximab)\n",
    "\n",
    "TCGACcetuximab = pd.read_csv(\"TCGA_CNA.Cetuximab.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "TCGACcetuximab.drop_duplicates(keep='last')\n",
    "TCGACcetuximab = pd.DataFrame.transpose(TCGACcetuximab)\n",
    "TCGACcetuximab = TCGACcetuximab.loc[:, ~TCGACcetuximab.columns.duplicated()]\n",
    "\n",
    "TCGACcetuximab.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variance threshold for GDSC expression dataset (EGFRi)\n",
    "\n",
    "selector = VarianceThreshold(0.05)\n",
    "selector.fit_transform(GDSCE)\n",
    "GDSCE = GDSCE[GDSCE.columns[selector.get_support(indices=True)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fill NA values and binarize GDSC CNA and mutation datasets (EGFRi)\n",
    "\n",
    "GDSCM = GDSCM.fillna(0)\n",
    "GDSCM[GDSCM != 0.0] = 1\n",
    "GDSCC = GDSCC.fillna(0)\n",
    "GDSCC[GDSCC != 0.0] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC, PDX & TCGA # of common genes: 11223\n"
     ]
    }
   ],
   "source": [
    "# select shared genes between GDSC, PDX, and TCGA expression, mutation and CNA datasets \n",
    "\n",
    "ls = GDSCE.columns.intersection(GDSCM.columns)\n",
    "ls = ls.intersection(GDSCC.columns)\n",
    "ls = ls.intersection(PDXEcetuximab.columns)\n",
    "ls = ls.intersection(PDXMcetuximab.columns)\n",
    "ls = ls.intersection(PDXCcetuximab.columns)\n",
    "ls = ls.intersection(TCGAEcetuximab.columns)\n",
    "ls = ls.intersection(TCGAMcetuximab.columns)\n",
    "ls = ls.intersection(TCGACcetuximab.columns)\n",
    "\n",
    "ls = pd.unique(ls)\n",
    "\n",
    "print(f\"GDSC, PDX & TCGA # of common genes: {ls.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common cell lines: 611\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between GDSC expression, mutation, CNA, and response datasets (EGFRi)\n",
    "\n",
    "ls2 = GDSCE.index.intersection(GDSCM.index)\n",
    "ls2 = ls2.intersection(GDSCC.index)\n",
    "ls2 = ls2.intersection(GDSCR.index)\n",
    "\n",
    "print(f\"GDSC # of common cell lines: {ls2.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (611, 11223)\n",
      "GDSC # of common samples and genes (Mutation): (611, 11223)\n",
      "GDSC # of common samples and genes (CNA): (611, 11223)\n",
      "GDSC # of samples (Response): 3803\n"
     ]
    }
   ],
   "source": [
    "# subset shared genes and samples in GDSC expression, mutation, CNA, and response datasets\n",
    "\n",
    "GDSCE = GDSCE.loc[ls2, ls]\n",
    "GDSCM = GDSCM.loc[ls2, ls]\n",
    "GDSCC = GDSCC.loc[ls2, ls]\n",
    "GDSCR = GDSCR.loc[ls2, :]\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCE.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (Mutation): {GDSCM.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCC.shape}\")\n",
    "print(f\"GDSC # of samples (Response): {GDSCR.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples (Cetuximab): 60\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between PDX expression, mutation, and CNA datasets (Cetuximab) \n",
    "\n",
    "ls3 = PDXEcetuximab.index.intersection(PDXMcetuximab.index)\n",
    "ls3 = ls3.intersection(PDXCcetuximab.index)\n",
    "\n",
    "print(f\"PDX # of common samples (Cetuximab): {ls3.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA # of common samples (Cetuximab): 9\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between TCGA expression, mutation, and CNA datasets (Cetuximab) \n",
    "\n",
    "ls4 = TCGAEcetuximab.index.intersection(TCGAMcetuximab.index)\n",
    "ls4 = ls4.intersection(TCGACcetuximab.index)\n",
    "\n",
    "print(f\"TCGA # of common samples (Cetuximab): {ls4.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples and genes for Expression (Cetuximab): (60, 11223)\n",
      "PDX # of common samples and genes for Mutation (Cetuximab): (60, 11223)\n",
      "PDX # of common samples and genes for CNA (Cetuximab): (60, 11223)\n",
      "\n",
      "TCGA # of common samples and genes for Expression (Cetuximab): (9, 11223)\n",
      "TCGA # of common samples and genes for Mutation (Cetuximab): (9, 11223)\n",
      "TCGA # of common samples and genes for CNA (Cetuximab): (9, 11223)\n"
     ]
    }
   ],
   "source": [
    "# select shared genes and samples between PDX and TCGA expression, mutation, and CNA datasets (Cetuximab)\n",
    "\n",
    "PDXEcetuximab = PDXEcetuximab.loc[ls3, ls]\n",
    "PDXMcetuximab = PDXMcetuximab.loc[ls3, ls]\n",
    "PDXCcetuximab = PDXCcetuximab.loc[ls3, ls]\n",
    "\n",
    "TCGAEcetuximab = TCGAEcetuximab.loc[ls4, ls]\n",
    "TCGAMcetuximab = TCGAMcetuximab.loc[ls4, ls]\n",
    "TCGACcetuximab = TCGACcetuximab.loc[ls4, ls]\n",
    "\n",
    "print(f\"PDX # of common samples and genes for Expression (Cetuximab): {PDXEcetuximab.shape}\")\n",
    "print(f\"PDX # of common samples and genes for Mutation (Cetuximab): {PDXMcetuximab.shape}\")\n",
    "print(f\"PDX # of common samples and genes for CNA (Cetuximab): {PDXCcetuximab.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for Expression (Cetuximab): {TCGAEcetuximab.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for Mutation (Cetuximab): {TCGAMcetuximab.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for CNA (Cetuximab): {TCGACcetuximab.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/EGFRi\n"
     ]
    }
   ],
   "source": [
    "# change directory to read DEGs (EGFRi)\n",
    "\n",
    "os.chdir(DEGs_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1028 DEGs (p-adjusted < 0.05 and |logFC| > 1) among EGFR signaling inhibitors\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYOF</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TM4SF1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KRT18</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WWC1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAV1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Frequency\n",
       "0        MYOF          9\n",
       "1      TM4SF1          8\n",
       "2       KRT18          8\n",
       "3        WWC1          8\n",
       "4        CAV1          8"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read DEGs (EGFRi)\n",
    "\n",
    "DEGs_filtered_data = pd.read_excel(\"EGFRi_Differentially_Expressed_Genes (EnsemblID).xlsx\",\n",
    "                                    sheet_name = \"Common DEGs\")\n",
    "print(f\"There are {DEGs_filtered_data.shape[0]} DEGs (p-adjusted < 0.05 and |logFC| > 1) among EGFR signaling inhibitors\\n\")\n",
    "\n",
    "DEGs_filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>N4BP2L1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>PTPN12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>TEX19</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>DOCK2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene Symbol  Frequency\n",
       "607     N4BP2L1          1\n",
       "608      PTPN12          1\n",
       "609     ARHGAP4          1\n",
       "610       TEX19          1\n",
       "611       DOCK2          1"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset only DEGs with Frequency = 1\n",
    "\n",
    "filter = DEGs_filtered_data[\"Frequency\"] == 1\n",
    "DEGs_freq_one =  DEGs_filtered_data[filter]\n",
    "DEGs_freq_one.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 1 DEGs unique to cetuximab, so they were removed from the DEGs list\n"
     ]
    }
   ],
   "source": [
    "# read DEGs (Cetuximab) and select cetuximab-unique DEGs\n",
    "\n",
    "DEGs_cetuximab_data = pd.read_excel(\"EGFRi_Differentially_Expressed_Genes (EnsemblID).xlsx\",\n",
    "                                     sheet_name = \"Cetuximab\")\n",
    "\n",
    "filter = DEGs_cetuximab_data[\"Gene.Symbol\"].isin(DEGs_freq_one[\"Gene Symbol\"])\n",
    "only_cetuximab_degs = DEGs_cetuximab_data.loc[filter, \"Gene.Symbol\"]\n",
    "\n",
    "print(f'There are {only_cetuximab_degs.shape[0]} DEGs unique to cetuximab, so they were removed from the DEGs list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYOF</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TM4SF1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KRT18</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WWC1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAV1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Frequency\n",
       "0        MYOF          9\n",
       "1      TM4SF1          8\n",
       "2       KRT18          8\n",
       "3        WWC1          8\n",
       "4        CAV1          8"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove cetuximab-unique DEGs from feature set (MITOSISi)\n",
    "\n",
    "filter = DEGs_filtered_data[\"Gene Symbol\"].isin(only_cetuximab_degs)\n",
    "DEGs_filtered_data = DEGs_filtered_data[~filter]\n",
    "\n",
    "DEGs_filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2 input query terms found dup hits:\t[('CLECL1P', 2), ('CAST', 2)]\n",
      "1 input query terms found no hit:\t['SLC22A18']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1027"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene symbol from mygene database\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "DEGs_entrez_id = mg.querymany(DEGs_filtered_data[\"Gene Symbol\"], \n",
    "                              species=\"human\", \n",
    "                              scopes = \"symbol\",\n",
    "                              field = \"entrezgene\",\n",
    "                              as_dataframe=True)[\"entrezgene\"]\n",
    "DEGs_entrez_id = pd.unique(DEGs_entrez_id.dropna())\n",
    "DEGs_entrez_id = pd.Series(DEGs_entrez_id)\n",
    "\n",
    "result = mg.query(\"SLC22A18\", species=\"human\", scopes=\"symbol\", fields=\"entrezgene\")\n",
    "DEGs_entrez_id = pd.concat([DEGs_entrez_id, pd.Series(str(result[\"hits\"][1][\"entrezgene\"]))], ignore_index=True)\n",
    "\n",
    "DEGs_entrez_id.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign GDSC datasets to new variables\n",
    "\n",
    "exprs_z = GDSCE\n",
    "mut = GDSCM\n",
    "cna = GDSCC\n",
    "responses = GDSCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>1.24811569418395</td>\n",
       "      <td>Afatinib</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>3.82039284182925</td>\n",
       "      <td>Afatinib (rescreen)</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>910688</th>\n",
       "      <td>0</td>\n",
       "      <td>-2.24580154812369</td>\n",
       "      <td>CUDC-101</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        response            logIC50                 drug  exprs  CNA  \\\n",
       "910688         0   1.24811569418395             Afatinib      1    1   \n",
       "910688         0   3.82039284182925  Afatinib (rescreen)      1    1   \n",
       "910688         0  -2.24580154812369             CUDC-101      1    1   \n",
       "\n",
       "        mutations  \n",
       "910688          1  \n",
       "910688          1  \n",
       "910688          1  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GDSC response dataset\n",
    "\n",
    "responses.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Afatinib',\n",
       " 'Afatinib (rescreen)',\n",
       " 'CUDC-101',\n",
       " 'EKB-569',\n",
       " 'Erlotinib',\n",
       " 'Gefitinib',\n",
       " 'HG-5-88-01',\n",
       " 'Lapatinib'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list drug names in GDSC response dataset\n",
    "\n",
    "drugs = set(responses[\"drug\"].values)\n",
    "drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert Entrez gene IDs to string\n",
    "\n",
    "GDSCE.index = GDSCE.index.astype(str)\n",
    "GDSCM.index = GDSCM.index.astype(str)\n",
    "GDSCC.index = GDSCC.index.astype(str)\n",
    "responses.index = responses.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC\\nTissue descriptor 1\n",
       "leukemia             472\n",
       "urogenital_system    455\n",
       "lymphoma             431\n",
       "nervous_system       322\n",
       "lung_SCLC            308\n",
       "skin                 290\n",
       "breast               258\n",
       "digestive_system     231\n",
       "bone                 227\n",
       "neuroblastoma        193\n",
       "pancreas             134\n",
       "kidney               119\n",
       "myeloma              100\n",
       "soft_tissue           96\n",
       "lung                  88\n",
       "thyroid               79\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge GDSC cell line details with response dataset\n",
    "\n",
    "dat = pd.merge(GDSC_cell_line_details,\n",
    "         responses,\n",
    "         left_index = True, \n",
    "         right_index=True,\n",
    "         how = \"inner\")\n",
    "\n",
    "dat[\"GDSC\\nTissue descriptor 1\"].value_counts()\n",
    "#dat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC\\nTissue\\ndescriptor 2\n",
       "lung_small_cell_carcinoma    308\n",
       "glioma                       297\n",
       "melanoma                     270\n",
       "breast                       258\n",
       "B_cell_lymphoma              194\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get tissue descriptors\n",
    "\n",
    "dat[\"GDSC\\nTissue\\ndescriptor 2\"].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list count of cell lines as to tissues treated with cetuximab\n",
    "\n",
    "filter = (dat[\"drug\"] == \"Cetuximab\")\n",
    "dat.loc[filter,\"GDSC\\nTissue descriptor 1\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (3803, 708)\n",
      "GDSC # of common samples and genes (Mutation): (3803, 708)\n",
      "GDSC # of common samples and genes (CNA): (3803, 708)\n",
      "GDSC # of common samples (Response): (3803, 6)\n",
      "\n",
      "PDX # of common samples and genes for Cetuximab (Expression): (60, 708)\n",
      "PDX # of common samples and genes for Cetuximab (Mutation): (60, 708)\n",
      "PDX # of common samples and genes for Cetuximab (CNA): (60, 708)\n",
      "\n",
      "TCGA # of common samples and genes for Cetuximab (Expression): (9, 708)\n",
      "TCGA # of common samples and genes for Cetuximab (Mutation): (9, 708)\n",
      "TCGA # of common samples and genes for Cetuximab (CNA): (9, 708)\n"
     ]
    }
   ],
   "source": [
    "# filter GDSC expression, mutation, and CNA datasets (EGFRi) as to drugs\n",
    "# subset selected DEGs\n",
    "\n",
    "expression_zscores = []\n",
    "CNA = []\n",
    "mutations = []\n",
    "for drug in drugs:\n",
    "    samples = responses.loc[responses[\"drug\"] == drug, :].index.values\n",
    "    e_z = exprs_z.loc[samples, :]\n",
    "    c = cna.loc[samples, :]\n",
    "    m = mut.loc[samples, :]\n",
    "\n",
    "    expression_zscores.append(e_z)\n",
    "    CNA.append(c)\n",
    "    mutations.append(m)\n",
    "\n",
    "GDSCEv2 = pd.concat(expression_zscores, axis=0)\n",
    "GDSCMv2 = pd.concat(mutations, axis=0)\n",
    "GDSCCv2 = pd.concat(CNA, axis=0)\n",
    "GDSCRv2 = responses\n",
    "\n",
    "ls5 = list(set(GDSCE.columns).intersection(set(DEGs_entrez_id.astype(int))))\n",
    "ls6 = GDSCEv2.index.intersection(GDSCMv2.index)\n",
    "ls6 = ls6.intersection(GDSCCv2.index)\n",
    "\n",
    "GDSCEv2 = GDSCEv2.loc[ls6, ls5]\n",
    "GDSCMv2 = GDSCMv2.loc[ls6, ls5]\n",
    "GDSCCv2 = GDSCCv2.loc[ls6, ls5]\n",
    "GDSCRv2 = GDSCRv2.loc[ls6, :]\n",
    "\n",
    "PDXEcetuximab = PDXEcetuximab.loc[:,ls5]\n",
    "PDXMcetuximab = PDXMcetuximab.loc[:,ls5]\n",
    "PDXCcetuximab = PDXCcetuximab.loc[:,ls5]\n",
    "\n",
    "TCGAEcetuximab = TCGAEcetuximab.loc[:,ls5]\n",
    "TCGAMcetuximab = TCGAMcetuximab.loc[:,ls5]\n",
    "TCGACcetuximab = TCGACcetuximab.loc[:,ls5]\n",
    "\n",
    "responses.index = responses.index.values + \"_\" + responses[\"drug\"].values\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCEv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (Mutation): {GDSCMv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCCv2.shape}\")\n",
    "print(f\"GDSC # of common samples (Response): {GDSCRv2.shape}\\n\")\n",
    "\n",
    "print(f\"PDX # of common samples and genes for Cetuximab (Expression): {PDXEcetuximab.shape}\")\n",
    "print(f\"PDX # of common samples and genes for Cetuximab (Mutation): {PDXMcetuximab.shape}\")\n",
    "print(f\"PDX # of common samples and genes for Cetuximab (CNA): {PDXCcetuximab.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for Cetuximab (Expression): {TCGAEcetuximab.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for Cetuximab (Mutation): {TCGAMcetuximab.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for Cetuximab (CNA): {TCGACcetuximab.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 0 ... 0 0 0]\n"
     ]
    }
   ],
   "source": [
    "# GDSC response values\n",
    "\n",
    "Y = GDSCRv2['response'].values\n",
    "print(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA response datasets (Cetuximab)\n",
    "\n",
    "os.chdir(dataset_dir + \"response\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 60 samples with response data in the PDX (Cetuximab)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX response dataset (Cetuximab)\n",
    "\n",
    "PDXRcetuximab = pd.read_csv(\"PDX_response.Cetuximab.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "PDXRcetuximab.dropna(inplace=True)\n",
    "PDXRcetuximab.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "PDXRcetuximab[\"response\"] = PDXRcetuximab.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytscetuximab_pdx = PDXRcetuximab[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytscetuximab_pdx.shape[0]} samples with response data in the PDX (Cetuximab)\\n\")\n",
    "Ytscetuximab_pdx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-0933</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1055</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1156</th>\n",
       "      <td>NSCLC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tumor Type\n",
       "Model            \n",
       "X-0933        CRC\n",
       "X-1027        CRC\n",
       "X-1055        CRC\n",
       "X-1119        CRC\n",
       "X-1156      NSCLC"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX dataset to select cohorts (Cetuximab)\n",
    "\n",
    "pdx_cohort = pd.read_excel(pdx_data_dir + \"/nm.3954-S2.xlsx\",\n",
    "                              \"PCT raw data\")\n",
    "pdx_cohort = pdx_cohort.loc[pdx_cohort[\"Treatment\"] == \"cetuximab\",]\n",
    "pdx_cohort = pdx_cohort.loc[:,[\"Model\",\"Tumor Type\"]]\n",
    "pdx_cohort = pdx_cohort.drop_duplicates()\n",
    "pdx_cohort.set_index(\"Model\", inplace = True) \n",
    "pdx_cohort.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tumor Type\n",
       "CRC      36\n",
       "NSCLC    24\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged PDX cohort table with PDX response dataset (Cetuximab)\n",
    "\n",
    "dat_pdx = pd.merge(pdx_cohort,\n",
    "                   PDXRcetuximab,\n",
    "                   left_index = True,\n",
    "                   right_index = True)\n",
    "dat_pdx[\"Tumor Type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['X-1027', 'X-1119', 'X-1167', 'X-1173', 'X-1270', 'X-1290', 'X-1303',\n",
       "       'X-1329', 'X-1443', 'X-1479', 'X-1500', 'X-1536', 'X-1538', 'X-2145',\n",
       "       'X-2182', 'X-2239', 'X-2403', 'X-2483', 'X-2484', 'X-2538', 'X-2573',\n",
       "       'X-2659', 'X-2822', 'X-2846', 'X-2861', 'X-3093', 'X-3205', 'X-3224',\n",
       "       'X-3671', 'X-3792', 'X-4087', 'X-5254', 'X-5405', 'X-5446', 'X-5495',\n",
       "       'X-5578'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter CRC cohort samples (Cetuximab)\n",
    "\n",
    "filter = (dat_pdx[\"Tumor Type\"] == \"CRC\")\n",
    "pdx_crc_samples = dat_pdx.loc[filter,].index\n",
    "pdx_crc_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0])"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of CRC cohort samples (Cetuximab)\n",
    "\n",
    "Ytscetuximab_pdx_crc = PDXRcetuximab.loc[pdx_crc_samples,\"response\"].values\n",
    "Ytscetuximab_pdx_crc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    32\n",
       "1     4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistant CRC samples (Cetuximab)\n",
    "\n",
    "PDXRcetuximab.loc[pdx_crc_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['X-1156', 'X-1172', 'X-1323', 'X-1442', 'X-1499', 'X-1586', 'X-1658',\n",
       "       'X-1725', 'X-1787', 'X-1823', 'X-1834', 'X-1835', 'X-1934', 'X-1980',\n",
       "       'X-1993', 'X-2017', 'X-2042', 'X-2082', 'X-2088', 'X-2094', 'X-3029',\n",
       "       'X-3237', 'X-3843', 'X-4819'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter NSCLC cohort samples (Cetuximab)\n",
    "\n",
    "filter = (dat_pdx[\"Tumor Type\"] == \"NSCLC\")\n",
    "pdx_nsclc_samples = dat_pdx.loc[filter,].index\n",
    "pdx_nsclc_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,\n",
       "       0, 0])"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of NSCLC cohort samples (Cetuximab)\n",
    "\n",
    "Ytscetuximab_pdx_nsclc = PDXRcetuximab.loc[pdx_nsclc_samples,\"response\"].values\n",
    "Ytscetuximab_pdx_nsclc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    23\n",
       "1     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistant NSCLC samples (Cetuximab)\n",
    "\n",
    "PDXRcetuximab.loc[pdx_nsclc_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 9 samples with response data in the TCGA (Cetuximab)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1, 0, 1, 1, 1, 1, 0, 0, 1])"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA response dataset (Cetuximab)\n",
    "\n",
    "TCGARcetuximab = pd.read_csv(\"TCGA_response.Cetuximab.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "TCGARcetuximab.dropna(inplace=True)\n",
    "TCGARcetuximab.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "TCGARcetuximab[\"response\"] = TCGARcetuximab.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytscetuximab_tcga = TCGARcetuximab[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytscetuximab_tcga.shape[0]} samples with response data in the TCGA (Cetuximab)\\n\")\n",
    "Ytscetuximab_tcga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cohort\n",
       "HNSC    9\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TCGA cohort table with TCGA response dataset (Cetuximab)\n",
    "\n",
    "dat_tcga = TCGARcetuximab[\"cohort\"]\n",
    "\n",
    "dat_tcga.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCGA-BA-A4IG-01A', 'TCGA-CN-4731-01A', 'TCGA-CN-A49C-01A',\n",
       "       'TCGA-CN-A6V1-01A', 'TCGA-CR-7404-01A', 'TCGA-DQ-7596-01A',\n",
       "       'TCGA-HD-8224-01A', 'TCGA-KU-A6H8-01A', 'TCGA-RS-A6TO-01A'],\n",
       "      dtype='object', name='sample')"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter HNSC cohort samples (Cetuximab)\n",
    "\n",
    "filter = (TCGARcetuximab[\"cohort\"] == \"HNSC\")\n",
    "tcga_hnsc_samples = dat_tcga.loc[filter,].index\n",
    "tcga_hnsc_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 0, 1, 1, 1, 1, 0, 0, 1])"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of HNSC cohort samples (Cetuximab)\n",
    "\n",
    "Ytscetuximab_tcga_hnsc = TCGARcetuximab.loc[tcga_hnsc_samples,\"response\"].values\n",
    "Ytscetuximab_tcga_hnsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "1    6\n",
       "0    3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistance HNSC samples (Cetuximab)\n",
    "\n",
    "TCGARcetuximab.loc[tcga_hnsc_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maximum iteration\n",
    "\n",
    "max_iter = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use count of nodes and dropout rates found in trained pan-drug dataset\n",
    "\n",
    "hdm1 = 1024\n",
    "hdm2 = 512\n",
    "hdm3 = 1024\n",
    "rate1 = 0.8\n",
    "rate2 = 0.8\n",
    "rate3 = 0.7\n",
    "rate4 = 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC - Training AUC (EGFRi): 0.84\n",
      "GDSC - Training AUCPR (EGFRi): 0.41\n",
      "PDX - Test AUC (Cetuximab): 0.7\n",
      "PDX - Test AUCPR (Cetuximab): 0.18\n",
      "PDX - Test AUC (Cetuximab, CRC): 0.74\n",
      "PDX - Test AUCPR (Cetuximab, CRC): 0.23\n",
      "PDX - Test AUC (Cetuximab, NSCLC): 1.0\n",
      "PDX - Test AUCPR (Cetuximab, NSCLC): 1.0\n",
      "TCGA - Test AUC (Cetuximab): 0.61\n",
      "TCGA - Test AUCPR (Cetuximab): 0.79\n",
      "TCGA - Test AUC (Cetuximab, HNSC): 0.61\n",
      "TCGA - Test AUCPR (Cetuximab, HNSC): 0.79\n",
      "\n",
      "GDSC - Training AUC (EGFRi, Finetuned): 0.83\n",
      "GDSC - Training AUCPR (EGFRi, Finetuned): 0.41\n",
      "PDX - Test AUC (Cetuximab, Finetuned): 0.73\n",
      "PDX - Test AUCPR (Cetuximab, Finetuned): 0.2\n",
      "\n",
      "GDSC - Training AUC (EGFRi, Finetuned): 0.83\n",
      "GDSC - Training AUCPR (EGFRi, Finetuned): 0.39\n",
      "PDX - Test AUC (Cetuximab, CRC, Finetuned): 0.77\n",
      "PDX - Test AUCPR (Cetuximab, CRC, Finetuned): 0.3\n",
      "\n",
      "GDSC - Training AUC (EGFRi, Finetuned): 0.84\n",
      "GDSC - Training AUCPR (EGFRi, Finetuned): 0.41\n",
      "PDX - Test AUC (Cetuximab, NSCLC, Finetuned): 1.0\n",
      "PDX - Test AUCPR (Cetuximab, NSCLC, Finetuned): 1.0\n",
      "\n",
      "GDSC - Training AUC (EGFRi, Finetuned): 0.84\n",
      "GDSC - Training AUCPR (EGFRi, Finetuned): 0.41\n",
      "TCGA - Test AUC (Cetuximab, Finetuned): 0.67\n",
      "TCGA - Test AUCPR (Cetuximab, Finetuned): 0.84\n",
      "\n",
      "GDSC - Training AUC (EGFRi, Finetuned): 0.84\n",
      "GDSC - Training AUCPR (EGFRi, Finetuned): 0.41\n",
      "TCGA - Test AUC (Cetuximab, HNSC, Finetuned): 0.67\n",
      "TCGA - Test AUCPR (Cetuximab, HNSC, Finetuned): 0.84\n",
      "\n",
      "\n",
      "Models for maximum finetuning (Cetuximab, PDX)\n",
      "Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_26.pt\n",
      "Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_26.pt\n",
      "CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_26.pt\n",
      "Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_26.pt\n",
      "\n",
      "Models for maximum finetuning (Cetuximab, PDX, CRC)\n",
      "Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_23.pt\n",
      "Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_23.pt\n",
      "CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_23.pt\n",
      "Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_23.pt\n",
      "\n",
      "Models for maximum finetuning (Cetuximab, PDX, NSCLC)\n",
      "Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_4.pt\n",
      "Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_4.pt\n",
      "CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_4.pt\n",
      "Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_4.pt\n",
      "\n",
      "Models for maximum finetuning (Cetuximab, TCGA)\n",
      "Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "\n",
      "Models for maximum finetuning (Cetuximab, TCGA, HNSC)\n",
      "Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n",
      "Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_44.pt\n"
     ]
    }
   ],
   "source": [
    "# load pre-trained modules and make predictions\n",
    "\n",
    "scalerGDSC = sk.StandardScaler()\n",
    "scalerGDSC.fit(GDSCEv2.values)\n",
    "X_trainE = scalerGDSC.transform(GDSCEv2.values)\n",
    "X_testEcetuximab_pdx = scalerGDSC.transform(PDXEcetuximab.values) \n",
    "X_testEcetuximab_pdx_crc = scalerGDSC.transform(PDXEcetuximab.loc[pdx_crc_samples,].values) \n",
    "X_testEcetuximab_pdx_nsclc = scalerGDSC.transform(PDXEcetuximab.loc[pdx_nsclc_samples,].values) \n",
    "\n",
    "X_testEcetuximab_tcga = scalerGDSC.transform(TCGAEcetuximab.values)    \n",
    "X_testEcetuximab_tcga_hnsc = scalerGDSC.transform(TCGAEcetuximab.loc[tcga_hnsc_samples,].values)    \n",
    "\n",
    "X_trainM = np.nan_to_num(GDSCMv2.values)\n",
    "X_testMcetuximab_pdx = np.nan_to_num(PDXMcetuximab.values.astype(np.float32))\n",
    "X_testMcetuximab_pdx_crc = np.nan_to_num(PDXMcetuximab.loc[pdx_crc_samples,].values.astype(np.float32))\n",
    "X_testMcetuximab_pdx_nsclc = np.nan_to_num(PDXMcetuximab.loc[pdx_nsclc_samples,].values.astype(np.float32))\n",
    "X_testMcetuximab_tcga = np.nan_to_num(TCGAMcetuximab.values.astype(np.float32))\n",
    "X_testMcetuximab_tcga_hnsc = np.nan_to_num(TCGAMcetuximab.loc[tcga_hnsc_samples,].values.astype(np.float32))\n",
    "\n",
    "X_trainC = np.nan_to_num(GDSCCv2.values)\n",
    "X_testCcetuximab_pdx = np.nan_to_num(PDXCcetuximab.values.astype(np.float32))\n",
    "X_testCcetuximab_pdx_crc = np.nan_to_num(PDXCcetuximab.loc[pdx_crc_samples,].values.astype(np.float32))\n",
    "X_testCcetuximab_pdx_nsclc = np.nan_to_num(PDXCcetuximab.loc[pdx_nsclc_samples,].values.astype(np.float32))\n",
    "X_testCcetuximab_tcga = np.nan_to_num(TCGACcetuximab.values)\n",
    "X_testCcetuximab_tcga_hnsc = np.nan_to_num(TCGACcetuximab.loc[tcga_hnsc_samples,].values)\n",
    "\n",
    "TX_testEcetuximab_pdx = torch.FloatTensor(X_testEcetuximab_pdx)\n",
    "TX_testEcetuximab_pdx_crc = torch.FloatTensor(X_testEcetuximab_pdx_crc)\n",
    "TX_testEcetuximab_pdx_nsclc = torch.FloatTensor(X_testEcetuximab_pdx_nsclc)\n",
    "\n",
    "TX_testMcetuximab_pdx = torch.FloatTensor(X_testMcetuximab_pdx.astype(np.float32))\n",
    "TX_testMcetuximab_pdx_crc = torch.FloatTensor(X_testMcetuximab_pdx_crc.astype(np.float32))\n",
    "TX_testMcetuximab_pdx_nsclc = torch.FloatTensor(X_testMcetuximab_pdx_nsclc.astype(np.float32))\n",
    "\n",
    "TX_testCcetuximab_pdx = torch.FloatTensor(X_testCcetuximab_pdx.astype(np.float32))\n",
    "TX_testCcetuximab_pdx_crc = torch.FloatTensor(X_testCcetuximab_pdx_crc.astype(np.float32))\n",
    "TX_testCcetuximab_pdx_nsclc = torch.FloatTensor(X_testCcetuximab_pdx_nsclc.astype(np.float32))\n",
    "\n",
    "ty_testEcetuximab_pdx = torch.FloatTensor(Ytscetuximab_pdx.astype(int))\n",
    "ty_testEcetuximab_pdx_crc = torch.FloatTensor(Ytscetuximab_pdx_crc.astype(int))\n",
    "ty_testEcetuximab_pdx_nsclc = torch.FloatTensor(Ytscetuximab_pdx_nsclc.astype(int))\n",
    "\n",
    "TX_testEcetuximab_tcga = torch.FloatTensor(X_testEcetuximab_tcga)\n",
    "TX_testEcetuximab_tcga_hnsc = torch.FloatTensor(X_testEcetuximab_tcga_hnsc)\n",
    "\n",
    "TX_testMcetuximab_tcga = torch.FloatTensor(X_testMcetuximab_tcga.astype(np.float32))\n",
    "TX_testMcetuximab_tcga_hnsc = torch.FloatTensor(X_testMcetuximab_tcga_hnsc.astype(np.float32))\n",
    "\n",
    "TX_testCcetuximab_tcga = torch.FloatTensor(X_testCcetuximab_tcga.astype(np.float32))\n",
    "TX_testCcetuximab_tcga_hnsc = torch.FloatTensor(X_testCcetuximab_tcga_hnsc.astype(np.float32))\n",
    "\n",
    "ty_testEcetuximab_tcga = torch.FloatTensor(Ytscetuximab_tcga.astype(int))\n",
    "ty_testEcetuximab_tcga_hnsc = torch.FloatTensor(Ytscetuximab_tcga_hnsc.astype(int))\n",
    "\n",
    "n_sampE, IE_dim = X_trainE.shape\n",
    "n_sampM, IM_dim = X_trainM.shape\n",
    "n_sampC, IC_dim = X_trainC.shape\n",
    "\n",
    "h_dim1 = hdm1\n",
    "h_dim2 = hdm2\n",
    "h_dim3 = hdm3        \n",
    "Z_in = h_dim1 + h_dim2 + h_dim3\n",
    "\n",
    "class AEE(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEE, self).__init__()\n",
    "        self.EnE = torch.nn.Sequential(\n",
    "            nn.Linear(IE_dim, h_dim1),\n",
    "            nn.BatchNorm1d(h_dim1),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate1))\n",
    "    def forward(self, x):\n",
    "        output = self.EnE(x)\n",
    "        return output  \n",
    "\n",
    "\n",
    "class AEM(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEM, self).__init__()\n",
    "        self.EnM = torch.nn.Sequential(\n",
    "            nn.Linear(IM_dim, h_dim2),\n",
    "            nn.BatchNorm1d(h_dim2),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate2))\n",
    "    def forward(self, x):\n",
    "        output = self.EnM(x)\n",
    "        return output       \n",
    "\n",
    "class AEC(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEC, self).__init__()\n",
    "        self.EnC = torch.nn.Sequential(\n",
    "            nn.Linear(IC_dim, h_dim3),\n",
    "            nn.BatchNorm1d(h_dim3),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate3))\n",
    "    def forward(self, x):\n",
    "        output = self.EnC(x)\n",
    "        return output \n",
    "        \n",
    "class Classifier(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Classifier, self).__init__()\n",
    "        self.FC = torch.nn.Sequential(\n",
    "            nn.Linear(Z_in, 1),\n",
    "            nn.Dropout(rate4),\n",
    "            nn.Sigmoid())\n",
    "    def forward(self, x):\n",
    "        return self.FC(x)\n",
    "\n",
    "torch.cuda.manual_seed_all(42)\n",
    "\n",
    "AutoencoderE = torch.load(models_dir + \"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy.pt\")\n",
    "AutoencoderM = torch.load(models_dir + \"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy.pt\")\n",
    "AutoencoderC = torch.load(models_dir + \"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy.pt\")\n",
    "\n",
    "Clas = torch.load(models_dir + \"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy.pt\")\n",
    "\n",
    "AutoencoderE.eval()\n",
    "AutoencoderM.eval()\n",
    "AutoencoderC.eval()\n",
    "Clas.eval()\n",
    "\n",
    "ZEX = AutoencoderE(torch.FloatTensor(X_trainE))\n",
    "ZMX = AutoencoderM(torch.FloatTensor(X_trainM))\n",
    "ZCX = AutoencoderC(torch.FloatTensor(X_trainC))\n",
    "ZTX = torch.cat((ZEX, ZMX, ZCX), 1)\n",
    "ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "PredX = Clas(ZTX)\n",
    "AUCT = roc_auc_score(Y, PredX.detach().numpy())\n",
    "AUCTaucpr = average_precision_score(Y, PredX.detach().numpy())\n",
    "\n",
    "ZETcetuximab_pdx = AutoencoderE(TX_testEcetuximab_pdx)\n",
    "ZMTcetuximab_pdx = AutoencoderM(TX_testMcetuximab_pdx)\n",
    "ZCTcetuximab_pdx = AutoencoderC(TX_testCcetuximab_pdx)\n",
    "ZTTcetuximab_pdx = torch.cat((ZETcetuximab_pdx, ZMTcetuximab_pdx, ZCTcetuximab_pdx), 1)\n",
    "ZTTcetuximab_pdx = F.normalize(ZTTcetuximab_pdx, p=2, dim=0)\n",
    "PredTcetuximab_pdx = Clas(ZTTcetuximab_pdx)\n",
    "AUCTcetuximab_pdx = roc_auc_score(Ytscetuximab_pdx.astype(int), PredTcetuximab_pdx.detach().numpy())\n",
    "AUCTcetuximab_pdx_aucpr = average_precision_score(Ytscetuximab_pdx.astype(int), PredTcetuximab_pdx.detach().numpy())\n",
    "\n",
    "ZETcetuximab_pdx_crc = AutoencoderE(TX_testEcetuximab_pdx_crc)\n",
    "ZMTcetuximab_pdx_crc = AutoencoderM(TX_testMcetuximab_pdx_crc)\n",
    "ZCTcetuximab_pdx_crc = AutoencoderC(TX_testCcetuximab_pdx_crc)\n",
    "ZTTcetuximab_pdx_crc = torch.cat((ZETcetuximab_pdx_crc, ZMTcetuximab_pdx_crc, ZCTcetuximab_pdx_crc), 1)\n",
    "ZTTcetuximab_pdx_crc = F.normalize(ZTTcetuximab_pdx_crc, p=2, dim=0)\n",
    "PredTcetuximab_pdx_crc = Clas(ZTTcetuximab_pdx_crc)\n",
    "AUCTcetuximab_pdx_crc = roc_auc_score(Ytscetuximab_pdx_crc.astype(int), PredTcetuximab_pdx_crc.detach().numpy())\n",
    "AUCTcetuximab_pdx_crc_aucpr = average_precision_score(Ytscetuximab_pdx_crc.astype(int), PredTcetuximab_pdx_crc.detach().numpy())\n",
    "\n",
    "ZETcetuximab_pdx_nsclc = AutoencoderE(TX_testEcetuximab_pdx_nsclc)\n",
    "ZMTcetuximab_pdx_nsclc = AutoencoderM(TX_testMcetuximab_pdx_nsclc)\n",
    "ZCTcetuximab_pdx_nsclc = AutoencoderC(TX_testCcetuximab_pdx_nsclc)\n",
    "ZTTcetuximab_pdx_nsclc = torch.cat((ZETcetuximab_pdx_nsclc, ZMTcetuximab_pdx_nsclc, ZCTcetuximab_pdx_nsclc), 1)\n",
    "ZTTcetuximab_pdx_nsclc = F.normalize(ZTTcetuximab_pdx_nsclc, p=2, dim=0)\n",
    "PredTcetuximab_pdx_nsclc = Clas(ZTTcetuximab_pdx_nsclc)\n",
    "AUCTcetuximab_pdx_nsclc = roc_auc_score(Ytscetuximab_pdx_nsclc.astype(int), PredTcetuximab_pdx_nsclc.detach().numpy())\n",
    "AUCTcetuximab_pdx_nsclc_aucpr = average_precision_score(Ytscetuximab_pdx_nsclc.astype(int), PredTcetuximab_pdx_nsclc.detach().numpy())\n",
    "\n",
    "ZETcetuximab_tcga = AutoencoderE(TX_testEcetuximab_tcga)\n",
    "ZMTcetuximab_tcga = AutoencoderM(TX_testMcetuximab_tcga)\n",
    "ZCTcetuximab_tcga = AutoencoderC(TX_testCcetuximab_tcga)\n",
    "ZTTcetuximab_tcga = torch.cat((ZETcetuximab_tcga, ZMTcetuximab_tcga, ZCTcetuximab_tcga), 1)\n",
    "ZTTcetuximab_tcga = F.normalize(ZTTcetuximab_tcga, p=2, dim=0)\n",
    "PredTcetuximab_tcga = Clas(ZTTcetuximab_tcga)\n",
    "AUCTcetuximab_tcga = roc_auc_score(Ytscetuximab_tcga.astype(int), PredTcetuximab_tcga.detach().numpy())\n",
    "AUCTcetuximab_tcga_aucpr = average_precision_score(Ytscetuximab_tcga.astype(int), PredTcetuximab_tcga.detach().numpy())\n",
    "\n",
    "ZETcetuximab_tcga_hnsc = AutoencoderE(TX_testEcetuximab_tcga_hnsc)\n",
    "ZMTcetuximab_tcga_hnsc = AutoencoderM(TX_testMcetuximab_tcga_hnsc)\n",
    "ZCTcetuximab_tcga_hnsc = AutoencoderC(TX_testCcetuximab_tcga_hnsc)\n",
    "ZTTcetuximab_tcga_hnsc = torch.cat((ZETcetuximab_tcga_hnsc, ZMTcetuximab_tcga_hnsc, ZCTcetuximab_tcga_hnsc), 1)\n",
    "ZTTcetuximab_tcga_hnsc = F.normalize(ZTTcetuximab_tcga_hnsc, p=2, dim=0)\n",
    "PredTcetuximab_tcga_hnsc = Clas(ZTTcetuximab_tcga_hnsc)\n",
    "AUCTcetuximab_tcga_hnsc = roc_auc_score(Ytscetuximab_tcga_hnsc.astype(int), PredTcetuximab_tcga_hnsc.detach().numpy())\n",
    "AUCTcetuximab_tcga_hnsc_aucpr = average_precision_score(Ytscetuximab_tcga_hnsc.astype(int), PredTcetuximab_tcga_hnsc.detach().numpy())\n",
    "\n",
    "AUCT_finetuned = dict()\n",
    "AUCTaucpr_finetuned = dict()\n",
    "AUCTcetuximab_pdx_finetuned = dict()\n",
    "AUCTcetuximab_tcga_finetuned = dict()\n",
    "AUCTcetuximab_pdx_crc_finetuned = dict()\n",
    "AUCTcetuximab_pdx_nsclc_finetuned = dict()\n",
    "AUCTcetuximab_tcga_hnsc_finetuned = dict()\n",
    "AUCTcetuximab_pdx_aucpr_finetuned = dict()\n",
    "AUCTcetuximab_tcga_aucpr_finetuned = dict()\n",
    "AUCTcetuximab_pdx_crc_aucpr_finetuned = dict()\n",
    "AUCTcetuximab_pdx_nsclc_aucpr_finetuned = dict()\n",
    "AUCTcetuximab_tcga_hnsc_aucpr_finetuned = dict()\n",
    "for iter in range(max_iter):\n",
    "   iter += 1  \n",
    "\n",
    "   AutoencoderE_finetuned = torch.load(finetuned_dir + f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy_Finetuned_{iter}.pt\")\n",
    "   AutoencoderM_finetuned = torch.load(finetuned_dir + f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy_Finetuned_{iter}.pt\")\n",
    "   AutoencoderC_finetuned = torch.load(finetuned_dir + f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy_Finetuned_{iter}.pt\")\n",
    "   Clas_finetuned = torch.load(finetuned_dir + f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Fourth_Strategy_Finetuned_{iter}.pt\")\n",
    "   \n",
    "   AutoencoderE_finetuned.eval()\n",
    "   AutoencoderM_finetuned.eval()\n",
    "   AutoencoderC_finetuned.eval()\n",
    "   Clas_finetuned.eval()\n",
    "      \n",
    "   ZEX = AutoencoderE_finetuned(torch.FloatTensor(X_trainE))\n",
    "   ZMX = AutoencoderM_finetuned(torch.FloatTensor(X_trainM))\n",
    "   ZCX = AutoencoderC_finetuned(torch.FloatTensor(X_trainC))\n",
    "   ZTX = torch.cat((ZEX, ZMX, ZCX), 1)\n",
    "   ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "   PredX = Clas_finetuned(ZTX)\n",
    "   AUCT_finetuned[iter] = roc_auc_score(Y, PredX.detach().numpy())\n",
    "   AUCTaucpr_finetuned[iter] = average_precision_score(Y, PredX.detach().numpy())\n",
    "   \n",
    "   ZETcetuximab_pdx = AutoencoderE_finetuned(TX_testEcetuximab_pdx)\n",
    "   ZMTcetuximab_pdx = AutoencoderM_finetuned(TX_testMcetuximab_pdx)\n",
    "   ZCTcetuximab_pdx = AutoencoderC_finetuned(TX_testCcetuximab_pdx)\n",
    "   ZTTcetuximab_pdx = torch.cat((ZETcetuximab_pdx, ZMTcetuximab_pdx, ZCTcetuximab_pdx), 1)\n",
    "   ZTTcetuximab_pdx = F.normalize(ZTTcetuximab_pdx, p=2, dim=0)\n",
    "   PredTcetuximab_pdx = Clas_finetuned(ZTTcetuximab_pdx)\n",
    "   AUCTcetuximab_pdx_finetuned[iter] = roc_auc_score(Ytscetuximab_pdx.astype(int), PredTcetuximab_pdx.detach().numpy())\n",
    "   AUCTcetuximab_pdx_aucpr_finetuned[iter] = average_precision_score(Ytscetuximab_pdx.astype(int), PredTcetuximab_pdx.detach().numpy())\n",
    "\n",
    "   ZETcetuximab_pdx_crc = AutoencoderE_finetuned(TX_testEcetuximab_pdx_crc)\n",
    "   ZMTcetuximab_pdx_crc = AutoencoderM_finetuned(TX_testMcetuximab_pdx_crc)\n",
    "   ZCTcetuximab_pdx_crc = AutoencoderC_finetuned(TX_testCcetuximab_pdx_crc)\n",
    "   ZTTcetuximab_pdx_crc = torch.cat((ZETcetuximab_pdx_crc, ZMTcetuximab_pdx_crc, ZCTcetuximab_pdx_crc), 1)\n",
    "   ZTTcetuximab_pdx_crc = F.normalize(ZTTcetuximab_pdx_crc, p=2, dim=0)\n",
    "   PredTcetuximab_pdx_crc = Clas_finetuned(ZTTcetuximab_pdx_crc)\n",
    "   AUCTcetuximab_pdx_crc_finetuned[iter] = roc_auc_score(Ytscetuximab_pdx_crc.astype(int), PredTcetuximab_pdx_crc.detach().numpy())\n",
    "   AUCTcetuximab_pdx_crc_aucpr_finetuned[iter] = average_precision_score(Ytscetuximab_pdx_crc.astype(int), PredTcetuximab_pdx_crc.detach().numpy())\n",
    "\n",
    "   ZETcetuximab_pdx_nsclc = AutoencoderE_finetuned(TX_testEcetuximab_pdx_nsclc)\n",
    "   ZMTcetuximab_pdx_nsclc = AutoencoderM_finetuned(TX_testMcetuximab_pdx_nsclc)\n",
    "   ZCTcetuximab_pdx_nsclc = AutoencoderC_finetuned(TX_testCcetuximab_pdx_nsclc)\n",
    "   ZTTcetuximab_pdx_nsclc = torch.cat((ZETcetuximab_pdx_nsclc, ZMTcetuximab_pdx_nsclc, ZCTcetuximab_pdx_nsclc), 1)\n",
    "   ZTTcetuximab_pdx_nsclc = F.normalize(ZTTcetuximab_pdx_nsclc, p=2, dim=0)\n",
    "   PredTcetuximab_pdx_nsclc = Clas_finetuned(ZTTcetuximab_pdx_nsclc)\n",
    "   AUCTcetuximab_pdx_nsclc_finetuned[iter] = roc_auc_score(Ytscetuximab_pdx_nsclc.astype(int), PredTcetuximab_pdx_nsclc.detach().numpy())\n",
    "   AUCTcetuximab_pdx_nsclc_aucpr_finetuned[iter] = average_precision_score(Ytscetuximab_pdx_nsclc.astype(int), PredTcetuximab_pdx_nsclc.detach().numpy())\n",
    "\n",
    "   ZETcetuximab_tcga = AutoencoderE_finetuned(TX_testEcetuximab_tcga)\n",
    "   ZMTcetuximab_tcga = AutoencoderM_finetuned(TX_testMcetuximab_tcga)\n",
    "   ZCTcetuximab_tcga = AutoencoderC_finetuned(TX_testCcetuximab_tcga)\n",
    "   ZTTcetuximab_tcga = torch.cat((ZETcetuximab_tcga, ZMTcetuximab_tcga, ZCTcetuximab_tcga), 1)\n",
    "   ZTTcetuximab_tcga = F.normalize(ZTTcetuximab_tcga, p=2, dim=0)\n",
    "   PredTcetuximab_tcga = Clas_finetuned(ZTTcetuximab_tcga)\n",
    "   AUCTcetuximab_tcga_finetuned[iter] = roc_auc_score(Ytscetuximab_tcga.astype(int), PredTcetuximab_tcga.detach().numpy())\n",
    "   AUCTcetuximab_tcga_aucpr_finetuned[iter] = average_precision_score(Ytscetuximab_tcga.astype(int), PredTcetuximab_tcga.detach().numpy())\n",
    "\n",
    "   ZETcetuximab_tcga_hnsc = AutoencoderE_finetuned(TX_testEcetuximab_tcga_hnsc)\n",
    "   ZMTcetuximab_tcga_hnsc = AutoencoderM_finetuned(TX_testMcetuximab_tcga_hnsc)\n",
    "   ZCTcetuximab_tcga_hnsc = AutoencoderC_finetuned(TX_testCcetuximab_tcga_hnsc)\n",
    "   ZTTcetuximab_tcga_hnsc = torch.cat((ZETcetuximab_tcga_hnsc, ZMTcetuximab_tcga_hnsc, ZCTcetuximab_tcga_hnsc), 1)\n",
    "   ZTTcetuximab_tcga_hnsc = F.normalize(ZTTcetuximab_tcga_hnsc, p=2, dim=0)\n",
    "   PredTcetuximab_tcga_hnsc = Clas_finetuned(ZTTcetuximab_tcga_hnsc)\n",
    "   AUCTcetuximab_tcga_hnsc_finetuned[iter] = roc_auc_score(Ytscetuximab_tcga_hnsc.astype(int), PredTcetuximab_tcga_hnsc.detach().numpy())\n",
    "   AUCTcetuximab_tcga_hnsc_aucpr_finetuned[iter] = average_precision_score(Ytscetuximab_tcga_hnsc.astype(int), PredTcetuximab_tcga_hnsc.detach().numpy())\n",
    "\n",
    "max_key_pdx = max(AUCTcetuximab_pdx_aucpr_finetuned, key=AUCTcetuximab_pdx_aucpr_finetuned.get)\n",
    "max_key_pdx_crc = max(AUCTcetuximab_pdx_crc_aucpr_finetuned, key=AUCTcetuximab_pdx_crc_aucpr_finetuned.get)\n",
    "max_key_pdx_nsclc = max(AUCTcetuximab_pdx_nsclc_aucpr_finetuned, key=AUCTcetuximab_pdx_nsclc_aucpr_finetuned.get)\n",
    "max_key_tcga = max(AUCTcetuximab_tcga_aucpr_finetuned, key=AUCTcetuximab_tcga_aucpr_finetuned.get)\n",
    "max_key_tcga_hnsc = max(AUCTcetuximab_tcga_hnsc_aucpr_finetuned, key=AUCTcetuximab_tcga_hnsc_aucpr_finetuned.get)\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi): {round(AUCT,2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi): {round(AUCTaucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab): {round(AUCTcetuximab_pdx,2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab): {round(AUCTcetuximab_pdx_aucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab, CRC): {round(AUCTcetuximab_pdx_crc,2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab, CRC): {round(AUCTcetuximab_pdx_crc_aucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab, NSCLC): {round(AUCTcetuximab_pdx_nsclc,2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab, NSCLC): {round(AUCTcetuximab_pdx_nsclc_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (Cetuximab): {round(AUCTcetuximab_tcga,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Cetuximab): {round(AUCTcetuximab_tcga_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (Cetuximab, HNSC): {round(AUCTcetuximab_tcga_hnsc,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Cetuximab, HNSC): {round(AUCTcetuximab_tcga_hnsc_aucpr,2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi, Finetuned): {round(AUCT_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab, Finetuned): {round(AUCTcetuximab_pdx_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab, Finetuned): {round(AUCTcetuximab_pdx_aucpr_finetuned[max_key_pdx],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi, Finetuned): {round(AUCT_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab, CRC, Finetuned): {round(AUCTcetuximab_pdx_crc_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab, CRC, Finetuned): {round(AUCTcetuximab_pdx_crc_aucpr_finetuned[max_key_pdx_crc],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi, Finetuned): {round(AUCT_finetuned[max_key_pdx_nsclc],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx_nsclc],2)}\")\n",
    "print(f\"PDX - Test AUC (Cetuximab, NSCLC, Finetuned): {round(AUCTcetuximab_pdx_nsclc_finetuned[max_key_pdx_nsclc],2)}\")\n",
    "print(f\"PDX - Test AUCPR (Cetuximab, NSCLC, Finetuned): {round(AUCTcetuximab_pdx_nsclc_aucpr_finetuned[max_key_pdx_nsclc],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi, Finetuned): {round(AUCT_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUC (Cetuximab, Finetuned): {round(AUCTcetuximab_tcga_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Cetuximab, Finetuned): {round(AUCTcetuximab_tcga_aucpr_finetuned[max_key_tcga],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (EGFRi, Finetuned): {round(AUCT_finetuned[max_key_tcga_hnsc],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (EGFRi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga_hnsc],2)}\")\n",
    "print(f\"TCGA - Test AUC (Cetuximab, HNSC, Finetuned): {round(AUCTcetuximab_tcga_hnsc_finetuned[max_key_tcga_hnsc],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Cetuximab, HNSC, Finetuned): {round(AUCTcetuximab_tcga_hnsc_aucpr_finetuned[max_key_tcga_hnsc],2)}\\n\\n\")\n",
    "\n",
    "print(\"Models for maximum finetuning (Cetuximab, PDX)\")\n",
    "print(f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Cetuximab, PDX, CRC)\")\n",
    "print(f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Cetuximab, PDX, NSCLC)\")\n",
    "print(f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_nsclc}.pt\")\n",
    "print(f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_nsclc}.pt\")\n",
    "print(f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_nsclc}.pt\")\n",
    "print(f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_pdx_nsclc}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Cetuximab, TCGA)\")\n",
    "print(f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Cetuximab, TCGA, HNSC)\")\n",
    "print(f\"Exprs_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga_hnsc}.pt\")\n",
    "print(f\"Mut_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga_hnsc}.pt\")\n",
    "print(f\"CNA_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga_hnsc}.pt\")\n",
    "print(f\"Class_Cetuximab_GDSC_PDX_TCGA_EGFRi_Expression_Mutation_CNA_Finetuned_{max_key_tcga_hnsc}.pt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>AUC (GDSC)</th>\n",
       "      <th>AUCPR (GDSC)</th>\n",
       "      <th>AUC (PDX)</th>\n",
       "      <th>AUCPR (PDX)</th>\n",
       "      <th>AUC (PDX-CRC)</th>\n",
       "      <th>AUCPR (PDX-CRC)</th>\n",
       "      <th>AUC (PDX-NSCLC)</th>\n",
       "      <th>AUCPR (PDX-NSCLC)</th>\n",
       "      <th>AUC (TCGA)</th>\n",
       "      <th>AUCPR (TCGA)</th>\n",
       "      <th>AUC (TCGA-HNSC)</th>\n",
       "      <th>AUCPR (TCGA-HNSC)</th>\n",
       "      <th>Sample Size (GDSC)</th>\n",
       "      <th>Feature Size (GDSC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Expr + Mut + CNA</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.23</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.61</td>\n",
       "      <td>0.79</td>\n",
       "      <td>3803</td>\n",
       "      <td>708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Expr + Mut + CNA (Fine-tuned)</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.73</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.77</td>\n",
       "      <td>0.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.84</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.84</td>\n",
       "      <td>3803</td>\n",
       "      <td>708</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Data AUC (GDSC) AUCPR (GDSC) AUC (PDX)  \\\n",
       "0               Expr + Mut + CNA       0.84         0.41       0.7   \n",
       "1  Expr + Mut + CNA (Fine-tuned)       0.83         0.41      0.73   \n",
       "\n",
       "  AUCPR (PDX) AUC (PDX-CRC) AUCPR (PDX-CRC) AUC (PDX-NSCLC) AUCPR (PDX-NSCLC)  \\\n",
       "0        0.18          0.74            0.23             1.0               1.0   \n",
       "1         0.2          0.77             0.3             1.0               1.0   \n",
       "\n",
       "  AUC (TCGA) AUCPR (TCGA) AUC (TCGA-HNSC) AUCPR (TCGA-HNSC)  \\\n",
       "0       0.61         0.79            0.61              0.79   \n",
       "1       0.67         0.84            0.67              0.84   \n",
       "\n",
       "  Sample Size (GDSC) Feature Size (GDSC)  \n",
       "0               3803                 708  \n",
       "1               3803                 708  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe with the prediction results \n",
    "\n",
    "emc  = {\"Data\": [\"Expr + Mut + CNA\", \"Expr + Mut + CNA (Fine-tuned)\"],\n",
    "       \"AUC (GDSC)\": [round(AUCT, 2), round(AUCT_finetuned[max_key_pdx], 2)],\n",
    "       \"AUCPR (GDSC)\": [round(AUCTaucpr,2), round(AUCTaucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX)\": [round(AUCTcetuximab_pdx,2), round(AUCTcetuximab_pdx_finetuned[max_key_pdx],2)],\n",
    "       \"AUCPR (PDX)\":  [round(AUCTcetuximab_pdx_aucpr,2), round(AUCTcetuximab_pdx_aucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX-CRC)\":  [round(AUCTcetuximab_pdx_crc,2), round(AUCTcetuximab_pdx_crc_finetuned[max_key_pdx_crc],2)],\n",
    "       \"AUCPR (PDX-CRC)\": [round(AUCTcetuximab_pdx_crc_aucpr,2), round(AUCTcetuximab_pdx_crc_aucpr_finetuned[max_key_pdx_crc],2)],\n",
    "       \"AUC (PDX-NSCLC)\":  [round(AUCTcetuximab_pdx_nsclc,2), round(AUCTcetuximab_pdx_nsclc_finetuned[max_key_pdx_nsclc],2)],\n",
    "       \"AUCPR (PDX-NSCLC)\": [round(AUCTcetuximab_pdx_nsclc_aucpr,2), round(AUCTcetuximab_pdx_nsclc_aucpr_finetuned[max_key_pdx_nsclc],2)],\n",
    "       \"AUC (TCGA)\": [round(AUCTcetuximab_tcga,2), round(AUCTcetuximab_tcga_finetuned[max_key_tcga],2)],\n",
    "       \"AUCPR (TCGA)\":  [round(AUCTcetuximab_tcga_aucpr,2), round(AUCTcetuximab_tcga_aucpr_finetuned[max_key_tcga],2)],\n",
    "       \"AUC (TCGA-HNSC)\":  [round(AUCTcetuximab_tcga_hnsc,2), round(AUCTcetuximab_tcga_hnsc_finetuned[max_key_tcga_hnsc],2)],\n",
    "       \"AUCPR (TCGA-HNSC)\": [round(AUCTcetuximab_tcga_hnsc_aucpr,2), round(AUCTcetuximab_tcga_hnsc_aucpr_finetuned[max_key_tcga_hnsc],2)],\n",
    "       \"Sample Size (GDSC)\": [GDSCEv2.shape[0], GDSCEv2.shape[0]],\n",
    "       \"Feature Size (GDSC)\": [GDSCEv2.shape[1], GDSCEv2.shape[1]] \n",
    "}\n",
    "\n",
    "emc_dataframe = pd.DataFrame.from_dict(emc, orient='index').transpose() \n",
    "os.makedirs(save_results_to, exist_ok = True)\n",
    "emc_dataframe.to_csv(save_results_to + \"GDSC_PDX_TCGA_Expression_Mutation_CNA_EGFRi_Cetuximab_Fourth_Strategy.tsv\",\n",
    "                     sep = \"\\t\",\n",
    "                     index = False)\n",
    "\n",
    "emc_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEE(\n",
       "  (EnE): Sequential(\n",
       "    (0): Linear(in_features=708, out_features=1024, bias=True)\n",
       "    (1): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.8, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show expression layer parameters\n",
    "\n",
    "AutoencoderE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEM(\n",
       "  (EnM): Sequential(\n",
       "    (0): Linear(in_features=708, out_features=512, bias=True)\n",
       "    (1): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.8, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show mutation layer parameters\n",
    "\n",
    "AutoencoderM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEC(\n",
       "  (EnC): Sequential(\n",
       "    (0): Linear(in_features=708, out_features=1024, bias=True)\n",
       "    (1): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.7, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show CNA layer parameters\n",
    "\n",
    "AutoencoderC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Classifier(\n",
       "  (FC): Sequential(\n",
       "    (0): Linear(in_features=2560, out_features=1, bias=True)\n",
       "    (1): Dropout(p=0.5, inplace=False)\n",
       "    (2): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show classification layer parameters\n",
    "\n",
    "Clas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/EGFRi/Expression_Mutation_CNA/Predictions/GDSC_PDX_TCGA_Fourth_Strategy\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(save_results_to)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save predictions\n",
    "\n",
    "file = open('GDSC_PDX_TCGA_EGFRi_Cetuximab_Predictions_Fourth_Strategy (Expression, Mutation, and CNA).txt', 'w')\n",
    "file.write(f\"GDSC Training (EMC) AUC (EGFRi): {round(AUCT,2)}\\n\")\n",
    "file.write(f\"GDSC Training (EMC) AUCPR (EGFRi): {round(AUCTaucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EMC) AUC (Cetuximab): {round(AUCTcetuximab_pdx,2)}\\n\")\n",
    "file.write(f\"PDX Test (EMC) AUCPR (Cetuximab): {round(AUCTcetuximab_pdx_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EMC) AUC (Cetuximab, CRC): {round(AUCTcetuximab_pdx_crc,2)}\\n\")\n",
    "file.write(f\"PDX Test (EMC) AUCPR (Cetuximab, CRC): {round(AUCTcetuximab_pdx_crc_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EMC) AUC (Cetuximab, NSCLC): {round(AUCTcetuximab_pdx_nsclc,2)}\\n\")\n",
    "file.write(f\"PDX Test (EMC) AUCPR (Cetuximab, NSCLC): {round(AUCTcetuximab_pdx_nsclc_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EMC) AUC (Cetuximab): {round(AUCTcetuximab_tcga,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EMC) AUCPR (Cetuximab): {round(AUCTcetuximab_tcga_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EMC) AUC (Cetuximab, HNSC): {round(AUCTcetuximab_tcga_hnsc,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EMC) AUCPR (Cetuximab, HNSC): {round(AUCTcetuximab_tcga_hnsc_aucpr,2)}\\n\")\n",
    "\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
