{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "\n",
    "import os\n",
    "import torch \n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import math\n",
    "import mygene\n",
    "import sklearn.preprocessing as sk\n",
    "import seaborn as sns\n",
    "from sklearn import metrics\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "from torch.utils.data.sampler import WeightedRandomSampler\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import average_precision_score, precision_recall_curve\n",
    "import random\n",
    "import warnings\n",
    "from random import randint\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define directories\n",
    "\n",
    "cell_line_dir = \"/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC/\"\n",
    "models_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_CNA/Models/GDSC_PDX_TCGA_Third_Strategy/\"\n",
    "finetuned_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_CNA/Models/GDSC_PDX_TCGA_Third_Strategy/Third_Strategy_Finetuned_Models/\"\n",
    "DEGs_dir = \"/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/DNA_REPi/\"\n",
    "dataset_dir = \"/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/\"\n",
    "pdx_data_dir = \"/Volumes/Expansion/Thesis Work/Datasets/PDX/Expression/\"\n",
    "save_results_to = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_CNA/Predictions/GDSC_PDX_TCGA_Third_Strategy/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set random seeds\n",
    "\n",
    "torch.manual_seed(42)\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC cell line details\n",
    "\n",
    "os.chdir(cell_line_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Whole Exome Sequencing (WES)</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>Drug\\nResponse</th>\n",
       "      <th>GDSC\\nTissue descriptor 1</th>\n",
       "      <th>GDSC\\nTissue\\ndescriptor 2</th>\n",
       "      <th>Cancer Type\\n(matching TCGA label)</th>\n",
       "      <th>Microsatellite \\ninstability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1299069</th>\n",
       "      <td>TMK-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>digestive_system</td>\n",
       "      <td>stomach</td>\n",
       "      <td>STAD</td>\n",
       "      <td></td>\n",
       "      <td>R</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1298218</th>\n",
       "      <td>KP-2</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>PAAD</td>\n",
       "      <td></td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330932</th>\n",
       "      <td>KO52</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>leukemia</td>\n",
       "      <td>acute_myeloid_leukaemia</td>\n",
       "      <td>LAML</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331030</th>\n",
       "      <td>SC-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>lymphoma</td>\n",
       "      <td>B_cell_lymphoma</td>\n",
       "      <td>DLBC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503373</th>\n",
       "      <td>U-CH2</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>bone</td>\n",
       "      <td>bone_other</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample Name Whole Exome Sequencing (WES)  \\\n",
       "COSMIC identifier                                            \n",
       "1299069                 TMK-1                            Y   \n",
       "1298218                  KP-2                            Y   \n",
       "1330932                  KO52                            Y   \n",
       "1331030                  SC-1                            Y   \n",
       "1503373                 U-CH2                            Y   \n",
       "\n",
       "                  Copy Number Alterations (CNA) Gene Expression Methylation  \\\n",
       "COSMIC identifier                                                             \n",
       "1299069                                       Y               N           N   \n",
       "1298218                                       N               Y           N   \n",
       "1330932                                       Y               N           N   \n",
       "1331030                                       Y               N           N   \n",
       "1503373                                       Y               N           N   \n",
       "\n",
       "                  Drug\\nResponse GDSC\\nTissue descriptor 1  \\\n",
       "COSMIC identifier                                            \n",
       "1299069                        Y          digestive_system   \n",
       "1298218                        Y                  pancreas   \n",
       "1330932                        N                  leukemia   \n",
       "1331030                        N                  lymphoma   \n",
       "1503373                        N                      bone   \n",
       "\n",
       "                  GDSC\\nTissue\\ndescriptor 2  \\\n",
       "COSMIC identifier                              \n",
       "1299069                              stomach   \n",
       "1298218                             pancreas   \n",
       "1330932              acute_myeloid_leukaemia   \n",
       "1331030                      B_cell_lymphoma   \n",
       "1503373                           bone_other   \n",
       "\n",
       "                  Cancer Type\\n(matching TCGA label)  \\\n",
       "COSMIC identifier                                      \n",
       "1299069                                         STAD   \n",
       "1298218                                         PAAD   \n",
       "1330932                                         LAML   \n",
       "1331030                                         DLBC   \n",
       "1503373                                                \n",
       "\n",
       "                  Microsatellite \\ninstability Status (MSI) Screen Medium  \\\n",
       "COSMIC identifier                                                           \n",
       "1299069                                                                 R   \n",
       "1298218                                                             D/F12   \n",
       "1330932                                           MSS/MSI-L         D/F12   \n",
       "1331030                                           MSS/MSI-L             R   \n",
       "1503373                                           MSS/MSI-L         D/F12   \n",
       "\n",
       "                  Growth Properties  \n",
       "COSMIC identifier                    \n",
       "1299069                    Adherent  \n",
       "1298218                    Adherent  \n",
       "1330932                          NA  \n",
       "1331030                          NA  \n",
       "1503373                    Adherent  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC cell line details table\n",
    "\n",
    "GDSC_cell_line_details = pd.read_excel(\"GDSC_Cell_Lines_Details.xlsx\", keep_default_na = False)\n",
    "GDSC_cell_line_details.set_index(\"COSMIC identifier\", inplace = True)\n",
    "GDSC_cell_line_details = GDSC_cell_line_details.iloc[:-1,]\n",
    "GDSC_cell_line_details.index = GDSC_cell_line_details.index.astype(str)\n",
    "GDSC_cell_line_details.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC expression dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/exprs/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>107984199</th>\n",
       "      <th>107984208</th>\n",
       "      <th>107984325</th>\n",
       "      <th>107984411</th>\n",
       "      <th>107984648</th>\n",
       "      <th>107984889</th>\n",
       "      <th>107984989</th>\n",
       "      <th>107986313</th>\n",
       "      <th>107986782</th>\n",
       "      <th>107986810</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>-0.523146241125996</td>\n",
       "      <td>-0.4774107580335559</td>\n",
       "      <td>1.1792180707204845</td>\n",
       "      <td>-0.1950133103016388</td>\n",
       "      <td>0.2086331356678636</td>\n",
       "      <td>0.1512743407965193</td>\n",
       "      <td>-1.0570300318558898</td>\n",
       "      <td>0.1189857462468134</td>\n",
       "      <td>-1.60237161712403</td>\n",
       "      <td>0.0522097313574845</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4777943283680773</td>\n",
       "      <td>1.7440832645129538</td>\n",
       "      <td>-0.1411845263288239</td>\n",
       "      <td>1.550275627132598</td>\n",
       "      <td>1.7625075247186934</td>\n",
       "      <td>-0.9885319787599142</td>\n",
       "      <td>-1.398962216984746</td>\n",
       "      <td>-1.1477981243123072</td>\n",
       "      <td>0.2309989847175704</td>\n",
       "      <td>0.1184597758225788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753531</th>\n",
       "      <td>-0.4224302754300224</td>\n",
       "      <td>-0.3242516223071998</td>\n",
       "      <td>-0.3208981588446875</td>\n",
       "      <td>-0.4397392179491251</td>\n",
       "      <td>3.025286158155485</td>\n",
       "      <td>-0.24438999503986</td>\n",
       "      <td>0.7515936357204973</td>\n",
       "      <td>0.0632614716673226</td>\n",
       "      <td>1.1869816204036372</td>\n",
       "      <td>0.910750698661622</td>\n",
       "      <td>...</td>\n",
       "      <td>0.5061734764924927</td>\n",
       "      <td>-1.0342661397324897</td>\n",
       "      <td>-0.4760165332952215</td>\n",
       "      <td>0.4559917131361604</td>\n",
       "      <td>-0.2809661852015279</td>\n",
       "      <td>0.7118421818299816</td>\n",
       "      <td>-0.346615229870122</td>\n",
       "      <td>-1.470020745651282</td>\n",
       "      <td>0.0534664475662629</td>\n",
       "      <td>-1.0214126092212672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753532</th>\n",
       "      <td>0.38120783287858</td>\n",
       "      <td>-0.216246760163162</td>\n",
       "      <td>-1.3542414608934643</td>\n",
       "      <td>-0.0288553253901928</td>\n",
       "      <td>-0.210475976908014</td>\n",
       "      <td>-0.2095800353031544</td>\n",
       "      <td>-0.6041666360277236</td>\n",
       "      <td>-0.3005598678634803</td>\n",
       "      <td>-1.0678335183424454</td>\n",
       "      <td>-0.3855263679950399</td>\n",
       "      <td>...</td>\n",
       "      <td>2.138001331301404</td>\n",
       "      <td>0.4656565668245515</td>\n",
       "      <td>-0.6052206632622278</td>\n",
       "      <td>2.411447606433391</td>\n",
       "      <td>0.2137155086353199</td>\n",
       "      <td>-1.1485572785697182</td>\n",
       "      <td>0.6297279321839303</td>\n",
       "      <td>-0.7263920504862745</td>\n",
       "      <td>0.5305975336134361</td>\n",
       "      <td>0.0576046675547845</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18597 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  1                    2                    9          \\\n",
       "906794   -0.523146241125996  -0.4774107580335559   1.1792180707204845   \n",
       "753531  -0.4224302754300224  -0.3242516223071998  -0.3208981588446875   \n",
       "753532     0.38120783287858   -0.216246760163162  -1.3542414608934643   \n",
       "\n",
       "                  10                  12                   13         \\\n",
       "906794  -0.1950133103016388  0.2086331356678636   0.1512743407965193   \n",
       "753531  -0.4397392179491251   3.025286158155485    -0.24438999503986   \n",
       "753532  -0.0288553253901928  -0.210475976908014  -0.2095800353031544   \n",
       "\n",
       "                  14                   15                   16         \\\n",
       "906794  -1.0570300318558898   0.1189857462468134    -1.60237161712403   \n",
       "753531   0.7515936357204973   0.0632614716673226   1.1869816204036372   \n",
       "753532  -0.6041666360277236  -0.3005598678634803  -1.0678335183424454   \n",
       "\n",
       "                  18         ...           107984199            107984208  \\\n",
       "906794   0.0522097313574845  ...  0.4777943283680773   1.7440832645129538   \n",
       "753531    0.910750698661622  ...  0.5061734764924927  -1.0342661397324897   \n",
       "753532  -0.3855263679950399  ...   2.138001331301404   0.4656565668245515   \n",
       "\n",
       "                  107984325           107984411            107984648  \\\n",
       "906794  -0.1411845263288239   1.550275627132598   1.7625075247186934   \n",
       "753531  -0.4760165332952215  0.4559917131361604  -0.2809661852015279   \n",
       "753532  -0.6052206632622278   2.411447606433391   0.2137155086353199   \n",
       "\n",
       "                  107984889           107984989            107986313  \\\n",
       "906794  -0.9885319787599142  -1.398962216984746  -1.1477981243123072   \n",
       "753531   0.7118421818299816  -0.346615229870122   -1.470020745651282   \n",
       "753532  -1.1485572785697182  0.6297279321839303  -0.7263920504862745   \n",
       "\n",
       "                 107986782            107986810  \n",
       "906794  0.2309989847175704   0.1184597758225788  \n",
       "753531  0.0534664475662629  -1.0214126092212672  \n",
       "753532  0.5305975336134361   0.0576046675547845  \n",
       "\n",
       "[3 rows x 18597 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC expression dataset (DNA_REPi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCE = pd.read_csv(\"GDSC_exprs.z.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index_col=0,\n",
    "                    decimal=\",\")\n",
    "GDSCE = pd.DataFrame.transpose(GDSCE)\n",
    "GDSCE = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCE,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCE[\"GDSC\\nTissue descriptor 1\"] != \"pancreas\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"PAAD\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"SARC\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"LIHC\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"PCPG\") & \\\n",
    "         (GDSCE[\"Cancer Type\\n(matching TCGA label)\"] != \"LUSC\")\n",
    "GDSCE = GDSCE.loc[filter, ]\n",
    "GDSCE.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCE.index = GDSCE.index.astype(int)\n",
    "\n",
    "GDSCE.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC CNA dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/CNA/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>107985535</th>\n",
       "      <th>107985759</th>\n",
       "      <th>107986588</th>\n",
       "      <th>107986809</th>\n",
       "      <th>107986898</th>\n",
       "      <th>107987337</th>\n",
       "      <th>107987341</th>\n",
       "      <th>109731405</th>\n",
       "      <th>112441434</th>\n",
       "      <th>121676927</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753531</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753532</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>753535</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>...</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1290724</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-4.320000</td>\n",
       "      <td>-4.32</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24402 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         1          2          9          10         12         13         \\\n",
       "906794   -0.584963  -0.584963   0.415037   0.415037  -0.584963   0.736966   \n",
       "753531    0.000000   0.000000  -1.000000  -1.000000   0.000000   0.584963   \n",
       "753532    0.000000   0.000000  -1.000000  -1.000000   0.000000   0.584963   \n",
       "753535    0.000000   0.000000  -0.584963  -0.584963   0.415037   0.000000   \n",
       "1290724   0.000000   0.000000  -1.000000  -1.000000   0.000000   0.000000   \n",
       "\n",
       "         14         15         16         18         ...  107985535  \\\n",
       "906794    0.000000  -0.584963  -0.584963   0.415037  ...   0.000000   \n",
       "753531    0.000000   0.584963   0.000000   0.000000  ...   0.000000   \n",
       "753532    0.000000   0.000000   0.000000   0.000000  ...   0.000000   \n",
       "753535   -0.584963   0.000000   0.000000  -0.584963  ...   0.415037   \n",
       "1290724  -1.000000   0.000000   0.000000   0.000000  ...   0.000000   \n",
       "\n",
       "         107985759  107986588  107986809  107986898  107987337  107987341  \\\n",
       "906794   -0.584963  -0.584963        0.0   0.415037  -4.320000      -4.32   \n",
       "753531    0.000000   0.000000       -1.0   0.584963  -4.320000      -4.32   \n",
       "753532    0.000000   0.000000        0.0   0.584963  -4.320000      -4.32   \n",
       "753535    0.000000   0.000000        0.0   0.415037  -0.584963      -4.32   \n",
       "1290724   0.000000   0.000000        0.0   0.584963  -4.320000      -4.32   \n",
       "\n",
       "         109731405  112441434  121676927  \n",
       "906794   -0.584963   0.000000   0.415037  \n",
       "753531    0.000000   0.000000   0.584963  \n",
       "753532    0.000000   0.584963   0.584963  \n",
       "753535    0.000000  -0.584963   0.415037  \n",
       "1290724   0.000000   0.000000   1.000000  \n",
       "\n",
       "[5 rows x 24402 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC CNA dataset (DNA_REPi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCC = pd.read_csv(\"GDSC_CNA.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\", index_col=0, decimal=\".\")\n",
    "GDSCC.drop_duplicates(keep='last')\n",
    "GDSCC = pd.DataFrame.transpose(GDSCC)\n",
    "GDSCC = GDSCC.loc[:, ~GDSCC.columns.duplicated()]\n",
    "GDSCC = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCC,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCC[\"GDSC\\nTissue descriptor 1\"] != \"pancreas\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"PAAD\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"SARC\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"LIHC\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"PCPG\") & \\\n",
    "         (GDSCC[\"Cancer Type\\n(matching TCGA label)\"] != \"LUSC\")\n",
    "GDSCC = GDSCC.loc[filter, ]\n",
    "GDSCC.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCC.index = GDSCC.index.astype(int)\n",
    "\n",
    "GDSCC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC response dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/response/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>3.25484543425934</td>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>-3.08563930086988</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>1</td>\n",
       "      <td>-2.64209579180854</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>-2.92273746018911</td>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>-1.0950578547037</td>\n",
       "      <td>Mitomycin C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   response            logIC50           drug  exprs  CNA  \\\n",
       "COSMIC identifier                                                           \n",
       "906794                    0   3.25484543425934  Pyrimethamine      1    1   \n",
       "906794                    0  -3.08563930086988    Doxorubicin      1    1   \n",
       "906794                    1  -2.64209579180854      Etoposide      1    1   \n",
       "906794                    0  -2.92273746018911    Gemcitabine      1    1   \n",
       "906794                    0   -1.0950578547037    Mitomycin C      1    1   \n",
       "\n",
       "                   mutations  \n",
       "COSMIC identifier             \n",
       "906794                     1  \n",
       "906794                     1  \n",
       "906794                     1  \n",
       "906794                     1  \n",
       "906794                     1  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC response dataset (DNA_REPi)\n",
    "# remove samples with shared cancer types from training dataset\n",
    "\n",
    "GDSCR = pd.read_csv(\"GDSC_response.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index_col=0,\n",
    "                    decimal=\",\")\n",
    "GDSCR.dropna(inplace=True)\n",
    "GDSCR.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "GDSCR[\"response\"] = GDSCR.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "GDSCR = pd.merge(GDSC_cell_line_details,\n",
    "                 GDSCR,\n",
    "                 left_index=True,\n",
    "                 right_index=True)\n",
    "filter = (GDSCR[\"GDSC\\nTissue descriptor 1\"] != \"pancreas\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"PAAD\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"SARC\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"LIHC\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"PCPG\") & \\\n",
    "         (GDSCR[\"Cancer Type\\n(matching TCGA label)\"] != \"LUSC\")\n",
    "GDSCR = GDSCR.loc[filter, ]\n",
    "GDSCR.drop(GDSC_cell_line_details.columns, axis=1, inplace=True)\n",
    "GDSCR.index = GDSCR.index.astype(int)\n",
    "\n",
    "GDSCR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "drug\n",
       "Etoposide         725\n",
       "SN-38             721\n",
       "Mitomycin C       717\n",
       "Doxorubicin       716\n",
       "5-Fluorouracil    714\n",
       "Gemcitabine       711\n",
       "Bleomycin         709\n",
       "Temozolomide      706\n",
       "Cisplatin         690\n",
       "Methotrexate      690\n",
       "Camptothecin      689\n",
       "Cytarabine        687\n",
       "Pyrimethamine     368\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print count of cell lines for each drug (DNA_REPi)\n",
    "\n",
    "GDSCR[\"drug\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs_homogenized\n"
     ]
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA expression datasets homogenized with GDSC expression dataset (Gemcitabine)\n",
    "\n",
    "os.chdir(dataset_dir + \"/exprs_homogenized/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100507206</th>\n",
       "      <th>100507254</th>\n",
       "      <th>100507436</th>\n",
       "      <th>100507472</th>\n",
       "      <th>100526773</th>\n",
       "      <th>100527978</th>\n",
       "      <th>100532746</th>\n",
       "      <th>100820829</th>\n",
       "      <th>102724473</th>\n",
       "      <th>105375355</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1199</th>\n",
       "      <td>4.01417756220044</td>\n",
       "      <td>3.55811958088587</td>\n",
       "      <td>6.83900390800765</td>\n",
       "      <td>4.38877376417435</td>\n",
       "      <td>1.50666127948843</td>\n",
       "      <td>5.02771257808434</td>\n",
       "      <td>8.43955592164473</td>\n",
       "      <td>2.85208997496268</td>\n",
       "      <td>9.00048066769525</td>\n",
       "      <td>5.52248167850218</td>\n",
       "      <td>...</td>\n",
       "      <td>2.83437923324963</td>\n",
       "      <td>3.18066968142204</td>\n",
       "      <td>6.86894950827138</td>\n",
       "      <td>3.09303561080421</td>\n",
       "      <td>3.06056009765469</td>\n",
       "      <td>3.25550719551845</td>\n",
       "      <td>2.66209753115624</td>\n",
       "      <td>4.89535625008263</td>\n",
       "      <td>0</td>\n",
       "      <td>2.53286170515559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1289</th>\n",
       "      <td>3.81163538029984</td>\n",
       "      <td>3.55811958088587</td>\n",
       "      <td>7.32657832557435</td>\n",
       "      <td>4.60329219527869</td>\n",
       "      <td>2.70668525625435</td>\n",
       "      <td>3.94530321859601</td>\n",
       "      <td>8.38639293838613</td>\n",
       "      <td>3.16570624417349</td>\n",
       "      <td>8.80528687662163</td>\n",
       "      <td>4.02249343941382</td>\n",
       "      <td>...</td>\n",
       "      <td>2.78756471299352</td>\n",
       "      <td>3.18066968142204</td>\n",
       "      <td>6.45652489684015</td>\n",
       "      <td>3.09303561080421</td>\n",
       "      <td>3.06056009765469</td>\n",
       "      <td>2.55755763454488</td>\n",
       "      <td>2.66209753115624</td>\n",
       "      <td>4.66362817768075</td>\n",
       "      <td>0</td>\n",
       "      <td>2.92817140721819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1317</th>\n",
       "      <td>6.56141894155303</td>\n",
       "      <td>9.93900865308036</td>\n",
       "      <td>5.9645748758066</td>\n",
       "      <td>2.89206965997821</td>\n",
       "      <td>4.24925937434505</td>\n",
       "      <td>2.43002821482342</td>\n",
       "      <td>8.19585380485642</td>\n",
       "      <td>3.24884065836222</td>\n",
       "      <td>9.13531900723866</td>\n",
       "      <td>3.39900580748116</td>\n",
       "      <td>...</td>\n",
       "      <td>2.84342078648607</td>\n",
       "      <td>4.76280820040358</td>\n",
       "      <td>6.724116861893</td>\n",
       "      <td>3.09303561080421</td>\n",
       "      <td>3.06056009765469</td>\n",
       "      <td>2.55755763454488</td>\n",
       "      <td>4.01252295145878</td>\n",
       "      <td>3.28462638352963</td>\n",
       "      <td>0</td>\n",
       "      <td>4.63700409301447</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18189 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID         1                 2                 9          \\\n",
       "X-1199    4.01417756220044  3.55811958088587  6.83900390800765   \n",
       "X-1289    3.81163538029984  3.55811958088587  7.32657832557435   \n",
       "X-1317    6.56141894155303  9.93900865308036   5.9645748758066   \n",
       "\n",
       "ENTREZID         10                12                13         \\\n",
       "X-1199    4.38877376417435  1.50666127948843  5.02771257808434   \n",
       "X-1289    4.60329219527869  2.70668525625435  3.94530321859601   \n",
       "X-1317    2.89206965997821  4.24925937434505  2.43002821482342   \n",
       "\n",
       "ENTREZID         14                15                16         \\\n",
       "X-1199    8.43955592164473  2.85208997496268  9.00048066769525   \n",
       "X-1289    8.38639293838613  3.16570624417349  8.80528687662163   \n",
       "X-1317    8.19585380485642  3.24884065836222  9.13531900723866   \n",
       "\n",
       "ENTREZID         18         ...         100507206         100507254  \\\n",
       "X-1199    5.52248167850218  ...  2.83437923324963  3.18066968142204   \n",
       "X-1289    4.02249343941382  ...  2.78756471299352  3.18066968142204   \n",
       "X-1317    3.39900580748116  ...  2.84342078648607  4.76280820040358   \n",
       "\n",
       "ENTREZID         100507436         100507472         100526773  \\\n",
       "X-1199    6.86894950827138  3.09303561080421  3.06056009765469   \n",
       "X-1289    6.45652489684015  3.09303561080421  3.06056009765469   \n",
       "X-1317      6.724116861893  3.09303561080421  3.06056009765469   \n",
       "\n",
       "ENTREZID         100527978         100532746         100820829 102724473  \\\n",
       "X-1199    3.25550719551845  2.66209753115624  4.89535625008263         0   \n",
       "X-1289    2.55755763454488  2.66209753115624  4.66362817768075         0   \n",
       "X-1317    2.55755763454488  4.01252295145878  3.28462638352963         0   \n",
       "\n",
       "ENTREZID         105375355  \n",
       "X-1199    2.53286170515559  \n",
       "X-1289    2.92817140721819  \n",
       "X-1317    4.63700409301447  \n",
       "\n",
       "[3 rows x 18189 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX expression dataset homogenized with GDSC expression dataset (Gemcitabine)\n",
    "\n",
    "PDXEgemcitabine = pd.read_csv(\"PDX_exprs.Gemcitabine.eb_with.GDSC_exprs.Gemcitabine.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "PDXEgemcitabine = pd.DataFrame.transpose(PDXEgemcitabine)\n",
    "PDXEgemcitabine.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100188893</th>\n",
       "      <th>100190940</th>\n",
       "      <th>100190949</th>\n",
       "      <th>100191040</th>\n",
       "      <th>100192386</th>\n",
       "      <th>100268168</th>\n",
       "      <th>100271715</th>\n",
       "      <th>100287718</th>\n",
       "      <th>100288778</th>\n",
       "      <th>100289635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQA-01A</th>\n",
       "      <td>3.98033768533371</td>\n",
       "      <td>2.62820652008909</td>\n",
       "      <td>7.03249429468481</td>\n",
       "      <td>3.08514834183615</td>\n",
       "      <td>3.34557990074588</td>\n",
       "      <td>3.78552975984518</td>\n",
       "      <td>8.73764964752829</td>\n",
       "      <td>2.48671779353536</td>\n",
       "      <td>8.21923206815042</td>\n",
       "      <td>3.76684655565602</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0875725194276</td>\n",
       "      <td>2.81337967290836</td>\n",
       "      <td>3.51332685491118</td>\n",
       "      <td>3.98785715416424</td>\n",
       "      <td>3.07567558045391</td>\n",
       "      <td>2.56126348424232</td>\n",
       "      <td>2.63693987323012</td>\n",
       "      <td>2.83133300992887</td>\n",
       "      <td>5.13739151604047</td>\n",
       "      <td>4.69382514336529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQE-01A</th>\n",
       "      <td>4.68614471637408</td>\n",
       "      <td>3.25092419132046</td>\n",
       "      <td>5.75542155466148</td>\n",
       "      <td>2.79866479035723</td>\n",
       "      <td>1.96329083991897</td>\n",
       "      <td>3.43063349827168</td>\n",
       "      <td>8.80671304478463</td>\n",
       "      <td>2.62779485919227</td>\n",
       "      <td>8.54483025692787</td>\n",
       "      <td>3.81532940417382</td>\n",
       "      <td>...</td>\n",
       "      <td>8.87561409312678</td>\n",
       "      <td>2.81337967290836</td>\n",
       "      <td>2.78991865702044</td>\n",
       "      <td>3.30688425304558</td>\n",
       "      <td>3.45865630140412</td>\n",
       "      <td>2.53490145426966</td>\n",
       "      <td>2.74899279856495</td>\n",
       "      <td>2.83133300992887</td>\n",
       "      <td>4.61757632178081</td>\n",
       "      <td>4.2235154535305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQI-01A</th>\n",
       "      <td>3.71023111103364</td>\n",
       "      <td>3.88843437623036</td>\n",
       "      <td>5.77580575110606</td>\n",
       "      <td>3.10749843796475</td>\n",
       "      <td>5.98690471000489</td>\n",
       "      <td>3.75736553725429</td>\n",
       "      <td>8.4255369043372</td>\n",
       "      <td>2.48671779353536</td>\n",
       "      <td>9.21052968908206</td>\n",
       "      <td>3.97685914269242</td>\n",
       "      <td>...</td>\n",
       "      <td>9.48642956588486</td>\n",
       "      <td>2.84555917176674</td>\n",
       "      <td>2.78991865702044</td>\n",
       "      <td>3.48989111572046</td>\n",
       "      <td>2.79620231354257</td>\n",
       "      <td>2.49624245190844</td>\n",
       "      <td>2.630937590025</td>\n",
       "      <td>2.83133300992887</td>\n",
       "      <td>4.38859435268046</td>\n",
       "      <td>3.64555981820771</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18197 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID                 1                 2                 9          \\\n",
       "TCGA-2L-AAQA-01A  3.98033768533371  2.62820652008909  7.03249429468481   \n",
       "TCGA-2L-AAQE-01A  4.68614471637408  3.25092419132046  5.75542155466148   \n",
       "TCGA-2L-AAQI-01A  3.71023111103364  3.88843437623036  5.77580575110606   \n",
       "\n",
       "ENTREZID                 10                12                13         \\\n",
       "TCGA-2L-AAQA-01A  3.08514834183615  3.34557990074588  3.78552975984518   \n",
       "TCGA-2L-AAQE-01A  2.79866479035723  1.96329083991897  3.43063349827168   \n",
       "TCGA-2L-AAQI-01A  3.10749843796475  5.98690471000489  3.75736553725429   \n",
       "\n",
       "ENTREZID                 14                15                16         \\\n",
       "TCGA-2L-AAQA-01A  8.73764964752829  2.48671779353536  8.21923206815042   \n",
       "TCGA-2L-AAQE-01A  8.80671304478463  2.62779485919227  8.54483025692787   \n",
       "TCGA-2L-AAQI-01A   8.4255369043372  2.48671779353536  9.21052968908206   \n",
       "\n",
       "ENTREZID                 18         ...         100188893         100190940  \\\n",
       "TCGA-2L-AAQA-01A  3.76684655565602  ...  10.0875725194276  2.81337967290836   \n",
       "TCGA-2L-AAQE-01A  3.81532940417382  ...  8.87561409312678  2.81337967290836   \n",
       "TCGA-2L-AAQI-01A  3.97685914269242  ...  9.48642956588486  2.84555917176674   \n",
       "\n",
       "ENTREZID                 100190949         100191040         100192386  \\\n",
       "TCGA-2L-AAQA-01A  3.51332685491118  3.98785715416424  3.07567558045391   \n",
       "TCGA-2L-AAQE-01A  2.78991865702044  3.30688425304558  3.45865630140412   \n",
       "TCGA-2L-AAQI-01A  2.78991865702044  3.48989111572046  2.79620231354257   \n",
       "\n",
       "ENTREZID                 100268168         100271715         100287718  \\\n",
       "TCGA-2L-AAQA-01A  2.56126348424232  2.63693987323012  2.83133300992887   \n",
       "TCGA-2L-AAQE-01A  2.53490145426966  2.74899279856495  2.83133300992887   \n",
       "TCGA-2L-AAQI-01A  2.49624245190844    2.630937590025  2.83133300992887   \n",
       "\n",
       "ENTREZID                 100288778         100289635  \n",
       "TCGA-2L-AAQA-01A  5.13739151604047  4.69382514336529  \n",
       "TCGA-2L-AAQE-01A  4.61757632178081   4.2235154535305  \n",
       "TCGA-2L-AAQI-01A  4.38859435268046  3.64555981820771  \n",
       "\n",
       "[3 rows x 18197 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA expression dataset homogenized with GDSC expression dataset (Gemcitabine)\n",
    "\n",
    "TCGAEgemcitabine = pd.read_csv(\"TCGA_exprs.Gemcitabine.eb_with.GDSC_exprs.Gemcitabine.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "TCGAEgemcitabine = pd.DataFrame.transpose(TCGAEgemcitabine)\n",
    "TCGAEgemcitabine.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA'"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX CNA dataset (Gemcitabine)\n",
    "\n",
    "os.chdir(dataset_dir + \"CNA\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>...</th>\n",
       "      <th>101340252</th>\n",
       "      <th>102723547</th>\n",
       "      <th>102724473</th>\n",
       "      <th>103091865</th>\n",
       "      <th>105375355</th>\n",
       "      <th>107126285</th>\n",
       "      <th>109623460</th>\n",
       "      <th>109731405</th>\n",
       "      <th>118126072</th>\n",
       "      <th>128385369</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1199</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.9434164716336324</td>\n",
       "      <td>-0.9159357352115256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.9573556625915064</td>\n",
       "      <td>-0.929610672108602</td>\n",
       "      <td>-0.9434164716336324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1289</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.7858751946471527</td>\n",
       "      <td>-0.7734914701913203</td>\n",
       "      <td>-0.7612131404128835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.8236772273595371</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1317</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6915341649192002</td>\n",
       "      <td>0.6915341649192002</td>\n",
       "      <td>0.4905701304462014</td>\n",
       "      <td>0.4905701304462014</td>\n",
       "      <td>0.3673710656485294</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4005379295837288</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5249151170512174</td>\n",
       "      <td>-1.377069649079823</td>\n",
       "      <td>-1.377069649079823</td>\n",
       "      <td>-0.4540316308947075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.415037499278844</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.6551715030025588</td>\n",
       "      <td>0.4005379295837288</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID 1                   2                   3          \\\n",
       "X-1199         0.0                 0.0                 0.0   \n",
       "X-1289         0.0                 0.0                 0.0   \n",
       "X-1317         0.0  0.6915341649192002  0.6915341649192002   \n",
       "\n",
       "ENTREZID            9                    10                  12         \\\n",
       "X-1199    -0.9434164716336324  -0.9159357352115256                 0.0   \n",
       "X-1289                    0.0                  0.0                 0.0   \n",
       "X-1317     0.4905701304462014   0.4905701304462014  0.3673710656485294   \n",
       "\n",
       "ENTREZID 13        14                  15        16         ...  \\\n",
       "X-1199         0.0       0.0                 0.0       0.0  ...   \n",
       "X-1289         0.0       0.0                 0.0       0.0  ...   \n",
       "X-1317         0.0       0.0  0.4005379295837288       0.0  ...   \n",
       "\n",
       "ENTREZID            101340252            102723547            102724473  \\\n",
       "X-1199    -0.9573556625915064   -0.929610672108602  -0.9434164716336324   \n",
       "X-1289                    0.0  -0.7858751946471527  -0.7734914701913203   \n",
       "X-1317    -0.5249151170512174   -1.377069649079823   -1.377069649079823   \n",
       "\n",
       "ENTREZID            103091865 105375355 107126285            109623460  \\\n",
       "X-1199                    0.0       0.0       0.0                  0.0   \n",
       "X-1289    -0.7612131404128835       0.0       0.0  -0.8236772273595371   \n",
       "X-1317    -0.4540316308947075       0.0       0.0   -1.415037499278844   \n",
       "\n",
       "ENTREZID 109731405            118126072           128385369  \n",
       "X-1199         0.0                  0.0                 0.0  \n",
       "X-1289         0.0                  0.0                 0.0  \n",
       "X-1317         0.0  -0.6551715030025588  0.4005379295837288  \n",
       "\n",
       "[3 rows x 23205 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX CNA dataset (Gemcitabine)\n",
    "\n",
    "PDXCgemcitabine = pd.read_csv(\"PDX_CNA.Gemcitabine.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "PDXCgemcitabine.drop_duplicates(keep='last')\n",
    "PDXCgemcitabine = pd.DataFrame.transpose(PDXCgemcitabine)\n",
    "PDXCgemcitabine = PDXCgemcitabine.loc[:, ~PDXCgemcitabine.columns.duplicated()]\n",
    "\n",
    "PDXCgemcitabine.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>13</th>\n",
       "      <th>24</th>\n",
       "      <th>30</th>\n",
       "      <th>34</th>\n",
       "      <th>39</th>\n",
       "      <th>55</th>\n",
       "      <th>68</th>\n",
       "      <th>73</th>\n",
       "      <th>86</th>\n",
       "      <th>93</th>\n",
       "      <th>...</th>\n",
       "      <th>100616290</th>\n",
       "      <th>100616357</th>\n",
       "      <th>100616366</th>\n",
       "      <th>100616420</th>\n",
       "      <th>100616434</th>\n",
       "      <th>100652770</th>\n",
       "      <th>100652778</th>\n",
       "      <th>100652797</th>\n",
       "      <th>100652854</th>\n",
       "      <th>100652917</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQA-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.5075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.5075</td>\n",
       "      <td>-0.5104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.5104</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>-0.521</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQE-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.3378</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.3346</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.6335</td>\n",
       "      <td>-0.6335</td>\n",
       "      <td>-0.6409</td>\n",
       "      <td>-0.6409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.6409</td>\n",
       "      <td>-0.6409</td>\n",
       "      <td>-0.6409</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.6633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-2L-AAQI-01A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 31922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 13        24        30        34        39        55         \\\n",
       "TCGA-2L-AAQA-01A       0.0   -0.5075       0.0   -0.5075   -0.5104       0.0   \n",
       "TCGA-2L-AAQE-01A       0.0       0.0       0.0       0.0   -0.3378       0.0   \n",
       "TCGA-2L-AAQI-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "                 68        73        86        93         ... 100616290  \\\n",
       "TCGA-2L-AAQA-01A   -0.5104       0.0       0.0       0.0  ...    -0.521   \n",
       "TCGA-2L-AAQE-01A   -0.3346       0.0       0.0       0.0  ...   -0.6335   \n",
       "TCGA-2L-AAQI-01A       0.0       0.0       0.0       0.0  ...       0.0   \n",
       "\n",
       "                 100616357 100616366 100616420 100616434 100652770 100652778  \\\n",
       "TCGA-2L-AAQA-01A    -0.521    -0.521    -0.521       0.0    -0.521    -0.521   \n",
       "TCGA-2L-AAQE-01A   -0.6335   -0.6409   -0.6409       0.0   -0.6409   -0.6409   \n",
       "TCGA-2L-AAQI-01A       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "                 100652797 100652854 100652917  \n",
       "TCGA-2L-AAQA-01A    -0.521       0.0    -0.521  \n",
       "TCGA-2L-AAQE-01A   -0.6409       0.0   -0.6633  \n",
       "TCGA-2L-AAQI-01A       0.0       0.0       0.0  \n",
       "\n",
       "[3 rows x 31922 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA CNA dataset (Gemcitabine)\n",
    "\n",
    "TCGACgemcitabine = pd.read_csv(\"TCGA_CNA.Gemcitabine.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "TCGACgemcitabine.drop_duplicates(keep='last')\n",
    "TCGACgemcitabine = pd.DataFrame.transpose(TCGACgemcitabine)\n",
    "TCGACgemcitabine = TCGACgemcitabine.loc[:, ~TCGACgemcitabine.columns.duplicated()]\n",
    "\n",
    "TCGACgemcitabine.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variance threshold for GDSC expression dataset (DNA_REPi)\n",
    "\n",
    "selector = VarianceThreshold(0.05)\n",
    "selector.fit_transform(GDSCE)\n",
    "GDSCE = GDSCE[GDSCE.columns[selector.get_support(indices=True)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fill NA values and binarize GDSC CNA dataset (DNA_REPi)\n",
    "\n",
    "GDSCC = GDSCC.fillna(0)\n",
    "GDSCC[GDSCC != 0.0] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC, PDX & TCGA # of common genes: 16367\n"
     ]
    }
   ],
   "source": [
    "# select shared genes between GDSC, PDX, and TCGA expression and CNA datasets \n",
    "\n",
    "ls = GDSCE.columns.intersection(GDSCC.columns)\n",
    "ls = ls.intersection(PDXEgemcitabine.columns)\n",
    "ls = ls.intersection(PDXCgemcitabine.columns)\n",
    "ls = ls.intersection(TCGAEgemcitabine.columns)\n",
    "ls = ls.intersection(TCGACgemcitabine.columns)\n",
    "\n",
    "ls = pd.unique(ls)\n",
    "\n",
    "print(f\"GDSC, PDX & TCGA # of common genes: {ls.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common cell lines: 751\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between GDSC expression, CNA, and response datasets (DNA_REPi)\n",
    "\n",
    "ls2 = GDSCE.index.intersection(GDSCC.index)\n",
    "ls2 = ls2.intersection(GDSCR.index)\n",
    "\n",
    "print(f\"GDSC # of common cell lines: {ls2.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (751, 16367)\n",
      "GDSC # of common samples and genes (CNA): (751, 16367)\n",
      "GDSC # of samples (Response): 8843\n"
     ]
    }
   ],
   "source": [
    "# subset shared genes and samples in GDSC expression, CNA, and response datasets\n",
    "\n",
    "GDSCE = GDSCE.loc[ls2, ls]\n",
    "GDSCC = GDSCC.loc[ls2, ls]\n",
    "GDSCR = GDSCR.loc[ls2, :]\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCE.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCC.shape}\")\n",
    "print(f\"GDSC # of samples (Response): {GDSCR.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples (Gemcitabine): 25\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between PDX expression and CNA datasets (Gemcitabine) \n",
    "\n",
    "ls3 = PDXEgemcitabine.index.intersection(PDXCgemcitabine.index)\n",
    "\n",
    "print(f\"PDX # of common samples (Gemcitabine): {ls3.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA # of common samples (Gemcitabine): 55\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between TCGA expression and CNA datasets (Gemcitabine) \n",
    "\n",
    "ls4 = TCGAEgemcitabine.index.intersection(TCGACgemcitabine.index)\n",
    "\n",
    "print(f\"TCGA # of common samples (Gemcitabine): {ls4.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples and genes for Expression (Gemcitabine): (25, 16367)\n",
      "PDX # of common samples and genes for CNA (Gemcitabine): (25, 16367)\n",
      "\n",
      "TCGA # of common samples and genes for Expression (Gemcitabine): (55, 16367)\n",
      "TCGA # of common samples and genes for CNA (Gemcitabine): (55, 16367)\n"
     ]
    }
   ],
   "source": [
    "# select shared genes and samples between PDX and TCGA expression and CNA datasets (Gemcitabine)\n",
    "\n",
    "PDXEgemcitabine = PDXEgemcitabine.loc[ls3, ls]\n",
    "PDXCgemcitabine = PDXCgemcitabine.loc[ls3, ls]\n",
    "\n",
    "TCGAEgemcitabine = TCGAEgemcitabine.loc[ls4, ls]\n",
    "TCGACgemcitabine = TCGACgemcitabine.loc[ls4, ls]\n",
    "\n",
    "print(f\"PDX # of common samples and genes for Expression (Gemcitabine): {PDXEgemcitabine.shape}\")\n",
    "print(f\"PDX # of common samples and genes for CNA (Gemcitabine): {PDXCgemcitabine.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for Expression (Gemcitabine): {TCGAEgemcitabine.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for CNA (Gemcitabine): {TCGACgemcitabine.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/DNA_REPi\n"
     ]
    }
   ],
   "source": [
    "# change directory to read DEGs (DNA_REPi)\n",
    "\n",
    "os.chdir(DEGs_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 858 DEGs (p-adjusted < 0.05 and |logFC| > 1) among DNA replication inhibitors\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLFN11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCKAP1L</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDC4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KRT18</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S100A14</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Frequency\n",
       "0      SLFN11          9\n",
       "1     NCKAP1L          8\n",
       "2        SDC4          7\n",
       "3       KRT18          7\n",
       "4     S100A14          7"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read DEGs (DNA_REPi)\n",
    "\n",
    "DEGs_filtered_data = pd.read_excel(\"DNA_REPi_Differentially_Expressed_Genes (EnsemblID).xlsx\",\n",
    "                                    sheet_name = \"Common DEGs\")\n",
    "print(f\"There are {DEGs_filtered_data.shape[0]} DEGs (p-adjusted < 0.05 and |logFC| > 1) among DNA replication inhibitors\\n\")\n",
    "\n",
    "DEGs_filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 input query terms found dup hits:\t[('CLECL1P', 2)]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "858"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene symbol from mygene database\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "DEGs_entrez_id = mg.querymany(DEGs_filtered_data[\"Gene Symbol\"], \n",
    "                              species=\"human\", \n",
    "                              scopes = \"symbol\",\n",
    "                              field = \"entrezgene\",\n",
    "                              as_dataframe=True)[\"entrezgene\"]\n",
    "DEGs_entrez_id.dropna(inplace=True)\n",
    "\n",
    "DEGs_entrez_id.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign GDSC datasets to new variables\n",
    "\n",
    "exprs_z = GDSCE\n",
    "cna = GDSCC\n",
    "responses = GDSCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>3.25484543425934</td>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>0</td>\n",
       "      <td>-3.08563930086988</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>1</td>\n",
       "      <td>-2.64209579180854</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        response            logIC50           drug  exprs  CNA  mutations\n",
       "906794         0   3.25484543425934  Pyrimethamine      1    1          1\n",
       "906794         0  -3.08563930086988    Doxorubicin      1    1          1\n",
       "906794         1  -2.64209579180854      Etoposide      1    1          1"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GDSC response dataset\n",
    "\n",
    "responses.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'5-Fluorouracil',\n",
       " 'Bleomycin',\n",
       " 'Camptothecin',\n",
       " 'Cisplatin',\n",
       " 'Cytarabine',\n",
       " 'Doxorubicin',\n",
       " 'Etoposide',\n",
       " 'Gemcitabine',\n",
       " 'Methotrexate',\n",
       " 'Mitomycin C',\n",
       " 'Pyrimethamine',\n",
       " 'SN-38',\n",
       " 'Temozolomide'}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list drug names in GDSC response dataset\n",
    "\n",
    "drugs = set(responses[\"drug\"].values)\n",
    "drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert Entrez gene IDs to string\n",
    "\n",
    "GDSCE.index = GDSCE.index.astype(str)\n",
    "GDSCC.index = GDSCC.index.astype(str)\n",
    "responses.index = responses.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Whole Exome Sequencing (WES)</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>Drug\\nResponse</th>\n",
       "      <th>GDSC\\nTissue descriptor 1</th>\n",
       "      <th>GDSC\\nTissue\\ndescriptor 2</th>\n",
       "      <th>Cancer Type\\n(matching TCGA label)</th>\n",
       "      <th>Microsatellite \\ninstability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>A253</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>0</td>\n",
       "      <td>3.25484543425934</td>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>A253</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>0</td>\n",
       "      <td>-3.08563930086988</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906794</th>\n",
       "      <td>A253</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>aero_dig_tract</td>\n",
       "      <td>head and neck</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "      <td>1</td>\n",
       "      <td>-2.64209579180854</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample Name Whole Exome Sequencing (WES)  \\\n",
       "COSMIC identifier                                            \n",
       "906794                   A253                            Y   \n",
       "906794                   A253                            Y   \n",
       "906794                   A253                            Y   \n",
       "\n",
       "                  Copy Number Alterations (CNA) Gene Expression Methylation  \\\n",
       "COSMIC identifier                                                             \n",
       "906794                                        Y               Y           Y   \n",
       "906794                                        Y               Y           Y   \n",
       "906794                                        Y               Y           Y   \n",
       "\n",
       "                  Drug\\nResponse GDSC\\nTissue descriptor 1  \\\n",
       "COSMIC identifier                                            \n",
       "906794                         Y            aero_dig_tract   \n",
       "906794                         Y            aero_dig_tract   \n",
       "906794                         Y            aero_dig_tract   \n",
       "\n",
       "                  GDSC\\nTissue\\ndescriptor 2  \\\n",
       "COSMIC identifier                              \n",
       "906794                         head and neck   \n",
       "906794                         head and neck   \n",
       "906794                         head and neck   \n",
       "\n",
       "                  Cancer Type\\n(matching TCGA label)  \\\n",
       "COSMIC identifier                                      \n",
       "906794                                                 \n",
       "906794                                                 \n",
       "906794                                                 \n",
       "\n",
       "                  Microsatellite \\ninstability Status (MSI) Screen Medium  \\\n",
       "COSMIC identifier                                                           \n",
       "906794                                            MSS/MSI-L         D/F12   \n",
       "906794                                            MSS/MSI-L         D/F12   \n",
       "906794                                            MSS/MSI-L         D/F12   \n",
       "\n",
       "                  Growth Properties  response            logIC50  \\\n",
       "COSMIC identifier                                                  \n",
       "906794                     Adherent         0   3.25484543425934   \n",
       "906794                     Adherent         0  -3.08563930086988   \n",
       "906794                     Adherent         1  -2.64209579180854   \n",
       "\n",
       "                            drug  exprs  CNA  mutations  \n",
       "COSMIC identifier                                        \n",
       "906794             Pyrimethamine      1    1          1  \n",
       "906794               Doxorubicin      1    1          1  \n",
       "906794                 Etoposide      1    1          1  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge GDSC cell line details with response dataset\n",
    "\n",
    "dat = pd.merge(GDSC_cell_line_details,\n",
    "         responses,\n",
    "         left_index = True, \n",
    "         right_index=True,\n",
    "         how = \"inner\")\n",
    "\n",
    "dat[\"GDSC\\nTissue descriptor 1\"].value_counts()\n",
    "dat.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC\\nTissue\\ndescriptor 2\n",
       "lung_NSCLC_adenocarcinoma    639\n",
       "lung_small_cell_carcinoma    569\n",
       "glioma                       558\n",
       "breast                       524\n",
       "melanoma                     522\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get tissue descriptors\n",
    "\n",
    "dat[\"GDSC\\nTissue\\ndescriptor 2\"].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GDSC\\nTissue descriptor 1\n",
       "urogenital_system    76\n",
       "lung_NSCLC           74\n",
       "leukemia             64\n",
       "aero_dig_tract       62\n",
       "lymphoma             59\n",
       "nervous_system       47\n",
       "skin                 47\n",
       "lung_SCLC            46\n",
       "breast               43\n",
       "large_intestine      35\n",
       "bone                 34\n",
       "digestive_system     25\n",
       "neuroblastoma        25\n",
       "kidney               17\n",
       "soft_tissue          16\n",
       "lung                 15\n",
       "myeloma              13\n",
       "thyroid              13\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list count of cell lines as to tissues treated with gemcitabine\n",
    "\n",
    "filter = (dat[\"drug\"] == \"Gemcitabine\")\n",
    "dat.loc[filter,\"GDSC\\nTissue descriptor 1\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['906794', '906794', '906794', '906794', '906794', '906794', '906794',\n",
       "       '906794', '906794', '906794',\n",
       "       ...\n",
       "       '1299061', '1299061', '1299061', '1299061', '1299061', '1299061',\n",
       "       '1299061', '1299061', '1299061', '1299061'],\n",
       "      dtype='object', length=8843)"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert Entrez gene IDs to string\n",
    "\n",
    "responses.index = responses.index.astype(str)\n",
    "responses.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (8843, 772)\n",
      "GDSC # of common samples and genes (CNA): (8843, 772)\n",
      "GDSC # of common samples and genes (Response): 8843\n",
      "\n",
      "PDX # of common samples and genes for Gemcitabine (Expression): (25, 772)\n",
      "PDX # of common samples and genes for Gemcitabine (CNA): (25, 772)\n",
      "\n",
      "TCGA # of common samples and genes for Gemcitabine (Expression): (55, 772)\n",
      "TCGA # of common samples and genes for Gemcitabine (CNA): (55, 772)\n"
     ]
    }
   ],
   "source": [
    "# filter GDSC expression and CNA datasets (DNA_REPi) as to drugs\n",
    "# subset selected DEGs\n",
    "\n",
    "expression_zscores = []\n",
    "CNA = []\n",
    "for drug in drugs:\n",
    "    samples = responses.loc[responses[\"drug\"] == drug, :].index.values\n",
    "    e_z = exprs_z.loc[samples, :]\n",
    "    c = cna.loc[samples, :]\n",
    "    expression_zscores.append(e_z)\n",
    "    CNA.append(c)\n",
    "\n",
    "GDSCEv2 = pd.concat(expression_zscores, axis=0)\n",
    "GDSCCv2 = pd.concat(CNA, axis=0)\n",
    "GDSCRv2 = responses\n",
    "\n",
    "ls5 = list(set(GDSCE.columns).intersection(set(DEGs_entrez_id.astype(int))))\n",
    "ls6 = GDSCEv2.index.intersection(GDSCCv2.index)\n",
    "\n",
    "GDSCEv2 = GDSCEv2.loc[ls6, ls5]\n",
    "GDSCCv2 = GDSCCv2.loc[ls6, ls5]\n",
    "GDSCRv2 = GDSCRv2.loc[ls6, :]\n",
    "\n",
    "PDXEgemcitabine = PDXEgemcitabine.loc[:,ls5]\n",
    "PDXCgemcitabine = PDXCgemcitabine.loc[:,ls5]\n",
    "TCGAEgemcitabine = TCGAEgemcitabine.loc[:,ls5]\n",
    "TCGACgemcitabine = TCGACgemcitabine.loc[:,ls5]\n",
    "\n",
    "responses.index = responses.index.values + \"_\" + responses[\"drug\"].values\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCEv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCCv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (Response): {GDSCRv2.shape[0]}\\n\")\n",
    "\n",
    "print(f\"PDX # of common samples and genes for Gemcitabine (Expression): {PDXEgemcitabine.shape}\")\n",
    "print(f\"PDX # of common samples and genes for Gemcitabine (CNA): {PDXCgemcitabine.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for Gemcitabine (Expression): {TCGAEgemcitabine.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for Gemcitabine (CNA): {TCGACgemcitabine.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 1 ... 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "# GDSC response values\n",
    "\n",
    "Y = GDSCRv2['response'].values\n",
    "print(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response'"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA response datasets (Gemcitabine)\n",
    "\n",
    "os.chdir(dataset_dir + \"response\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 25 samples with response data in the PDX (Gemcitabine)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0,\n",
       "       0, 1, 1])"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX response dataset (Gemcitabine)\n",
    "\n",
    "PDXRgemcitabine = pd.read_csv(\"PDX_response.Gemcitabine.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "PDXRgemcitabine.dropna(inplace=True)\n",
    "PDXRgemcitabine.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "PDXRgemcitabine[\"response\"] = PDXRgemcitabine.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytsgemcitabine_pdx = PDXRgemcitabine[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytsgemcitabine_pdx.shape[0]} samples with response data in the PDX (Gemcitabine)\\n\")\n",
    "Ytsgemcitabine_pdx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-0991</th>\n",
       "      <td>PDAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1199</th>\n",
       "      <td>PDAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1289</th>\n",
       "      <td>PDAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1317</th>\n",
       "      <td>PDAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1362</th>\n",
       "      <td>PDAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tumor Type\n",
       "Model            \n",
       "X-0991       PDAC\n",
       "X-1199       PDAC\n",
       "X-1289       PDAC\n",
       "X-1317       PDAC\n",
       "X-1362       PDAC"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX dataset to select cohorts (Gemcitabine)\n",
    "\n",
    "pdx_cohort = pd.read_excel(pdx_data_dir + \"/nm.3954-S2.xlsx\",\n",
    "                              \"PCT raw data\")\n",
    "pdx_cohort = pdx_cohort.loc[pdx_cohort[\"Treatment\"] == \"gemcitabine-50mpk\",]\n",
    "pdx_cohort = pdx_cohort.loc[:,[\"Model\",\"Tumor Type\"]]\n",
    "pdx_cohort = pdx_cohort.drop_duplicates()\n",
    "pdx_cohort.set_index(\"Model\", inplace = True) \n",
    "pdx_cohort.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tumor Type\n",
       "PDAC    25\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged PDX cohort table with PDX response dataset (Gemcitabine)\n",
    "\n",
    "dat_pdx = pd.merge(pdx_cohort,\n",
    "                   PDXRgemcitabine,\n",
    "                   left_index = True,\n",
    "                   right_index = True)\n",
    "dat_pdx[\"Tumor Type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['X-1199', 'X-1289', 'X-1317', 'X-1362', 'X-2026', 'X-2043', 'X-2081',\n",
       "       'X-2283', 'X-2339', 'X-2428', 'X-3052', 'X-3268', 'X-3782', 'X-3800',\n",
       "       'X-3816', 'X-3846', 'X-3947', 'X-3990', 'X-4015', 'X-4226', 'X-4316',\n",
       "       'X-4378', 'X-4649', 'X-4676', 'X-5205'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter PDAC cohort samples (Gemcitabine)\n",
    "\n",
    "filter = (dat_pdx[\"Tumor Type\"] == \"PDAC\")\n",
    "pdx_pdac_samples = dat_pdx.loc[filter,].index\n",
    "pdx_pdac_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0,\n",
       "       0, 1, 1])"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of PDAC cohort samples (Gemcitabine)\n",
    "\n",
    "Ytsgemcitabine_pdx_pdac = PDXRgemcitabine.loc[pdx_pdac_samples,\"response\"].values\n",
    "Ytsgemcitabine_pdx_pdac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    18\n",
       "1     7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistant PDAC samples (Gemcitabine)\n",
    "\n",
    "PDXRgemcitabine.loc[pdx_pdac_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 55 samples with response data in the TCGA (Gemcitabine)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0,\n",
       "       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1,\n",
       "       0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0])"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA response dataset (Gemcitabine)\n",
    "\n",
    "TCGARgemcitabine = pd.read_csv(\"TCGA_response.Gemcitabine.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "TCGARgemcitabine.dropna(inplace=True)\n",
    "TCGARgemcitabine.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "TCGARgemcitabine[\"response\"] = TCGARgemcitabine.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytsgemcitabine_tcga = TCGARgemcitabine[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytsgemcitabine_tcga.shape[0]} samples with response data in the TCGA (Gemcitabine)\\n\")\n",
    "Ytsgemcitabine_tcga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cohort\n",
       "PAAD    48\n",
       "SARC     3\n",
       "LIHC     2\n",
       "LUSC     1\n",
       "PCPG     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TCGA cohort table with TCGA response dataset (Gemcitabine)\n",
    "\n",
    "dat_tcga = TCGARgemcitabine[\"cohort\"]\n",
    "\n",
    "dat_tcga.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCGA-2L-AAQA-01A', 'TCGA-2L-AAQE-01A', 'TCGA-2L-AAQI-01A',\n",
       "       'TCGA-2L-AAQJ-01A', 'TCGA-3A-A9I9-01A', 'TCGA-3A-A9IH-01A',\n",
       "       'TCGA-3A-A9IX-01A', 'TCGA-3A-A9IZ-01A', 'TCGA-3A-A9J0-01A',\n",
       "       'TCGA-F2-6879-01A', 'TCGA-F2-6880-01A', 'TCGA-F2-7273-01A',\n",
       "       'TCGA-F2-7276-01A', 'TCGA-F2-A44G-01A', 'TCGA-F2-A44H-01A',\n",
       "       'TCGA-F2-A8YN-01A', 'TCGA-FB-A5VM-01A', 'TCGA-FB-A78T-01A',\n",
       "       'TCGA-FB-AAPP-01A', 'TCGA-FB-AAPY-01A', 'TCGA-FB-AAQ6-01A',\n",
       "       'TCGA-H8-A6C1-01A', 'TCGA-HZ-7924-01A', 'TCGA-HZ-8317-01A',\n",
       "       'TCGA-HZ-A49G-01A', 'TCGA-HZ-A49H-01A', 'TCGA-HZ-A77O-01A',\n",
       "       'TCGA-HZ-A77P-01A', 'TCGA-IB-7644-01A', 'TCGA-IB-7647-01A',\n",
       "       'TCGA-IB-7651-01A', 'TCGA-IB-7652-01A', 'TCGA-IB-7885-01A',\n",
       "       'TCGA-IB-7889-01A', 'TCGA-IB-7891-01A', 'TCGA-IB-8127-01A',\n",
       "       'TCGA-IB-A5SO-01A', 'TCGA-IB-A5SP-01A', 'TCGA-IB-A5SS-01A',\n",
       "       'TCGA-IB-A6UF-01A', 'TCGA-IB-A7LX-01A', 'TCGA-IB-AAUO-01A',\n",
       "       'TCGA-LB-A8F3-01A', 'TCGA-Q3-A5QY-01A', 'TCGA-RB-AA9M-01A',\n",
       "       'TCGA-S4-A8RM-01A', 'TCGA-US-A774-01A', 'TCGA-XD-AAUI-01A'],\n",
       "      dtype='object', name='sample')"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter PAAD cohort samples (Gemcitabine)\n",
    "\n",
    "filter = (TCGARgemcitabine[\"cohort\"] == \"PAAD\")\n",
    "tcga_paad_samples = dat_tcga.loc[filter,].index\n",
    "tcga_paad_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,\n",
       "       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,\n",
       "       0, 1, 0, 0])"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of PAAD cohort samples (Gemcitabine)\n",
    "\n",
    "Ytsgemcitabine_tcga_paad = TCGARgemcitabine.loc[tcga_paad_samples,\"response\"].values\n",
    "Ytsgemcitabine_tcga_paad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    28\n",
       "1    20\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistance PAAD samples (Gemcitabine)\n",
    "\n",
    "TCGARgemcitabine.loc[tcga_paad_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maximum iteration\n",
    "\n",
    "max_iter = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use count of nodes and dropout rates found in trained pan-drug dataset\n",
    "\n",
    "hdm1 = 16\n",
    "hdm2 = 16\n",
    "rate1 = 0.4\n",
    "rate2 = 0.3\n",
    "rate3 = 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC - Training AUC (DNA_REPi): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi): 0.32\n",
      "PDX - Test AUC (Gemcitabine): 0.42\n",
      "PDX - Test AUCPR (Gemcitabine): 0.33\n",
      "PDX - Test AUC (Gemcitabine, PDAC): 0.42\n",
      "PDX - Test AUCPR (Gemcitabine, PDAC): 0.33\n",
      "TCGA - Test AUC (Gemcitabine): 0.59\n",
      "TCGA - Test AUCPR (Gemcitabine): 0.42\n",
      "TCGA - Test AUC (Gemcitabine, PAAD): 0.58\n",
      "TCGA - Test AUCPR (Gemcitabine, PAAD): 0.48\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "PDX - Test AUC (Gemcitabine, Finetuned): 0.46\n",
      "PDX - Test AUCPR (Gemcitabine, Finetuned): 0.36\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "PDX - Test AUC (Gemcitabine, PDAC, Finetuned): 0.46\n",
      "PDX - Test AUCPR (Gemcitabine, PDAC, Finetuned): 0.36\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "TCGA - Test AUC (Gemcitabine, Finetuned): 0.59\n",
      "TCGA - Test AUCPR (Gemcitabine, Finetuned): 0.44\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "TCGA - Test AUC (Gemcitabine, PAAD, Finetuned): 0.59\n",
      "TCGA - Test AUCPR (Gemcitabine, PAAD, Finetuned): 0.51\n",
      "\n",
      "\n",
      "Models for maximum finetuning (Gemcitabine, PDX)\n",
      "Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "\n",
      "Models for maximum finetuning (Gemcitabine, PDX, PDAC)\n",
      "Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_17.pt\n",
      "\n",
      "Models for maximum finetuning (Gemcitabine, TCGA)\n",
      "Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_13.pt\n",
      "CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_13.pt\n",
      "Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_13.pt\n",
      "\n",
      "Models for maximum finetuning (Gemcitabine, TCGA, PAAD)\n",
      "Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_15.pt\n",
      "CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_15.pt\n",
      "Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_15.pt\n"
     ]
    }
   ],
   "source": [
    "# load pre-trained modules and make predictions\n",
    "\n",
    "scalerGDSC = sk.StandardScaler()\n",
    "scalerGDSC.fit(GDSCEv2.values)\n",
    "X_trainE = scalerGDSC.transform(GDSCEv2.values)\n",
    "X_testEgemcitabine_pdx = scalerGDSC.transform(PDXEgemcitabine.values) \n",
    "X_testEgemcitabine_pdx_pdac = scalerGDSC.transform(PDXEgemcitabine.loc[pdx_pdac_samples,].values) \n",
    "\n",
    "X_testEgemcitabine_tcga = scalerGDSC.transform(TCGAEgemcitabine.values)    \n",
    "X_testEgemcitabine_tcga_paad = scalerGDSC.transform(TCGAEgemcitabine.loc[tcga_paad_samples,].values)    \n",
    "\n",
    "X_trainC = np.nan_to_num(GDSCCv2.values)\n",
    "X_testCgemcitabine_pdx = np.nan_to_num(PDXCgemcitabine.values.astype(np.float32))\n",
    "X_testCgemcitabine_pdx_pdac = np.nan_to_num(PDXCgemcitabine.loc[pdx_pdac_samples,].values.astype(np.float32))\n",
    "X_testCgemcitabine_tcga = np.nan_to_num(TCGACgemcitabine.values)\n",
    "X_testCgemcitabine_tcga_paad = np.nan_to_num(TCGACgemcitabine.loc[tcga_paad_samples,].values)\n",
    "\n",
    "TX_testEgemcitabine_pdx = torch.FloatTensor(X_testEgemcitabine_pdx)\n",
    "TX_testEgemcitabine_pdx_pdac = torch.FloatTensor(X_testEgemcitabine_pdx_pdac)\n",
    "\n",
    "TX_testCgemcitabine_pdx = torch.FloatTensor(X_testCgemcitabine_pdx.astype(np.float32))\n",
    "TX_testCgemcitabine_pdx_pdac = torch.FloatTensor(X_testCgemcitabine_pdx_pdac.astype(np.float32))\n",
    "\n",
    "ty_testEgemcitabine_pdx = torch.FloatTensor(Ytsgemcitabine_pdx.astype(int))\n",
    "ty_testEgemcitabine_pdx_pdac = torch.FloatTensor(Ytsgemcitabine_pdx_pdac.astype(int))\n",
    "\n",
    "TX_testEgemcitabine_tcga = torch.FloatTensor(X_testEgemcitabine_tcga)\n",
    "TX_testEgemcitabine_tcga_paad = torch.FloatTensor(X_testEgemcitabine_tcga_paad)\n",
    "\n",
    "TX_testCgemcitabine_tcga = torch.FloatTensor(X_testCgemcitabine_tcga.astype(np.float32))\n",
    "TX_testCgemcitabine_tcga_paad = torch.FloatTensor(X_testCgemcitabine_tcga_paad.astype(np.float32))\n",
    "\n",
    "ty_testEgemcitabine_tcga = torch.FloatTensor(Ytsgemcitabine_tcga.astype(int))\n",
    "ty_testEgemcitabine_tcga_paad = torch.FloatTensor(Ytsgemcitabine_tcga_paad.astype(int))\n",
    "\n",
    "n_sampE, IE_dim = X_trainE.shape\n",
    "n_sampC, IC_dim = X_trainC.shape\n",
    "\n",
    "h_dim1 = hdm1\n",
    "h_dim2 = hdm2\n",
    "Z_in = h_dim1 + h_dim2\n",
    "\n",
    "class AEE(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEE, self).__init__()\n",
    "        self.EnE = torch.nn.Sequential(\n",
    "            nn.Linear(IE_dim, h_dim1),\n",
    "            nn.BatchNorm1d(h_dim1),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate1))\n",
    "    def forward(self, x):\n",
    "        output = self.EnE(x)\n",
    "        return output  \n",
    "    \n",
    "\n",
    "class AEC(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEC, self).__init__()\n",
    "        self.EnC = torch.nn.Sequential(\n",
    "            nn.Linear(IC_dim, h_dim2),\n",
    "            nn.BatchNorm1d(h_dim2),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate2))\n",
    "    def forward(self, x):\n",
    "        output = self.EnC(x)\n",
    "        return output \n",
    "        \n",
    "class Classifier(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Classifier, self).__init__()\n",
    "        self.FC = torch.nn.Sequential(\n",
    "            nn.Linear(Z_in, 1),\n",
    "            nn.Dropout(rate3),\n",
    "            nn.Sigmoid())\n",
    "    def forward(self, x):\n",
    "        return self.FC(x)\n",
    "\n",
    "torch.cuda.manual_seed_all(42)\n",
    "\n",
    "AutoencoderE = torch.load(models_dir + \"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy.pt\")\n",
    "AutoencoderC = torch.load(models_dir + \"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy.pt\")\n",
    "\n",
    "Clas = torch.load(models_dir + \"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy.pt\")\n",
    "\n",
    "AutoencoderE.eval()\n",
    "AutoencoderC.eval()\n",
    "Clas.eval()\n",
    "\n",
    "ZEX = AutoencoderE(torch.FloatTensor(X_trainE))\n",
    "ZCX = AutoencoderC(torch.FloatTensor(X_trainC))\n",
    "ZTX = torch.cat((ZEX, ZCX), 1)\n",
    "ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "PredX = Clas(ZTX)\n",
    "AUCT = roc_auc_score(Y, PredX.detach().numpy())\n",
    "AUCTaucpr = average_precision_score(Y, PredX.detach().numpy())\n",
    "\n",
    "ZETgemcitabine_pdx = AutoencoderE(TX_testEgemcitabine_pdx)\n",
    "ZCTgemcitabine_pdx = AutoencoderC(TX_testCgemcitabine_pdx)\n",
    "ZTTgemcitabine_pdx = torch.cat((ZETgemcitabine_pdx, ZCTgemcitabine_pdx), 1)\n",
    "ZTTgemcitabine_pdx = F.normalize(ZTTgemcitabine_pdx, p=2, dim=0)\n",
    "PredTgemcitabine_pdx = Clas(ZTTgemcitabine_pdx)\n",
    "AUCTgemcitabine_pdx = roc_auc_score(Ytsgemcitabine_pdx.astype(int), PredTgemcitabine_pdx.detach().numpy())\n",
    "AUCTgemcitabine_pdx_aucpr = average_precision_score(Ytsgemcitabine_pdx.astype(int), PredTgemcitabine_pdx.detach().numpy())\n",
    "\n",
    "ZETgemcitabine_pdx_pdac = AutoencoderE(TX_testEgemcitabine_pdx_pdac)\n",
    "ZCTgemcitabine_pdx_pdac = AutoencoderC(TX_testCgemcitabine_pdx_pdac)\n",
    "ZTTgemcitabine_pdx_pdac = torch.cat((ZETgemcitabine_pdx_pdac, ZCTgemcitabine_pdx_pdac), 1)\n",
    "ZTTgemcitabine_pdx_pdac = F.normalize(ZTTgemcitabine_pdx_pdac, p=2, dim=0)\n",
    "PredTgemcitabine_pdx_pdac = Clas(ZTTgemcitabine_pdx_pdac)\n",
    "AUCTgemcitabine_pdx_pdac = roc_auc_score(Ytsgemcitabine_pdx_pdac.astype(int), PredTgemcitabine_pdx_pdac.detach().numpy())\n",
    "AUCTgemcitabine_pdx_pdac_aucpr = average_precision_score(Ytsgemcitabine_pdx_pdac.astype(int), PredTgemcitabine_pdx_pdac.detach().numpy())\n",
    "\n",
    "ZETgemcitabine_tcga = AutoencoderE(TX_testEgemcitabine_tcga)\n",
    "ZCTgemcitabine_tcga = AutoencoderC(TX_testCgemcitabine_tcga)\n",
    "ZTTgemcitabine_tcga = torch.cat((ZETgemcitabine_tcga, ZCTgemcitabine_tcga), 1)\n",
    "ZTTgemcitabine_tcga = F.normalize(ZTTgemcitabine_tcga, p=2, dim=0)\n",
    "PredTgemcitabine_tcga = Clas(ZTTgemcitabine_tcga)\n",
    "AUCTgemcitabine_tcga = roc_auc_score(Ytsgemcitabine_tcga.astype(int), PredTgemcitabine_tcga.detach().numpy())\n",
    "AUCTgemcitabine_tcga_aucpr = average_precision_score(Ytsgemcitabine_tcga.astype(int), PredTgemcitabine_tcga.detach().numpy())\n",
    "\n",
    "ZETgemcitabine_tcga_paad = AutoencoderE(TX_testEgemcitabine_tcga_paad)\n",
    "ZCTgemcitabine_tcga_paad = AutoencoderC(TX_testCgemcitabine_tcga_paad)\n",
    "ZTTgemcitabine_tcga_paad = torch.cat((ZETgemcitabine_tcga_paad, ZCTgemcitabine_tcga_paad), 1)\n",
    "ZTTgemcitabine_tcga_paad = F.normalize(ZTTgemcitabine_tcga_paad, p=2, dim=0)\n",
    "PredTgemcitabine_tcga_paad = Clas(ZTTgemcitabine_tcga_paad)\n",
    "AUCTgemcitabine_tcga_paad = roc_auc_score(Ytsgemcitabine_tcga_paad.astype(int), PredTgemcitabine_tcga_paad.detach().numpy())\n",
    "AUCTgemcitabine_tcga_paad_aucpr = average_precision_score(Ytsgemcitabine_tcga_paad.astype(int), PredTgemcitabine_tcga_paad.detach().numpy())\n",
    "\n",
    "AUCT_finetuned = dict()\n",
    "AUCTaucpr_finetuned = dict()\n",
    "AUCTgemcitabine_pdx_finetuned = dict()\n",
    "AUCTgemcitabine_tcga_finetuned = dict()\n",
    "AUCTgemcitabine_pdx_pdac_finetuned = dict()\n",
    "AUCTgemcitabine_tcga_paad_finetuned = dict()\n",
    "AUCTgemcitabine_pdx_aucpr_finetuned = dict()\n",
    "AUCTgemcitabine_tcga_aucpr_finetuned = dict()\n",
    "AUCTgemcitabine_pdx_pdac_aucpr_finetuned = dict()\n",
    "AUCTgemcitabine_tcga_paad_aucpr_finetuned = dict()\n",
    "for iter in range(max_iter):\n",
    "   iter += 1  \n",
    "\n",
    "   AutoencoderE_finetuned = torch.load(finetuned_dir + f\"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{iter}.pt\")\n",
    "   AutoencoderC_finetuned = torch.load(finetuned_dir + f\"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{iter}.pt\")\n",
    "   Clas_finetuned = torch.load(finetuned_dir + f\"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{iter}.pt\")\n",
    "   \n",
    "   AutoencoderE_finetuned.eval()\n",
    "   AutoencoderC_finetuned.eval()\n",
    "   Clas_finetuned.eval()\n",
    "      \n",
    "   ZEX = AutoencoderE_finetuned(torch.FloatTensor(X_trainE))\n",
    "   ZCX = AutoencoderC_finetuned(torch.FloatTensor(X_trainC))\n",
    "   ZTX = torch.cat((ZEX, ZCX), 1)\n",
    "   ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "   PredX = Clas_finetuned(ZTX)\n",
    "   AUCT_finetuned[iter] = roc_auc_score(Y, PredX.detach().numpy())\n",
    "   AUCTaucpr_finetuned[iter] = average_precision_score(Y, PredX.detach().numpy())\n",
    "   \n",
    "   ZETgemcitabine_pdx = AutoencoderE_finetuned(TX_testEgemcitabine_pdx)\n",
    "   ZCTgemcitabine_pdx = AutoencoderC_finetuned(TX_testCgemcitabine_pdx)\n",
    "   ZTTgemcitabine_pdx = torch.cat((ZETgemcitabine_pdx, ZCTgemcitabine_pdx), 1)\n",
    "   ZTTgemcitabine_pdx = F.normalize(ZTTgemcitabine_pdx, p=2, dim=0)\n",
    "   PredTgemcitabine_pdx = Clas_finetuned(ZTTgemcitabine_pdx)\n",
    "   AUCTgemcitabine_pdx_finetuned[iter] = roc_auc_score(Ytsgemcitabine_pdx.astype(int), PredTgemcitabine_pdx.detach().numpy())\n",
    "   AUCTgemcitabine_pdx_aucpr_finetuned[iter] = average_precision_score(Ytsgemcitabine_pdx.astype(int), PredTgemcitabine_pdx.detach().numpy())\n",
    "\n",
    "   ZETgemcitabine_pdx_pdac = AutoencoderE_finetuned(TX_testEgemcitabine_pdx_pdac)\n",
    "   ZCTgemcitabine_pdx_pdac = AutoencoderC_finetuned(TX_testCgemcitabine_pdx_pdac)\n",
    "   ZTTgemcitabine_pdx_pdac = torch.cat((ZETgemcitabine_pdx_pdac, ZCTgemcitabine_pdx_pdac), 1)\n",
    "   ZTTgemcitabine_pdx_pdac = F.normalize(ZTTgemcitabine_pdx_pdac, p=2, dim=0)\n",
    "   PredTgemcitabine_pdx_pdac = Clas_finetuned(ZTTgemcitabine_pdx_pdac)\n",
    "   AUCTgemcitabine_pdx_pdac_finetuned[iter] = roc_auc_score(Ytsgemcitabine_pdx_pdac.astype(int), PredTgemcitabine_pdx_pdac.detach().numpy())\n",
    "   AUCTgemcitabine_pdx_pdac_aucpr_finetuned[iter] = average_precision_score(Ytsgemcitabine_pdx_pdac.astype(int), PredTgemcitabine_pdx_pdac.detach().numpy())\n",
    "\n",
    "   ZETgemcitabine_tcga = AutoencoderE_finetuned(TX_testEgemcitabine_tcga)\n",
    "   ZCTgemcitabine_tcga = AutoencoderC_finetuned(TX_testCgemcitabine_tcga)\n",
    "   ZTTgemcitabine_tcga = torch.cat((ZETgemcitabine_tcga, ZCTgemcitabine_tcga), 1)\n",
    "   ZTTgemcitabine_tcga = F.normalize(ZTTgemcitabine_tcga, p=2, dim=0)\n",
    "   PredTgemcitabine_tcga = Clas_finetuned(ZTTgemcitabine_tcga)\n",
    "   AUCTgemcitabine_tcga_finetuned[iter] = roc_auc_score(Ytsgemcitabine_tcga.astype(int), PredTgemcitabine_tcga.detach().numpy())\n",
    "   AUCTgemcitabine_tcga_aucpr_finetuned[iter] = average_precision_score(Ytsgemcitabine_tcga.astype(int), PredTgemcitabine_tcga.detach().numpy())\n",
    "\n",
    "   ZETgemcitabine_tcga_paad = AutoencoderE_finetuned(TX_testEgemcitabine_tcga_paad)\n",
    "   ZCTgemcitabine_tcga_paad = AutoencoderC_finetuned(TX_testCgemcitabine_tcga_paad)\n",
    "   ZTTgemcitabine_tcga_paad = torch.cat((ZETgemcitabine_tcga_paad, ZCTgemcitabine_tcga_paad), 1)\n",
    "   ZTTgemcitabine_tcga_paad = F.normalize(ZTTgemcitabine_tcga_paad, p=2, dim=0)\n",
    "   PredTgemcitabine_tcga_paad = Clas_finetuned(ZTTgemcitabine_tcga_paad)\n",
    "   AUCTgemcitabine_tcga_paad_finetuned[iter] = roc_auc_score(Ytsgemcitabine_tcga_paad.astype(int), PredTgemcitabine_tcga_paad.detach().numpy())\n",
    "   AUCTgemcitabine_tcga_paad_aucpr_finetuned[iter] = average_precision_score(Ytsgemcitabine_tcga_paad.astype(int), PredTgemcitabine_tcga_paad.detach().numpy())\n",
    "\n",
    "max_key_pdx = max(AUCTgemcitabine_pdx_aucpr_finetuned, key=AUCTgemcitabine_pdx_aucpr_finetuned.get)\n",
    "max_key_pdx_pdac = max(AUCTgemcitabine_pdx_pdac_aucpr_finetuned, key=AUCTgemcitabine_pdx_pdac_aucpr_finetuned.get)\n",
    "max_key_tcga = max(AUCTgemcitabine_tcga_aucpr_finetuned, key=AUCTgemcitabine_tcga_aucpr_finetuned.get)\n",
    "max_key_tcga_paad = max(AUCTgemcitabine_tcga_paad_aucpr_finetuned, key=AUCTgemcitabine_tcga_paad_aucpr_finetuned.get)\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi): {round(AUCT,2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi): {round(AUCTaucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (Gemcitabine): {round(AUCTgemcitabine_pdx,2)}\")\n",
    "print(f\"PDX - Test AUCPR (Gemcitabine): {round(AUCTgemcitabine_pdx_aucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (Gemcitabine, PDAC): {round(AUCTgemcitabine_pdx_pdac,2)}\")\n",
    "print(f\"PDX - Test AUCPR (Gemcitabine, PDAC): {round(AUCTgemcitabine_pdx_pdac_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (Gemcitabine): {round(AUCTgemcitabine_tcga,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Gemcitabine): {round(AUCTgemcitabine_tcga_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (Gemcitabine, PAAD): {round(AUCTgemcitabine_tcga_paad,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Gemcitabine, PAAD): {round(AUCTgemcitabine_tcga_paad_aucpr,2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUC (Gemcitabine, Finetuned): {round(AUCTgemcitabine_pdx_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUCPR (Gemcitabine, Finetuned): {round(AUCTgemcitabine_pdx_aucpr_finetuned[max_key_pdx],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_pdx_pdac],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx_pdac],2)}\")\n",
    "print(f\"PDX - Test AUC (Gemcitabine, PDAC, Finetuned): {round(AUCTgemcitabine_pdx_pdac_finetuned[max_key_pdx_pdac],2)}\")\n",
    "print(f\"PDX - Test AUCPR (Gemcitabine, PDAC, Finetuned): {round(AUCTgemcitabine_pdx_pdac_aucpr_finetuned[max_key_pdx_pdac],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUC (Gemcitabine, Finetuned): {round(AUCTgemcitabine_tcga_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Gemcitabine, Finetuned): {round(AUCTgemcitabine_tcga_aucpr_finetuned[max_key_tcga],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"TCGA - Test AUC (Gemcitabine, PAAD, Finetuned): {round(AUCTgemcitabine_tcga_paad_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (Gemcitabine, PAAD, Finetuned): {round(AUCTgemcitabine_tcga_paad_aucpr_finetuned[max_key_tcga_paad],2)}\\n\\n\")\n",
    "\n",
    "print(\"Models for maximum finetuning (Gemcitabine, PDX)\")\n",
    "print(f\"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Gemcitabine, PDX, PDAC)\")\n",
    "print(f\"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx_pdac}.pt\")\n",
    "print(f\"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx_pdac}.pt\")\n",
    "print(f\"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_pdx_pdac}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Gemcitabine, TCGA)\")\n",
    "print(f\"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (Gemcitabine, TCGA, PAAD)\")\n",
    "print(f\"Exprs_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n",
    "print(f\"CNA_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n",
    "print(f\"Class_Gemcitabine_GDSC_PDX_TCGA_DNA_REPi_Expression_CNA_Third_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEE(\n",
       "  (EnE): Sequential(\n",
       "    (0): Linear(in_features=772, out_features=16, bias=True)\n",
       "    (1): BatchNorm1d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.4, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show expression layer parameters\n",
    "\n",
    "AutoencoderE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEC(\n",
       "  (EnC): Sequential(\n",
       "    (0): Linear(in_features=772, out_features=16, bias=True)\n",
       "    (1): BatchNorm1d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.3, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show CNA layer parameters\n",
    "\n",
    "AutoencoderC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Classifier(\n",
       "  (FC): Sequential(\n",
       "    (0): Linear(in_features=32, out_features=1, bias=True)\n",
       "    (1): Dropout(p=0.3, inplace=False)\n",
       "    (2): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show classification layer parameters\n",
    "\n",
    "Clas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>AUC (GDSC)</th>\n",
       "      <th>AUCPR (GDSC)</th>\n",
       "      <th>AUC (PDX)</th>\n",
       "      <th>AUCPR (PDX)</th>\n",
       "      <th>AUC (PDX-PDAC)</th>\n",
       "      <th>AUCPR (PDX-PDAC)</th>\n",
       "      <th>AUC (TCGA)</th>\n",
       "      <th>AUCPR (TCGA)</th>\n",
       "      <th>AUC (TCGA-PAAD)</th>\n",
       "      <th>AUCPR (TCGA-PAAD)</th>\n",
       "      <th>Sample Size (GDSC)</th>\n",
       "      <th>Feature Size (GDSC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Expr + CNA</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.42</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.48</td>\n",
       "      <td>8843</td>\n",
       "      <td>772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Expr + CNA (Fine-tuned)</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.51</td>\n",
       "      <td>8843</td>\n",
       "      <td>772</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Data AUC (GDSC) AUCPR (GDSC) AUC (PDX) AUCPR (PDX)  \\\n",
       "0               Expr + CNA       0.81         0.32      0.42        0.33   \n",
       "1  Expr + CNA (Fine-tuned)       0.81         0.32      0.46        0.36   \n",
       "\n",
       "  AUC (PDX-PDAC) AUCPR (PDX-PDAC) AUC (TCGA) AUCPR (TCGA) AUC (TCGA-PAAD)  \\\n",
       "0           0.42             0.33       0.59         0.42            0.58   \n",
       "1           0.46             0.36       0.59         0.44            0.59   \n",
       "\n",
       "  AUCPR (TCGA-PAAD) Sample Size (GDSC) Feature Size (GDSC)  \n",
       "0              0.48               8843                 772  \n",
       "1              0.51               8843                 772  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe with the prediction results \n",
    "\n",
    "ec   = {\"Data\": [\"Expr + CNA\", \"Expr + CNA (Fine-tuned)\"],\n",
    "       \"AUC (GDSC)\": [round(AUCT, 2), round(AUCT_finetuned[max_key_pdx], 2)],\n",
    "       \"AUCPR (GDSC)\": [round(AUCTaucpr,2), round(AUCTaucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX)\": [round(AUCTgemcitabine_pdx,2), round(AUCTgemcitabine_pdx_finetuned[max_key_pdx],2)],\n",
    "       \"AUCPR (PDX)\":  [round(AUCTgemcitabine_pdx_aucpr,2), round(AUCTgemcitabine_pdx_aucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX-PDAC)\":  [round(AUCTgemcitabine_pdx_pdac,2), round(AUCTgemcitabine_pdx_pdac_finetuned[max_key_pdx_pdac],2)],\n",
    "       \"AUCPR (PDX-PDAC)\": [round(AUCTgemcitabine_pdx_pdac_aucpr,2), round(AUCTgemcitabine_pdx_pdac_aucpr_finetuned[max_key_pdx_pdac],2)],\n",
    "       \"AUC (TCGA)\": [round(AUCTgemcitabine_tcga,2), round(AUCTgemcitabine_tcga_finetuned[max_key_tcga],2)],\n",
    "       \"AUCPR (TCGA)\":  [round(AUCTgemcitabine_tcga_aucpr,2), round(AUCTgemcitabine_tcga_aucpr_finetuned[max_key_tcga],2)],\n",
    "       \"AUC (TCGA-PAAD)\":  [round(AUCTgemcitabine_tcga_paad,2), round(AUCTgemcitabine_tcga_paad_finetuned[max_key_tcga_paad],2)],\n",
    "       \"AUCPR (TCGA-PAAD)\": [round(AUCTgemcitabine_tcga_paad_aucpr,2), round(AUCTgemcitabine_tcga_paad_aucpr_finetuned[max_key_tcga_paad],2)],\n",
    "       \"Sample Size (GDSC)\": [GDSCEv2.shape[0], GDSCEv2.shape[0]],\n",
    "       \"Feature Size (GDSC)\": [GDSCEv2.shape[1], GDSCEv2.shape[1]] \n",
    "}\n",
    "\n",
    "ec_dataframe = pd.DataFrame.from_dict(ec, orient='index').transpose() \n",
    "os.makedirs(save_results_to, exist_ok = True)\n",
    "ec_dataframe.to_csv(save_results_to + \"GDSC_PDX_TCGA_Expression_CNA_DNA_REPi_Gemcitabine_Third_Strategy.tsv\",\n",
    "                     sep = \"\\t\",\n",
    "                     index = False)\n",
    "\n",
    "ec_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_CNA/Predictions/GDSC_PDX_TCGA_Third_Strategy\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(save_results_to)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save predictions\n",
    "\n",
    "file = open('GDSC_PDX_TCGA_DNA_REPi_Gemcitabine_Predictions_Third_Strategy (Expression and CNA).txt', 'w')\n",
    "file.write(f\"GDSC Training (EC) AUC (DNA_REPi): {round(AUCT,2)}\\n\")\n",
    "file.write(f\"GDSC Training (EC) AUCPR (DNA_REPi): {round(AUCTaucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EC) AUC (Gemcitabine): {round(AUCTgemcitabine_pdx,2)}\\n\")\n",
    "file.write(f\"PDX Test (EC) AUCPR (Gemcitabine): {round(AUCTgemcitabine_pdx_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EC) AUC (Gemcitabine, PDAC): {round(AUCTgemcitabine_pdx_pdac,2)}\\n\")\n",
    "file.write(f\"PDX Test (EC) AUCPR (Gemcitabine, PDAC): {round(AUCTgemcitabine_pdx_pdac_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EC) AUC (Gemcitabine): {round(AUCTgemcitabine_tcga,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EC) AUCPR (Gemcitabine): {round(AUCTgemcitabine_tcga_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EC) AUC (Gemcitabine, PAAD): {round(AUCTgemcitabine_tcga_paad,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EC) AUCPR (Gemcitabine, PAAD): {round(AUCTgemcitabine_tcga_paad_aucpr,2)}\\n\")\n",
    "\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
