{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9652d21c-3f74-45cb-bfba-22e89b1abddb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Files\n",
      "/Volumes/Expansion/Thesis Work/Codes\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n",
    "os.chdir(\"/Volumes/Expansion/Thesis Work/Codes\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d481a98d-f944-4468-9500-cc1cfffc8d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import pandas as pd\n",
    "import sys\n",
    "import pybedtools as pbt\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "%matplotlib inline\n",
    "from mapper import expand, parse_mapping_table, apply_mappers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5b3502aa-cd86-4c2a-a272-aa4eb735abfa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: '1',\n",
       " 2: '2',\n",
       " 3: '3',\n",
       " 4: '4',\n",
       " 5: '5',\n",
       " 6: '6',\n",
       " 7: '7',\n",
       " 8: '8',\n",
       " 9: '9',\n",
       " 10: '10',\n",
       " 11: '11',\n",
       " 12: '12',\n",
       " 13: '13',\n",
       " 14: '14',\n",
       " 15: '15',\n",
       " 16: '16',\n",
       " 17: '17',\n",
       " 18: '18',\n",
       " 19: '19',\n",
       " 20: '20',\n",
       " 21: '21'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_dict = dict(zip(range(1, 22), map(str, range(1,22))))\n",
    "chr_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5c84830-719a-42b3-ab41-4bf8b757d0be",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: '1',\n",
       " 2: '2',\n",
       " 3: '3',\n",
       " 4: '4',\n",
       " 5: '5',\n",
       " 6: '6',\n",
       " 7: '7',\n",
       " 8: '8',\n",
       " 9: '9',\n",
       " 10: '10',\n",
       " 11: '11',\n",
       " 12: '12',\n",
       " 13: '13',\n",
       " 14: '14',\n",
       " 15: '15',\n",
       " 16: '16',\n",
       " 17: '17',\n",
       " 18: '18',\n",
       " 19: '19',\n",
       " 20: '20',\n",
       " 21: '21',\n",
       " 22: 'X',\n",
       " 23: 'Y'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_dict.update({22: 'X', 23: \"Y\"})\n",
    "chr_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d2321ba-38c9-4029-9a7a-c67d3a1846af",
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = \"/Volumes/Expansion/Thesis Work/Supplementary Files/GRCh37/\"\n",
    "gene_coords_file = root_dir + \"ref_GRCh37.p5_top_level.gff3.bed\"  # must contain chromosome, start, end and Entrez Gene ID for hg19"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24b41e3b-6d1e-43d1-839e-8a59eb63fb88",
   "metadata": {},
   "source": [
    "# TCGA \n",
    "\n",
    "Assume that segmentation files from GDAC : http://gdac.broadinstitute.org/runs/stddata__2015_08_21/data/*/*snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.seg.txt are dowmnoaded\n",
    "\n",
    "1) Filtering segments:\n",
    " - segments containing less than 5 probes removed\n",
    " - keep only segments with segment mean below -0.23 or above 0.2. This means that one copy gains and losses are detectable when their CCF (canncer cell fraction) is 0.3 or higher. \n",
    " \n",
    "TODO: remove segements overlapping with germline CNA forund in normals (add this as the first step)\n",
    "2). For each samples aggregte to gene-level:\n",
    " - rename chromosomes 22 and 23 to X and Y\n",
    " - overpal segemntation file with Entrez gene coordinates for hg19\n",
    " - if a gene overlaps by multiple segments, keep the one with most extreme values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ab4db08-3f1f-4f69-87e3-6d10378827cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_marker_thr = 5\n",
    "# to detect 1 copy gains or losses presenting at CCF >= 0.3\n",
    "pos_seg_mean_thr = 0.3\n",
    "neg_seg_mean_thr = -0.23\n",
    "\n",
    "preprocessed_dir = \"/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c3eb8aca-94ad-48df-a94b-e01c06f57f12",
   "metadata": {},
   "outputs": [],
   "source": [
    "######## functions for TCGA and CCLE #################################\n",
    "\n",
    "def filter_lowconf_segments(df, num_marker_thr, pos_seg_mean_thr, neg_seg_mean_thr):\n",
    "    # filter low-confidence segments with too few probes\n",
    "    df = df[df[\"Num_Probes\"] >= num_marker_thr]\n",
    "    # filter low-confidence segments with Segment_Mean too close to zero:\n",
    "    df = df[(df[\"Segment_Mean\"] >= pos_seg_mean_thr) | (df[\"Segment_Mean\"] <= neg_seg_mean_thr)]  \n",
    "    return df\n",
    "\n",
    "def sample_type(barcode):\n",
    "    if barcode[13:16] in [\"10A\",\"10B\",\"11A\",\"11B\",\"10C\",\"11C\"]:\n",
    "        return \"Normal\"\n",
    "    else:\n",
    "        return \"Tumor\"\n",
    "    \n",
    "def find_matching_normal(tumor_barcode, barcodes_list):\n",
    "    patient_id = tumor_barcode[:12]\n",
    "    normal_barcodes = []\n",
    "    for barcode in barcodes_list:\n",
    "        if barcode.startswith(patient_id) and sample_type(barcode) == \"Normal\":\n",
    "            normal_barcodes.append(barcode)\n",
    "    return normal_barcodes\n",
    "            \n",
    "def cnv2bed(seg):\n",
    "    # cnv_bed = seg[[\"Chromosome\",\"Start\",\"End\",\"Segment_Mean\"]]\n",
    "    # cnv_bed.columns = [\"chrom\",\"start\",\"stop\",\"Segment_Mean\"]\n",
    "    cnv_bed = seg.rename({\"Chromosome\":\"chrom\", \n",
    "                          \"Start\":\"start\",\n",
    "                          \"End\":\"stop\"}, axis = \"columns\")\n",
    "    cnv_bed = cnv_bed.loc[:,[\"chrom\",\"start\",\"stop\",\"Segment_Mean\",\"Sample\",\"Num_Probes\"]]\n",
    "    return pbt.BedTool.from_dataframe(cnv_bed)\n",
    "\n",
    "def bed2cnv(cnv_bed):\n",
    "    cnv_bed = str(cnv_bed)\n",
    "    if len(cnv_bed) > 0:\n",
    "        seg = pd.read_csv(StringIO(cnv_bed), sep = \"\\t\", header = None)\n",
    "        seg.columns = [\"Chromosome\",\"Start\",\"End\",\"Segment_Mean\",\"Sample\",\"Num_Probes\"]\n",
    "        seg = seg.loc[:,[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "    else:\n",
    "        seg = pd.DataFrame(columns = [\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"])\n",
    "    return seg\n",
    "\n",
    "def remove_overlapping_segments(tumor, normal, sample_name):\n",
    "    tumor_bed = cnv2bed(tumor)\n",
    "    normal_bed = cnv2bed(normal)\n",
    "    tumor_wo_germline = tumor_bed.subtract(normal_bed, r=True, f= 0.8, A = True)\n",
    "    tumor_wo_germline = bed2cnv(tumor_wo_germline)\n",
    "    n_segs_removed = tumor.shape[0] - tumor_wo_germline.shape[0]\n",
    "    if n_segs_removed * 1.0 / tumor.shape[0] > 0.5 and n_segs_removed > 5:\n",
    "        print(n_segs_removed,\"of\",tumor.shape[0],\"segments removed in\",sample_name,\"due to overlap with normal\",file = sys.stderr)\n",
    "    return tumor_wo_germline\n",
    "\n",
    "def cnv2genelevel(cnv_bed, gene_intervals_bed, sample_name, verbose = True, sorted_index = \"\"):\n",
    "    cnv2gene = str(gene_intervals_bed.intersect(cnv_bed, wa = True, wb = True))\n",
    "    if len(cnv2gene) == 0:\n",
    "        print(sample_name,\"has no genes with altered CN\",file = sys.stderr)\n",
    "        return pd.DataFrame(columns = [sample])\n",
    "    cnv2gene = pd.read_csv(StringIO(cnv2gene), sep = \"\\t\", header = None)\n",
    "    cnv2gene = cnv2gene[[3,7]].copy()\n",
    "    cnv2gene.columns = [\"gene\",\"Segment_Mean\"]\n",
    "    \n",
    "    # find genes overlapping with more than one segment:\n",
    "    # take the most exterme segement_mean value\n",
    "    \n",
    "    dups = cnv2gene.loc[cnv2gene.duplicated(subset = [\"gene\"], keep = False),]\n",
    "    if dups.shape[0] > 0:\n",
    "        cnv2gene = cnv2gene.drop_duplicates(subset = [\"gene\"], keep = False)\n",
    "        dups[\"abs_seg_mean\"] = abs(dups[\"Segment_Mean\"])\n",
    "        if verbose:\n",
    "            print(sample_name,\"contain \",len(set(dups[\"gene\"].values)),\"genes overlapped with more than one segment\",file=sys.stderr)\n",
    "            #print(dups.head(10),file=sys.stderr)            \n",
    "        dups = dups.groupby(['gene'], group_keys=False).apply(lambda row: row.loc[row['abs_seg_mean'].idxmax()])\n",
    "        cnv2gene = pd.concat([cnv2gene,dups],sort=False)\n",
    "        \n",
    "    cnv2gene = cnv2gene[[\"gene\",\"Segment_Mean\"]]\n",
    "    cnv2gene.set_index(\"gene\",inplace=True,drop=True)\n",
    "    cnv2gene.rename(int,axis=0,inplace=True)\n",
    "    # add copy-neutral genes with 0s\n",
    "    \n",
    "    cnv2gene = cnv2gene.loc[sorted(set(cnv2gene.index) & set(sorted_index)),:]\n",
    "    cnv2gene.columns = [sample_name]\n",
    "    return cnv2gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "51c0963a-2047-4a37-bc4a-ed60d19dc999",
   "metadata": {},
   "outputs": [],
   "source": [
    "### functions for GDSC and PDX #################################\n",
    "\n",
    "def CN2log2R(col, median_ploidy=2 ):\n",
    "    # this is for GDSC only\n",
    "    lRs = []\n",
    "    genes = col.index.values\n",
    "    for code in col.values:\n",
    "        if not code == \"-1,-1,-,-\":\n",
    "            [max_cn,min_cn,zygosity,disruption] = code.split(\",\")\n",
    "            if int(max_cn) == 0:\n",
    "                lRs.append(-4.32) # CN=0 with 95% purity\n",
    "            else:\n",
    "                max_lR = np.log2(float(max_cn)/median_ploidy)\n",
    "                if not disruption == \"D\":\n",
    "                    lRs.append(max_lR)\n",
    "                else:\n",
    "                    if int(min_cn) == 0:\n",
    "                        min_lR = -4.32\n",
    "                    else:\n",
    "                        min_lR = np.log2(float(min_cn)/median_ploidy)\n",
    "                    if abs(min_lR) > abs(max_lR):\n",
    "                        lRs.append(min_lR)\n",
    "                    else:\n",
    "                        lRs.append(max_lR)\n",
    "                \n",
    "        else:\n",
    "            lRs.append(np.NaN)\n",
    "    return pd.Series(dict(zip(genes, lRs)))\n",
    "\n",
    "def define_avg_ploidy(col):\n",
    "    n,pl = 0,0\n",
    "    CN_non_disrupted = []\n",
    "    for code in col.values:\n",
    "        if not code == \"-1,-1,-,-\":\n",
    "            [max_cn,min_cn,zygosity,disruption] = code.split(\",\")\n",
    "            n+=1\n",
    "            cn = (int(max_cn)+int(min_cn))*0.5\n",
    "            pl += cn\n",
    "            if not disruption == \"D\":\n",
    "                CN_non_disrupted.append((cn))\n",
    "    return pd.Series({\"avg_pl\":pl/n , \"median_pl\":np.median(CN_non_disrupted)})\n",
    "\n",
    "def clean_logR(logR_value, pos_seg_mean_thr, neg_seg_mean_thr):\n",
    "    if logR_value >= pos_seg_mean_thr:\n",
    "        return logR_value \n",
    "    elif logR_value <= neg_seg_mean_thr:\n",
    "            return logR_value \n",
    "    else:\n",
    "        return 0\n",
    "    \n",
    "def handle_dups(df,corr_thr = 0.75):\n",
    "    '''Detect dupliated row IDs. Merge 2 or more rows with the same ID, \n",
    "    if averaged correlation in all pairvise comparision is >= corr_thhr;\\n\n",
    "    otherwise drop all duplicates.  Keeps abs. max value (negative preferred).'''\n",
    "    dups = df.index\n",
    "    dups = list(set(dups[dups.duplicated()]))\n",
    "    if len(dups)==0:\n",
    "        print(\"No duplicated row IDs. Do nothing.\")\n",
    "        return df\n",
    "    print(len(dups), \"duplicated IDs in\",df.loc[dups,:].shape[0],\"rows found.\")\n",
    "    dups_merge = [] # if corr > corr_thr\n",
    "    dups_remove = [] # corr < \n",
    "    for dup in dups:\n",
    "        r = df.loc[dup,:].T.corr()\n",
    "        n_dups = df.loc[dup,:].shape[0]\n",
    "        r_avg = []\n",
    "        for i in range(0,n_dups):\n",
    "            for j in range(i+1,n_dups):\n",
    "                r_avg.append(r.iloc[i,j])\n",
    "        if np.average(r_avg) < corr_thr :\n",
    "            #print(dup,r_avg, n_dups)\n",
    "            dups_remove.append(dup)\n",
    "        else:\n",
    "            dups_merge.append(dup)\n",
    "    \n",
    "    # remove not similar duplicates\n",
    "    df_size = df.shape[0]\n",
    "    df = df.loc[~df.index.isin(dups_remove),:]\n",
    "    print(\"duplicate rows removed due to low correlation of duplicated profiles\", df_size - df.shape[0])\n",
    "    df_size = df.shape[0]\n",
    "    \n",
    "    # merge simialr duplicates\n",
    "    d1 = df.loc[~df.index.isin(dups_merge),:]\n",
    "    d2 = df.loc[dups_merge,:]\n",
    "    d2 = d2.groupby(d2.index).agg(lambda x: -max(-x.max(),-x.min(),key= abs))\n",
    "    df = pd.concat([d1,d2])\n",
    "    df.sort_index(inplace=True)\n",
    "    print(\"Merged \",df_size-df.shape[0]+len(dups_merge),\"duplicated rows into\",len(dups_merge),\"rows\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2abce6-20ef-40f2-8dc4-e2b01441da2c",
   "metadata": {
    "tags": []
   },
   "source": [
    "### next few tabs demonstrate necessity of removing low-confidence and germline segments:  \n",
    "\n",
    "(e.g. fragment 11:126596926-127130276 presents in both tumor and normal\n",
    "therefore, it is germline; see chr11:126596926-12713027 in UCSC browser - it covers part of KIRELL3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "166a7c86-64dd-4448-a9fa-f8f94ead7a11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Codes\n",
      "/Volumes/Expansion/Thesis Work/Datasets/TCGA/CNA/data__2016_01_28\n"
     ]
    }
   ],
   "source": [
    "print(os.getcwd())\n",
    "os.chdir(\"/Volumes/Expansion/Thesis Work/Datasets/TCGA/CNA/data__2016_01_28\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "02daa612-a1db-4c38-af04-a7aa2ec105c9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57858</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3218610</td>\n",
       "      <td>48833552</td>\n",
       "      <td>24063.0</td>\n",
       "      <td>0.0205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57859</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>1</td>\n",
       "      <td>48861267</td>\n",
       "      <td>80199565</td>\n",
       "      <td>19329.0</td>\n",
       "      <td>-0.5717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57860</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>1</td>\n",
       "      <td>80200131</td>\n",
       "      <td>80200369</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.3683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57861</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>1</td>\n",
       "      <td>80205990</td>\n",
       "      <td>85979932</td>\n",
       "      <td>3694.0</td>\n",
       "      <td>-0.5614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57862</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>1</td>\n",
       "      <td>85980350</td>\n",
       "      <td>86006009</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome     Start       End  \\\n",
       "57858  TCGA-ZJ-AAXJ-01A-11D-A42N-01           1   3218610  48833552   \n",
       "57859  TCGA-ZJ-AAXJ-01A-11D-A42N-01           1  48861267  80199565   \n",
       "57860  TCGA-ZJ-AAXJ-01A-11D-A42N-01           1  80200131  80200369   \n",
       "57861  TCGA-ZJ-AAXJ-01A-11D-A42N-01           1  80205990  85979932   \n",
       "57862  TCGA-ZJ-AAXJ-01A-11D-A42N-01           1  85980350  86006009   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57858     24063.0        0.0205  \n",
       "57859     19329.0       -0.5717  \n",
       "57860         2.0       -2.3683  \n",
       "57861      3694.0       -0.5614  \n",
       "57862        12.0        0.0854  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"gdac.broadinstitute.org_CESC.Merge_snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.Level_3.2016012800.0.0/CESC.snp__genome_wide_snp_6__broad_mit_edu__Level_3__segmented_scna_minus_germline_cnv_hg19__seg.seg.txt\"\n",
    "df = pd.read_csv(file_path, sep = \"\\t\")\n",
    "tumor_barcode = \"TCGA-ZJ-AAXJ-01A-11D-A42N-01\"\n",
    "t = df.loc[df[\"Sample\"]==tumor_barcode,:]\n",
    "t_shape = t.shape[0]\n",
    "t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ea983282-d8df-4bec-9c16-422278dab7f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TCGA-ZJ-AAXJ-10A-01D-A42Q-01']\n",
      "segments in tumor 204 segments in normal 121\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57737</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>1</td>\n",
       "      <td>3218610</td>\n",
       "      <td>31396170</td>\n",
       "      <td>14758.0</td>\n",
       "      <td>-0.0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57738</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>1</td>\n",
       "      <td>31399900</td>\n",
       "      <td>31402148</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-1.8255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57739</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>1</td>\n",
       "      <td>31403343</td>\n",
       "      <td>247813706</td>\n",
       "      <td>113946.0</td>\n",
       "      <td>0.0024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57740</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>2</td>\n",
       "      <td>484222</td>\n",
       "      <td>242476062</td>\n",
       "      <td>130939.0</td>\n",
       "      <td>0.0057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57741</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>3</td>\n",
       "      <td>2212571</td>\n",
       "      <td>13912850</td>\n",
       "      <td>7013.0</td>\n",
       "      <td>0.0036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome     Start        End  \\\n",
       "57737  TCGA-ZJ-AAXJ-10A-01D-A42Q-01           1   3218610   31396170   \n",
       "57738  TCGA-ZJ-AAXJ-10A-01D-A42Q-01           1  31399900   31402148   \n",
       "57739  TCGA-ZJ-AAXJ-10A-01D-A42Q-01           1  31403343  247813706   \n",
       "57740  TCGA-ZJ-AAXJ-10A-01D-A42Q-01           2    484222  242476062   \n",
       "57741  TCGA-ZJ-AAXJ-10A-01D-A42Q-01           3   2212571   13912850   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57737     14758.0       -0.0003  \n",
       "57738         2.0       -1.8255  \n",
       "57739    113946.0        0.0024  \n",
       "57740    130939.0        0.0057  \n",
       "57741      7013.0        0.0036  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = find_matching_normal(tumor_barcode,list(set(df[\"Sample\"].values)))\n",
    "print(n)\n",
    "\n",
    "n = df.loc[df[\"Sample\"]==n[0],:]\n",
    "print(\"segments in tumor\",t.shape[0],\"segments in normal\",n.shape[0])\n",
    "n.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9fe01174-523b-49e4-b754-720984663da0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57803</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>8896255</td>\n",
       "      <td>4489.0</td>\n",
       "      <td>-0.0113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57804</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8899400</td>\n",
       "      <td>8899668</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.3344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57805</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8900394</td>\n",
       "      <td>126596817</td>\n",
       "      <td>67487.0</td>\n",
       "      <td>0.0010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57806</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>453.0</td>\n",
       "      <td>-1.0306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57807</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>128342803</td>\n",
       "      <td>864.0</td>\n",
       "      <td>-0.0031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57808</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>128342819</td>\n",
       "      <td>128350888</td>\n",
       "      <td>44.0</td>\n",
       "      <td>0.2824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57809</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>128353007</td>\n",
       "      <td>134142530</td>\n",
       "      <td>3708.0</td>\n",
       "      <td>0.0082</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57803  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11     456120    8896255   \n",
       "57804  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8899400    8899668   \n",
       "57805  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8900394  126596817   \n",
       "57806  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  126596926  127130276   \n",
       "57807  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  127132920  128342803   \n",
       "57808  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  128342819  128350888   \n",
       "57809  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  128353007  134142530   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57803      4489.0       -0.0113  \n",
       "57804         3.0       -1.3344  \n",
       "57805     67487.0        0.0010  \n",
       "57806       453.0       -1.0306  \n",
       "57807       864.0       -0.0031  \n",
       "57808        44.0        0.2824  \n",
       "57809      3708.0        0.0082  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n.loc[n['Chromosome']==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ba561e3e-0384-4c12-ae7f-628926004083",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57960</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>64200041</td>\n",
       "      <td>34710.0</td>\n",
       "      <td>0.0054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57961</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64208988</td>\n",
       "      <td>64319750</td>\n",
       "      <td>61.0</td>\n",
       "      <td>-0.6748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57962</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64325209</td>\n",
       "      <td>126596817</td>\n",
       "      <td>37207.0</td>\n",
       "      <td>0.0571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57963</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>454.0</td>\n",
       "      <td>-1.0760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57964</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>132080656</td>\n",
       "      <td>3591.0</td>\n",
       "      <td>0.0449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57965</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132080885</td>\n",
       "      <td>132099465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.6123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57966</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132099856</td>\n",
       "      <td>134142530</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>0.0483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57960  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11     456120   64200041   \n",
       "57961  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64208988   64319750   \n",
       "57962  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64325209  126596817   \n",
       "57963  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  126596926  127130276   \n",
       "57964  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  127132920  132080656   \n",
       "57965  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132080885  132099465   \n",
       "57966  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132099856  134142530   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57960     34710.0        0.0054  \n",
       "57961        61.0       -0.6748  \n",
       "57962     37207.0        0.0571  \n",
       "57963       454.0       -1.0760  \n",
       "57964      3591.0        0.0449  \n",
       "57965        15.0       -0.6123  \n",
       "57966      1010.0        0.0483  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t.loc[t[\"Chromosome\"] ==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "97a32a93-c345-425b-b592-3506db2ddf32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segemtns in normal after dropping low.conf.: 38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57804</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>8899400</td>\n",
       "      <td>8899668</td>\n",
       "      <td>3.0</td>\n",
       "      <td>-1.3344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57806</th>\n",
       "      <td>TCGA-ZJ-AAXJ-10A-01D-A42Q-01</td>\n",
       "      <td>11</td>\n",
       "      <td>126596926</td>\n",
       "      <td>127130276</td>\n",
       "      <td>453.0</td>\n",
       "      <td>-1.0306</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             Sample  Chromosome      Start        End  \\\n",
       "57804  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11    8899400    8899668   \n",
       "57806  TCGA-ZJ-AAXJ-10A-01D-A42Q-01          11  126596926  127130276   \n",
       "\n",
       "       Num_Probes  Segment_Mean  \n",
       "57804         3.0       -1.3344  \n",
       "57806       453.0       -1.0306  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = filter_lowconf_segments(n, 0, 0.46,-0.68)\n",
    "print(\"segemtns in normal after dropping low.conf.:\",n.shape[0])\n",
    "n.loc[n['Chromosome']==11,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b9c27078-7408-4e0b-afee-4a0fde63e11f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "segemtns in tumor after dropping low.conf.: 194\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Num_Probes</th>\n",
       "      <th>Segment_Mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>456120</td>\n",
       "      <td>64200041</td>\n",
       "      <td>34710.0</td>\n",
       "      <td>0.0054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64208988</td>\n",
       "      <td>64319750</td>\n",
       "      <td>61.0</td>\n",
       "      <td>-0.6748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>64325209</td>\n",
       "      <td>126596817</td>\n",
       "      <td>37207.0</td>\n",
       "      <td>0.0571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>127132920</td>\n",
       "      <td>132080656</td>\n",
       "      <td>3591.0</td>\n",
       "      <td>0.0449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132080885</td>\n",
       "      <td>132099465</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-0.6123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>TCGA-ZJ-AAXJ-01A-11D-A42N-01</td>\n",
       "      <td>11</td>\n",
       "      <td>132099856</td>\n",
       "      <td>134142530</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>0.0483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Sample  Chromosome      Start        End  \\\n",
       "96   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11     456120   64200041   \n",
       "97   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64208988   64319750   \n",
       "98   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11   64325209  126596817   \n",
       "99   TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  127132920  132080656   \n",
       "100  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132080885  132099465   \n",
       "101  TCGA-ZJ-AAXJ-01A-11D-A42N-01          11  132099856  134142530   \n",
       "\n",
       "     Num_Probes  Segment_Mean  \n",
       "96      34710.0        0.0054  \n",
       "97         61.0       -0.6748  \n",
       "98      37207.0        0.0571  \n",
       "99       3591.0        0.0449  \n",
       "100        15.0       -0.6123  \n",
       "101      1010.0        0.0483  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = remove_overlapping_segments(t, n, tumor_barcode)\n",
    "print(\"segemtns in tumor after dropping low.conf.:\",t.shape[0])\n",
    "t.loc[t[\"Chromosome\"] == 11,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "661e38df-b428-4399-bbeb-27366d957d1b",
   "metadata": {},
   "source": [
    "# TCGA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f12f710a-cb7a-4b55-ac62-a747eee10ceb",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRP samples: 590 CNA events per sample on avg.: 79.51525423728813\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 590 tumors: 288 normals: 302\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 268\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KIRP samples: 283 Segments per sample on avg.: 20.15547703180212\n",
      "KIRP samples: 590 CNA events per sample on avg.: 79.51525423728813\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 590 tumors: 288 normals: 302\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 268\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KIRP samples: 283 Segments per sample on avg.: 20.15547703180212\n",
      "ESCA samples: 373 CNA events per sample on avg.: 163.01072386058982\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 373 tumors: 185 normals: 188\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 181\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ESCA samples: 184 Segments per sample on avg.: 124.59239130434783\n",
      "ESCA samples: 373 CNA events per sample on avg.: 163.01072386058982\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 373 tumors: 185 normals: 188\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 181\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ESCA samples: 184 Segments per sample on avg.: 124.59239130434783\n",
      "STAD samples: 904 CNA events per sample on avg.: 130.9612831858407\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 904 tumors: 442 normals: 462\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 409\n",
      "\ttumors without any somatic CNA 7\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STAD samples: 435 Segments per sample on avg.: 85.9816091954023\n",
      "STAD samples: 904 CNA events per sample on avg.: 130.9612831858407\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 904 tumors: 442 normals: 462\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 409\n",
      "\ttumors without any somatic CNA 7\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STAD samples: 435 Segments per sample on avg.: 85.9816091954023\n",
      "GBM samples: 1104 CNA events per sample on avg.: 133.018115942029\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1104 tumors: 590 normals: 514\n",
      "\ttumors without matched normal 78\n",
      "\ttumors with at least one sCNA 511\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "GBM samples: 589 Segments per sample on avg.: 66.48217317487267\n",
      "GBM samples: 1104 CNA events per sample on avg.: 133.018115942029\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1104 tumors: 590 normals: 514\n",
      "\ttumors without matched normal 78\n",
      "\ttumors with at least one sCNA 511\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "GBM samples: 589 Segments per sample on avg.: 66.48217317487267\n",
      "UVM samples: 160 CNA events per sample on avg.: 81.08125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 160 tumors: 80 normals: 80\n",
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 80\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 160 tumors: 80 normals: 80\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UVM samples: 80 Segments per sample on avg.: 37.625\n",
      "UVM samples: 160 CNA events per sample on avg.: 81.08125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 80\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UVM samples: 80 Segments per sample on avg.: 37.625\n",
      "UCEC samples: 4 CNA events per sample on avg.: 349.5\n",
      "after filtering\n",
      "UCEC samples: 4 Segments per sample on avg.: 87.25\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 4 tumors: 4 normals: 0\n",
      "\ttumors without matched normal 4\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SARC samples: 513 CNA events per sample on avg.: 208.0682261208577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 513 tumors: 263 normals: 250\n",
      "\ttumors without matched normal 17\n",
      "\ttumors with at least one sCNA 245\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SARC samples: 262 Segments per sample on avg.: 171.58396946564886\n",
      "SARC samples: 513 CNA events per sample on avg.: 208.0682261208577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 513 tumors: 263 normals: 250\n",
      "\ttumors without matched normal 17\n",
      "\ttumors with at least one sCNA 245\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SARC samples: 262 Segments per sample on avg.: 171.58396946564886\n",
      "THYM samples: 248 CNA events per sample on avg.: 62.78629032258065\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 248 tumors: 125 normals: 123\n",
      "\ttumors without matched normal 5\n",
      "\ttumors with at least one sCNA 104\n",
      "\ttumors without any somatic CNA 16\n",
      "total samples: 248 tumors: 125 normals: 123\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THYM samples: 109 Segments per sample on avg.: 8.559633027522937\n",
      "THYM samples: 248 CNA events per sample on avg.: 62.78629032258065\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 5\n",
      "\ttumors with at least one sCNA 104\n",
      "\ttumors without any somatic CNA 16\n",
      "total samples: 7 tumors: 7 normals: 0\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 12 tumors: 12 normals: 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THYM samples: 109 Segments per sample on avg.: 8.559633027522937\n",
      "BRCA samples: 7 CNA events per sample on avg.: 677.5714285714286\n",
      "after filtering\n",
      "BRCA samples: 7 Segments per sample on avg.: 179.71428571428572\n",
      "LUAD samples: 12 CNA events per sample on avg.: 739.8333333333334\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LUAD samples: 12 Segments per sample on avg.: 183.25\n",
      "COAD samples: 12 CNA events per sample on avg.: 379.5\n",
      "after filtering\n",
      "COAD samples: 12 Segments per sample on avg.: 98.41666666666667\n",
      "SKCM samples: 937 CNA events per sample on avg.: 115.35112059765208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 937 tumors: 472 normals: 465\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 460\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SKCM samples: 467 Segments per sample on avg.: 74.92933618843684\n",
      "SKCM samples: 937 CNA events per sample on avg.: 115.35112059765208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 937 tumors: 472 normals: 465\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 460\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "SKCM samples: 467 Segments per sample on avg.: 74.92933618843684\n",
      "GBMLGG samples: 2119 CNA events per sample on avg.: 106.95752713544124\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 2119 tumors: 1120 normals: 999\n",
      "\ttumors without matched normal 111\n",
      "\ttumors with at least one sCNA 1004\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "GBMLGG samples: 1115 Segments per sample on avg.: 47.96233183856502\n",
      "GBMLGG samples: 2119 CNA events per sample on avg.: 106.95752713544124\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 2119 tumors: 1120 normals: 999\n",
      "\ttumors without matched normal 111\n",
      "\ttumors with at least one sCNA 1004\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "GBMLGG samples: 1115 Segments per sample on avg.: 47.96233183856502\n",
      "CHOL samples: 85 CNA events per sample on avg.: 89.05882352941177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 85 tumors: 36 normals: 49\n",
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 36\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 85 tumors: 36 normals: 49\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CHOL samples: 36 Segments per sample on avg.: 49.416666666666664\n",
      "CHOL samples: 85 CNA events per sample on avg.: 89.05882352941177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 36\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CHOL samples: 36 Segments per sample on avg.: 49.416666666666664\n",
      "LGG samples: 1015 CNA events per sample on avg.: 78.61182266009853\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1015 tumors: 530 normals: 485\n",
      "\ttumors without matched normal 33\n",
      "\ttumors with at least one sCNA 493\n",
      "\ttumors without any somatic CNA 4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LGG samples: 526 Segments per sample on avg.: 27.224334600760457\n",
      "LGG samples: 1015 CNA events per sample on avg.: 78.61182266009853\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1015 tumors: 530 normals: 485\n",
      "\ttumors without matched normal 33\n",
      "\ttumors with at least one sCNA 493\n",
      "\ttumors without any somatic CNA 4\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LGG samples: 526 Segments per sample on avg.: 27.224334600760457\n",
      "MESO samples: 172 CNA events per sample on avg.: 106.59883720930233\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 172 tumors: 87 normals: 85\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 83\n",
      "\ttumors without any somatic CNA 2\n",
      "total samples: 172 tumors: 87 normals: 85\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "MESO samples: 85 Segments per sample on avg.: 53.858823529411765\n",
      "MESO samples: 172 CNA events per sample on avg.: 106.59883720930233\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 83\n",
      "\ttumors without any somatic CNA 2\n",
      "total samples: 180 tumors: 90 normals: 90\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "MESO samples: 85 Segments per sample on avg.: 53.858823529411765\n",
      "ACC samples: 180 CNA events per sample on avg.: 116.95555555555555\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 88\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 180 tumors: 90 normals: 90\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ACC samples: 90 Segments per sample on avg.: 94.72222222222223\n",
      "ACC samples: 180 CNA events per sample on avg.: 116.95555555555555\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 88\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 3 tumors: 3 normals: 0\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 96 tumors: 52 normals: 44\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "ACC samples: 90 Segments per sample on avg.: 94.72222222222223\n",
      "KIPAN samples: 3 CNA events per sample on avg.: 284.0\n",
      "after filtering\n",
      "KIPAN samples: 3 Segments per sample on avg.: 58.666666666666664\n",
      "DLBC samples: 96 CNA events per sample on avg.: 97.32291666666667\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 40\n",
      "\ttumors without any somatic CNA 2\n",
      "total samples: 96 tumors: 52 normals: 44\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "DLBC samples: 50 Segments per sample on avg.: 40.42\n",
      "DLBC samples: 96 CNA events per sample on avg.: 97.32291666666667\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 40\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "DLBC samples: 50 Segments per sample on avg.: 40.42\n",
      "CESC samples: 586 CNA events per sample on avg.: 101.45051194539249\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 586 tumors: 297 normals: 289\n",
      "\ttumors without matched normal 16\n",
      "\ttumors with at least one sCNA 280\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CESC samples: 296 Segments per sample on avg.: 53.1554054054054\n",
      "CESC samples: 586 CNA events per sample on avg.: 101.45051194539249\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 586 tumors: 297 normals: 289\n",
      "\ttumors without matched normal 16\n",
      "\ttumors with at least one sCNA 280\n",
      "\ttumors without any somatic CNA 1\n",
      "total samples: 6 tumors: 6 normals: 0\n",
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "CESC samples: 296 Segments per sample on avg.: 53.1554054054054\n",
      "PRAD samples: 6 CNA events per sample on avg.: 453.1666666666667\n",
      "after filtering\n",
      "PRAD samples: 6 Segments per sample on avg.: 109.5\n",
      "TGCT samples: 304 CNA events per sample on avg.: 83.8125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 304 tumors: 156 normals: 148\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 154\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "TGCT samples: 156 Segments per sample on avg.: 32.59615384615385\n",
      "TGCT samples: 304 CNA events per sample on avg.: 83.8125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 304 tumors: 156 normals: 148\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 154\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "TGCT samples: 156 Segments per sample on avg.: 32.59615384615385\n",
      "STES samples: 1277 CNA events per sample on avg.: 140.32263116679718\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1277 tumors: 627 normals: 650\n",
      "\ttumors without matched normal 29\n",
      "\ttumors with at least one sCNA 590\n",
      "\ttumors without any somatic CNA 8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STES samples: 619 Segments per sample on avg.: 97.45880452342487\n",
      "STES samples: 1277 CNA events per sample on avg.: 140.32263116679718\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1277 tumors: 627 normals: 650\n",
      "\ttumors without matched normal 29\n",
      "\ttumors with at least one sCNA 590\n",
      "\ttumors without any somatic CNA 8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "STES samples: 619 Segments per sample on avg.: 97.45880452342487\n",
      "COADREAD samples: 12 CNA events per sample on avg.: 379.5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "COADREAD samples: 12 Segments per sample on avg.: 98.41666666666667\n",
      "LIHC samples: 760 CNA events per sample on avg.: 122.8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 760 tumors: 373 normals: 387\n",
      "\ttumors without matched normal 21\n",
      "\ttumors with at least one sCNA 349\n",
      "\ttumors without any somatic CNA 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LIHC samples: 370 Segments per sample on avg.: 73.3\n",
      "LIHC samples: 760 CNA events per sample on avg.: 122.8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 760 tumors: 373 normals: 387\n",
      "\ttumors without matched normal 21\n",
      "\ttumors with at least one sCNA 349\n",
      "\ttumors without any somatic CNA 3\n",
      "total samples: 3 tumors: 3 normals: 0\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LIHC samples: 370 Segments per sample on avg.: 73.3\n",
      "KIRC samples: 3 CNA events per sample on avg.: 284.0\n",
      "after filtering\n",
      "KIRC samples: 3 Segments per sample on avg.: 58.666666666666664\n",
      "PCPG samples: 346 CNA events per sample on avg.: 90.33526011560694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 346 tumors: 168 normals: 178\n",
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 1\n",
      "total samples: 346 tumors: 168 normals: 178\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PCPG samples: 167 Segments per sample on avg.: 41.01796407185629\n",
      "PCPG samples: 346 CNA events per sample on avg.: 90.33526011560694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PCPG samples: 167 Segments per sample on avg.: 41.01796407185629\n",
      "LUSC samples: 1032 CNA events per sample on avg.: 130.68217054263565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1032 tumors: 501 normals: 531\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 475\n",
      "\ttumors without any somatic CNA 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LUSC samples: 498 Segments per sample on avg.: 85.24698795180723\n",
      "LUSC samples: 1032 CNA events per sample on avg.: 130.68217054263565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1032 tumors: 501 normals: 531\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 475\n",
      "\ttumors without any somatic CNA 3\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LUSC samples: 498 Segments per sample on avg.: 85.24698795180723\n",
      "KICH samples: 132 CNA events per sample on avg.: 77.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 132 tumors: 66 normals: 66\n",
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 66\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 132 tumors: 66 normals: 66\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KICH samples: 66 Segments per sample on avg.: 42.333333333333336\n",
      "KICH samples: 132 CNA events per sample on avg.: 77.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 0\n",
      "\ttumors with at least one sCNA 66\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "KICH samples: 66 Segments per sample on avg.: 42.333333333333336\n",
      "OV samples: 1168 CNA events per sample on avg.: 224.04109589041096\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1168 tumors: 597 normals: 571\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 571\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "OV samples: 597 Segments per sample on avg.: 184.85427135678393\n",
      "OV samples: 1168 CNA events per sample on avg.: 224.04109589041096\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1168 tumors: 597 normals: 571\n",
      "\ttumors without matched normal 26\n",
      "\ttumors with at least one sCNA 571\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "OV samples: 597 Segments per sample on avg.: 184.85427135678393\n",
      "PAAD samples: 365 CNA events per sample on avg.: 95.36438356164383\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 365 tumors: 185 normals: 180\n",
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 14\n",
      "total samples: 365 tumors: 185 normals: 180\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PAAD samples: 171 Segments per sample on avg.: 28.339181286549707\n",
      "PAAD samples: 365 CNA events per sample on avg.: 95.36438356164383\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 10\n",
      "\ttumors with at least one sCNA 161\n",
      "\ttumors without any somatic CNA 14\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "PAAD samples: 171 Segments per sample on avg.: 28.339181286549707\n",
      "LAML samples: 380 CNA events per sample on avg.: 74.53684210526316\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 380 tumors: 191 normals: 189\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 172\n",
      "\ttumors without any somatic CNA 16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LAML samples: 175 Segments per sample on avg.: 6.8914285714285715\n",
      "LAML samples: 380 CNA events per sample on avg.: 74.53684210526316\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 380 tumors: 191 normals: 189\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 172\n",
      "\ttumors without any somatic CNA 16\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "LAML samples: 175 Segments per sample on avg.: 6.8914285714285715\n",
      "READ samples: 316 CNA events per sample on avg.: 113.18037974683544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 316 tumors: 166 normals: 150\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 141\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "READ samples: 164 Segments per sample on avg.: 63.23780487804878\n",
      "READ samples: 316 CNA events per sample on avg.: 113.18037974683544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 316 tumors: 166 normals: 150\n",
      "\ttumors without matched normal 23\n",
      "\ttumors with at least one sCNA 141\n",
      "\ttumors without any somatic CNA 2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "READ samples: 164 Segments per sample on avg.: 63.23780487804878\n",
      "THCA samples: 1013 CNA events per sample on avg.: 54.409674234945705\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1013 tumors: 506 normals: 507\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 387\n",
      "\ttumors without any somatic CNA 104\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THCA samples: 402 Segments per sample on avg.: 3.6766169154228856\n",
      "THCA samples: 1013 CNA events per sample on avg.: 54.409674234945705\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1013 tumors: 506 normals: 507\n",
      "\ttumors without matched normal 15\n",
      "\ttumors with at least one sCNA 387\n",
      "\ttumors without any somatic CNA 104\n",
      "total samples: 111 tumors: 56 normals: 55\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "THCA samples: 402 Segments per sample on avg.: 3.6766169154228856\n",
      "UCS samples: 111 CNA events per sample on avg.: 173.85585585585585\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 54\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 111 tumors: 56 normals: 55\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UCS samples: 56 Segments per sample on avg.: 159.69642857142858\n",
      "UCS samples: 111 CNA events per sample on avg.: 173.85585585585585\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 54\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 2 tumors: 2 normals: 0\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "UCS samples: 56 Segments per sample on avg.: 159.69642857142858\n",
      "BLCA samples: 2 CNA events per sample on avg.: 1064.0\n",
      "after filtering\n",
      "BLCA samples: 2 Segments per sample on avg.: 354.5\n",
      "HNSC samples: 1089 CNA events per sample on avg.: 101.27548209366391\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1089 tumors: 530 normals: 559\n",
      "\ttumors without matched normal 28\n",
      "\ttumors with at least one sCNA 497\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "HNSC samples: 525 Segments per sample on avg.: 54.73714285714286\n",
      "HNSC samples: 1089 CNA events per sample on avg.: 101.27548209366391\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 1089 tumors: 530 normals: 559\n",
      "\ttumors without matched normal 28\n",
      "\ttumors with at least one sCNA 497\n",
      "\ttumors without any somatic CNA 5\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "HNSC samples: 525 Segments per sample on avg.: 54.73714285714286\n",
      "BLCA samples: 2 CNA events per sample on avg.: 1064.0\n",
      "after filtering\n",
      "BLCA samples: 2 Segments per sample on avg.: 354.5\n",
      "BRCA samples: 7 CNA events per sample on avg.: 677.5714285714286\n",
      "after filtering\n",
      "BRCA samples: 7 Segments per sample on avg.: 179.71428571428572\n",
      "COADREAD samples: 12 CNA events per sample on avg.: 379.5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 2 tumors: 2 normals: 0\n",
      "\ttumors without matched normal 2\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 7 tumors: 7 normals: 0\n",
      "\ttumors without matched normal 7\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "after filtering\n",
      "COADREAD samples: 12 Segments per sample on avg.: 98.41666666666667\n",
      "COAD samples: 12 CNA events per sample on avg.: 379.5\n",
      "after filtering\n",
      "COAD samples: 12 Segments per sample on avg.: 98.41666666666667\n",
      "KIRC samples: 3 CNA events per sample on avg.: 284.0\n",
      "after filtering\n",
      "KIRC samples: 3 Segments per sample on avg.: 58.666666666666664\n",
      "KIPAN samples: 3 CNA events per sample on avg.: 284.0\n",
      "after filtering\n",
      "KIPAN samples: 3 Segments per sample on avg.: 58.666666666666664\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 3 tumors: 3 normals: 0\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 3 tumors: 3 normals: 0\n",
      "\ttumors without matched normal 3\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LUAD samples: 12 CNA events per sample on avg.: 739.8333333333334\n",
      "after filtering\n",
      "LUAD samples: 12 Segments per sample on avg.: 183.25\n",
      "PRAD samples: 6 CNA events per sample on avg.: 453.1666666666667\n",
      "after filtering\n",
      "PRAD samples: 6 Segments per sample on avg.: 109.5\n",
      "UCEC samples: 4 CNA events per sample on avg.: 349.5\n",
      "after filtering\n",
      "UCEC samples: 4 Segments per sample on avg.: 87.25\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "total samples: 12 tumors: 12 normals: 0\n",
      "\ttumors without matched normal 12\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 6 tumors: 6 normals: 0\n",
      "\ttumors without matched normal 6\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n",
      "total samples: 4 tumors: 4 normals: 0\n",
      "\ttumors without matched normal 4\n",
      "\ttumors with at least one sCNA 0\n",
      "\ttumors without any somatic CNA 0\n"
     ]
    }
   ],
   "source": [
    "dfs = {}\n",
    "dfs_normals = {}\n",
    "tumors_without_CNA = {}\n",
    "for f in os.listdir():\n",
    "    if f.endswith(\"tar.gz\"):\n",
    "        fp = f.replace(\".tar.gz\",\"\")\n",
    "        fp = fp.replace(\"._gdac.broadinstitute.org\",\"gdac.broadinstitute.org\")\n",
    "        cohort = fp.split(\".\")[2].replace(\"org_\",\"\")\n",
    "        file_path = fp+\"/\"+cohort+\".\"+fp.split(\".\")[3].replace(\"Merge_\",\"\")+\".seg.txt\"\n",
    "        df = pd.read_csv(file_path, sep = \"\\t\")\n",
    "        \n",
    "        df[\"Chromosome\"] = df[\"Chromosome\"].map(chr_dict)\n",
    "        print(cohort,\"samples:\",len(set(df[\"Sample\"].values)),\n",
    "              \"CNA events per sample on avg.:\",float(df.shape[0])/len(set(df[\"Sample\"].values)))\n",
    "\n",
    "        #### remove segments overlapping with segemnts in normals by 80% or more reciprocally ####\n",
    "        df[\"type\"] = df[\"Sample\"].apply(sample_type)\n",
    "        df_normals = df.loc[df[\"type\"]== \"Normal\",[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "        df_tumors = df.loc[df[\"type\"]== \"Tumor\",[\"Sample\",\"Chromosome\",\"Start\",\"End\",\"Num_Probes\",\"Segment_Mean\"]]\n",
    "        normal_samples = list(set(df_normals[\"Sample\"].values))\n",
    "        tumor_samples = list(set(df_tumors[\"Sample\"].values))\n",
    "        print(\"total samples:\", len(set(df[\"Sample\"].values)),\n",
    "              \"tumors:\",len(tumor_samples),\"normals:\",len(normal_samples),file= sys.stderr)\n",
    "        \n",
    "        tumors_without_somatic_CNA = []\n",
    "        tumors_germline_removed = []\n",
    "        tumors_without_matching_normal = []\n",
    "        filtered_normals = []\n",
    "        for tumor_sample in tumor_samples:\n",
    "            #print(tumor_sample, find_matching_normal(sample,list(set(d[\"Sample\"]))))\n",
    "            tumor = df_tumors.loc[df_tumors [\"Sample\"]== tumor_sample,:]\n",
    "            matching_normals = find_matching_normal(tumor_sample,normal_samples)\n",
    "            if len(matching_normals) >0:\n",
    "                n_segs = tumor.shape[0]\n",
    "                for normal_sample in matching_normals:\n",
    "                    normal  =  df_normals.loc[df_normals[\"Sample\"]== normal_sample,:]\n",
    "                    # thresholds for +1 and -1 copy in 75% of normal cell;\n",
    "                    # this is to retain segments appeared due to slight tumor contamination\n",
    "                    normal =  filter_lowconf_segments(normal,0,0.46, -0.68 )\n",
    "                    filtered_normals.append(normal)\n",
    "                    tumor = remove_overlapping_segments(tumor, normal,tumor_sample)\n",
    "                #if n_segs > tumor.shape[0]:\n",
    "                #    print(n_segs - tumor.shape[0],\"segments removed in sample\",tumor_sample,\n",
    "                #    tumor.shape[0],\"remained\",file= sys.stderr)\n",
    "                tumor = filter_lowconf_segments(tumor,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "                if tumor.shape[0] == 0:\n",
    "                    tumors_without_somatic_CNA.append(tumor_sample)\n",
    "                else:\n",
    "                    tumors_germline_removed.append(tumor)\n",
    "            else:\n",
    "                tumor = filter_lowconf_segments(tumor,num_marker_thr,pos_seg_mean_thr, neg_seg_mean_thr )\n",
    "                if tumor.shape[0] == 0:\n",
    "                    tumors_without_somatic_CNA.append(tumor_sample)\n",
    "                else:\n",
    "                    tumors_without_matching_normal.append(tumor)\n",
    "\n",
    "        print(\"\\ttumors without matched normal\",len(tumors_without_matching_normal),file= sys.stderr)\n",
    "        print(\"\\ttumors with at least one sCNA\",len(tumors_germline_removed),file= sys.stderr)\n",
    "        print(\"\\ttumors without any somatic CNA\",len(tumors_without_somatic_CNA),file= sys.stderr)\n",
    "        #dfs[cohort] = df\n",
    "        filtered_tumors = pd.concat(tumors_germline_removed+tumors_without_matching_normal)\n",
    "        dfs[cohort] = filtered_tumors\n",
    "        \n",
    "        if(len(filtered_normals) > 0):\n",
    "            filtered_normals = pd.concat(filtered_normals)\n",
    "            dfs_normals[cohort] = filtered_normals\n",
    "            \n",
    "        tumors_without_CNA[cohort] = tumors_without_somatic_CNA\n",
    "        print(\"after filtering\")\n",
    "        print(cohort,\"samples:\",len(set(filtered_tumors[\"Sample\"].values)),\n",
    "              \"Segments per sample on avg.:\",float(filtered_tumors.shape[0])/len(set(filtered_tumors[\"Sample\"].values)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2286b0dc-11c0-48e4-a088-b22cc7765450",
   "metadata": {},
   "source": [
    "### Aggregating to gene-level\n",
    "\n",
    "Gene annotation must be:\n",
    " - with Entrez gene IDs \n",
    " - in hg19 coordinates\n",
    " - with columns \"chrom\",\"start\",\"stop\",\"gene\" (this is foru-column bed format)\n",
    " \n",
    "wget ftp://ftp.ncbi.nih.gov/genomes/H_sapiens/ARCHIVE/BUILD.37.3/GFF/ref_GRCh37.p5_top_level.gff3.gz\n",
    "\n",
    "echo -e \"chrom\\tstart\\tstop\\tgene\\tname\"  > ref_GRCh37.p5_top_level.gff3.bed;\n",
    "zcat < ref_GRCh37.p5_top_level.gff3.gz  | awk '$3==\"gene\"' | cut -f 1,4,5,9| sed -e 's/;/\\t/g'| cut -f 1-3,5,6 | grep  GeneID |  sed -re 's/(Dbxref=GeneID:[0-9]*),.*/\\1/' | sed -e 's/Name=//' -e 's/Dbxref=GeneID://' | awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$5\"\\t\"$4}'  >> ref_GRCh37.p5_top_level.gff3.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a8100bf7-1535-408d-b11f-c53d6c6ca3ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_chroms = {\"NC_000001.10\":1,\"NC_000002.11\":2,\"NC_000003.11\":3,\"NC_000004.11\":4,\n",
    "                 \"NC_000005.9\":5,\"NC_000006.11\":6,\"NC_000007.13\":7,\"NC_000008.10\":8,\n",
    "                 \"NC_000009.11\":9,\"NC_000010.10\":10,\"NC_000011.9\":11,\"NC_000012.11\":12,\"NC_000013.10\":13,\n",
    "                 \"NC_000014.8\":14,\"NC_000015.9\":15,\"NC_000016.9\":16,\"NC_000017.10\":17,\n",
    "                 \"NC_000018.9\":18,\"NC_000019.9\":19,\"NC_000020.10\":20,\"NC_000021.8\":21,\n",
    "                 \"NC_000022.10\":22,\"NC_000023.10\":23,\"NC_000024.9\":24}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f9e3807b-1c75-4203-8911-be9c7e3b02a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>10954</td>\n",
       "      <td>11507</td>\n",
       "      <td>100506145</td>\n",
       "      <td>LOC100506145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>12190</td>\n",
       "      <td>13639</td>\n",
       "      <td>100652771</td>\n",
       "      <td>LOC100652771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>653635</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>100302278</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>34611</td>\n",
       "      <td>36081</td>\n",
       "      <td>645520</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>52453</td>\n",
       "      <td>53396</td>\n",
       "      <td>79504</td>\n",
       "      <td>OR4G4P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>63016</td>\n",
       "      <td>63885</td>\n",
       "      <td>403263</td>\n",
       "      <td>OR4G11P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>69091</td>\n",
       "      <td>70008</td>\n",
       "      <td>79501</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>131125</td>\n",
       "      <td>135677</td>\n",
       "      <td>100420257</td>\n",
       "      <td>LOC100420257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_000001.10</td>\n",
       "      <td>136698</td>\n",
       "      <td>140566</td>\n",
       "      <td>729737</td>\n",
       "      <td>LOC729737</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom   start    stop       gene          name\n",
       "0  NC_000001.10   10954   11507  100506145  LOC100506145\n",
       "1  NC_000001.10   12190   13639  100652771  LOC100652771\n",
       "2  NC_000001.10   14362   29370     653635        WASH7P\n",
       "3  NC_000001.10   30366   30503  100302278     MIR1302-2\n",
       "4  NC_000001.10   34611   36081     645520       FAM138A\n",
       "5  NC_000001.10   52453   53396      79504        OR4G4P\n",
       "6  NC_000001.10   63016   63885     403263       OR4G11P\n",
       "7  NC_000001.10   69091   70008      79501         OR4F5\n",
       "8  NC_000001.10  131125  135677  100420257  LOC100420257\n",
       "9  NC_000001.10  136698  140566     729737     LOC729737"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intervals = pd.read_csv(gene_coords_file, sep = \"\\t\")\n",
    "gene_intervals.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d61b81b5-2800-4753-8f95-aa3a2f0bb3db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosomes: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]\n",
      "(36019, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gene</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10954</td>\n",
       "      <td>11507</td>\n",
       "      <td>100506145</td>\n",
       "      <td>LOC100506145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>12190</td>\n",
       "      <td>13639</td>\n",
       "      <td>100652771</td>\n",
       "      <td>LOC100652771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>653635</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>100302278</td>\n",
       "      <td>MIR1302-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom  start   stop       gene          name\n",
       "0      1  10954  11507  100506145  LOC100506145\n",
       "1      1  12190  13639  100652771  LOC100652771\n",
       "2      1  14362  29370     653635        WASH7P\n",
       "3      1  30366  30503  100302278     MIR1302-2"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_intervals = pd.read_csv(gene_coords_file, sep = \"\\t\")\n",
    "gene_intervals = gene_intervals.loc[gene_intervals[\"chrom\"].isin(rename_chroms.keys()),:]\n",
    "gene_intervals[\"chrom\"] = gene_intervals[\"chrom\"].apply(lambda x : rename_chroms[x])\n",
    "gene_intervals.head()\n",
    "print(\"chromosomes:\",list(set(gene_intervals[\"chrom\"].values)))\n",
    "gene_intervals = gene_intervals.sort_values(by=[\"chrom\",\"start\",\"stop\"],ascending=True)\n",
    "gene_intervals.to_csv(root_dir + \"ref_GRCh37.p5_top_level.gff3.chroms_renamed.bed\",sep = \"\\t\",index=False)\n",
    "print(gene_intervals.shape)\n",
    "gene_intervals.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "99e24399-10f3-48d0-bcfa-7e1319043611",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_intervals_bed = pbt.BedTool.from_dataframe(gene_intervals[[\"chrom\",\"start\",\"stop\",\"gene\"]])\n",
    "# prepare copy-neutral table\n",
    "cnv_baseline = gene_intervals.copy()\n",
    "cnv_baseline[\"Segment_Mean\"] = [0]*cnv_baseline.shape[0]\n",
    "cnv_baseline = cnv_baseline[[\"gene\",\"Segment_Mean\"]]\n",
    "cnv_baseline.set_index(\"gene\",inplace=True,drop=True)\n",
    "cnv_baseline.sort_index(inplace=True)\n",
    "sorted_index = list(cnv_baseline.index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "21938a0a-c0d7-4f8a-8dd3-b3bbaaf66e2a",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KIRP\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-A4-7828-01A-11D-2135-01 has no genes with altered CN\n",
      "TCGA-2Z-A9J5-01A-21D-A381-01 has no genes with altered CN\n",
      "TCGA-IA-A83S-01A-11D-A34Y-01 has no genes with altered CN\n",
      "TCGA-DW-7838-01A-11D-2135-01 has no genes with altered CN\n",
      "TCGA-BQ-7062-01A-11D-1959-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-4A-A93Y-01A-11D-A36W-01 has no genes with altered CN\n",
      "TCGA-GL-A4EM-01A-11D-A253-01 has no genes with altered CN\n",
      "TCGA-Y8-A8S1-01A-11D-A36W-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRP (31922, 288)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ESCA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ESCA (31922, 185)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STAD\n",
      "TCGA-MX-A5UG-01A-21D-A31K-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BR-7957-01A-11D-2200-01 has no genes with altered CN\n",
      "TCGA-D7-A6ET-01A-32D-A32M-01 has no genes with altered CN\n",
      "TCGA-CG-5722-01A-21D-1599-01 has no genes with altered CN\n",
      "TCGA-BR-A4J9-01A-12D-A253-01 has no genes with altered CN\n",
      "TCGA-BR-7196-01A-11D-2052-01 has no genes with altered CN\n",
      "TCGA-D7-6522-01A-11D-1799-01 has no genes with altered CN\n",
      "TCGA-HU-A4GJ-01A-11D-A253-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-RD-A8NB-01A-12D-A396-01 has no genes with altered CN\n",
      "TCGA-VQ-A91W-01A-11D-A40Z-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BR-6563-01A-13D-2052-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STAD (31922, 442)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBM\n",
      "TCGA-06-5410-01A-01D-1694-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-0119-01A-08D-0214-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-0165-01A-01D-0236-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n",
      "... 400 processed.\n",
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GBM (31922, 590)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UVM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UVM (31843, 80)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UCEC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCEC (13188, 4)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SARC\n",
      "TCGA-WK-A8Y0-10D-01D-A419-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-WK-A8XS-10E-01D-A37E-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-QQ-A5V2-01A-11D-A32H-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SARC (31922, 263)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "THYM\n",
      "TCGA-ZB-A96O-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-4V-A9QQ-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XU-A92X-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8M4-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96B-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-3S-AAYX-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96G-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-XM-AAZ3-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8DC-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XH-A853-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XU-AAXZ-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZB-A96M-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8DB-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96E-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8D8-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-XM-A8RB-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-X7-A8M8-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96R-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZT-A8OM-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-XU-AAY1-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-YT-A95E-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZB-A96A-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-3Q-A9WF-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZC-AAAA-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-ZC-AAAH-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-X7-A8M1-01A-11D-A422-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-X7-A8M7-01A-11D-A422-01 has no genes with altered CN\n",
      "TCGA-ZB-A96Q-01A-11D-A427-01 has no genes with altered CN\n",
      "TCGA-4X-A9F9-01A-11D-A422-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THYM (31922, 125)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BRCA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BRCA (14141, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LUAD\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LUAD (10348, 12)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "COAD\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COAD (8910, 12)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SKCM\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EB-A3XD-01A-22D-A238-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EB-A4OZ-01A-12D-A25P-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-ER-A19A-06A-21D-A191-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SKCM (31922, 472)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "GBMLGG\n",
      "TCGA-TM-A84B-12A-01D-A366-01 has no genes with altered CN\n",
      "TCGA-P5-A5EY-01A-11D-A27J-01 has no genes with altered CN\n",
      "TCGA-HT-7680-01A-11D-2252-01 has no genes with altered CN\n",
      "TCGA-DU-7011-01A-11D-2023-01 has no genes with altered CN\n",
      "TCGA-06-5410-01A-01D-1694-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-FG-7643-01A-11D-2085-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-06-0119-01A-08D-0214-01 has no genes with altered CN\n",
      "TCGA-FG-8189-01B-11D-A288-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-DU-5872-02A-21D-A36N-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n",
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-S9-A6WI-01A-21D-A33S-01 has no genes with altered CN\n",
      "TCGA-06-0165-01A-01D-0236-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 600 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-HT-7602-01A-21D-2085-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 700 processed.\n",
      "... 800 processed.\n",
      "... 900 processed.\n",
      "... 1000 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-CS-6669-01A-11D-1892-01 has no genes with altered CN\n",
      "TCGA-HT-8106-01A-11D-2391-01 has no genes with altered CN\n",
      "TCGA-FG-8181-01A-11D-2252-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 1100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GBMLGG (31922, 1120)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CHOL\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHOL (31577, 36)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LGG\n",
      "TCGA-TM-A84B-12A-01D-A366-01 has no genes with altered CN\n",
      "TCGA-P5-A5EY-01A-11D-A27J-01 has no genes with altered CN\n",
      "TCGA-HT-7680-01A-11D-2252-01 has no genes with altered CN\n",
      "TCGA-DU-7011-01A-11D-2023-01 has no genes with altered CN\n",
      "TCGA-FG-7643-01A-11D-2085-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-FG-8189-01B-11D-A288-01 has no genes with altered CN\n",
      "TCGA-DU-5872-02A-21D-A36N-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-S9-A6WI-01A-21D-A33S-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-HT-7602-01A-21D-2085-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-CS-6669-01A-11D-1892-01 has no genes with altered CN\n",
      "TCGA-HT-8106-01A-11D-2391-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-FG-8181-01A-11D-2252-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LGG (31922, 530)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "MESO\n",
      "TCGA-LK-A4O6-01A-11D-A34B-01 has no genes with altered CN\n",
      "TCGA-TS-A7P8-01A-11D-A34B-01 has no genes with altered CN\n",
      "TCGA-3H-AB3O-01A-11D-A39Q-01 has no genes with altered CN\n",
      "TCGA-TS-A8AS-01A-21D-A39Q-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MESO (31922, 87)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ACC\n",
      "TCGA-OR-A5KV-01A-11D-A29H-01 has no genes with altered CN\n",
      "TCGA-OR-A5KQ-01A-11D-A309-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACC (31922, 90)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KIPAN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIPAN (2460, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DLBC\n",
      "TCGA-FF-8062-01A-11D-2209-01 has no genes with altered CN\n",
      "TCGA-GR-A4D5-01A-11D-A31W-01 has no genes with altered CN\n",
      "TCGA-G8-6914-14A-01D-2209-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DLBC (31806, 52)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CESC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CESC (31922, 297)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PRAD\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRAD (2148, 6)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TGCT\n",
      "TCGA-YU-A90S-01A-11D-A434-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "TGCT (31922, 156)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "STES\n",
      "TCGA-MX-A5UG-01A-21D-A31K-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BR-7957-01A-11D-2200-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D7-A6ET-01A-32D-A32M-01 has no genes with altered CN\n",
      "TCGA-CG-5722-01A-21D-1599-01 has no genes with altered CN\n",
      "TCGA-BR-A4J9-01A-12D-A253-01 has no genes with altered CN\n",
      "TCGA-BR-7196-01A-11D-2052-01 has no genes with altered CN\n",
      "TCGA-D7-6522-01A-11D-1799-01 has no genes with altered CN\n",
      "TCGA-HU-A4GJ-01A-11D-A253-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n",
      "... 400 processed.\n",
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-RD-A8NB-01A-12D-A396-01 has no genes with altered CN\n",
      "TCGA-VQ-A91W-01A-11D-A40Z-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 600 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BR-6563-01A-13D-2052-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STES (31922, 627)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "COADREAD\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "COADREAD (8910, 12)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LIHC\n",
      "TCGA-2V-A95S-10D-01D-A36Z-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-DD-A4NL-01A-11D-A28W-01 has no genes with altered CN\n",
      "TCGA-G3-A25V-01A-11D-A16U-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-UB-AA0V-01A-11D-A381-01 has no genes with altered CN\n",
      "TCGA-CC-A9FV-01A-11D-A36W-01 has no genes with altered CN\n",
      "TCGA-DD-A3A6-01A-11D-A22E-01 has no genes with altered CN\n",
      "TCGA-MR-A520-01A-11D-A25U-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-BC-A10X-01A-11D-A12Y-01 has no genes with altered CN\n",
      "TCGA-ED-A5KG-01A-11D-A27H-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LIHC (31922, 373)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KIRC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KIRC (2460, 3)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PCPG\n",
      "TCGA-SA-A6C2-01A-11D-A35H-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-WB-A817-01A-11D-A35H-01 has no genes with altered CN\n",
      "TCGA-RW-A7CZ-01A-11D-A35C-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PCPG (31922, 168)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LUSC\n",
      "TCGA-98-A53H-01A-12D-A25M-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-56-8623-01A-11D-2391-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n",
      "... 300 processed.\n",
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LUSC (31922, 501)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "KICH\n",
      "TCGA-KN-8437-01A-11D-2308-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KICH (31922, 66)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "OV\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n",
      "... 200 processed.\n",
      "... 300 processed.\n",
      "... 400 processed.\n",
      "... 500 processed.\n",
      "OV (31922, 597)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAAD\n",
      "TCGA-HZ-8002-01A-11D-2200-01 has no genes with altered CN\n",
      "TCGA-IB-AAUS-01A-12D-A38F-01 has no genes with altered CN\n",
      "TCGA-IB-AAUR-01A-21D-A38F-01 has no genes with altered CN\n",
      "TCGA-IB-A5SQ-01A-11D-A32M-01 has no genes with altered CN\n",
      "TCGA-F2-A44H-01A-11D-A26H-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-Z5-AAPL-01A-12D-A40V-01 has no genes with altered CN\n",
      "TCGA-XD-AAUG-01A-61D-A40V-01 has no genes with altered CN\n",
      "TCGA-IB-7885-01A-11D-2153-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PAAD (31922, 185)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "LAML\n",
      "TCGA-AB-2840-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2947-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2978-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2845-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2808-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2991-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2842-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2826-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2844-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2824-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2932-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2922-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2940-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2854-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-3006-03A-01D-0756-21 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-AB-2969-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2836-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2998-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2871-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2837-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2880-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2931-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2851-03A-01D-0756-21 has no genes with altered CN\n",
      "TCGA-AB-2884-03A-01D-0756-21 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LAML (29424, 191)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "READ\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "READ (31922, 166)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "THCA\n",
      "TCGA-EL-A3TB-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-ET-A3DW-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-CE-A3ME-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-EL-A4JZ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V2-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A3GX-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-EM-A3FO-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-DE-A4M8-01A-21D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A3D0-01A-12D-A201-01 has no genes with altered CN\n",
      "TCGA-L6-A4EQ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UO-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-E8-A433-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EL-A3CL-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E3-A3DZ-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-BJ-A2P4-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-EM-A2CN-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EL-A3MY-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-EM-A3AP-01A-12D-A20A-01 has no genes with altered CN\n",
      "TCGA-EL-A4K2-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FY-A3TY-01A-11D-A22Y-01 has no genes with altered CN\n",
      "TCGA-EL-A3TA-01A-12D-A22C-01 has no genes with altered CN\n",
      "TCGA-EL-A4K1-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-ET-A3BU-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-ET-A39J-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-BJ-A0Z9-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-DJ-A2Q3-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-DJ-A2Q2-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-ET-A39O-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A3AL-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-ET-A2N1-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-DE-A4M9-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A13X-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-EL-A3H8-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-E8-A415-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A40Q-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EM-A4FO-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-ET-A39M-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A2CU-01A-12D-A17S-01 has no genes with altered CN\n",
      "TCGA-DJ-A1QM-01A-11D-A14V-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EL-A3GW-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VK-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-J8-A3O2-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-ET-A39T-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-FY-A3I4-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-FY-A4B4-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VJ-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-BJ-A0ZG-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-ET-A2MZ-01A-12D-A19I-01 has no genes with altered CN\n",
      "TCGA-ET-A3BX-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-E3-A3E5-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-BJ-A28S-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UT-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-ET-A3DP-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UN-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UK-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-EL-A4K4-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A4KD-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DE-A0XZ-01A-11D-A17S-01 has no genes with altered CN\n",
      "TCGA-CE-A483-01A-11D-A23T-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EM-A3O7-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-EM-A3FM-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-EM-A1CU-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-J8-A3O2-06A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EM-A22K-01A-11D-A17S-01 has no genes with altered CN\n",
      "TCGA-EL-A3D1-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EL-A3D4-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-GE-A2C6-01A-11D-A16M-01 has no genes with altered CN\n",
      "TCGA-EM-A3FK-01A-11D-A219-01 has no genes with altered CN\n",
      "TCGA-BJ-A18Y-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PP-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-H2-A26U-01A-11D-A16M-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PO-01A-21D-A19I-01 has no genes with altered CN\n",
      "TCGA-FY-A3RA-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-DE-A0Y2-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-EL-A3ZT-01A-12D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UY-01A-21D-A22C-01 has no genes with altered CN\n",
      "TCGA-E8-A437-01A-12D-A23T-01 has no genes with altered CN\n",
      "TCGA-DJ-A3UX-01A-11D-A22C-01 has no genes with altered CN\n",
      "TCGA-E8-A418-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-J8-A3YH-01A-11D-A22Y-01 has no genes with altered CN\n",
      "TCGA-ET-A25N-01A-11D-A16M-01 has no genes with altered CN\n",
      "TCGA-ET-A39L-01A-12D-A19I-01 has no genes with altered CN\n",
      "TCGA-FY-A3R8-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-E3-A3E1-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-KS-A4I9-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EM-A4FQ-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V5-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-BJ-A0ZJ-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-EM-A1CS-01A-11D-A13V-01 has no genes with altered CN\n",
      "TCGA-ET-A40T-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EL-A3T6-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-EM-A1YD-01A-11D-A14V-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-EM-A3O8-01A-11D-A21Y-01 has no genes with altered CN\n",
      "TCGA-EL-A3ZR-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V4-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EL-A3CX-01A-11D-A19I-01 has no genes with altered CN\n",
      "TCGA-EM-A1CW-01A-21D-A13V-01 has no genes with altered CN\n",
      "TCGA-DJ-A2PS-01A-11D-A18E-01 has no genes with altered CN\n",
      "TCGA-IM-A41Z-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-EL-A4K7-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-BJ-A18Z-01A-21D-A13V-01 has no genes with altered CN\n",
      "TCGA-EM-A2P3-01A-11D-A201-01 has no genes with altered CN\n",
      "TCGA-MK-A4N7-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DJ-A4V0-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-DE-A69J-01A-21D-A396-01 has no genes with altered CN\n",
      "TCGA-FY-A4B3-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-E8-A414-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-DJ-A4UT-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-KS-A4I7-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-EM-A4FM-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-CE-A485-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-E3-A3E3-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-EL-A3N2-01A-11D-A20A-01 has no genes with altered CN\n",
      "TCGA-ET-A3DV-01A-12D-A201-01 has no genes with altered CN\n",
      "TCGA-DJ-A3VE-01A-11D-A23L-01 has no genes with altered CN\n",
      "TCGA-BJ-A0Z5-01A-11D-A10T-01 has no genes with altered CN\n",
      "TCGA-FY-A76V-01A-11D-A396-01 has no genes with altered CN\n",
      "TCGA-BJ-A45D-01A-11D-A23T-01 has no genes with altered CN\n",
      "TCGA-DE-A4MA-01A-11D-A256-01 has no genes with altered CN\n",
      "TCGA-FE-A3PC-01A-11D-A21Y-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 400 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-KS-A4I5-01A-11D-A256-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THCA (31843, 506)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UCS\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCS (31922, 56)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BLCA\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLCA (6675, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "HNSC\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 100 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-D6-6824-01A-11D-1910-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 200 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-F7-8489-01A-31D-2390-01 has no genes with altered CN\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... 300 processed.\n",
      "... 400 processed.\n",
      "... 500 processed.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TCGA-CR-6481-01A-11D-1869-01 has no genes with altered CN\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n",
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_773/142371080.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  cna_table[sample] = 0\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HNSC (31922, 530)\n"
     ]
    }
   ],
   "source": [
    "for cohort in dfs.keys():\n",
    "    print(cohort, file=sys.stderr)\n",
    "    df = dfs[cohort]\n",
    "    cna_table = []\n",
    "    n_samples = 0\n",
    "    for sample in list(set(df.Sample.values)):\n",
    "        n_samples +=1\n",
    "        cnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample].astype({'Start': 'int'})),\n",
    "                                 gene_intervals_bed,\n",
    "                                 sample,\n",
    "                                 verbose = False,\n",
    "                                 sorted_index = sorted_index)\n",
    "        \n",
    "        cna_table.append(cnv2gene)\n",
    "        if n_samples % 100 == 0:\n",
    "            print(\"...\",n_samples, \"processed.\")\n",
    "    cna_table = pd.concat(cna_table,axis =1)\n",
    "    \n",
    "\n",
    "    for sample in tumors_without_CNA[cohort]:\n",
    "        cna_table[sample] = 0\n",
    "    \n",
    "    cna_table.fillna(0, inplace = True)\n",
    "    cna_table.to_csv(preprocessed_dir+\"/TCGA-\"+cohort+\".Segment_Mean.CNA.tsv\",\n",
    "                     sep = \"\\t\",header=True,index=True)\n",
    "    print(cohort,cna_table.shape)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ae8fd0f0-48dd-4025-9773-0b9747594144",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'t = time.time()\\ncnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample]),gene_intervals_bed,sample,\\n                         verbose = False,sorted_index = sorted_index)\\nprint( time.time() - t)\\ncnv2gene'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"t = time.time()\n",
    "cnv2gene = cnv2genelevel(cnv2bed(df[df.Sample == sample]),gene_intervals_bed,sample,\n",
    "                         verbose = False,sorted_index = sorted_index)\n",
    "print( time.time() - t)\n",
    "cnv2gene\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f451d4e7-c89a-4dc2-b534-f0a84af041aa",
   "metadata": {},
   "source": [
    "# GDSC\n",
    "Assume that supplementary file with gene-level CN is downloaded :\n",
    "\n",
    "wget \n",
    "\n",
    "GDSC provides gene-level integer estimated CN, max. and min. CN over all segments covering a gene. In order to make it comparable with TCGA and CCLE, we  divide estimated CN by CN of copy-neutral state and log2-transform it. \n",
    "\n",
    "1) Copy-neutral state was defined from average ploiy, as median of integer CN values in non-disrupted genes.\n",
    "\n",
    "2) Compute log2(CN/neutral-CN) for min and max CN; keep the value with most extreme estimate\n",
    "\n",
    "3) Replace estimates below thresholds with zeroes. \n",
    "\n",
    "\n",
    "DGSC uses 4 comma-separated values for gene-level CN (max_cn,min_cn,zygosity,disruption): e.g. (from \"legend\" tab)\n",
    "\n",
    "2,2,H,-\tGene resides on a single genomic segment in a diploid region of the genome.\n",
    "2,0,L,D\tGene spans multiple segments, higest copy number is 2 but part of the coding sequence is homozygously deleted, the gene is disrupted.\n",
    "13,13,H,-\tGene resides on a single genomic segment of copy number 13 in a heterozygous part of the genome (amplification).\n",
    "14,12,L,D\tGene spans multiple genomic segments all of which are amplified to 12 or more copies, some or all segments have LOH, the gene is disrupted.\n",
    "0,0,0,-\tComplete gene sequence falls within a homozygous deletion.\n",
    "-1,-1,-,- gene level CN not assigned\n",
    "\n",
    "* min and max CN are integers \n",
    "* zygosity -  can be L (LOH in any overlapping segment) or H (heterozygous) or 0 (homozygous deleteion of the whole gene) or - (undefined)\n",
    "* disruption - D (if disrupted) or \"-\" (not disrupted) \n",
    "\n",
    "Average ploidies of cell lines were downloaded from COSMIC:\n",
    "\n",
    "wget https://cog.sanger.ac.uk/cosmic/GRCh37/cell_lines/v86/PICNIC_average_ploidies.tsv?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1540792525&Signature=mcSB6oFv%2BXCF4%2Fezm4a3Ds1JXo4%3D\n",
    "\n",
    "echo 'alpsoy@tau.edu.tr:2595.SaSa' | base64 \n",
    "\n",
    "curl -H \"Authorization: Basic YWxwc295QHRhdS5lZHUudHI6MjU5NS5TYVNhCg==\" \\\n",
    "  https://cancer.sanger.ac.uk/cosmic/file_download/GRCh37/cosmic/latest/ascat_acf_ploidy.tsv\n",
    "  \n",
    "curl -o ascat_acf_ploidy.tsv \"https://cog.sanger.ac.uk/cosmic/GRCh37/cosmic/v97/ascat_acf_ploidy.tsv?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1679695775&Signature=TEG7VBK0gXD8gonYq8S2Hmfwah0%3D\"\n",
    "\n",
    "curl -o classification.csv \"https://cog.sanger.ac.uk/cosmic/GRCh37/cosmic/v97/classification.csv?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1679649115&Signature=SvR3vixUfet%2FAvQtqzXwy%2FCHZqM%3D\"\n",
    "\n",
    "wget ftp:// ftp.sanger.ac.uk/pub/project/cancerrxgene/releases/release-7.0/Gene_level_CN.xlsx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0ce5db65-ccc0-4190-b078-9e49ea5dd1bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>201T</th>\n",
       "      <th>22RV1</th>\n",
       "      <th>23132-87</th>\n",
       "      <th>42-MG-BA</th>\n",
       "      <th>451Lu</th>\n",
       "      <th>5637</th>\n",
       "      <th>...</th>\n",
       "      <th>WSU-NHL</th>\n",
       "      <th>YAPC</th>\n",
       "      <th>YH-13</th>\n",
       "      <th>YKG-1</th>\n",
       "      <th>YMB-1-E</th>\n",
       "      <th>YT</th>\n",
       "      <th>ZR-75-30</th>\n",
       "      <th>huH-1</th>\n",
       "      <th>no-10</th>\n",
       "      <th>no-11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1287381</td>\n",
       "      <td>924100</td>\n",
       "      <td>910924</td>\n",
       "      <td>687561</td>\n",
       "      <td>1287706</td>\n",
       "      <td>687452</td>\n",
       "      <td>...</td>\n",
       "      <td>909785</td>\n",
       "      <td>909904</td>\n",
       "      <td>909905</td>\n",
       "      <td>687592</td>\n",
       "      <td>1303911</td>\n",
       "      <td>946358</td>\n",
       "      <td>909907</td>\n",
       "      <td>1298146</td>\n",
       "      <td>908452</td>\n",
       "      <td>908450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>1</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14412.0</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>...</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>1</td>\n",
       "      <td>14363.0</td>\n",
       "      <td>29806.0</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>...</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "      <td>-1,-1,-,-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  chr    start     stop       201T      22RV1   23132-87   42-MG-BA  \\\n",
       "0      NaN  NaN      NaN      NaN    1287381     924100     910924     687561   \n",
       "1  DDX11L1    1  11869.0  14412.0  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "2   WASH7P    1  14363.0  29806.0  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "\n",
       "       451Lu       5637  ...    WSU-NHL       YAPC      YH-13      YKG-1  \\\n",
       "0    1287706     687452  ...     909785     909904     909905     687592   \n",
       "1  -1,-1,-,-  -1,-1,-,-  ...  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "2  -1,-1,-,-  -1,-1,-,-  ...  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-   \n",
       "\n",
       "     YMB-1-E         YT   ZR-75-30      huH-1      no-10      no-11  \n",
       "0    1303911     946358     909907    1298146     908452     908450  \n",
       "1  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  \n",
       "2  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  -1,-1,-,-  \n",
       "\n",
       "[3 rows x 1000 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GDSC_CNA = \"/Volumes/Expansion/Thesis Work/Datasets/GDSC/CNA/GDSC_Gene_level_CN.xlsx\"\n",
    "\n",
    "gdsc = pd.read_excel(GDSC_CNA,\"Gene_level_CN\")\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "46a481ae-5603-4f6d-be91-adb607241447",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25 gene IDs excluded due to string to datetime conversion in Excel.\n",
      "Strings containing duplicated gene IDs: 0\n"
     ]
    }
   ],
   "source": [
    "gdsc.set_index(\"gene\",inplace = True)\n",
    "gdsc.drop([\"chr\",\"start\",\"stop\"],inplace=True,axis=1)\n",
    "gdsc.columns = gdsc.iloc[0,:]\n",
    "gdsc = gdsc.iloc[1:,:]\n",
    "gdsc.columns.name = None\n",
    "# replace 2001-12-01 with DEC1 and get remove gene names converted to datetimes\n",
    "gdsc.index.values[37778] = \"DEC1\"\n",
    "df_size = gdsc.shape[0]\n",
    "ndxs=pd.Series(gdsc.index).apply(lambda x : type(x) == str or type(x) == bytes)\n",
    "gdsc = gdsc.loc[gdsc.index.values[ndxs[ndxs].index],:]\n",
    "print(df_size - gdsc.shape[0],\"gene IDs excluded due to string to datetime conversion in Excel.\")\n",
    "\n",
    "gdsc.index.name = \"gene_id\"\n",
    "ids = gdsc.index\n",
    "ids = list(set(ids[ids.duplicated()]))\n",
    "print(\"Strings containing duplicated gene IDs:\",gdsc.loc[ids,:].shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00003b1-58da-490c-9d1b-ddf9ad712aa7",
   "metadata": {
    "tags": []
   },
   "source": [
    "### distribution of averaged ploidies in GDSC\n",
    "\n",
    "we compared average ploidies reported in PICNIC_average_ploidies.tsv provided by COSMIC with  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d5039ead-f860-4e1f-b7fd-e7daee0045d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1020\n",
      "1016\n"
     ]
    }
   ],
   "source": [
    "GDSC_Ploidies = \"/Volumes/Expansion/Thesis Work/Datasets/GDSC/CNA/GDSC_PICNIC_average_ploidies.tsv\"\n",
    "GDSC_Ploidies = pd.read_csv(GDSC_Ploidies,sep = \"\\t\")\n",
    "GDSC_Ploidies.drop(\"#sample_name\",axis = 1, inplace= True)\n",
    "GDSC_Ploidies.set_index(\"sample_id\",inplace=True)\n",
    "print(GDSC_Ploidies.shape[0])\n",
    "GDSC_Ploidies.dropna(inplace=True)\n",
    "print(GDSC_Ploidies.shape[0])\n",
    "\n",
    "est_ploidies = gdsc.apply(define_avg_ploidy).T\n",
    "df_ploidies = pd.DataFrame.from_dict({\"est. avg. ploidy from CN profile\":est_ploidies[\"avg_pl\"],\n",
    "                                      \"PICNIC avg. pl.\":GDSC_Ploidies[\"average_ploidy\"],\n",
    "                                     \"est. median. ploidy\":est_ploidies[\"median_pl\"]})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "779929ff-7d7c-400b-b87a-771dbbee44bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABkYAAAHBCAYAAAA8WQcTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABl20lEQVR4nO3de1hVZf7//9dOYCMIKCJsSEQqtRIzlUrtAIqHSK2kUrMD5iHzVA46JpqJTYlZmQ2mTTPm2ezblFZaeUpxHHUGTynaQQsLJ4gyBSHbIq7fH/3Yn7aAAgIbWM/Hda3rYt3rXmu/773Q/Wa9972WxTAMQwAAAAAAAAAAACZwhasDAAAAAAAAAAAAqCkURgAAAAAAAAAAgGlQGAEAAAAAAAAAAKZBYQQAAAAAAAAAAJgGhREAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACmQWEEAAAAAAAAAACYBoURAAAAAAAAAABgGhRGUOvs2LFDSUlJOnXqlKtDqTWSkpJksVgqte+QIUPUsmXLcvW1WCxKSkqq1OuUZvPmzYqMjJS3t7csFovWrFlTZceuLt9++63Gjh2r1q1bq2HDhvLy8lLbtm31zDPP6H//+5+j35AhQ2SxWNS2bVsVFRWVOI7FYtHYsWNrMvRyeeaZZ9SiRQu5ubmpcePGkqTo6GhFR0c79avq3wUAQM1avHixLBaLY3Fzc1Pz5s312GOPOX2ebd26VRaLRf/85z9LHOPAgQN67LHHFB4eLk9PTzVq1EgdO3bU7Nmz9csvvzj6RUdHy2Kx6M477yxxjGPHjslisejll1+u0tc0g9I+nwEAqIvqy3Weli1basiQIY714pxm69atLoupPC6MuyLKe22gOPc8duxYpV4HcAU3VwcAXGjHjh2aMWOGhgwZ4rhwi8qbNm2annrqqRp/XcMwNGDAALVu3VoffvihvL291aZNmxqPoyLWrl2rQYMGKSAgQGPHjlWHDh1ksVh08OBBvfXWW1q3bp327dvntM/hw4e1ePFiDRs2zEVRl98HH3ygF154QVOnTlVsbKysVqskaf78+S6ODABQXRYtWqRrr71WZ86c0bZt25ScnKzU1FQdPHhQ3t7eZe7397//XaNHj1abNm305z//Wddff70KCwu1e/duvfHGG9q5c6dWr17ttM/69ev12WefqXv37pWKtTKvCQAAar/6ep2nY8eO2rlzp66//npXh1Jtdu7cqebNm7s6DKBaUBgB6rmrr77aJa/7ww8/6JdfflH//v0VExNz0b6//vqrvLy8aiiy0mVkZGjQoEFq3bq1tmzZIj8/P8e27t2768knnyxxMcbb21sdO3bU9OnTNXjwYDVs2LBGYz5z5ow8PT3LPZsoPT1dkvTkk08qMDDQ0V6fkzgAMLuIiAhFRkZKkrp166aioiL95S9/0Zo1a/TQQw+Vus/OnTs1atQo9ezZU2vWrHEU0iWpZ8+emjBhgj799FOnfVq3bq1z585p0qRJSktLq/BM18q8JgAAgCv5+vqqc+fOrg6jWtX38cHcuJUWqsyRI0c0ePBgBQYGymq16rrrrtPrr7/u1Of8+fN6/vnn1aZNGzVs2FCNGzfWDTfcoNdee03S77eM+vOf/yxJCg8Pd9z+oaLTEn/66SeNHj1a119/vRo1aqTAwEB1795d//rXvxx9CgsLFRgYqEceeaTE/qdOnVLDhg2VkJDgaDt06JB69eolLy8vNWvWTGPGjNG6desqPW2yeMrl8uXLlZCQIJvNpoYNGyoqKqrErITSnD9/XrNnz9a1114rq9WqwMBAPfroozp+/LhTv9JupZWXl6cRI0aoadOmatSoke688059/fXXTn3+9a9/yWKx6O233y7x2kuXLpXFYlFaWlqpsSUlJTm+UfD000/LYrE4Yii+LdjevXt1//33q0mTJo7izW+//abExESFh4fLw8NDV155pcaMGVNium3Lli3Vt29frV27Vh06dFDDhg113XXXae3atZJ+n8J53XXXydvbWzfffLN27959yfdzzpw5Kigo0Pz5852KIsUsFovi4uJKtL/44ov63//+5/gdrqjiW2797W9/U+vWrWW1WnX99ddr1apVTv2Kp6Vu2LBBQ4cOVbNmzeTl5SW73V6u34WWLVvqmWeekSQFBQU5TYct7606srOzNXLkSDVv3lweHh4KDw/XjBkzdO7cuUqNHQBQ84r/uP3uu+/K7DNz5kxZLBa9+eabTgWKYh4eHrr77rud2tzd3fXCCy9oz549eueddyocV2Ve80K7d+/WoEGD1LJlSzVs2FAtW7bUgw8+6DTWzz//XBaLRQsXLiyx/yeffCKLxaIPP/zQ0fbBBx/ohhtukNVq1VVXXaXXXnvtsm5xWrzvvn37FBcXJ19fX/n5+enhhx/WTz/9VKljAgBQXWrTdZ7iv4k/++wzx/UMX19fPfrooyooKFB2drYGDBigxo0bKzg4WBMnTlRhYaHTMc6ePavnn3/e8bdzs2bN9Nhjj5X4DC4sLNSkSZNks9nk5eWl2267Tf/9739LxFTarbTKk4/8cTxbtmzRqFGjFBAQoKZNmyouLk4//PDDJd+PIUOGqFGjRjp06JBiYmLk7e2tZs2aaezYsfr1118vuf/333+vhx9+2OncvvLKKzp//rxTv9JupbVr1y7deuut8vT0VEhIiBITE0u818OGDZO/v3+psXTv3l1t27a9ZIxAdWPGCKrE4cOH1bVrV7Vo0UKvvPKKbDab1q9fryeffFI///yzpk+fLkmaPXu2kpKS9Mwzz+iOO+5QYWGhvvzyS8eF7+HDh+uXX35RSkqK3n//fQUHB0uq+Dfai+9BPX36dNlsNuXn52v16tWKjo7W5s2bFR0dLXd3dz388MN644039Prrr8vX19ex/9tvv63ffvtNjz32mCQpKytLUVFR8vb21oIFCxQYGKi33367Sp4hMWXKFHXs2FH/+Mc/lJubq6SkJEVHR2vfvn266qqrytxv1KhRevPNNzV27Fj17dtXx44d07Rp07R161bt3btXAQEBpe5nGIbuvfde7dixQ88++6xuuukm/fvf/1ZsbKxTv9tvv10dOnTQ66+/rgcffNBp27x583TTTTfppptuKvU1hg8frvbt2ysuLk7jxo3T4MGDS1zkiIuL06BBg/TEE0+ooKDAEdfmzZuVmJio22+/XQcOHND06dO1c+dO7dy50+kYn3/+uRITEzV16lT5+flpxowZiouLU2JiojZv3uy4wPL000+rb9++ysjIuOiMjg0bNigoKKjC34bo0qWL+vfvrxdffFGPP/64/P39K7S/JH344YfasmWLnnvuOXl7e2v+/Pl68MEH5ebmpvvvv9+p79ChQ9WnTx8tW7ZMBQUFcnd3L9fvwurVq/X6669r4cKF+vTTT+Xn51eh6bDZ2dm6+eabdcUVV+jZZ5/V1VdfrZ07d+r555/XsWPHtGjRogqPGwBQ844ePSpJatasWanbi4qK9Nlnn6lTp04KDQ2t0LEHDhyol19+Wc8884zuu+8+ubu7l2u/y3nNPzp27JjatGmjQYMGyd/fX1lZWVqwYIFuuukmHT58WAEBAWrfvr06dOigRYsWlbgN5uLFixUYGKi77rpLkvTpp58qLi5Od9xxh9555x2dO3dOL7/8sn788cdKx1isf//+GjBggJ544gkdOnRI06ZN0+HDh/Wf//yn3O8bAADVqbZd5yk2fPhwxcXFadWqVdq3b5+mTJmic+fO6auvvlJcXJwef/xxbdq0SS+++KJCQkIcX3g9f/687rnnHv3rX//SpEmT1LVrV3333XeaPn26oqOjtXv3bsc1gxEjRmjp0qWaOHGievbsqfT0dMXFxen06dOXjK88+ciF4+nTp49WrlypzMxM/fnPf9bDDz+szz777JKvVVhYqLvuuksjR47U5MmTtWPHDj3//PP67rvv9NFHH5W5308//aSuXbvq7Nmz+stf/qKWLVtq7dq1mjhxor755puL3m778OHDiomJUcuWLbV48WJ5eXlp/vz5WrlypVO/p556Sm+99ZZWrlyp4cOHO+2/ZcuWEgU2wCUMoAr07t3baN68uZGbm+vUPnbsWMPT09P45ZdfDMMwjL59+xo33njjRY/10ksvGZKMjIyMKovv3LlzRmFhoRETE2P079/f0X7gwAFDkvHmm2869b/55puNTp06Odb//Oc/GxaLxTh06JBTv969exuSjC1btlQ4pi1bthiSjI4dOxrnz593tB87dsxwd3c3hg8f7mibPn268cd/rl988YUhyRg9erTTMf/zn/8YkowpU6Y42uLj442wsDDH+ieffGJIMl577TWnfV944QVDkjF9+nRH26JFiwxJxr59+xxt//3vfw1JxpIlSy46voyMDEOS8dJLLzm1F4/l2WefdWr/9NNPDUnG7NmzndrfeeedEucoLCzMaNiwoXH8+HFH2/79+w1JRnBwsFFQUOBoX7NmjSHJ+PDDDy8ar6enp9G5c+eL9vmj+Ph4w9vb2zAMw/jyyy+NBg0aGBMmTHBsl2SMGTPmkseRZDRs2NDIzs52tJ07d8649tprjWuuucbRVnwuHn30Uaf9K/K7UPze//TTT059o6KijKioqBJx/fF3YeTIkUajRo2M7777zqnfyy+/bEgq8W8DAOBaxZ8bu3btMgoLC43Tp08ba9euNZo1a2b4+Pg4PneK85F3333XMAzDyM7ONiQZgwYNKvdrRUVFGW3btjUMwzA2bdpkSDJSUlIMwyg9H6iK1yyPc+fOGfn5+Ya3t7dT3vPXv/7VkGR89dVXjrZffvnFsFqtTp/lN910kxEaGmrY7XZH2+nTp42mTZs65WUVUfxZ/Kc//cmpfcWKFYYkY/ny5Y620j6fAQCoKbXtOk9xbjNu3Din9nvvvdeQZMyZM8ep/cYbbzQ6duzoWH/77bcNScZ7773n1C8tLc2QZMyfP98wjP/7G7usz+r4+HhHW3FOc7FrQmXlI8XjufBv+dmzZxuSjKysrLLfDOP3axIXu7azfft2R1tYWJhT3JMnTzYkGf/5z3+c9h01apRhsViccqQLrw0MHDiwzGsYF57jqKioEr8bo0aNMnx9fY3Tp09fdHxATeBWWrhsv/32mzZv3qz+/fvLy8tL586dcyx33XWXfvvtN+3atUuSdPPNN+vzzz/X6NGjtX79euXl5VVbXG+88YY6duwoT09Pubm5yd3dXZs3b9YXX3zh6NOuXTt16tTJ6dvuX3zxhf773/9q6NChjrbU1FRFRESU+EbDhTMpKmPw4MFOt2MICwtT165dtWXLljL3Kd42ZMgQp/abb75Z1113nTZv3nzJfS+8r/jgwYNL9H3wwQcVGBjoVMlPSUlRs2bNNHDgwLIHVQ733Xef03rxtyEuHNMDDzwgb2/vEmO68cYbdeWVVzrWr7vuOkm/3xbqj88rKW6/2C1DLlebNm00bNgwzZs3T99//32F94+JiVFQUJBjvUGDBho4cKCOHj1a4tZoF75vl/O7UBFr165Vt27dFBIS4vRvvHimUWpqapW8DgCganXu3Fnu7u7y8fFR3759ZbPZ9Mknnzh97lSlmJgY9erVS88991y5vlVZlfLz8/X000/rmmuukZubm9zc3NSoUSMVFBQ45X8PPfSQrFarFi9e7Gh7++23ZbfbHbOFCwoKtHv3bt17773y8PBw9GvUqJH69et32bFemIcNGDBAbm5uF83/AACoKbX1Oo8k9e3b12m9+G/+Pn36lGj/43WAtWvXqnHjxurXr5/TeG688UbZbDbH7bDKumZS/Fl9KeXNR4pdeKvQG264QVL5r2GUdW3nYjnFZ599puuvv14333yzU/uQIUNkGMZFZ6ts2bKlzGsYF3rqqae0f/9+/fvf/5b0+23dly1bpvj4eDVq1OjSgwOqGYURXLYTJ07o3LlzSklJkbu7u9NSfCuCn3/+WZKUmJiol19+Wbt27VJsbKyaNm2qmJiYcj0DoiLmzJmjUaNG6ZZbbtF7772nXbt2KS0tTXfeeafOnDnj1Hfo0KHauXOnvvzyS0nSokWLZLVanYoeJ06cKPUCQlVcVLDZbKW2nThxosx9ircVT0H9o5CQkEvu6+bmpqZNm14yDqvVqpEjR2rlypU6deqUfvrpJ/2///f/NHz48FLv/10RF8ZeHNeFt/awWCylvh8X3rKq+KJFWe2//fbbReNp0aKFMjIyyj+ACyQlJalBgwaaNm1ahfct63dAUolxl/a+ldYuXfp3oSJ+/PFHffTRRyX+jRffF7T43zgAoHZZunSp0tLStG/fPv3www86cOCAbr311jL7BwQEyMvL67I+E1988UX9/PPPevnll8vVvypeU/r9QsC8efM0fPhwrV+/Xv/973+VlpamZs2aOeV//v7+uvvuu7V06VIVFRVJ+v02WjfffLPjc+3kyZMyDKPG8r/i3KyqPrcBALgctfE6T7GKXAv443WAH3/8UadOnZKHh0eJMWVnZzvGU/xZXNZn9aWUNx8pduExi6+1lNb3Qhe7tnOp60JlXUMoz74Xu4bxR/fcc49atmzp+LLt4sWLVVBQoDFjxpR5fKAm8YwRXLYmTZqoQYMGeuSRR8r8zy08PFzS7/9pJyQkKCEhQadOndKmTZs0ZcoU9e7dW5mZmU7f9L8cy5cvV3R0tBYsWODUXto3Fx988EElJCRo8eLFeuGFF7Rs2TLde++9atKkiaNP06ZNS72fdHZ29mXHWtoxsrOzL/qBW7wtKyurxHMifvjhhzKfL1K877lz53TixAmn1yhrLKNGjdKsWbP01ltv6bffftO5c+f0xBNPXHRM5XHhQ0uL4/rpp5+ciiOGYSg7O7vM55lUld69eyslJUW7du2q8HNGpN8LE+PHj9esWbM0YcKECu1b1u+AVDJJKu19kyr3u1ARAQEBuuGGG/TCCy+Uur04gQIA1C7XXXedIiMjy92/QYMGiomJ0SeffKLjx49X6HlUxW688UY9+OCDmjNnjuPiSXW/Zm5urtauXavp06dr8uTJjna73e549twfPfbYY3r33Xe1ceNGtWjRQmlpaU55Y5MmTWSxWKo1//vjzNfScjMAAFylNl7nuVzFDzf/9NNPS93u4+Mj6f/+xi7rs/piKpqPXK6LXdu51DWlrKysEu3FD32/1DWli13D+KMrrrhCY8aM0ZQpU/TKK69o/vz5iomJUZs2bcoeFFCDmDGCy+bl5aVu3bpp3759uuGGGxQZGVliKe0/5MaNG+v+++/XmDFj9Msvv+jYsWOSKlYdL4vFYikxo+HAgQPauXNnib5NmjTRvffeq6VLl2rt2rXKzs52uo2WJEVFRSk9PV2HDx92al+1alWlYyz29ttvyzAMx/p3332nHTt2KDo6usx9unfvLun3AtAfpaWl6YsvvlBMTEyZ+3br1k2StGLFCqf2Cx+UVSw4OFgPPPCA5s+frzfeeEP9+vVTixYtLjqmyiiO+cIxvffeeyooKLjomKrCn/70J3l7e2v06NHKzc0tsd0wDK1evfqix3j66afl7+/vlACVx+bNm50uvBQVFemdd97R1VdffcmLQ5fzu1ARffv2VXp6uq6++upS/41TGAGA+iMxMVGGYWjEiBE6e/Zsie2FhYUXfaCnJD3//PM6e/asZsyYUSOvabFYZBhGifzvH//4h2NWyB/16tVLV155pRYtWqRFixbJ09PTabawt7e3IiMjtWbNGqd48vPztXbt2nKN6WIuzMP+3//7fzp37txF8z8AAGpKbbzOc7n69u2rEydOqKioqNTxFF+sL/4sLuuz+mIqmo9UhbKu7Vwsp4iJidHhw4e1d+9ep/alS5fKYrE4rhuVplu3bmVewyjN8OHD5eHhoYceekhfffWVxo4de6khATWGGSOoEq+99ppuu+023X777Ro1apRatmyp06dP6+jRo/roo48c9yfs16+fIiIiFBkZqWbNmum7777T3LlzFRYWplatWkn6/bkfxceMj4+Xu7u72rRpIx8fHw0bNkxLlizRN998o7CwsDLj6du3r/7yl79o+vTpioqK0ldffaXnnntO4eHhpX6QDR06VO+8847Gjh2r5s2bq0ePHk7bx48fr7feekuxsbF67rnnFBQUpJUrVzpuv3XFFf9XY3zuuef03HPPafPmzYqKirrke5eTk6P+/ftrxIgRys3N1fTp0+Xp6anExMQy92nTpo0ef/xxpaSk6IorrlBsbKyOHTumadOmKTQ0VH/605/K3LdXr1664447NGnSJBUUFCgyMlL//ve/tWzZsjL3eeqpp3TLLbdIktPzWKpSz5491bt3bz399NPKy8vTrbfeqgMHDmj69Onq0KGDHnnkkWp53WLh4eFatWqVBg4cqBtvvFFjx45Vhw4dJEmHDx/WW2+9JcMw1L9//zKP4evrq6lTp170/S9NQECAunfvrmnTpsnb21vz58/Xl19+Wa7C2+X8LlTEc889p40bN6pr16568skn1aZNG/322286duyYPv74Y73xxhuV+oYvAKD26dKlixYsWKDRo0erU6dOGjVqlNq2bavCwkLt27dPb775piIiIi76rI3w8HCNGjVKr732Wo28pq+vr+644w699NJLCggIUMuWLZWamqqFCxeqcePGJfo3aNBAjz76qObMmSNfX1/FxcXJz8/Pqc9zzz2nPn36qHfv3nrqqadUVFSkl156SY0aNSrxrc+YmBilpqZe8oJJsffff19ubm7q2bOnDh06pGnTpql9+/YaMGDARfe75pprJElHjx4t1+sAAFBZte06z+UaNGiQVqxYobvuuktPPfWUbr75Zrm7u+v48ePasmWL7rnnHvXv31/XXXedHn74Yc2dO1fu7u7q0aOH0tPT9fLLL8vX1/eir1HRfORyeXh46JVXXlF+fr5uuukm7dixQ88//7xiY2N12223lbnfn/70Jy1dulR9+vTRc889p7CwMK1bt07z58/XqFGj1Lp16zL3feaZZ/Thhx+qe/fuevbZZ+Xl5aXXX39dBQUFpfZv3LixHn30US1YsEBhYWFV8qw2oMq46qnvqH8yMjKMoUOHGldeeaXh7u5uNGvWzOjatavx/PPPO/q88sorRteuXY2AgADDw8PDaNGihTFs2DDj2LFjTsdKTEw0QkJCjCuuuMKQZGzZssUwDMOIj483JBkZGRkXjcVutxsTJ040rrzySsPT09Po2LGjsWbNGiM+Pt4ICwsr0b+oqMgIDQ01JBlTp04t9Zjp6elGjx49DE9PT8Pf398YNmyYsWTJEkOS8fnnnzv6TZ8+3SnmsmzZssWQZCxbtsx48sknjWbNmhlWq9W4/fbbjd27dzv1LT7mhTG/+OKLRuvWrQ13d3cjICDAePjhh43MzEynfqWN+dSpU8bQoUONxo0bG15eXkbPnj2NL7/80pBkTJ8+vdR4W7ZsaVx33XUXHdMfZWRkGJKMl156qdSx/PTTTyX2OXPmjPH0008bYWFhhru7uxEcHGyMGjXKOHnypFO/sLAwo0+fPiX2l2SMGTOmXHGU5ZtvvjFGjx5tXHPNNYbVajUaNmxoXH/99UZCQoLT7118fLzh7e1dYn+73W6Eh4eXGktpivvNnz/fuPrqqw13d3fj2muvNVasWOHUb9GiRYYkIy0trcQxyvu7UNZ7HxUVZURFRZWI68LfhZ9++sl48sknjfDwcMPd3d3w9/c3OnXqZEydOtXIz8+/5FgBADXnYp8bf1Scj7z77rsltu3fv9+Ij483WrRoYXh4eBje3t5Ghw4djGeffdbIyclx9IuKijLatm1bYv+ffvrJ8PX1LfE5XBWvWZrjx48b9913n9GkSRPDx8fHuPPOO4309HQjLCzMiI+PL9H/66+/NiQZkoyNGzeWeszVq1cb7dq1c+Sts2bNMp588kmjSZMmTv2ioqJK5GqlKf4s3rNnj9GvXz+jUaNGho+Pj/Hggw8aP/74Y4ljXvj5HBYWVmouCwBAdahN13nKym3K+ju3tL/ZCwsLjZdfftlo37694enpaTRq1Mi49tprjZEjRxpHjhxx9LPb7caECROMwMBAw9PT0+jcubOxc+fOEjlFcU7zx+s/5c1HyhpPaccsTfH4Dhw4YERHRxsNGzY0/P39jVGjRpX4+7y0XOi7774zBg8ebDRt2tRwd3c32rRpY7z00ktGUVGRU7/Srg38+9//Njp37mxYrVbDZrMZf/7zn40333yzzPO4detWQ5Ixa9asi44JqGkWw/jDPXwAVMjjjz+ut99+WydOnHA88Ku8tm7dqm7duundd9/V/fffX00RVo0DBw6offv2ev311zV69GhXh1OvWCwWjRkzRvPmzXN1KAAA4BIKCwt144036sorr9SGDRsqvH9SUpJmzJihn376qcqeAwYAAMxnyJAh+uc//6n8/HxXh3JJEyZM0IIFC5SZmcnz1FCrcCstoJyee+45hYSE6KqrrnLcX/of//iHnnnmmQoXReqKb775Rt99952mTJmi4OBgDRkyxNUhAQAA1Jhhw4apZ8+eCg4OVnZ2tt544w198cUX5b5FGAAAgFnt2rVLX3/9tebPn6+RI0dSFEGtQ2EEKCd3d3e99NJLOn78uM6dO6dWrVppzpw5euqpp1wdWrX5y1/+omXLlum6667Tu+++Ky8vL1eHBAAAUGNOnz6tiRMn6qeffpK7u7s6duyojz/+uMTz6AAAAOCsS5cu8vLyUt++ffX888+7OhygBG6lBQAAAAAAAAAATOMKVwcAAAAAAAAAAABQUyiMAAAAAAAAAAAA06AwAgAAAAAAAAAATKNOPnz9/Pnz+uGHH+Tj4yOLxeLqcAAAqBUMw9Dp06cVEhKiK67guw/ViVwEAICSyEVqDrkIAAAlVSQXqZOFkR9++EGhoaGuDgMAgFopMzNTzZs3d3UY9Rq5CAAAZSMXqX7kIgAAlK08uUidLIz4+PhI+n2Avr6+Lo4GAIDaIS8vT6GhoY7PSVQfchEAAEoiF6k55CIAAJRUkVykThZGiqeJ+vr6kgAAAHABbqdQ/chFAAAoG7lI9SMXAQCgbOXJRbjpJwAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEzDzdUBmFnLyesqtd+xWX2qOBIAAGBG5CIAAOCPyA0AAGbBjBEAAFCnbNu2Tf369VNISIgsFovWrFnjtN1isZS6vPTSS44+0dHRJbYPGjSohkcCAAAAAABcgcIIAACoUwoKCtS+fXvNmzev1O1ZWVlOy1tvvSWLxaL77rvPqd+IESOc+v3tb3+rifABAAAAAICLcSstAABQp8TGxio2NrbM7TabzWn9gw8+ULdu3XTVVVc5tXt5eZXoCwAAAAAA6j9mjAAAgHrrxx9/1Lp16zRs2LAS21asWKGAgAC1bdtWEydO1OnTp10QIQAAAAAAqGnMGAEAAPXWkiVL5OPjo7i4OKf2hx56SOHh4bLZbEpPT1diYqI+//xzbdy4sdTj2O122e12x3peXl61xg0AAAAAAKoPhREAAFBvvfXWW3rooYfk6enp1D5ixAjHzxEREWrVqpUiIyO1d+9edezYscRxkpOTNWPGjGqPFwAAAAAAVD9upQUAAOqlf/3rX/rqq680fPjwS/bt2LGj3N3ddeTIkVK3JyYmKjc317FkZmZWdbgAAAAAAKCGMGMEAADUSwsXLlSnTp3Uvn37S/Y9dOiQCgsLFRwcXOp2q9Uqq9Va1SECAAAAAAAXoDACAADqlPz8fB09etSxnpGRof3798vf318tWrSQ9PszQN5991298sorJfb/5ptvtGLFCt11110KCAjQ4cOHNWHCBHXo0EG33nprjY0DAAAAAAC4BoURAABQp+zevVvdunVzrCckJEiS4uPjtXjxYknSqlWrZBiGHnzwwRL7e3h4aPPmzXrttdeUn5+v0NBQ9enTR9OnT1eDBg1qZAwAAAAAAMB1KIwAAIA6JTo6WoZhXLTP448/rscff7zUbaGhoUpNTa2O0AAAAAAAQB3Aw9cBAAAAAAAAAIBpUBgBAAAAAAAAAACmQWEEAAAAAAAAAACYRoULI9u2bVO/fv0UEhIii8WiNWvWOG23WCylLi+99JKjT3R0dIntgwYNuuzBAAAAAAAAAAAAXEyFCyMFBQVq37695s2bV+r2rKwsp+Wtt96SxWLRfffd59RvxIgRTv3+9re/VW4EAAAAAAAAAAAA5eRW0R1iY2MVGxtb5nabzea0/sEHH6hbt2666qqrnNq9vLxK9AUAAAAAAAAAAKhO1fqMkR9//FHr1q3TsGHDSmxbsWKFAgIC1LZtW02cOFGnT58u8zh2u115eXlOCwAAAAAAAAAAQEVVeMZIRSxZskQ+Pj6Ki4tzan/ooYcUHh4um82m9PR0JSYm6vPPP9fGjRtLPU5ycrJmzJhRnaECAAAAAAAAAAATqNbCyFtvvaWHHnpInp6eTu0jRoxw/BwREaFWrVopMjJSe/fuVceOHUscJzExUQkJCY71vLw8hYaGVl/gAAAAAAAAAACgXqq2wsi//vUvffXVV3rnnXcu2bdjx45yd3fXkSNHSi2MWK1WWa3W6ggTAAAAAAAAAACYSLU9Y2ThwoXq1KmT2rdvf8m+hw4dUmFhoYKDg6srHAAAAAAAAAAAgIrPGMnPz9fRo0cd6xkZGdq/f7/8/f3VokULSb/f6urdd9/VK6+8UmL/b775RitWrNBdd92lgIAAHT58WBMmTFCHDh106623XsZQAAAAAAAAAAAALq7ChZHdu3erW7dujvXiZ3/Ex8dr8eLFkqRVq1bJMAw9+OCDJfb38PDQ5s2b9dprryk/P1+hoaHq06ePpk+frgYNGlRyGAAAAAAAAAAAAJdW4cJIdHS0DMO4aJ/HH39cjz/+eKnbQkNDlZqaWtGXBQAAAAAAAAAAuGzV9owRAAAAAACA+iw5OVkWi0Xjx493tBmGoaSkJIWEhKhhw4aKjo7WoUOHnPaz2+0aN26cAgIC5O3trbvvvlvHjx+v4egBADAvCiMAAAAAAAAVlJaWpjfffFM33HCDU/vs2bM1Z84czZs3T2lpabLZbOrZs6dOnz7t6DN+/HitXr1aq1at0vbt25Wfn6++ffuqqKiopocBAIApURgBAAAAAACogPz8fD300EP6+9//riZNmjjaDcPQ3LlzNXXqVMXFxSkiIkJLlizRr7/+qpUrV0qScnNztXDhQr3yyivq0aOHOnTooOXLl+vgwYPatGmTq4YEAICpUBgBAAAAAACogDFjxqhPnz7q0aOHU3tGRoays7PVq1cvR5vValVUVJR27NghSdqzZ48KCwud+oSEhCgiIsLRBwAAVK8KP3wdAAAAAADArFatWqW9e/cqLS2txLbs7GxJUlBQkFN7UFCQvvvuO0cfDw8Pp5kmxX2K97+Q3W6X3W53rOfl5V3WGAAAMDtmjAAAAAAAAJRDZmamnnrqKS1fvlyenp5l9rNYLE7rhmGUaLvQxfokJyfLz8/PsYSGhlY8eAAA4EBhBAAAAAAAoBz27NmjnJwcderUSW5ubnJzc1Nqaqr++te/ys3NzTFT5MKZHzk5OY5tNptNZ8+e1cmTJ8vsc6HExETl5uY6lszMzGoYHQAA5kFhBAAAAAAAoBxiYmJ08OBB7d+/37FERkbqoYce0v79+3XVVVfJZrNp48aNjn3Onj2r1NRUde3aVZLUqVMnubu7O/XJyspSenq6o8+FrFarfH19nRYAAFB5PGMEAAAAAACgHHx8fBQREeHU5u3traZNmzrax48fr5kzZ6pVq1Zq1aqVZs6cKS8vLw0ePFiS5Ofnp2HDhmnChAlq2rSp/P39NXHiRLVr167Ew9wBAED1oDACAAAAAABQRSZNmqQzZ85o9OjROnnypG655RZt2LBBPj4+jj6vvvqq3NzcNGDAAJ05c0YxMTFavHixGjRo4MLIAQAwDwojAAAAAAAAlbR161andYvFoqSkJCUlJZW5j6enp1JSUpSSklK9wQEAgFLxjBEAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACmQWEEAAAAAAAAAACYBoURAAAAAAAAAABgGhRGAAAAAAAAAACAaVAYAQAAAAAAAAAApkFhBAAAAAAAAAAAmAaFEQAAAAAAAAAAYBoURgAAAAAAAAAAgGlQGAEAAAAAAAAAAKbh5uoAAAAAAAAAAEgtJ6+r1H7HZvWp4kgAoH5jxggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAACoU7Zt26Z+/fopJCREFotFa9ascdo+ZMgQWSwWp6Vz585Ofex2u8aNG6eAgAB5e3vr7rvv1vHjx2twFAAAAAAAwFUojAAAgDqloKBA7du317x588rsc+eddyorK8uxfPzxx07bx48fr9WrV2vVqlXavn278vPz1bdvXxUVFVV3+AAAAAAAwMXcXB0AAABARcTGxio2NvaifaxWq2w2W6nbcnNztXDhQi1btkw9evSQJC1fvlyhoaHatGmTevfuXeUxAwAAAACA2oMZIwAAoN7ZunWrAgMD1bp1a40YMUI5OTmObXv27FFhYaF69erlaAsJCVFERIR27NhR6vHsdrvy8vKcFgAAAAAAUDdRGAEAAPVKbGysVqxYoc8++0yvvPKK0tLS1L17d9ntdklSdna2PDw81KRJE6f9goKClJ2dXeoxk5OT5efn51hCQ0OrfRwAAAAAAKB6cCstAABQrwwcONDxc0REhCIjIxUWFqZ169YpLi6uzP0Mw5DFYil1W2JiohISEhzreXl5FEcAAAAAAKijmDECAADqteDgYIWFhenIkSOSJJvNprNnz+rkyZNO/XJychQUFFTqMaxWq3x9fZ0WAAAAAABQN1EYAQAA9dqJEyeUmZmp4OBgSVKnTp3k7u6ujRs3OvpkZWUpPT1dXbt2dVWYAAAAAACghnArLQAAUKfk5+fr6NGjjvWMjAzt379f/v7+8vf3V1JSku677z4FBwfr2LFjmjJligICAtS/f39Jkp+fn4YNG6YJEyaoadOm8vf318SJE9WuXTv16NHDVcMCAAAAAAA1hMIIAACoU3bv3q1u3bo51ouf/REfH68FCxbo4MGDWrp0qU6dOqXg4GB169ZN77zzjnx8fBz7vPrqq3Jzc9OAAQN05swZxcTEaPHixWrQoEGNjwcAAAAAANQsCiMAAKBOiY6OlmEYZW5fv379JY/h6emplJQUpaSkVGVoAAAAAACgDuAZIwAAAAAAAAAAwDQqXBjZtm2b+vXrp5CQEFksFq1Zs8Zp+5AhQ2SxWJyWzp07O/Wx2+0aN26cAgIC5O3trbvvvlvHjx+/rIEAAAAAAAAAAABcSoULIwUFBWrfvr3mzZtXZp8777xTWVlZjuXjjz922j5+/HitXr1aq1at0vbt25Wfn6++ffuqqKio4iMAAAAAAAAAAAAopwo/YyQ2NlaxsbEX7WO1WmWz2Urdlpubq4ULF2rZsmXq0aOHJGn58uUKDQ3Vpk2b1Lt374qGBAAAAAAAAAAAUC7V8oyRrVu3KjAwUK1bt9aIESOUk5Pj2LZnzx4VFhaqV69ejraQkBBFRERox44dpR7PbrcrLy/PaQEAAAAAAAAAAKioKi+MxMbGasWKFfrss8/0yiuvKC0tTd27d5fdbpckZWdny8PDQ02aNHHaLygoSNnZ2aUeMzk5WX5+fo4lNDS0qsMGAAAAAAAAAAAmUOFbaV3KwIEDHT9HREQoMjJSYWFhWrduneLi4srczzAMWSyWUrclJiYqISHBsZ6Xl0dxBAAAAAAAAAAAVFi13Errj4KDgxUWFqYjR45Ikmw2m86ePauTJ0869cvJyVFQUFCpx7BarfL19XVaAAAAAAAAAAAAKqraCyMnTpxQZmamgoODJUmdOnWSu7u7Nm7c6OiTlZWl9PR0de3atbrDAQAAAAAAAAAAJlbhW2nl5+fr6NGjjvWMjAzt379f/v7+8vf3V1JSku677z4FBwfr2LFjmjJligICAtS/f39Jkp+fn4YNG6YJEyaoadOm8vf318SJE9WuXTv16NGj6kYGAAAAAAAAAABwgQoXRnbv3q1u3bo51ouf/REfH68FCxbo4MGDWrp0qU6dOqXg4GB169ZN77zzjnx8fBz7vPrqq3Jzc9OAAQN05swZxcTEaPHixWrQoEEVDAkAAAAAAAAAAKB0FS6MREdHyzCMMrevX7/+ksfw9PRUSkqKUlJSKvryAAAAAAAAAAAAlVbtzxgBAAAAAAAAAACoLSiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAoE7Ztm2b+vXrp5CQEFksFq1Zs8axrbCwUE8//bTatWsnb29vhYSE6NFHH9UPP/zgdIzo6GhZLBanZdCgQTU8EgAAAAAA4AoURgAAQJ1SUFCg9u3ba968eSW2/frrr9q7d6+mTZumvXv36v3339fXX3+tu+++u0TfESNGKCsry7H87W9/q4nwAQAAAACAi7m5OgAAAICKiI2NVWxsbKnb/Pz8tHHjRqe2lJQU3Xzzzfr+++/VokULR7uXl5dsNlu1xgoAAAAAAGofZowAAIB6LTc3VxaLRY0bN3ZqX7FihQICAtS2bVtNnDhRp0+fdk2AAAAAAACgRjFjBAAA1Fu//fabJk+erMGDB8vX19fR/tBDDyk8PFw2m03p6elKTEzU559/XmK2STG73S673e5Yz8vLq/bYAQAAAABA9aAwAgAA6qXCwkINGjRI58+f1/z58522jRgxwvFzRESEWrVqpcjISO3du1cdO3Yscazk5GTNmDGj2mMGAAAAAADVj1tpAQCAeqewsFADBgxQRkaGNm7c6DRbpDQdO3aUu7u7jhw5Uur2xMRE5ebmOpbMzMzqCBsAANRyCxYs0A033CBfX1/5+vqqS5cu+uSTTxzbDcNQUlKSQkJC1LBhQ0VHR+vQoUNOx7Db7Ro3bpwCAgLk7e2tu+++W8ePH6/poQAAYGoURgAAQL1SXBQ5cuSINm3apKZNm15yn0OHDqmwsFDBwcGlbrdarY4LIMULAAAwn+bNm2vWrFnavXu3du/ere7du+uee+5xFD9mz56tOXPmaN68eUpLS5PNZlPPnj2dnmU2fvx4rV69WqtWrdL27duVn5+vvn37qqioyFXDAgDAdLiVFgAAqFPy8/N19OhRx3pGRob2798vf39/hYSE6P7779fevXu1du1aFRUVKTs7W5Lk7+8vDw8PffPNN1qxYoXuuusuBQQE6PDhw5owYYI6dOigW2+91VXDAgAAdUC/fv2c1l944QUtWLBAu3bt0vXXX6+5c+dq6tSpiouLkyQtWbJEQUFBWrlypUaOHKnc3FwtXLhQy5YtU48ePSRJy5cvV2hoqDZt2qTevXvX+JgAADAjZowAAIA6Zffu3erQoYM6dOggSUpISFCHDh307LPP6vjx4/rwww91/Phx3XjjjQoODnYsO3bskCR5eHho8+bN6t27t9q0aaMnn3xSvXr10qZNm9SgQQNXDg0AANQhRUVFWrVqlQoKCtSlSxdlZGQoOztbvXr1cvSxWq2Kiopy5CF79uxRYWGhU5+QkBBFREQ4+gAAgOpX4cLItm3b1K9fP4WEhMhisWjNmjWObYWFhXr66afVrl07eXt7KyQkRI8++qh++OEHp2NER0fLYrE4LYMGDbrswQAAgPovOjpahmGUWBYvXqyWLVuWus0wDEVHR0uSQkNDlZqaqhMnTshut+vo0aN67bXX5O/v79qBAQCAOuHgwYNq1KiRrFarnnjiCa1evVrXX3+9Y5ZqUFCQU/+goCDHtuzsbHl4eKhJkyZl9imN3W5XXl6e0wIAACqvwoWRgoICtW/fXvPmzSux7ddff9XevXs1bdo07d27V++//76+/vpr3X333SX6jhgxQllZWY7lb3/7W+VGAAAAAAAAUEPatGmj/fv3a9euXRo1apTi4+N1+PBhx3aLxeLU3zCMEm0XulSf5ORk+fn5OZbQ0NDLGwQAACZX4WeMxMbGKjY2ttRtfn5+2rhxo1NbSkqKbr75Zn3//fdq0aKFo93Ly0s2m62iLw8AAAAAAOAyHh4euuaaayRJkZGRSktL02uvvaann35a0u+zQoKDgx39c3JyHLNIbDabzp49q5MnTzrNGsnJyVHXrl3LfM3ExEQlJCQ41vPy8iiOAABwGar9GSO5ubmyWCxq3LixU/uKFSsUEBCgtm3bauLEiTp9+nR1hwIAAAAAAFClDMOQ3W5XeHi4bDab0xdGz549q9TUVEfRo1OnTnJ3d3fqk5WVpfT09IsWRqxWq3x9fZ0WAABQeRWeMVIRv/32myZPnqzBgwc7fWg/9NBDjoQhPT1diYmJ+vzzz0vMNilmt9tlt9sd69xLEwAAAAAA1LQpU6YoNjZWoaGhOn36tFatWqWtW7fq008/lcVi0fjx4zVz5ky1atVKrVq10syZM+Xl5aXBgwdL+v1OG8OGDdOECRPUtGlT+fv7a+LEiWrXrp169Ojh4tEBAGAe1VYYKSws1KBBg3T+/HnNnz/faduIESMcP0dERKhVq1aKjIzU3r171bFjxxLHSk5O1owZM6orVAAAAAAAgEv68ccf9cgjjygrK0t+fn664YYb9Omnn6pnz56SpEmTJunMmTMaPXq0Tp48qVtuuUUbNmyQj4+P4xivvvqq3NzcNGDAAJ05c0YxMTFavHixGjRo4KphAQBgOtVSGCksLNSAAQOUkZGhzz777JJTPDt27Ch3d3cdOXKk1MII99IEAAAAAACutnDhwotut1gsSkpKUlJSUpl9PD09lZKSopSUlCqODgAAlFeVF0aKiyJHjhzRli1b1LRp00vuc+jQIRUWFjo9nOyPrFarrFZrVYcKAAAAAAAAAABMpsKFkfz8fB09etSxnpGRof3798vf318hISG6//77tXfvXq1du1ZFRUXKzs6WJPn7+8vDw0PffPONVqxYobvuuksBAQE6fPiwJkyYoA4dOujWW2+tupEBAAAAAAAAAABcoMKFkd27d6tbt26O9eJbXMXHxyspKUkffvihJOnGG2902m/Lli2Kjo6Wh4eHNm/erNdee035+fkKDQ1Vnz59NH36dO6nCQAAAAAAAAAAqlWFCyPR0dEyDKPM7RfbJkmhoaFKTU2t6MsCAAAAAAAAAABctitcHQAAAAAAAAAAAEBNoTACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAoE7Ztm2b+vXrp5CQEFksFq1Zs8Zpu2EYSkpKUkhIiBo2bKjo6GgdOnTIqY/dbte4ceMUEBAgb29v3X333Tp+/HgNjgIAAAAAALgKhREAAFCnFBQUqH379po3b16p22fPnq05c+Zo3rx5SktLk81mU8+ePXX69GlHn/Hjx2v16tVatWqVtm/frvz8fPXt21dFRUU1NQwAAAAAAOAibq4OAAAAoCJiY2MVGxtb6jbDMDR37lxNnTpVcXFxkqQlS5YoKChIK1eu1MiRI5Wbm6uFCxdq2bJl6tGjhyRp+fLlCg0N1aZNm9S7d+8aGwsAAAAAAKh5zBgBAAD1RkZGhrKzs9WrVy9Hm9VqVVRUlHbs2CFJ2rNnjwoLC536hISEKCIiwtHnQna7XXl5eU4LAAAAAAComyiMAACAeiM7O1uSFBQU5NQeFBTk2JadnS0PDw81adKkzD4XSk5Olp+fn2MJDQ2thugBAAAAAEBNoDACAADqHYvF4rRuGEaJtgtdrE9iYqJyc3MdS2ZmZpXFCgAAAAAAahaFEQAAUG/YbDZJKjHzIycnxzGLxGaz6ezZszp58mSZfS5ktVrl6+vrtAAAAAAAgLqJwggAAKg3wsPDZbPZtHHjRkfb2bNnlZqaqq5du0qSOnXqJHd3d6c+WVlZSk9Pd/QBAAAAAAD1V4ULI9u2bVO/fv0UEhIii8WiNWvWOG03DENJSUkKCQlRw4YNFR0drUOHDjn1sdvtGjdunAICAuTt7a27775bx48fv6yBAAAAc8jPz9f+/fu1f/9+Sb8/cH3//v36/vvvZbFYNH78eM2cOVOrV69Wenq6hgwZIi8vLw0ePFiS5Ofnp2HDhmnChAnavHmz9u3bp4cffljt2rVTjx49XDgyAAAAAABQEypcGCkoKFD79u01b968UrfPnj1bc+bM0bx585SWliabzaaePXvq9OnTjj7jx4/X6tWrtWrVKm3fvl35+fnq27evioqKKj8SAABgCrt371aHDh3UoUMHSVJCQoI6dOigZ599VpI0adIkjR8/XqNHj1ZkZKT+97//acOGDfLx8XEc49VXX9W9996rAQMG6NZbb5WXl5c++ugjNWjQwCVjAgAAAAAANcetojvExsYqNja21G2GYWju3LmaOnWq4uLiJElLlixRUFCQVq5cqZEjRyo3N1cLFy7UsmXLHN/KXL58uUJDQ7Vp0yb17t37MoYDAADqu+joaBmGUeZ2i8WipKQkJSUlldnH09NTKSkpSklJqYYIAQAAAABAbValzxjJyMhQdna2evXq5WizWq2KiorSjh07JEl79uxRYWGhU5+QkBBFREQ4+lzIbrcrLy/PaQEAAAAAAAAAAKioKi2MZGdnS5KCgoKc2oOCghzbsrOz5eHhoSZNmpTZ50LJycny8/NzLKGhoVUZNgAAAAAAAAAAMIkqLYwUs1gsTuuGYZRou9DF+iQmJio3N9exZGZmVlmsAAAAAAAAAADAPKq0MGKz2SSpxMyPnJwcxywSm82ms2fP6uTJk2X2uZDVapWvr6/TAgAAAAAAAAAAUFFVWhgJDw+XzWbTxo0bHW1nz55VamqqunbtKknq1KmT3N3dnfpkZWUpPT3d0QcAAAAAAAAAAKA6uFV0h/z8fB09etSxnpGRof3798vf318tWrTQ+PHjNXPmTLVq1UqtWrXSzJkz5eXlpcGDB0uS/Pz8NGzYME2YMEFNmzaVv7+/Jk6cqHbt2qlHjx5VNzIAAAAAAAAAAIALVLgwsnv3bnXr1s2xnpCQIEmKj4/X4sWLNWnSJJ05c0ajR4/WyZMndcstt2jDhg3y8fFx7PPqq6/Kzc1NAwYM0JkzZxQTE6PFixerQYMGVTAkAAAAAAAAAACA0lW4MBIdHS3DMMrcbrFYlJSUpKSkpDL7eHp6KiUlRSkpKRV9eQAAAAAAAAAAgEqr0meMAAAAAAAAAAAA1GYURgAAAAAAAAAAgGlQGAEAAAAAAAAAAKZBYQQAAAAAAAAAAJgGhREAAAAAAAAAAGAaFEYAAAAAAAAAAIBpuLk6AAAAANQtLSevq/A+x2b1qYZIAAAAAACoOGaMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAlENycrJuuukm+fj4KDAwUPfee6+++uorpz6GYSgpKUkhISFq2LChoqOjdejQIac+drtd48aNU0BAgLy9vXX33Xfr+PHjNTkUAABMjcIIAAAAAABAOaSmpmrMmDHatWuXNm7cqHPnzqlXr14qKChw9Jk9e7bmzJmjefPmKS0tTTabTT179tTp06cdfcaPH6/Vq1dr1apV2r59u/Lz89W3b18VFRW5YlgAAJiOm6sDAAAAAAAAqAs+/fRTp/VFixYpMDBQe/bs0R133CHDMDR37lxNnTpVcXFxkqQlS5YoKChIK1eu1MiRI5Wbm6uFCxdq2bJl6tGjhyRp+fLlCg0N1aZNm9S7d+8aHxcAAGbDjBEAAAAAAIBKyM3NlST5+/tLkjIyMpSdna1evXo5+litVkVFRWnHjh2SpD179qiwsNCpT0hIiCIiIhx9LmS325WXl+e0AACAyqMwAgAAAAAAUEGGYSghIUG33XabIiIiJEnZ2dmSpKCgIKe+QUFBjm3Z2dny8PBQkyZNyuxzoeTkZPn5+TmW0NDQqh4OAACmwq20AAAAABdoOXldpfY7NqtPFUcCAKiMsWPH6sCBA9q+fXuJbRaLxWndMIwSbRe6WJ/ExEQlJCQ41vPy8iiOAABwGZgxAgAAAAAAUAHjxo3Thx9+qC1btqh58+aOdpvNJkklZn7k5OQ4ZpHYbDadPXtWJ0+eLLPPhaxWq3x9fZ0WAABQeRRGAAAAAAAAysEwDI0dO1bvv/++PvvsM4WHhzttDw8Pl81m08aNGx1tZ8+eVWpqqrp27SpJ6tSpk9zd3Z36ZGVlKT093dEHAABUL26lBQAAAAAAUA5jxozRypUr9cEHH8jHx8cxM8TPz08NGzaUxWLR+PHjNXPmTLVq1UqtWrXSzJkz5eXlpcGDBzv6Dhs2TBMmTFDTpk3l7++viRMnql27durRo4crhwcAgGlQGAEAAAAAACiHBQsWSJKio6Od2hctWqQhQ4ZIkiZNmqQzZ85o9OjROnnypG655RZt2LBBPj4+jv6vvvqq3NzcNGDAAJ05c0YxMTFavHixGjRoUFNDAQDA1CiMAAAAAAAAlINhGJfsY7FYlJSUpKSkpDL7eHp6KiUlRSkpKVUYHQAAKC+eMQIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAKBeadmypSwWS4llzJgxkqQhQ4aU2Na5c2cXRw0AAAAAAGoKzxgBAAD1SlpamoqKihzr6enp6tmzpx544AFH25133qlFixY51j08PGo0RgAAAAAA4DoURgAAQL3SrFkzp/VZs2bp6quvVlRUlKPNarXKZrPVdGgAAAAAAKAW4FZaAACg3jp79qyWL1+uoUOHymKxONq3bt2qwMBAtW7dWiNGjFBOTo4LowQAAAAAADWJGSMAAKDeWrNmjU6dOqUhQ4Y42mJjY/XAAw8oLCxMGRkZmjZtmrp37649e/bIarWWehy73S673e5Yz8vLq+7QAQAAAABANaEwAgAA6q2FCxcqNjZWISEhjraBAwc6fo6IiFBkZKTCwsK0bt06xcXFlXqc5ORkzZgxo9rjBQAAAAAA1a/Kb6XVsmVLWSyWEsuYMWMkSUOGDCmxrXPnzlUdBgAAMLnvvvtOmzZt0vDhwy/aLzg4WGFhYTpy5EiZfRITE5Wbm+tYMjMzqzpcAAAAAABQQ6p8xkhaWpqKiooc6+np6erZs6ceeOABR9udd96pRYsWOdY9PDyqOgwAAGByixYtUmBgoPr06XPRfidOnFBmZqaCg4PL7GO1Wsu8zRYAAAAAAKhbqrww0qxZM6f1WbNm6eqrr1ZUVJSjzWq1ymazVfVLAwAASJLOnz+vRYsWKT4+Xm5u/5fu5OfnKykpSffdd5+Cg4N17NgxTZkyRQEBAerfv78LIwYAAAAAADWlym+l9Udnz57V8uXLNXToUFksFkf71q1bFRgYqNatW2vEiBHKycmpzjAAAIDJbNq0Sd9//72GDh3q1N6gQQMdPHhQ99xzj1q3bq34+Hi1bt1aO3fulI+Pj4uiBQAAAAAANalaH76+Zs0anTp1SkOGDHG0xcbG6oEHHlBYWJgyMjI0bdo0de/eXXv27CnzFhV2u112u92xnpeXV51hAwCAOq5Xr14yDKNEe8OGDbV+/XoXRAQAAAAAAGqLai2MLFy4ULGxsQoJCXG0DRw40PFzRESEIiMjFRYWpnXr1ikuLq7U4yQnJ2vGjBnVGSoAAAAAAAAAADCBaruV1nfffadNmzZp+PDhF+0XHByssLAwHTlypMw+iYmJys3NdSyZmZlVHS4AAAAAAAAAADCBapsxsmjRIgUGBqpPnz4X7XfixAllZmYqODi4zD5Wq7XM22wBAAAAVaXl5HWV2u/YrIvnvAAAAACA2qNaZoycP39eixYtUnx8vNzc/q/2kp+fr4kTJ2rnzp06duyYtm7dqn79+ikgIED9+/evjlAAAAAAAAAAAAAcqmXGyKZNm/T9999r6NChTu0NGjTQwYMHtXTpUp06dUrBwcHq1q2b3nnnHfn4+FRHKECZ+EYoAAAAAAAAAJhPtRRGevXqJcMwSrQ3bNhQ69evr46XBAAAAAAAAAAAuKRqe8YIAAAAYBaVnYkKAAAAAKh51fKMEQAAAAAAAAAAgNqIwggAAAAAAAAAADANbqUFAAAAAAAAAJVU2duqHpvVp4ojAVBeFEZQq/BBAgAAAAAAAACoTtxKCwAAAAAAAAAAmAaFEQAAAAAAAAAAYBoURgAAAAAAAAAAgGlQGAEAAAAAAAAAAKZBYQQAAAAAAAAAAJgGhREAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACm4ebqAICq0HLyOleHAAAAAAAAAACoA5gxAgAAAAAAAAAATIPCCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDR4xkgdVJnnaRyb1acaIgEAAAAAAAAAoG5hxggAAAAAAAAAADANZowANaQyM30kZvsAAAAAAAAAQFVixggAAAAAAAAAADANZowAAAAAdQizUAEAAADg8jBjBAAAAAAAAAAAmAaFEQAAAAAAAAAAYBoURgAAAAAAAAAAgGnwjBEAAADUO5V9Dkd9xrNJAAAAAOB3zBgBAAAAAAAAAACmwYyRKsA3EgEAAAAAAAAAqBuYMQIAAAAAAAAAAEyDGSOoFsyiAQAAAAAAAADURswYAQAAAAAAAAAApkFhBAAAAAAAAAAAmAaFEQAAAAAAAAAAYBo8YwSoIJ6fAgAAAAAAAAB1FzNGAAAAAAAAAACAaVAYAQAA9UpSUpIsFovTYrPZHNsNw1BSUpJCQkLUsGFDRUdH69ChQy6MGAAAAAAA1CQKIwAAoN5p27atsrKyHMvBgwcd22bPnq05c+Zo3rx5SktLk81mU8+ePXX69GkXRgwAAAAAAGoKzxgBAAD1jpubm9MskWKGYWju3LmaOnWq4uLiJElLlixRUFCQVq5cqZEjR9Z0qLgEnu0FAAAAAKhqzBgBAAD1zpEjRxQSEqLw8HANGjRI3377rSQpIyND2dnZ6tWrl6Ov1WpVVFSUduzYUebx7Ha78vLynBYAAAAAAFA3VXlhhPt6AwAAV7rlllu0dOlSrV+/Xn//+9+VnZ2trl276sSJE8rOzpYkBQUFOe0TFBTk2Faa5ORk+fn5OZbQ0NBqHQMAAAAAAKg+1TJjhPt6AwAAV4mNjdV9992ndu3aqUePHlq37vdbMS1ZssTRx2KxOO1jGEaJtj9KTExUbm6uY8nMzKye4AEAAAAAQLWrlsJI8X29i5dmzZpJKnlf74iICC1ZskS//vqrVq5cWR2hAAAAk/P29la7du105MgRxyzWC2eH5OTklJhF8kdWq1W+vr5OCwAAAAAAqJuqpTDCfb0BAEBtYbfb9cUXXyg4OFjh4eGy2WzauHGjY/vZs2eVmpqqrl27ujBKAAAAAABQU6q8MMJ9vQEAgCtNnDhRqampysjI0H/+8x/df//9ysvLU3x8vCwWi8aPH6+ZM2dq9erVSk9P15AhQ+Tl5aXBgwe7OnQAAAAAAFAD3Kr6gLGxsY6f27Vrpy5duujqq6/WkiVL1LlzZ0mVu693QkKCYz0vL4/iCAAAKNXx48f14IMP6ueff1azZs3UuXNn7dq1S2FhYZKkSZMm6cyZMxo9erROnjypW265RRs2bJCPj4+LIwcAAAAAADWhygsjF/rjfb3vvfdeSb/f1zs4ONjRpzz39bZardUdKkrRcvI6V4cAAECFrFq16qLbLRaLkpKSlJSUVDMBAQAAAACAWqVanjHyR9zXGwAAAAAAAAAA1BZVPmNk4sSJ6tevn1q0aKGcnBw9//zzpd7Xu1WrVmrVqpVmzpzJfb0BAAAAAAAAAECNqPLCCPf1BgAAAAAAAAAAtVWVF0a4rzcAAAAAAAAAAKitqv3h6wAAAABQXi0nr6vwPsdm9amGSAAAAADUV9X+8HUAAAAAAAAAAIDagsIIAAAAAAAAAAAwDQojAAAAAAAAAADANHjGCAAAQB1XmWcyAACAytm2bZteeukl7dmzR1lZWVq9erXuvfdex3bDMDRjxgy9+eabOnnypG655Ra9/vrratu2raOP3W7XxIkT9fbbb+vMmTOKiYnR/Pnz1bx5cxeMCAAA82HGCAAAAAAAQDkVFBSoffv2mjdvXqnbZ8+erTlz5mjevHlKS0uTzWZTz549dfr0aUef8ePHa/Xq1Vq1apW2b9+u/Px89e3bV0VFRTU1DAAATI0ZIwAAAAAAAOUUGxur2NjYUrcZhqG5c+dq6tSpiouLkyQtWbJEQUFBWrlypUaOHKnc3FwtXLhQy5YtU48ePSRJy5cvV2hoqDZt2qTevXvX2FgAADArZowAAAAAAABUgYyMDGVnZ6tXr16ONqvVqqioKO3YsUOStGfPHhUWFjr1CQkJUUREhKPPhex2u/Ly8pwWAABQeRRGAAAAAAAAqkB2drYkKSgoyKk9KCjIsS07O1seHh5q0qRJmX0ulJycLD8/P8cSGhpaDdEDAGAeFEYAAAAAAACqkMVicVo3DKNE24Uu1icxMVG5ubmOJTMzs8piBQDAjCiMAAAAAAAAVAGbzSZJJWZ+5OTkOGaR2Gw2nT17VidPniyzz4WsVqt8fX2dFgAAUHkURgAAAAAAAKpAeHi4bDabNm7c6Gg7e/asUlNT1bVrV0lSp06d5O7u7tQnKytL6enpjj4AAKB6ubk6AAAAAAAAgLoiPz9fR48edaxnZGRo//798vf3V4sWLTR+/HjNnDlTrVq1UqtWrTRz5kx5eXlp8ODBkiQ/Pz8NGzZMEyZMUNOmTeXv76+JEyeqXbt26tGjh6uGBQCAqVAYAWq5lpPXVWq/Y7P61OrXAgAAAIC6aPfu3erWrZtjPSEhQZIUHx+vxYsXa9KkSTpz5oxGjx6tkydP6pZbbtGGDRvk4+Pj2OfVV1+Vm5ubBgwYoDNnzigmJkaLFy9WgwYNanw8AACYEYURAAAAAGXiixMA4Cw6OlqGYZS53WKxKCkpSUlJSWX28fT0VEpKilJSUqohQgAAcCk8YwQAAAAAAAAAAJgGM0YAAAAAVLnKzjQBAAAAgOrGjBEAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACmQWEEAAAAAAAAAACYBg9fB+opHngKAAAAAAAAACUxYwQAAAAAAAAAAJgGhREAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACmwcPXAQAAUO1aTl7n6hAAAAAAAJBEYcQ0uBgBAAAAAAAAAACFEQAAAAB1XGW/BHRsVp8qjgQAAABAXcAzRgAAAAAAAAAAgGlQGAEAAAAAAAAAAKbBrbQAXDZuXwEAAAAAAACgrmDGCAAAAAAAAAAAMA0KIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAKhXkpOTddNNN8nHx0eBgYG699579dVXXzn1GTJkiCwWi9PSuXNnF0UMAAAAAABqEoURAABQr6SmpmrMmDHatWuXNm7cqHPnzqlXr14qKChw6nfnnXcqKyvLsXz88ccuihgAAAAAANSkKi+M8C1NAADgSp9++qmGDBmitm3bqn379lq0aJG+//577dmzx6mf1WqVzWZzLP7+/i6KGAAAAAAA1KQqL4zwLU0AAFCb5ObmSlKJwsfWrVsVGBio1q1ba8SIEcrJyXFFeAAAAAAAoIa5VfUBP/30U6f1RYsWKTAwUHv27NEdd9zhaC/+liYAAEB1MQxDCQkJuu222xQREeFoj42N1QMPPKCwsDBlZGRo2rRp6t69u/bs2SOr1VriOHa7XXa73bGel5dXI/EDAAAAAICqV+WFkQtd6luajRs3VlRUlF544QUFBgZWdzgAAMBExo4dqwMHDmj79u1O7QMHDnT8HBERocjISIWFhWndunWKi4srcZzk5GTNmDGj2uMFAAAAAADVr1ofvn6xb2muWLFCn332mV555RWlpaWpe/fuTt/E/CO73a68vDynBQAA4GLGjRunDz/8UFu2bFHz5s0v2jc4OFhhYWE6cuRIqdsTExOVm5vrWDIzM6sjZAAAAAAAUAOqdcYI39IEAAA1zTAMjRs3TqtXr9bWrVsVHh5+yX1OnDihzMxMBQcHl7rdarWWeostAAAAAABQ91TbjBG+pQkAAFxhzJgxWr58uVauXCkfHx9lZ2crOztbZ86ckSTl5+dr4sSJ2rlzp44dO6atW7eqX79+CggIUP/+/V0cPQAAAAAAqG5VPmOEb2kCAABXWrBggSQpOjraqX3RokUaMmSIGjRooIMHD2rp0qU6deqUgoOD1a1bN73zzjvy8fFxQcQAAAAAAKAmVXlhZMyYMVq5cqU++OADx7c0JcnPz08NGzZUfn6+kpKSdN999yk4OFjHjh3TlClT+JYmAACoEoZhXHR7w4YNtX79+hqKBgAAAAAA1DZVXhjhW5oAAAAAAAAAAKC2qpZbaV0M39IEAAAAAAAAAACuUm0PXwcAAAAAAAAAAKhtKIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwjSp/+DoAlFfLyesqtd+xWX2qOBIAAAAAAAAAZsGMEQAAAAAAAAAAYBrMGAEAAABgSsxeBQAAAMyJGSMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg2eM/EFl7zEMAAAAAAAAAADqBmaMAAAAAAAAAAAA02DGCAAAQC3B7FUAAAAAAKofM0YAAAAAAAAAAIBpMGMEAAAAAGpAZWeFHZvVp4ojAQAAAMyNGSMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8IIAAAAAAAAAAAwDQojAAAAAAAAAADANCiMAAAAAAAAAAAA06AwAgAAAAAAAAAATIPCCAAAAAAAAAAAMA03VwcAADWl5eR1ldrv2Kw+VRwJAAAAAAAAAFdhxggAAAAAAAAAADANZowAwCVUZqYJs0wAAAAAAACA2okZIwAAAAAAAAAAwDQojAAAAAAAAAAAANOgMAIAAAAAAAAAAEyDwggAAAAAAAAAADANCiMAAAAAAAAAAMA03FwdAABUVMvJ61wdAgAAMLGazkUq83rHZvWpsde6nNcDAAAAXIEZIwAAAAAAAAAAwDQojAAAAAAAAAAAANPgVloAAAAAAAAAgDqL24GioiiMAEA1qOkPZBIAAAAAAAAAoHy4lRYAAAAAAAAAADANZowAQC1S2ZkfAAAAf0ROAQAAAJSNGSMAAAAAAAAAAMA0XDpjZP78+XrppZeUlZWltm3bau7cubr99ttdGRIAoBwq8y3U+v48E57zUjeRiwBA1ajpGSp8fpZELlI3kYsAAOAaLpsx8s4772j8+PGaOnWq9u3bp9tvv12xsbH6/vvvXRUSAAAwEXIRAADgSuQiAAC4jstmjMyZM0fDhg3T8OHDJUlz587V+vXrtWDBAiUnJ7sqLAAwlfp8/3G+NYlLIRcBgLqrLuQw5BS4FHIRAEB9Uteuw7ikMHL27Fnt2bNHkydPdmrv1auXduzYUaK/3W6X3W53rOfm5kqS8vLyqjSu8/Zfq/R4AID/U9X/Z19KZf9Pr2ycNf16FzuWYRhVdsz6ilwEAFDd6nvuc7FjkYtcWn3LRWr6970+4xzUTZw31+McuF5tOAcVyUVcUhj5+eefVVRUpKCgIKf2oKAgZWdnl+ifnJysGTNmlGgPDQ2tthgBAFXLb66rIyifmo6zOl7v9OnT8vPzq/oD1yPkIgCA6mbm3Idc5NLqWy5SV37f6zPOQd3EeXM9zoHruSoXcenD1y0Wi9O6YRgl2iQpMTFRCQkJjvXz58/rl19+UdOmTUvtX5fl5eUpNDRUmZmZ8vX1dXU4psP771q8/67HOXCty33/DcPQ6dOnFRISUg3R1U/kIpfG/wuuxfvvWrz/rsX771qVef/JRSquNuUi9e3fHOOp3erTeOrTWCTGU9sxnourSC7iksJIQECAGjRoUOJbEDk5OSW+LSFJVqtVVqvVqa1x48bVGaLL+fr61otf7rqK99+1eP9dj3PgWpfz/vPtzPIhF6k4/l9wLd5/1+L9dy3ef9eq6PtPLlI+tTkXqW//5hhP7VafxlOfxiIxntqO8ZStvLnIFVXyahXk4eGhTp06aePGjU7tGzduVNeuXV0REgAAMBFyEQAA4ErkIgAAuJbLbqWVkJCgRx55RJGRkerSpYvefPNNff/993riiSdcFRIAADARchEAAOBK5CIAALiOywojAwcO1IkTJ/Tcc88pKytLERER+vjjjxUWFuaqkGoFq9Wq6dOnl5gii5rB++9avP+uxzlwLd7/mkUuUj78XroW779r8f67Fu+/a/H+V7/alovUt3POeGq3+jSe+jQWifHUdoyn6lgMwzBq/FUBAAAAAAAAAABcwCXPGAEAAAAAAAAAAHAFCiMAAAAAAAAAAMA0KIwAAAAAAAAAAADToDACAAAAAAAAAABMg8JILZCcnKybbrpJPj4+CgwM1L333quvvvrK1WGZyoIFC3TDDTfI19dXvr6+6tKliz755BNXh2VaycnJslgsGj9+vKtDMY2kpCRZLBanxWazuTosU/nf//6nhx9+WE2bNpWXl5duvPFG7dmzx9VhwaTITVyP3KT2IC+peeQlrkdeUv9U9rM9NTVVnTp1kqenp6666iq98cYbNRDtpVVmPFu3bi3xf4vFYtGXX35ZQ1GXrTKf+7X13EgVH09tPjcXKm9eUJvPzx+VZzy1+fxUJmeozeemouOpzeemWGVyipo6R27VclRUSGpqqsaMGaObbrpJ586d09SpU9WrVy8dPnxY3t7erg7PFJo3b65Zs2bpmmuukSQtWbJE99xzj/bt26e2bdu6ODpzSUtL05tvvqkbbrjB1aGYTtu2bbVp0ybHeoMGDVwYjbmcPHlSt956q7p166ZPPvlEgYGB+uabb9S4cWNXhwaTIjdxPXKT2oG8xHXIS1yHvKR+qsxne0ZGhu666y6NGDFCy5cv17///W+NHj1azZo103333VfDI3B2ObnKV199JV9fX8d6s2bNqjvcS6ro535tPjdS5fOY2nhu/qi8eUFtPz/FKprn1NbzU5GcoS6cm8rkQLX13FQmp6jRc2Sg1snJyTEkGampqa4OxdSaNGli/OMf/3B1GKZy+vRpo1WrVsbGjRuNqKgo46mnnnJ1SKYxffp0o3379q4Ow7Sefvpp47bbbnN1GECZyE1qB3KTmkVe4jrkJa5FXmIO5flsnzRpknHttdc6tY0cOdLo3LlzdYdXYeUZz5YtWwxJxsmTJ2susMtwsc/9unRuil1sPHXh3FQkL6gL56ci46nN56eiOUNtPzcVHU9tPjeGUbmcoibPEbfSqoVyc3MlSf7+/i6OxJyKioq0atUqFRQUqEuXLq4Ox1TGjBmjPn36qEePHq4OxZSOHDmikJAQhYeHa9CgQfr2229dHZJpfPjhh4qMjNQDDzygwMBAdejQQX//+99dHRbgQG7iWuQmrkFe4lrkJa5DXmIO5fls37lzp3r16uXU1rt3b+3evVuFhYXVGl9FVSRX6dChg4KDgxUTE6MtW7ZUd2gVVp7P/bp0biqSx9Tmc1ORvKAunJ/K5Dm19fxUJGeoC+emMjlQbT03lckpavIcURipZQzDUEJCgm677TZFRES4OhxTOXjwoBo1aiSr1aonnnhCq1ev1vXXX+/qsExj1apV2rt3r5KTk10diindcsstWrp0qdavX6+///3vys7OVteuXXXixAlXh2YK3377rRYsWKBWrVpp/fr1euKJJ/Tkk09q6dKlrg4NIDdxIXIT1yEvcS3yEtciL6n/yvvZnp2draCgIKe2oKAgnTt3Tj///HN1h1lu5R1PcHCw3nzzTb333nt6//331aZNG8XExGjbtm01GG3ZKvK5XxfOTUXGU9vPTUXzgtp+fio6ntp8fiqaM9T2c1PR8dTmcyNVLqeoyXPEM0ZqmbFjx+rAgQPavn27q0MxnTZt2mj//v06deqU3nvvPcXHxys1NZULEDUgMzNTTz31lDZs2CBPT09Xh2NKsbGxjp/btWunLl266Oqrr9aSJUuUkJDgwsjM4fz584qMjNTMmTMl/f5tj0OHDmnBggV69NFHXRwdzI7cxHXITVyDvMT1yEtci7yk/qvIZ7vFYnFaNwyj1HZXKu942rRpozZt2jjWu3TposzMTL388su64447qjvMS6ro535tPzcVGU9tPjeVzQtq6/mpzHhq8/mpTM5QW8+NVPHx1OZzI1U+p6ipc8SMkVpk3Lhx+vDDD7VlyxY1b97c1eGYjoeHh6655hpFRkYqOTlZ7du312uvvebqsExhz549ysnJUadOneTm5iY3Nzelpqbqr3/9q9zc3FRUVOTqEE3H29tb7dq105EjR1wdiikEBweX+APhuuuu0/fff++iiIDfkZu4FrmJa5CX1D7kJTWLvKR+q8hnu81mU3Z2tlNbTk6O3Nzc1LRp0+oMs9wuN1fp3Llzrfm/pSKf+3Xh3FxuHlNbzk1l8oLafH6qKs+pLefnQpfKGWrzuSlNZXKg2nRuKpNT1OQ5YsZILWAYhsaNG6fVq1dr69atCg8Pd3VI0O/nxW63uzoMU4iJidHBgwed2h577DFde+21evrpp9WgQQMXRWZedrtdX3zxhW6//XZXh2IKt956q7766iuntq+//lphYWEuighmR25SO5Gb1AzyktqHvKRmkZfUT5X5bO/SpYs++ugjp7YNGzYoMjJS7u7u1RVquVRVrrJv3z4FBwdXcXRV42Kf+7X53JSlonlMbTk3lckLavP5qao8p7acnwtdKmeozeemNJXJgWrTualMTlGj56jKH+eOChs1apTh5+dnbN261cjKynIsv/76q6tDM43ExERj27ZtRkZGhnHgwAFjypQpxhVXXGFs2LDB1aGZVlRUlPHUU0+5OgzTmDBhgrF161bj22+/NXbt2mX07dvX8PHxMY4dO+bq0Ezhv//9r+Hm5ma88MILxpEjR4wVK1YYXl5exvLly10dGkyK3MT1yE1qF/KSmkVe4lrkJfVTeT7bJ0+ebDzyyCOO9W+//dbw8vIy/vSnPxmHDx82Fi5caLi7uxv//Oc/XTEEJ5UZz6uvvmqsXr3a+Prrr4309HRj8uTJhiTjvffec8UQnFzqc78unRvDqPh4avO5Kc2FeUFdOz8XutR4avP5uVTOUNfOTUXHU5vPjWGUL6dw5TmiMFILSCp1WbRokatDM42hQ4caYWFhhoeHh9GsWTMjJiaGCw8uxgWImjVw4EAjODjYcHd3N0JCQoy4uDjj0KFDrg7LVD766CMjIiLCsFqtxrXXXmu8+eabrg4JJkZu4nrkJrULeUnNIi9xPfKS+qc8n+3x8fFGVFSU035bt241OnToYHh4eBgtW7Y0FixYULOBl6Ey43nxxReNq6++2vD09DSaNGli3Hbbbca6detqPvhSXOpzvy6dG8Oo+Hhq87kpzYV5QV07Pxe61Hhq8/m5VM5Q185NRcdTm89NsUvlFK48RxbD+P+fXgIAAAAAAAAAAFDP8fB1AAAAAAAAAABgGhRGAAAAAAAAAACAaVAYAQAAAAAAAAAApkFhBAAAAAAAAAAAmAaFEQAAAAAAAAAAYBoURgAAAAAAAAAAgGlQGAEAAAAAAAAAAKZBYQQAAAAAAAAAAJgGhREAAAAAAAAAAGAaFEYAAAAAAAAAAIBpUBgBAAAAAAAAAACmQWEEAAAAAAAAAACYxv8H8mf4JQZwHgIAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 2000x500 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGwCAYAAACHJU4LAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABDvUlEQVR4nO3de1xUdf7H8fdwEQVhRMDMQBTBjPJWQirKVttlNyvNXxfN+1q7lUFbaWbXbbvY7qN2K3MtqzXJxNpKs9rsspuamrcUNS0TATG1BOQisIHC+f3RMsvAADMww1x4PR8PH484c+acj07Km+/5fj9fk2EYhgAAADyQn7sLAAAAaApBBQAAeCyCCgAA8FgEFQAA4LEIKgAAwGMRVAAAgMciqAAAAI8V4O4C2qK2tlZHjx5VaGioTCaTu8sBAAB2MAxDJ0+eVK9eveTn1/yYiVcHlaNHjyomJsbdZQAAgFY4fPiwoqOjmz3Hq4NKaGiopJ9/o2FhYW6uBgAA2KOsrEwxMTGW7+PN8eqgUve4JywsjKACAICXsWfaBpNpAQCAxyKoAAAAj0VQAQAAHougAgAAPJbbg8qRI0c0efJkRUREKDg4WEOGDNFXX33l7rIAAIAHcOuqn+LiYqWkpOjiiy/WRx99pB49eujgwYPq1q2bO8sCAAAewq1B5U9/+pNiYmK0ZMkSy7E+ffq4ryAAAOBR3ProZ/Xq1Ro2bJiuv/569ejRQ0OHDtXLL7/c5PlVVVUqKyuz+gUAAHyXW4NKTk6OFi1apISEBH388ce69dZblZ6eroyMDJvnz58/X2az2fKL9vkAAPg2k2EYhrtu3qlTJw0bNkybNm2yHEtPT9e2bdv05ZdfNjq/qqpKVVVVlq/rWvCWlpbSmRYAAC9RVlYms9ls1/dvt46onHnmmUpMTLQ6ds455yg/P9/m+UFBQZZ2+bTNBwDAtXIKyvX5/uPKLaxwWw1unUybkpKi/fv3Wx377rvvFBsb66aKAADwTTkF5Tp0olJ9IkLUNzKk2XNLKquVnpml9QcKLMdSE6K0YOJQmYMDXV2qFbcGlbvuuksjR47Uk08+qRtuuEFbt27V4sWLtXjxYneWBQCAz2hN6EjPzNLG7EKrYxuzC5WWuVMZM5NdWm9Dbn30k5SUpJUrVyozM1PnnXeeHnvsMT377LOaNGmSO8sCAMBnNBc6bMkpKNf6AwWqaTCFtcYwtP5AQbs/BnLriIokXXXVVbrqqqvcXQYAAD6nLnQ0VD90NHwMdOhEZbPXzCtq/B5XcnsLfQAA4Br2hI6GYrsHN/uePhHtF1IkggoAAD6rNaEjLqqrUhOi5G8yWR33N5mUmhDVrqMpEkEFAACf1drQsWDiUKXER1odS4mP1IKJQ11Wa1Pc2vCtrRxpGAMAQEdUWnlKaZk7W7XUOLewQnlFFXYtaXaEI9+/CSoAAHQArgodreHI92+3r/oBAACu1zfS/QGlNZijAgAAPBZBBQAAeCyCCgAA8FgEFQAA4LEIKgAAwGMRVAAAgMciqAAAAI9FUAEAAB6LoAIAADwWQQUAAHgsggoAAPBYBBUAAOCxCCoAAMBjEVQAAIDHIqgAAACPRVABAAAei6ACAAA8FkEFAAB4LIIKAADwWAQVAADgsQgqAADAYxFUAACAxyKoAAAAj0VQAQAAHivA3QUAANBR5BSU69CJSvWJCFHfyBB3l+MVCCoAALhYSWW10jOztP5AgeVYakKUFkwcKnNwoBsr83w8+gEAwMXSM7O0MbvQ6tjG7EKlZe50U0Xeg6ACAIAL5RSUa/2BAtUYhtXxGsPQ+gMFyi2scFNl3oGgAgCACx06Udns63lFBJXmEFQAAHCh2O7Bzb7eJ4JJtc0hqAAA4EJxUV2VmhAlf5PJ6ri/yaTUhChW/7SAoAIAgIstmDhUKfGRVsdS4iO1YOJQN1XkPVieDACAi5mDA5UxM1m5hRXKK6qgj4oDCCoAALSTvpEEFEfx6AcAAHgsggoAAPBYBBUAAOCxCCoAAMBjEVQAAIDHIqgAAACPxfJkAIDPyyko16ETlfQv8UIEFQCAzyqprFZ6ZpbWHyiwHEtNiNKCiUNlDg50Y2WwF49+AAA+Kz0zSxuzC62ObcwuVFrmTjdVBEcRVAAAPimnoFzrDxSoxjCsjtcYhtYfKFBuYYWbKoMjCCoAAJ906ERls6/nFRFUvAFBBQDgk2K7Bzf7ep8IJtV6A4IKAMAnxUV1VWpClPxNJqvj/iaTUhOiWP3jJQgqAACftWDiUKXER1odS4mP1IKJQ91UERzF8mQAgM8yBwcqY2aycgsrlFdUQR8VL0RQAQD4vL6RBBRvxaMfAADgsQgqAADAYxFUAACAxyKoAAAAj0VQAQAAHougAgAAPBbLkwEAPienoFyHTlTSN8UHEFQAAD6jpLJa6ZlZWn+gwHIsNSFKCyYOlTk40I2VobV49AMA8BnpmVnamF1odWxjdqHSMne6qSK0FUEFAOATcgrKtf5AgWoMw+p4jWFo/YEC5RZWuKkytAVBBQDgEw6dqGz29bwigoo3cmtQ+cMf/iCTyWT1q2fPnu4sCQDgRjkF5fp8//FWjX7Edg9u9vU+EUyq9UZun0x77rnn6rPPPrN87e/v78ZqAADu4IxJsHFRXZWaEKWN2YVWj3/8TSalxEey+sdLuf3RT0BAgHr27Gn5FRUV5e6SAADtzFmTYBdMHKqU+EirYynxkVowcWiba4R7uH1E5cCBA+rVq5eCgoJ04YUX6sknn1RcXJzNc6uqqlRVVWX5uqysrL3KBAA4Uf0+J8Z/J7s2VH8SrL2jIebgQGXMTFZuYYXyiiroo+ID3BpULrzwQmVkZKh///768ccf9fjjj2vkyJHau3evIiIiGp0/f/58Pfroo26oFADgDLYe8Zx3Vliz78krsj+o1OkbSUDxFSbDaLCOy40qKirUr18/3Xvvvbr77rsbvW5rRCUmJkalpaUKC2v+f3QAgPtNfXVrozkkfiaptpnvRJ/PvojQ4WPKyspkNpvt+v7t9kc/9YWEhGjgwIE6cOCAzdeDgoIUFBTUzlUBAJyhrs9JQ3UhxU9Sbb3jTIKF5AGTaeurqqrSN998ozPPPNPdpQAAnKylPieJvax/smYSLCQ3j6jMnj1bV199tXr37q3jx4/r8ccfV1lZmaZNm+bOsgAALtBSn5MFN50vSUyChRW3BpXvv/9eEydOVGFhoaKiojR8+HBt3rxZsbGx7iwLAOAC9vY5IaCgPo+aTOsoRybjAADcr7TylNIyd7K7cQfntZNpAQC+jT4ncBRBBQDQ7uhzAnt51KofAACA+ggqAADAYxFUAACAxyKoAAAAj8VkWgCAw9btP66s70t0fu9wjU6Icnc58GEEFQCA3Q4VVWjcwo0qrjxlORYeHKjVs0YpJqL5zrNAa/DoBwBgt4YhRZKKK0/p18+vV25hhZuqgi8jqAAA7LJu//FGIaVOeVWNLn56raa+ulWlTZwDtAZBBQBgl6zvS1o8Z2N2odIyd7q+GHQYBBUAgF2GRHdr8Zwaw9D6AwU8BoLTEFQAAHb5xdk9FG7nxoF5RQQVOAdBBQBgkVNQrs/3H29yRGT1rFF2hZU+EezjA+dgeTIAQCWV1UrPzNL6AwWWY6kJUVowcajM9YJJTESwdj58ub44UKAHV+5R/on/yKh3HX+TSSnxkWw4CKdhRAUAoPTMLG3MLrQ61tzE2NEJUVp9x+hGzd7O791NCyYOdVmd6HgIKgDQweUUlGv9gQLVGIbV8ZYmxpqDA/X8xCFK6hNuObbtULHSMneyRBlOQ1ABAB/R0vySphw6Udns681NjL156XZ9lVdsdYwlynAm5qgAgJezd35JU2K7N9/63tbE2JLKat2SsV3bDxU3eq3+SAxzVdBWjKgAgJdzdH5JQ3FRXZWaECV/k8nquL/JpNSEKJthIz0zS1/ZCCn1sUQZzkBQAQAv1tL8ki/qjbI0Z8HEoUqJj7Q6lhIfaXNibN09a41GL1lhiTKcgUc/AODFWppfMuXVrXY9BjIHBypjZrJyCyuUV1ShPhEhTT62aemefpJGNTESAziKERUA8GItzS+RHHsM1DcyRBef3aPZkNHSPS+IDWeJMpyGoAIAXqyp+SX1OfoYqLX39DNJSX3C9Y/bRto1iRewB0EFALycrfkltkx5daumvrrVKT1ObN1zVHyUXpma1OZrA/WZDMNoYTqU5yorK5PZbFZpaanCwsLcXQ4AuNX67wo09e9bmz2nrsV9xsxkST9PjD10orLZOSnNsWdOC9CQI9+/mUwLAD4itX+UUhOitDG7sNEqoDp1j4F2HS7RM5981+reK3X6RhJQ4Fo8+gEAH2LvY6AHVu1pU+8VoL0QVADAh9QtM874TfNzRb4+Uubw3j6AOxBUAMAHpfbv0WS32fN6NT8ngI6y8CQEFQDwUU11m33i2oHNvo+OsvAkTKYFAB/VXLdZW5Nu61YEMTkWnoQRFQDwcba6zTqytw/gToyoAICHa2uvE1sc2dsHcCeCCgB4oJyCcu07Wqalm/K07VCx5bitXidtCTL0QYGnI6gAgAcpqaxWemaWVSO2+up6nWTMTLZ5bmuatgGejDkqAOBB0jOzGjViq69+rxNb59K0Db6GoAIAHiKnoFzrDxQ02f6+vs05hTbPpWkbfA2PfgDAzermmPxQ+pMD7zI1+2peUQVzT+ATCCoA4CYtzUexpa7XyYV9uzd7Hk3b4Ct49AMAbtLSfBRb6nqdxEV1bbJFfmpCFKMp8BlOCyqXXnqp4uLinHU5APBpjsxHqfPU+IHKmJlsWdFD0zZ0BE579HPttdeqsNCxnwwAoKM6dKLS4fdcGBdh9TVN29AROC2ozJo1y1mXAgCfF9s92KHzzzsrrMkQQtM2+DLmqACAGzQ1x6Qp118Q7eKKAM9k14jK+PHj7b7gu+++2+piAKAjWTBxqG574yttOljU4rkl/znVDhUBnseuoGI2m11dBwD4DHv33jEHByrAz09+kmpbuOb5vcOdWiPgLewKKkuWLHF1HQDg9Rzde6du5U9LwoMDNTohyqm1At6i1XNUjh8/ri+++EIbNmzQ8ePHnVkTAHilm5du14YGwWNDdoHV3js5BeX6fP9x5RZW2L3yZ+HE851aJ+BNHF71U1ZWplmzZmnFihWqqamRJPn7++vGG2/UwoULeUwEoMMpqazWlFe3aM+Rskav1RrS+gMF2nigQC+tz7UaQUnqY9/jnKralh4MAb7L4RGVm2++WVu2bNEHH3ygkpISlZaW6oMPPtD27dt1yy23uKJGAPBYJZXVuvjptTZDSn2/fX17oy60Ow6VKDw4UH4tLPyhHT46ModHVD788EN9/PHHGjVqlOXYFVdcoZdfflm/+tWvnFocAHi6Ka9uUXFlyytyKqobj4rUGIaKK08pKTZc2w4VN3q9bl8feqSgI3N4RCUiIsLm4x2z2azwcGalA+gYSiqrdf2Lm1ocSbHH7ZfEa/WsFJ3XK8zqOO3wgVaMqDz44IO6++67lZGRoTPPPFOS9MMPP2jOnDl66KGHnF4gAHii9MwsfWVjFKQ16pYxf5A+mnb4QAMOB5VFixYpOztbsbGx6t27tyQpPz9fQUFBKigo0EsvvWQ5d8eOHc6rFAA8hL3LihsySaq/BaGtRzu0wwesORxUxo0b54IyAMB7tGZDwc4BfvrptPU8lQvjuvNoB2iBw0HlkUcecUUdANAu7O0a25zWNKCqahBS/CQF+PnZbAQH4H+ctnsyAHgyR7vG2nsNexkNvq7Vz/1VcgsreNQDNIPdkwF0COmZWY36mGzMLrTqGtuaa9Rn7tI48JzTM7TZa+YVVdh9f6AjYkQFgM9ravJrjWHYParR0gTaFyefr1+dd6ZyCyu0OadIJkkXxkXIMAxd8sy6Jt9HMzegeQQVAD6vpcmveUUtB5WWrvHC59n61Xln2ly1k5oQpY3Zhaox/vcAiGZugH149APA58V2D2729ZZGNUoqq7Xw39nNnvP1kTLlFtp+jLNg4lClxEdaHaOZG2Afp42ovPfeeyotLdXUqVOddUkAcIq4qK5tGtVIz8zSzvySFu/T1MiMOThQGTOTaeYGtILTRlTmzp2rGTNmOOtyAOBUrR3VqJubUj/gNKWlkZm+kSG6+OwehBTAAU4bUfn222+ddSkAcLrWjmrY09yN+SaA6zBHBUCHYtgxMlJfS/NbJOabAK7k8IjK7t27bR43mUzq3LmzevfuraCgoDYXBgDOZKtZ23lnhenJawdqUHS3Jt/X1PwWP0mJvcK04KbzGUkBXMhkOPjjhZ+fn0wmU5OvBwYG6sYbb9RLL72kzp07233d+fPn6/7779edd96pZ5991q73lJWVyWw2q7S0VGFhYS2/AUCHUb9VvmEYSs/cqX1Hy1Rr49yWOtSWVp5SWubONnW1BfA/jnz/dnhEZeXKlZo7d67mzJmj5ORkGYahbdu26ZlnntEjjzyi06dP67777tODDz6op59+2q5rbtu2TYsXL9agQYMcLQcArLSmzf36AwW67sVNWjx1GKt2AA/jcFB54okn9Nxzz+mKK66wHBs0aJCio6P10EMPaevWrQoJCdE999xjV1ApLy/XpEmT9PLLL+vxxx9v9tyqqipVVVVZvi4rK3O0fAA+Lj0zSxtasRfPgePluvjptc2OlNhq5gbAtRyeTLtnzx7FxsY2Oh4bG6s9e/ZIkoYMGaJjx47Zdb1Zs2ZpzJgxuvTSS1s8d/78+TKbzZZfMTExjhUPwKfVLSW29XjHXo7u/wPAtRwOKgMGDNBTTz2l6upqy7FTp07pqaee0oABAyRJR44c0RlnnNHitVasWKEdO3Zo/vz5dt173rx5Ki0ttfw6fPiwo+UD8GH2LCVuSf39fwC4n8OPfhYuXKhrrrlG0dHRGjRokEwmk3bv3q2amhp98MEHkqScnBzdfvvtzV7n8OHDuvPOO/XJJ5/YPek2KCiIFUUAmmTPUmJ72bP/DwDXc3jVj/TzvJJly5bpu+++k2EYGjBggG666SaFhja/nXl9q1at0rXXXit/f3/LsZqaGplMJvn5+amqqsrqNVtY9QOgTlZ+sR5872t9fcQ5c9dWz0rRoJhuTrkWAGuOfP92OKhUVlYqOLjtP7WcPHlShw4dsjo2Y8YMDRgwQHPnztV5553X4jUIKgBas8rHHqkJUcqYmezUawL4mUuXJ/fo0UPjxo3TlClTdNlll8nPr3XNbUNDQxuFkZCQEEVERNgVUgBAkm5btkNf5hQ5/bp181R4/AO4l8MpIyMjQ1VVVbr22mvVq1cv3Xnnndq2bZsragMAi5yCcn2+/7jVJNecgnKXhJQ6eUVMqAXczeERlfHjx2v8+PE6efKk3n77bWVmZmrkyJHq27evJk+erIcffrjVxaxdu7bV7wXgm2w92qnrdbIl94RL793SbsgAXK/VmxKGhoZqxowZ+uSTT7Rr1y6FhITo0UcfdWZtAKD0zCxtzC60OrbhQIEmvbJZhSermnhX2/ibTEpNiOKxD+ABHB5RqfPTTz9p9erVWr58udasWaMePXpo9uzZzqwNQAfUcI8eW5NkayV9fbRMXx91TXdqdkMGPIfDQeWTTz7RG2+8oVWrVsnf31/XXXedPv74Y/3iF79wRX0AOghbj3gSenRt9zpen5ms0QlR7X5fALY5HFTGjRunMWPGaOnSpRozZowCA9k5FEDb/fyIx3r05MDx8jZf9/yYbgrwN2lrXnGz5/mbTEqJjySkAB7G4aDyww8/0LMEgFPV7dHjbPVHR3ILK7Q5p0gmSYlnhunpT76zuiePewDP5HBQqR9S/vOf/+jUqVNNvg4A9nDGHj22RIf/rzllw52PM2YmK7ewQnlFFeoTwa7IgKdyOKhUVFRo7ty5euutt1RU1Lh/QU1NjVMKA9BxOHOPnvpa2q+nYXgB4HkcXp5877336t///rf+9re/KSgoSK+88ooeffRR9erVSxkZGa6oEYCP6x7SSWGdW70IsUn0QQG8n8P/Mrz//vvKyMjQRRddpN/85jcaPXq04uPjFRsbqzfeeEOTJk1yRZ0AfNjtb+xQ2U+n3V0GAA/k8IjKiRMn1LdvX0k/z0c5ceLnzpCjRo3S+vXrnVsdAJ9Uvx1+Vn6xNh10TRt8WuAD3s/hEZW4uDjl5eUpNjZWiYmJeuutt5ScnKz3339f3bp1c0GJAHyFrV4pwZ38XXY/Hv0A3s/hEZUZM2Zo165dkqR58+ZZ5qrcddddmjNnjtMLBOB5bG0QaA9b7fArq9s2AT/Az9ToHzJa4AO+w2QYhtGWC+Tn52v79u3q16+fBg8e7Ky67FJWViaz2azS0lKWRQPtoLkNAs3BzTd/zCko1yXPrHNqPQlnhOjvU5P1wKqvW1UTAPdw5Pt3m6fZ9+7dW717927rZQB4AVsjIhuzC5WWuVN/uCbRskePrZGMrXnO3+l48ZQkxUQE0xMF8GHOXw8IwCc11T225r8bB9YfLWk4olFSWa0H3t3j1HqSYsOtAgk9UQDfRFABYBdHusduyC7QpFc3694rBqjGMPT4B/tU06aHzNa6BvnroasTnXdBAB6LoALALo50j601pK+PlGnq37e6pJbyqhpd88JG5qIAHYDDq34AdExxUV2VmhAlf5PJ3aVY1M2PAeC77A4qR48e1ezZs1VWVtbotdLSUs2ZM0c//vijU4sD4FkWTByqlPhId5dhUTc/xtFl0gC8h91B5S9/+YvKyspsLiMym806efKk/vKXvzi1OACexRwcqIyZyfp89kVaMiNJn8++SKkJUe4uiw60gA+zO6isWbNGU6dObfL1qVOn6oMPPnBKUQA8W9/IEF18dg/1jQzRb1P7uLscOtACPszuoJKbm9tsv5To6Gjl5eU5oyYAXuSpNfvb5T7hwYF0oAU6ILtX/XTp0kV5eXlNhpW8vDx16dLFaYUB8Hw5BeX6+kjjeWuOSOjRVQeOl1u+TuoTrmkj++jcXmZJsjRx6x7cSWmZO616uaTER2rBxKFtuj8Az2Z3ULnwwgv1+uuvKzU11ebrGRkZSk5OdlphADyfI71VbEmKDdc/bhvZbFfZ+l/TgRboeOwOKrNnz9Zll10ms9msOXPm6IwzzpAk/fjjj/rzn/+s1157TZ988onLCgXgeRzprdJQ5wCTXpmWJMmxrrJ0oAU6FruDysUXX6yFCxfqzjvv1F//+leFhYXJZDKptLRUgYGBWrBggS655BJX1gqgHeUUlGtLbpEkk4bHRVjCQU5BuWVPn7ioroqNCNahIsdHVub+agCN2gC0yKHOtL/73e901VVX6a233lJ2drYMw1D//v113XXXKTo62lU1AmhHuw4X6643dymnQW+SYbHh6hTgp00HiyzHUhOiFN2tc6uCykUDzmhzrQB8n8Mt9M866yzdddddrqgFgBuVVFbrtmU79GVOkc3Xtx8qbnRsQ3aB/DyoUy0A32N3UFm9erVd511zzTWtLgaA+9ySsV3b8hqHkebUGlKt0brdBvOKKphrAqBFdgeVcePGtXiOyWRSTU1NW+oB0M5KKqt189LtNkdMXIkmbQDsYXdQqa2tdWUdANwkPTNLO9oxpPibTEqJj2Q0BYBdHJ6jAsB35BSUWzVQaw80aQPgCLuDyvr16+06r6mGcAA8T1sbttnr9ZnJOl1r0KQNgMPsDioXXXRRk6+Z/jvr32Qy6fTp020uCkD7aEvDNnvUPeYZ7QE7LAPwTnYHleJi28+wKysr9dxzz+n5559XXFyc0woD4Fp1jdtCOvmroto1k+B5zAOgrewOKmaz2err2tpa/f3vf9ejjz4qPz8/LVy4UNOmTXN6gQDarn432fDgQKVnZrlkbkpSn3A9dFWiiiqqecwDwClaNZn23Xff1f3336+CggLNmzdPaWlpCgoKcnZtANqopLK6USgJDw5U6X9OOe0eJknn9grTgpvOJ5gAcDo/R05et26dhg8frilTpmj8+PHKycnR7NmzCSmQ9PNP7Z/vP67cBq3X4T7pmVnamF1oday48pRqW9ejzabRCVF64+bhhBQALmH3iMqVV16pf/3rX5oxY4ZWrVqlnj17urIueBFbP7WnJkRpwcShbDrnRq5eevzwVefo4gFnEFAAuJTJMOzrf+3n56eAgACFhIRYVvnYcuLECacV15KysjKZzWaVlpYqLCys3e4La1Nf3aqN2YWqqfe/Ut1qj4yZyW6srGP7fP9xzViyzSXXPu+sMH2QNtol1wbg+xz5/m33iMqSJUvaXBh8T1M/tdcYhtYfKFBuIfu5uIsrlx4/ee1Al10bAOqzO6iwoge2tNQwjI3n3CcuqqtSE6K04UCBnLkBRlJsuAZFd3PiFQGgaXZPpi0uLtaCBQtUVlbW6LXS0tImX4Nva+mndjaecw1bE5frjq3/7rjltd+l9v15WY4TTR/Zx7kXBIBm2D2i8sILL2j37t1KS0tr9JrZbNYXX3yhsrIyPfDAA04tEJ6t7qf2puaoMJriXLYmLo+Ii5DJJG06WNQuNSSeZW75JABwErtHVN555x3deuutTb7+u9/9Tm+//bZTioJ3WTBxqFLiI62O0ZHUNWwtN/4yp6hdQoq/6efVXIRPAO3J7hGVgwcPKiEhocnXExISdPDgQacUBe9iDg5Uxsxk5RZWKK+ogo6kLuKOnY7rS4mPInwCaHd2BxV/f38dPXpUvXv3tvn60aNH5efnUP84+Ji+kQQUV2qvnY79TFJirzAtmHi+JBE+AbiV3UFl6NChWrVqlYYPH27z9ZUrV2roUH7aAlzF1Tsd1xkVb92sj4ACwJ3sDip33HGHJkyYoOjoaN12223y9/eXJNXU1Ohvf/ub/vrXv2r58uUuKxTo6OKiumpojFk7D5c69br+JpPO791Nt18Sz8gJAI9jd1D5v//7P917771KT0/XAw88oLi4OJlMJh08eFDl5eWaM2eOrrvuOlfWCnR4u753bkiR/jfxme0OAHgiu1vo19m6daveeOMNZWdnyzAM9e/fXzfddJOSk9u/VTot9NFRbDhQoN8s2apqJ3VuG3hWmO6+/GxGUAC4hUta6NdJTk52SygBOiJbfVOc4YlrB9JdFoBXsDuo7N69267zBg0a1OpiAFi7/Y0dLumRUlRR7fRrAoAr2B1UhgwZIpPJpOaeFJlMJtXU1DilMKCjW7f/uMsaubG1AQBvYXdQyc3NdWUdQIeRU1CuQycqm5wfUlJZrduW7dCXOc4PKSZJo+kuC8CL2B1UYmNjXVkH4PNszTdJTfhfz5KcgnJtyS3SC//O1pGSn1xSQ3yPrnSXBeBV7A4qlZWVmjNnjlatWqVTp07p0ksv1fPPP6/IyMiW3wzA5j49Gw4UaPrft0gmk3YeLnF5DYunDmMZMgCvYnfP+0ceeUSvvfaaxowZowkTJujTTz/Vbbfd5sraAJ9Rt09PTYM5XrWSdn5f2i4hZWS/CB75APA6do+ovPvuu3r11Vc1YcIESdLkyZOVkpKimpoaS5daALa11z49Tal7xAQA3sbuoHL48GGNHj3a8nVycrICAgJ09OhRxcTEuKQ4wFe01z49dVITojT7iv4qqqimqRsAr2Z3UKmpqVGnTp2s3xwQoNOnTzu9KMDbNVzZExfVVcNiw7X9ULFL7zugZ6gWTb6AYALAZ9gdVAzD0PTp0xUUFGQ59tNPP+nWW29VSMj//lF89913nVsh4EVsrewZERchk0kuDyn1VxABgK+wO6hMmzat0bHJkyc7tRjA29la2eOKfij13XVZgq4ZfBajKAB8kt1BZcmSJa6sA/BqdT1QnL0njz0IKQB8mcObEgL4H1dtGmgPf5NJKfGRhBQAPs3uPioAGrP1qMdVzF2s556kxEey5BiAz2NEBWiluiZu7eH1mckanRCl3MIK5RVVsOQYQIfh1hGVRYsWadCgQQoLC1NYWJhGjBihjz76yJ0lAXZrjyZufqafV/OMToiSJPWNDNHFZ/cgpADoMNwaVKKjo/XUU09p+/bt2r59uy655BKNHTtWe/fudWdZgJWcgnJ9vv+4cgsrrI63RxO3UfF0lAXQsbn10c/VV19t9fUTTzyhRYsWafPmzTr33HMbnV9VVaWqqirL12VlZS6vER1XS7sdS1JQgJ+qTtc6/d4Deobqz/83SINiujn92gDgTTxmjkpNTY3+8Y9/qKKiQiNGjLB5zvz58/Xoo4+2c2XoqGxNlN2YXahbl32ln07VuGQjwYQeIVo8NYlHOwDwXybDaLCdazvbs2ePRowYoZ9++kldu3bV8uXLdeWVV9o819aISkxMjEpLSxUWFtZeJaMDyCko1yXPrGvXeybFhuuVaUl0lgXg88rKymQ2m+36/u32EZWzzz5bWVlZKikp0TvvvKNp06Zp3bp1SkxMbHRuUFCQVQt/wFXae7fjoTFm/eO2ke16TwDwBm4PKp06dVJ8fLwkadiwYdq2bZuee+45vfTSS26uDB1Ze+52XDfvBQDQmNuDSkOGYVg93gHaW0lltf6wep/L7zN//EANj4tgPgoANMOtQeX+++/Xr3/9a8XExOjkyZNasWKF1q5dqzVr1rizLHQwOQXlOnSi0tJE7bcZ27U1z3U7HfuZfl52PDG5t8vuAQC+wq1B5ccff9SUKVN07Ngxmc1mDRo0SGvWrNFll13mzrLQQdhafpzUJ1zbXBhSJHqjAIAj3L7qpy0cmTUMNDT11a3akF2g2nb6G3BerzAtuOl8HvUA6PC8atUP4A67Dhe3647HI/tFaNGkC1h6DAAOIqigQ6mbj/Lkh9+0y/3iokL07A1D6DALAK1EUEGHYGs+iqvV7XgMAGg9t25KCLQXW+3wXe10e01+AQAfxogKfF5OQXm7jqTU6RPBpFkAaCuCCnxGw34oddq7Hb6/SUqJj2J1DwA4AUEFXs/W/JPUhCjdc3mCTlSekr/J1K71pNAnBQCchqACr2dr/sn6AwVWwSWkk78qq2vkilkjSbHheuiqRBVVVjcazQEAtA1BBV7N3vknFdU1Tr/3w1cl6uIBPQgmAOBCBBV4tfaefyJJJkmjE6L0m1F92/3eANDRsDwZXi22e3C733N0AnNQAKC9MKICrxYX1VWpCVEuX35sknQue/UAQLtjRAVe757LE1x+j9EJUXrj5uGEFABoZ4yowOvtO1bm9GuGBwfq3dtTlFdUwUoeAHAjggq8VklltW5btkNf5hQ59brmLgFaPWuUYiKCCSgA4GYEFXitm5du1/ZDxU67Xo/QQD1zw1A2EgQAD0JQgdfJyi/WPf/YpYMFFU67ZnhwoD6962KZgwOddk0AQNsRVOA1SiqrdUvGdm3Lc94oiiQNjTHrtRkXElIAwAMRVOAVdh0u1k0vb1ZFda3TrplwRogWT0liHgoAeDCCCjyarQ0HneWZ64cQUgDAw9FHBR4tPTNLG7Jd08ytqKLaJdcFADgPIyrwKDkF5Tp0olJ9IkK072ipSzvO9olgNAUAPB1BBR7BlY94GvI3mZQSH8ljHwDwAjz6gUdIz8zSxuzCdrlXSnwkmwoCgJdgRAVul1NQ7vKRlKTYcE0f2UeJZ5kZSQEAL0JQgdsdOlHpsms/NX6gLoyLIJwAgJciqMCtSiqr9dyn37nk2qkJUZqQ3Nsl1wYAtA/mqMBtcgrKNfmVLcr6vrRN1+kZFtTo2Ii4COahAIAPYEQF7W7X4WI9sPJrfX20zCnXm/9/g9QnIkRbcopkSBrOox4A8BkEFbQbVy1B7hMRor6RIYQTAPBBBBW0m9vf2KFNB4ucdj0/SaMSoggoAODDmKOCdpFTUO7UkCJJF8SGMw8FAHwcIypwmbp2+HuPlGrN18eceu2kPuH6x60jnXpNAIDnIajAqXIKyrXvaJmWbsrTtkPFLrnHyH4RWjTpApdcGwDgWQgqcIqSymrdkrFd2/JcE04kKaFHVz1zw2ANiu7msnsAADwLQQVtVlJZrYufXqviylMuuf55vcK04KbzmTQLAB0QQQVtdvPS7S4LKZ0D/LRo0gWKiQh2yfUBAJ6NVT9ok5yCcm130VwUSao6XasHVn3tsusDADwbQQVt4soNBSXJkLT+QIFyCytceh8AgGfi0Q9apW7p8btf5bfL/fKKKpijAgAdEEEFDnFVG/ykPuHNrhjqE0FIAYCOiKACh9y2bIe+zGl7h9mh0Wb9ZcJQ5RVVWPbqmfrqVm3MLlSNYVjO8zeZlBIfyWgKAHRQBBW0qO4xj7/J5JSQMiIuQi9OvkDm4ECrALJg4lClZe60Gq1JiY+kTT4AdGAEFTTJ2Y95Es4I0TPXD2myYZs5OFAZM5OVW1hhNdICAOi4CCpoUnpmlr5wUkh5fWayRidE2XVu30gCCgDgZwQV2JSVX+y0kZTUhCi7QwoAAPXRRwU2zXl7t1OuMzTGzBwTAECrMaICKyWV1bp56XYdOF7e5msN7BWmlbNGOaEqAEBHxYgKrNy8dLu+ckJL/PDgQC27ebgTKgIAdGSMqEDr9h/X5twirc46qiMlP7X5ekl9wvXK1CSZgwOdUB0AoCMjqHRgh4oqNG7hRqftfPzIVefoogFnsGIHAOA0BJUO7OoFG1T20+k2X8fPJI2Kj9KMUXFOqAoAgP9hjkoHVFJZraue/8IpIUX6OaSwsgcA4AqMqHQwOQXlmrFkqw6d+E+br3XXZQm6ZvBZPOoBALgMQaWDKKms1pRXt2jPkTKnXZOQAgBwNYJKB5CVX6zJr25ReVWNU67HjsYAgPZCUPEhdbsc123ml5VfrAff+1pfO3EURWJHYwBA+yGo+ABbuxyHdQ5w2mTZ83qF6cnxA1VUUc2OxgCAdkVQ8QHpmVnamF1odcxZIUWSFtx0PuEEAOAWBBUvl1NQ7rRdjhtiLgoAwN3oo+Ll9h5z7vyT+piLAgBwN0ZUvFzGpjynX/Op8QN1YVwEIykAALcjqHixJRtytC2v7Tsd1zckppsmJPd26jUBAGgtgooXOlRUoaue/0InndQXpU54cKCWzkh26jUBAGgLgoqXySko15jnv9B/TtU69bpJfcL1ytQkmYMDnXpdAADagqDiBXIKyrX3WJn+/kWudh4ucdp1X5+ZrNO1Br1RAAAei6DiwUoqq3VLxnanz0ORpBFxERqdEOX06wIA4ExuXZ48f/58JSUlKTQ0VD169NC4ceO0f/9+d5bkMbLyi5Xyp3+7JKSkJkTpxckXOP26AAA4m1tHVNatW6dZs2YpKSlJp0+f1gMPPKDLL79c+/btU0hIx3wUYasdvjP0PyNE865M5DEPAMCrmAzDMNxdRJ2CggL16NFD69atU2pqaqPXq6qqVFVVZfm6rKxMMTExKi0tVVhYWHuW6jJTX93qkk6zq2elaFBMN6dfFwAAR5WVlclsNtv1/dujOtOWlpZKkrp3727z9fnz58tsNlt+xcTEtGd5LufKdvhFldUuuS4AAK7kMSMqhmFo7NixKi4u1hdffGHzHF8eUSmprNaElzfr22MnXXL9z2dfxCMfAIBHcGRExWNW/dxxxx3avXu3NmzY0OQ5QUFBCgoKaseqXC+noFyHTlRq/of79N3xCqdfn40FAQDezCOCSlpamlavXq3169crOjra3eW0C1dMmg0KMGlwdDdtrbdSiI0FAQDezK1BxTAMpaWlaeXKlVq7dq369u3rznLa1e1v7NCmg0VOu96j1yRq2sif//xyCyuUV1TBCh8AgNdza1CZNWuWli9frvfee0+hoaH64YcfJElms1ldunRxZ2kutetwsdNCiknS6IQoS0iRpL6RBBQAgG9w62Rak8lk8/iSJUs0ffr0Ft/vyGQcT1FSWa0R8//ltL16UhOitGDiUPboAQB4Da+ZTOshC47a1fUvbmpzSLn70ngNjAnn0Q4AwOd5xGTajuBQUYXGPP+Fyqtq2nSdN28Zrgv7RTipKgAAPBtBxcVyCsq172iZ0lfsVG0bB5AevTqRkAIA6FAIKi6SlV+sOW/v1oHj5U67ZurZPZx2LQAAvAFBxclKKqt127Id+jLHeUuPpZ8nzTIfBQDQ0XjUXj++ID0zy+khZURcBE3bAAAdEiMqTuTMTQUTenTVxOTeunhAD0ZSAAAdFkHFSbLyi3X78h1tusbsy/vr3LPMLDsGAOC/CCpt5Mw5KWMG9SKgAABQD3NU2shZc1KS+oQTUgAAaICg0gZvbs13ypyU8OBAvTI1yQkVAQDgW3j00wqHiio0buFGFVeeavO1kmLD9cq0JPbqAQDABoKKA3IKynXoRKXueGOHKqrb1gr/vLPC9OS1AzUouptzigMAwAcRVOxQUlmt9Mwspy09fn1mskYnRDnlWgAA+DLmqNghPTNLG7MLnXa9023d9AcAgA6CEZUWOLOJW50+EazuAQDAHgSVZuQUlOv93Ueddj1/k0kp8ZEsQwYAwE4EFRucMSdlRFyETCZp08H/9VhJiY9kzx4AABxAULGhtXNSkmLDdfsl8VYt8HMLK5RXVEFbfAAAWoGg0kBr56SMiIvQi5MvaNQPpW8kAQUAgNYiqDRw6ESlQ+cn9OiqxVOHEUYAAHABgkoD3R3oEDuwV5iW3TycrrIAALgIQaWBZz45YNd5YZ0D9H76aBdXAwBAx0bDt3rsnZ9i7hKgD9MIKQAAuBojKvW0ND/l0nN6aNrIPrS/BwCgnRBU6ontHtzs6w+MSWTSLAAA7YhHP/XERXVVakKU/E0mq+P+JpNSE6IIKQAAtDOCSgMLJg5VSnyk1TE6ygIA4B48+mnAHByojJnJdJQFAMADEFSaQEdZAADcj0c/AADAYxFUAACAxyKoAAAAj0VQAQAAHougAgAAPBZBBQAAeCyCCgAA8FgEFQAA4LEIKgAAwGMRVAAAgMfy6hb6hmFIksrKytxcCQAAsFfd9+267+PN8eqgcvLkSUlSTEyMmysBAACOOnnypMxmc7PnmAx74oyHqq2t1dGjRxUaGiqTyeTuclyqrKxMMTExOnz4sMLCwtxdDmzgM/J8fEaej8/I8znjMzIMQydPnlSvXr3k59f8LBSvHlHx8/NTdHS0u8toV2FhYfzl9XB8Rp6Pz8jz8Rl5vrZ+Ri2NpNRhMi0AAPBYBBUAAOCxCCpeIigoSI888oiCgoLcXQqawGfk+fiMPB+fkedr78/IqyfTAgAA38aICgAA8FgEFQAA4LEIKgAAwGMRVAAAgMciqHi4+fPnKykpSaGhoerRo4fGjRun/fv3u7ss1LNo0SINGjTI0vxoxIgR+uijj9xdFpowf/58mUwm/f73v3d3KajnD3/4g0wmk9Wvnj17urssNHDkyBFNnjxZERERCg4O1pAhQ/TVV1+59J4EFQ+3bt06zZo1S5s3b9ann36q06dP6/LLL1dFRYW7S8N/RUdH66mnntL27du1fft2XXLJJRo7dqz27t3r7tLQwLZt27R48WINGjTI3aXAhnPPPVfHjh2z/NqzZ4+7S0I9xcXFSklJUWBgoD766CPt27dPzzzzjLp16+bS+3p1C/2OYM2aNVZfL1myRD169NBXX32l1NRUN1WF+q6++mqrr5944gktWrRImzdv1rnnnuumqtBQeXm5Jk2apJdfflmPP/64u8uBDQEBAYyieLA//elPiomJ0ZIlSyzH+vTp4/L7MqLiZUpLSyVJ3bt3d3MlsKWmpkYrVqxQRUWFRowY4e5yUM+sWbM0ZswYXXrppe4uBU04cOCAevXqpb59+2rChAnKyclxd0moZ/Xq1Ro2bJiuv/569ejRQ0OHDtXLL7/s8vsSVLyIYRi6++67NWrUKJ133nnuLgf17NmzR127dlVQUJBuvfVWrVy5UomJie4uC/+1YsUK7dixQ/Pnz3d3KWjChRdeqIyMDH388cd6+eWX9cMPP2jkyJEqKipyd2n4r5ycHC1atEgJCQn6+OOPdeuttyo9PV0ZGRkuvS+dab3IrFmz9OGHH2rDhg0dbtdoT1ddXa38/HyVlJTonXfe0SuvvKJ169YRVjzA4cOHNWzYMH3yyScaPHiwJOmiiy7SkCFD9Oyzz7q3ODSpoqJC/fr107333qu7777b3eVAUqdOnTRs2DBt2rTJciw9PV3btm3Tl19+6bL7MqLiJdLS0rR69Wp9/vnnhBQP1KlTJ8XHx2vYsGGaP3++Bg8erOeee87dZUHSV199pePHj+uCCy5QQECAAgICtG7dOj3//PMKCAhQTU2Nu0uEDSEhIRo4cKAOHDjg7lLwX2eeeWajH77OOecc5efnu/S+TKb1cIZhKC0tTStXrtTatWvVt29fd5cEOxiGoaqqKneXAUm//OUvG60emTFjhgYMGKC5c+fK39/fTZWhOVVVVfrmm280evRod5eC/0pJSWnUHuO7775TbGysS+9LUPFws2bN0vLly/Xee+8pNDRUP/zwgyTJbDarS5cubq4OknT//ffr17/+tWJiYnTy5EmtWLFCa9eubbRiC+4RGhraaE5XSEiIIiIimOvlQWbPnq2rr75avXv31vHjx/X444+rrKxM06ZNc3dp+K+77rpLI0eO1JNPPqkbbrhBW7du1eLFi7V48WKX3peg4uEWLVok6edn6vUtWbJE06dPb/+C0MiPP/6oKVOm6NixYzKbzRo0aJDWrFmjyy67zN2lAV7j+++/18SJE1VYWKioqCgNHz5cmzdvdvlP67BfUlKSVq5cqXnz5umPf/yj+vbtq2effVaTJk1y6X2ZTAsAADwWk2kBAIDHIqgAAACPRVABAAAei6ACAAA8FkEFAAB4LIIKAADwWAQVAADgsQgqAADAYxFUADTyhz/8QUOGDLH7/Ly8PJlMJmVlZTV5ztq1a2UymVRSUtLqugzD0G9/+1t17969xfvBPosXL1ZMTIz8/Pz07LPPNvrsp0+frnHjxrmtPoCgAjjAnm/IHVFMTIyOHTvm8r1z1qxZo9dee00ffPBBu9zPXtnZ2ZoxY4aio6MVFBSkvn37auLEidq+fbvlHJPJpM6dO+vQoUNW7x03bpzbtsMoKyvTHXfcoblz5+rIkSP67W9/q9mzZ+tf//qXW+oBbCGoAGgzf39/9ezZUwEBrt0+7ODBgzrzzDM1cuTIJu9XXV3t0hoa2r59uy644AJ99913eumll7Rv3z6tXLlSAwYM0D333GN1rslk0sMPP+zymk6dOmXXefn5+Tp16pTGjBmjM888U8HBweratasiIiJcXCFgP4IKOhzDMPTnP/9ZcXFx6tKliwYPHqy3337b8npxcbEmTZqkqKgodenSRQkJCVqyZIkkqW/fvpKkoUOHymQyNdossjnLli3TsGHDFBoaqp49e+qmm27S8ePHJUm1tbWKjo7Wiy++aPWeHTt2yGQyKScnR5L07bffatSoUercubMSExP12WefyWQyadWqVXbX8dprr6lbt25atWqV+vfvr86dO+uyyy7T4cOHm3xPbW2t/vjHP1pGDIYMGWK1O7StkaZ//vOf6t+/v7p06aKLL75YeXl5ltcqKioUFhZm9ecuSe+//75CQkJ08uTJRjVMnz5daWlpys/Pl8lkUp8+fST9vGHnHXfcobvvvluRkZGWzSDXrVun5ORkBQUF6cwzz9R9992n06dPW6530UUXKS0tTb///e8VHh6uM844Q4sXL1ZFRYVmzJih0NBQ9evXTx999FGTfy6GYWj69OlKSEjQF198oTFjxqhfv34aMmSIHnnkEb333ntW56elpWnZsmXas2dPk9dsyJ7Pq+5xzd///nfFxcUpKChIhmEoPz9fY8eOVdeuXRUWFqYbbrhBP/74o+W6AwcOlCTFxcXJZDIpLy+vxcd+Lf39AZyNoIIO58EHH9SSJUu0aNEi7d27V3fddZcmT56sdevWSZIeeugh7du3Tx999JG++eYbLVq0SJGRkZKkrVu3SpI+++wzHTt2TO+++67d962urtZjjz2mXbt2adWqVcrNzbUM+fv5+WnChAl64403rN6zfPlyjRgxQnFxcaqtrdW4ceMUHBysLVu2aPHixXrggQda9WdQWVmpJ554QkuXLtXGjRtVVlamCRMmNHn+c889p2eeeUZPP/20du/erSuuuELXXHONDhw4YPP8w4cPa/z48bryyiuVlZWlm2++Wffdd5/l9ZCQEE2YMMESAOssWbJE1113nUJDQ23WUBeWjh07pm3btlleW7p0qQICArRx40a99NJLOnLkiK688kolJSVp165dWrRokV599VU9/vjjVtdcunSpIiMjtXXrVqWlpem2227T9ddfr5EjR2rHjh264oorNGXKFFVWVtr8fWZlZWnv3r2655575OfX+J/Tbt26WX09cuRIXXXVVZo3b57N6zXFns8rOztbb731lt555x1LYBw3bpxOnDihdevW6dNPP9XBgwd14403SpJuvPFGffbZZ5J+/v/62LFjiomJabGWlv7+AE5nAB1IeXm50blzZ2PTpk1Wx2fOnGlMnDjRMAzDuPrqq40ZM2bYfH9ubq4hydi5c2eba9m6dashyTh58qRhGIaxY8cOw2QyGXl5eYZhGEZNTY1x1llnGQsXLjQMwzA++ugjIyAgwDh27JjlGp9++qkhyVi5cqXd912yZIkhydi8ebPl2DfffGNIMrZs2WIYhmE88sgjxuDBgy2v9+rVy3jiiSesrpOUlGTcfvvthmE0/nOZN2+ecc455xi1tbWW8+fOnWtIMoqLiw3DMIwtW7YY/v7+xpEjRwzDMIyCggIjMDDQWLt2bZO1//WvfzViY2Otjv3iF78whgwZYnXs/vvvN84++2yr+y9cuNDo2rWrUVNTY3nfqFGjLK+fPn3aCAkJMaZMmWI5duzYMUOS8eWXX9qs58033zQkGTt27Giy5jp1n9PevXsNf39/Y/369YZhGMbYsWONadOmNfk+ez+vwMBA4/jx45ZzPvnkE8Pf39/Iz8+3HNu7d68hydi6dathGIaxc+dOQ5KRm5trOafhZz9t2jRj7NixhmHY9/cHcDZGVNCh7Nu3Tz/99JMuu+wyde3a1fIrIyNDBw8elCTddtttWrFihYYMGaJ7771XmzZtcsq9d+7cqbFjxyo2NlahoaGWx0b5+fmSfn6cNGDAAGVmZkr6+dHF8ePHdcMNN0iS9u/fr5iYGPXs2dNyzeTk5FbVEhAQoGHDhlm+HjBggLp166Zvvvmm0bllZWU6evSoUlJSrI6npKTYPF+SvvnmGw0fPlwmk8lybMSIEVbnJCcn69xzz1VGRoYk6fXXX1fv3r2Vmprq8O+n/u+l7v4jRoywun9KSorKy8v1/fffW44NGjTI8t/+/v6KiIiwPA6RpDPOOEOSLI/oGjIMQ5Ks7tOSxMRETZ06VXPnzrX7PfZ8XrGxsYqKirJ8/c033ygmJsZqlCQxMbHJz9ke9vz9AZzNtTPfAA9TW1srSfrwww911llnWb0WFBQkSfr1r3+tQ4cO6cMPP9Rnn32mX/7yl5o1a5aefvrpVt+3oqJCl19+uS6//HItW7ZMUVFRys/P1xVXXGE1+XPSpElavny57rvvPi1fvlxXXHGF5bGTYRgOfUNsia1rNXf9hq81V0/dN/CW3HzzzXrhhRd03333acmSJZoxY0arfo8hISEt1mYrVAQGBlqdYzKZrI7VnVv3/01D/fv3l/RzKHBkOfejjz6q/v37OzS3qKXPy54/g+aO28Oevz+AszGigg4lMTFRQUFBys/PV3x8vNWv+j95RkVFafr06Vq2bJmeffZZLV68WJLUqVMnSVJNTY1D9/32229VWFiop556SqNHj9aAAQNs/pR+0003ac+ePfrqq6/09ttva9KkSZbXBgwYoPz8fMtkSElW8zQccfr0aauls/v371dJSYkGDBjQ6NywsDD16tVLGzZssDq+adMmnXPOOTavn5iYqM2bN1sda/i1JE2ePFn5+fl6/vnntXfvXk2bNq01vx2b99+0aZNVYNq0aZNCQ0MbfYNtiyFDhigxMVHPPPOMzTDTVM+YmJgY3XHHHbr//vvt+n/Jkc+rTmJiovLz860m3e7bt0+lpaVNfm4tsffvD+BMBBV0KKGhoZo9e7buuusuLV26VAcPHtTOnTu1cOFCLV26VJL08MMP67333lN2drb27t2rDz74wPIPe48ePdSlSxetWbNGP/74o0pLSyXJshy1Kb1791anTp20YMEC5eTkaPXq1Xrssccande3b1+NHDlSM2fO1OnTpzV27FjLa5dddpn69eunadOmaffu3dq4caNlMm39n5B/+ctf6oUXXmj2zyEwMFBpaWnasmWLduzYoRkzZmj48OFNPkqaM2eO/vSnP+nNN9/U/v37dd999ykrK0t33nmnzfNvvfVWHTx4UHfffbf279+v5cuX67XXXmt0Xnh4uMaPH685c+bo8ssvV3R0dLN12+v222/X4cOHlZaWpm+//VbvvfeeHnnkEd199902J722lslk0pIlS/Tdd98pNTVV//znP5WTk6Pdu3friSeesPr8Gpo3b56OHj1qmdDaHEc/L0m69NJLNWjQIE2aNEk7duzQ1q1bNXXqVP3iF79o9KjMXvb8/QGcjaCCDuexxx7Tww8/rPnz5+ucc87RFVdcoffff9+y9LhTp06aN2+eBg0apNTUVPn7+2vFihWSfp4r8Pzzz+ull15Sr169LN+ISktLtX///ibvGRUVpddee03/+Mc/lJiYqKeeeqrJR0mTJk3Srl27NH78eHXp0sVy3N/fX6tWrVJ5ebmSkpJ0880368EHH5Qkde7c2XLewYMHVVhY2OyfQXBwsObOnaubbrpJI0aMUJcuXSy/R1vS09N1zz336J577tHAgQO1Zs0arV69WgkJCTbP7927t9555x29//77Gjx4sF588UU9+eSTNs+dOXOmqqur9Zvf/KbZmh1x1lln6Z///Ke2bt2qwYMH69Zbb9XMmTMtf17OlJycrO3bt6tfv3665ZZbdM455+iaa67R3r179eyzzzb5vu7du2vu3Ln66aefWryHo5+XJMuy9fDwcKWmpurSSy9VXFyc3nzzTUd/i1Za+vsDOJvJsPdhMgCPs3HjRo0aNUrZ2dnq16+fXe957bXX9Pvf/75Nreyd6Y033tCdd96po0ePWh6t4X887fMC2huTaQEvsnLlSnXt2lUJCQnKzs7WnXfeqZSUFLtDiieprKxUbm6u5s+fr9/97neEFAA28egH8CInT57U7bffrgEDBmj69OlKSkpq1P3UW/z5z3/WkCFDdMYZZzjcAA1Ax8GjHwAA4LEYUQEAAB6LoAIAADwWQQUAAHgsggoAAPBYBBUAAOCxCCoAAMBjEVQAAIDHIqgAAACP9f+QGKgqVvPWagAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure(figsize=(20,5))\n",
    "plt.subplot(131)\n",
    "tmp = plt.hist(est_ploidies[\"avg_pl\"],bins=30)\n",
    "plt.title(\"est. avg. ploidy from CN profile\")\n",
    "plt.subplot(132)\n",
    "tmp = plt.hist(sorted(list(GDSC_Ploidies[\"average_ploidy\"].values)),bins=30)\n",
    "plt.title(\"PICNIC avg. pl.\")\n",
    "plt.subplot(133)\n",
    "tmp = plt.hist(est_ploidies[\"median_pl\"],bins=30)\n",
    "plt.title(\"est. median ploidy\")\n",
    "\n",
    "tmp = df_ploidies.plot.scatter(x = \"est. avg. ploidy from CN profile\",y=\"PICNIC avg. pl.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c5f0be52-6d41-4dec-8d04-b7d93ad8a24d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiYAAAHNCAYAAAAqpbCaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABMz0lEQVR4nO3deVxU5f4H8M/ADMMMmwoiiAgoCSq4gQuZgSLuppEtmmtZN3KpXNPr2jW9rlm30tRCu7klIiq5loKUkECWK5qEuaGoV0EWaQae3x/G/BxZBz1wgM/79eJVc+Z5zvM98wzDx7PMUQghBIiIiIhkwKy6CyAiIiIqwmBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYFJHrF+/HgqFwuinYcOGCAoKQnR0dHWXZ+Du7o7Ro0eb3C83Nxfz5s1DTEzME6+ptgoKCkJQUFC57RQKBcaPHy99QU/ItWvXMG/ePPz666/VXcpjGT16NNzd3Y2WKRQKzJs3r1rqeRxFnz8XL140ue+8efOgUCgq1Laynx8kLwwmdUx4eDji4+Nx9OhRrFmzBubm5hg4cCB2795d3aU9ltzcXMyfP5/BhHDt2jXMnz+/xgeTksTHx2Ps2LHVXUaVGjt2LOLj46u7DKpCyuougKqWj48P/P39DY/79OmD+vXrY/PmzRg4cGA1Vlaz5ObmQqvVVncZVMd06dKlukuock2aNEGTJk2quwyqQtxjUsdZWlrCwsICKpXKaPn//vc/vP3223BxcYGFhQWaNWuGf/7zn8jPzwcA3L9/H+3bt4enpycyMzMN/a5fvw4nJycEBQWhoKAAwINd0tbW1jh9+jSCg4NhZWWFhg0bYvz48cjNzS23xkuXLmH48OFwdHSEWq1Gy5YtsXz5chQWFgIALl68iIYNGwIA5s+fbzhUVd4u3dOnT6NXr17QarVo2LAhxo0bh++++w4KhcJoz0tQUBB8fHxw5MgRPP3009BqtXjttdcqVBsAxMTEFFtnUd0KhQLr1683LDPltRJC4PPPP0e7du2g0WhQv359DBkyBH/88UexdkuWLIGbmxssLS3RoUMH7N27t9zX/VFffPEFWrRoAbVajVatWmHLli1G26JUKrFo0aJi/Y4cOQKFQoFt27aVuf6srCxMmTIFHh4esLCwgIuLC959913k5OQYtdu2bRs6d+4MOzs7aLVaNGvWzDAfMTEx6NixIwBgzJgxhveCqYc/iuZm6dKlWLx4Mdzd3aHRaBAUFITz589Dp9Ph/fffR+PGjWFnZ4fnn38eGRkZxdazdetWBAQEwMrKCtbW1ujduzeOHz9erN369evh5eVleA99/fXXJdb16LbcvHkTb7/9Nlq1agVra2s4OjqiR48eiIuLK3F7li1bhhUrVsDDwwPW1tYICAhAQkKCSa/No+tcsmQJPvzwQzRt2hSWlpbw9/fHDz/8UKF1fPXVV2jbti0sLS3RoEEDPP/88zh79qxRm5IO5eh0OkybNg1OTk7QarV45plncOzYsWL1Pe57kqqJoDohPDxcABAJCQlCp9OJv/76S1y+fFlMnDhRmJmZiX379hna5uXliTZt2ggrKyuxbNkyceDAATF79myhVCpFv379DO3Onz8vbGxsRGhoqBBCiIKCAtGjRw/h6Ogorl27Zmg3atQoYWFhIZo2bSo+/PBDceDAATFv3jyhVCrFgAEDjOp0c3MTo0aNMjzOyMgQLi4uomHDhmL16tVi3759Yvz48QKACAsLE0IIcf/+fbFv3z4BQLz++usiPj5exMfHiwsXLpT6ely7dk3Y29uLpk2bivXr14s9e/aIESNGCHd3dwFAHD582NA2MDBQNGjQQLi6uor//Oc/4vDhwyI2NrZCtQkhxOHDh4utUwgh0tLSBAARHh5eqdfqjTfeECqVSkyePFns27dPbNq0SXh7e4tGjRqJ69evG9rNnTvX8Nrs3btXrFmzRri4uAgnJycRGBhY6mtUBIBwdXUVrVq1Eps3bxa7du0Sffr0EQDEtm3bDO2ef/550bRpU6HX6436v/jii6Jx48ZCp9OVOkZOTo5o166dcHBwECtWrBDff/+9+Pjjj4WdnZ3o0aOHKCwsFEIIcfToUaFQKMQrr7wi9uzZIw4dOiTCw8PFiBEjhBBCZGZmGt7rs2bNMrwXLl++XO52Pqxobtzc3MTAgQNFdHS0+Oabb0SjRo1EixYtxIgRI8Rrr70m9u7dK1avXi2sra3FwIEDjdbx4YcfCoVCIV577TURHR0tIiMjRUBAgLCyshKnT582tCuqd9CgQWL37t3im2++EZ6ensLV1VW4ubkVm4u5c+caHqekpIiwsDCxZcsWERMTI6Kjo8Xrr78uzMzMjN5vRdvj7u4u+vTpI6KiokRUVJTw9fUV9evXF3fv3jXp9Xl4na6uruKZZ54R27dvF9u2bRMdO3YUKpVKHD16tNg2pqWlGZYtXLhQABBDhw4V3333nfj6669Fs2bNhJ2dnTh//ryhXdH792GjRo0SCoVCTJ06VRw4cECsWLFCuLi4CFtbW6PPj8d5T1L1YTCpI4o+GB79UavV4vPPPzdqu3r1agFAfPvtt0bLFy9eLACIAwcOGJZt3bpVABArV64Uc+bMEWZmZkbPC/HgQwSA+Pjjj42Wf/jhhwKA+PHHHw3LHg0m77//vgAgfv75Z6O+YWFhQqFQiHPnzgkhhLh582axD+2yTJ06VSgUCqM/EEII0bt37xKDCQDxww8/GLWtaG2mBpOKvFbx8fECgFi+fLlRu8uXLwuNRiOmTZsmhBDizp07wtLSUjz//PNG7X766ScBoMLBRKPRGIUdvV4vvL29haenp2FZ0Xbu2LHDsOzq1atCqVSK+fPnlznGokWLhJmZmUhMTDRaHhERIQCIPXv2CCGEWLZsmQBQ5h/SxMTEYq+rqYrmpm3btqKgoMCwfOXKlQKAeO6554zav/vuuwKAyMzMFEIIcenSJaFUKsWECROM2t27d084OTmJl156SQjxIMw3btxYdOjQwRC+hBDi4sWLQqVSlRtMHqXX64VOpxPBwcFGc160Pb6+vkZ/pI8dOyYAiM2bN1fshXlI0TobN24s8vLyDMuzsrJEgwYNRM+ePQ3LHg0md+7cERqNxugfOkI8eN3UarUYNmyYYdmjweTs2bMCgHjvvfeM+m7cuFEAMPr8eJz3JFUfHsqpY77++mskJiYiMTERe/fuxahRozBu3Dh8+umnhjaHDh2ClZUVhgwZYtS36NDIw7tpX3rpJYSFhWHq1KlYsGABZs6ciZCQkBLHfvXVV40eDxs2DABw+PDhUus9dOgQWrVqhU6dOhWrRQiBQ4cOlb/RJYiNjYWPjw9atWpltHzo0KEltq9fvz569OhRJbUB5b9W0dHRUCgUGD58OPR6veHHyckJbdu2NRw2io+Px/3794ut7+mnn4abm1uF6wkODkajRo0Mj83NzfHyyy/jwoULuHLlCoAHh7zatm2Lzz77zNBu9erVUCgUePPNN8tcf3R0NHx8fNCuXTuj7endu7fRYbCiwzQvvfQSvv32W1y9erXC21AZ/fr1g5nZ/39MtmzZEgDQv39/o3ZFyy9dugQA2L9/P/R6PUaOHGm0PZaWlggMDDRsz7lz53Dt2jUMGzbM6HCFm5sbnn766QrVuHr1anTo0AGWlpZQKpVQqVT44Ycfih0SKarb3Nzc8LhNmzYAgD///LNCY5UkNDQUlpaWhsc2NjYYOHAgjhw5Yjic+6j4+Hjk5eUVO9zq6uqKHj16lHkoqOh34NH39EsvvQSl0vi0ycd5T1L1YTCpY1q2bAl/f3/4+/ujT58++OKLL9CrVy9MmzYNd+/eBQDcvn0bTk5OxY7rOjo6QqlU4vbt20bLX3vtNeh0OiiVSkycOLHEcZVKJezt7Y2WOTk5GcYrze3bt+Hs7FxseePGjcvtW5bbt28b/aEtUtIyACXWIFVtFXmtbty4ASEEGjVqBJVKZfSTkJCAW7duGbUv6l/SOiuirP4Pb+fEiRPxww8/4Ny5c9DpdFi7di2GDBlS7lg3btzAiRMnim2LjY0NhBCG7Xn22WcRFRVl+KPfpEkT+Pj4YPPmzRXeFlM0aNDA6LGFhUWZy+/fv2/YHuBBkHp0m7Zu3frE5mfFihUICwtD586dsX37diQkJCAxMRF9+vRBXl5esfaPvq/UajUAlNi2okqr/a+//kJ2dnaJfYq2u7Tfn/I+E0oat6TfG6Dy70mqPrwqh9CmTRvs378f58+fR6dOnWBvb4+ff/4ZQgijcJKRkQG9Xg8HBwfDspycHIwYMQItWrTAjRs3MHbsWOzcubPYGHq9Hrdv3zb64Lh+/TqA4h+WD7O3t0d6enqx5deuXQMAo1pMYW9vb/jj8bCimh5V0vcoVLS2on9NFp04XKToj9OjKvJaOTg4QKFQIC4uzvDH5WFFy4ral7Rd169fL/Y9GaUprf/DYwAP9uxMnz4dn332Gbp06YLr169j3Lhx5a7fwcEBGo0GX331VanPFxk0aBAGDRqE/Px8JCQkYNGiRRg2bBjc3d0REBBQoe2RWlG9ERERZe6ZKm9+yvPNN98gKCgIq1atMlp+7949U8p9LKXVbmFhAWtr6xL7FG13ab8/Zf1eP/yaubi4GJYX/d48qrLvSao+3GNChu97KLqyJTg4GNnZ2YiKijJqV3SlQHBwsGHZW2+9hUuXLiEyMhJffvkldu3ahY8++qjEcTZu3Gj0eNOmTQBQ5pd8BQcH48yZM/jll1+K1aJQKNC9e3cApv/LLzAwEKdOncKZM2eMlj98pUl5Klpb0R//EydOGLXbtWtXqesu77UaMGAAhBC4evWqYQ/Ywz++vr4AHlxeamlpWWx9R48eNWn3/Q8//GAU5AoKCrB161Y0b97c6FJOS0tLvPnmm9iwYQNWrFiBdu3aoWvXruWuf8CAAUhNTYW9vX2J21NSgFKr1QgMDMTixYsBwHC1y5PYC/C4evfuDaVSidTU1BK3p+iSfS8vLzg7O2Pz5s0QQhj6//nnnzh69Gi54ygUimLB9MSJE1X6vR+RkZGGPUXAg1C0e/dudOvWzeiw0cMCAgKg0WjwzTffGC2/cuUKDh06ZPQZ86ii34FH39Pffvst9Hp9sfaVfU9SNarG81uoChWdfBYeHm64UiE6Olq89tprAoDRiXJFV+XY2NiIFStWiIMHD4q5c+cKlUpldLLa2rVri51kOH78eKFSqYxOCC3rSpO+ffsa1VnaVTlOTk5izZo1Yv/+/WLixIlCoVCIt99+u1hfLy8vsX//fpGYmGh0BcCjrl69anRVzt69e8WIESOEm5ubACBiY2MNbQMDA0Xr1q2LrcOU2nr27Cnq168v1q5dKw4cOCCmT58unnrqKZOuynn0tXrzzTeFVqsVU6dOFbt37xaHDh0SGzduFGFhYUYnNM+aNctwVc6+ffvE2rVrn9hVOVu2bCnW/sqVK0KpVAoAYt26deWuXwghsrOzRfv27UWTJk3E8uXLxcGDB8X+/fvF2rVrxYsvvigSEhKEEELMnj1bjBkzRnzzzTciJiZGREVFie7duwuVSiVOnTolhHhwhY9GoxFdu3YVhw8fFomJieLq1atCCCE2bNggzM3NxYYNG8qsp+jEzqVLlxotLzqZ8uGrkYT4/9+vh0/eXbhwoVAqleIf//iH2LFjh4iJiRFbt24VkydPFnPmzDG0W7duneGqnKKrfyp6Vc6cOXOEQqEQc+bMET/88IP4/PPPhZOTk2jevLlR39K2p6R1mvoaFV2VExkZKSIiIkTHjh2FUqk0Oqm9rKtyRowYIfbs2SP++9//Ck9PzwpdlTN8+HChUCjEtGnTDFflNG7cuNhVOUUq856k6sNgUkeUdFWOnZ2daNeunVixYoW4f/++Ufvbt2+Lt956Szg7OwulUinc3NzEjBkzDO1OnDghNBpNsQ+B+/fvCz8/P+Hu7i7u3LkjhHjwx9bKykqcOHFCBAUFCY1GIxo0aCDCwsJEdna2Uf9Hg4kQQvz5559i2LBhwt7eXqhUKuHl5SWWLl1qdLWEEEJ8//33on379kKtVhc7O78kp06dEj179hSWlpaiQYMG4vXXXxcbNmwQAMRvv/1maFdaMDGltvT0dDFkyBDRoEEDYWdnJ4YPHy6SkpJKDCYVfa2EEOKrr74SnTt3FlZWVkKj0YjmzZuLkSNHiqSkJEObwsJCsWjRIuHq6iosLCxEmzZtxO7du0VgYGCFg8m4cePE559/Lpo3by5UKpXw9vYWGzduLLVPUFCQaNCggcjNzS13/UWys7PFrFmzhJeXl7CwsBB2dnbC19dXvPfee4YrgqKjo0Xfvn2Fi4uLsLCwEI6OjqJfv34iLi7OaF2bN28W3t7eQqVSGf3hfTigl+VJBBMhhCE42draCrVaLdzc3MSQIUPE999/b9Ru3bp14qmnnhIWFhaiRYsW4quvvhKjRo0qN5jk5+eLKVOmCBcXF2FpaSk6dOggoqKiivU1JZiY+hotXrxYzJ8/XzRp0kRYWFiI9u3bi/3795f4+jz6j4V169aJNm3aGOZ70KBBxa6UKymY5Ofni8mTJwtHR0dhaWkpunTpIuLj40v8/ChSmfckVQ+FEA/tPySSwOjRoxEREVHqiXBy8uabb2Lz5s24ffu24YTGqlSTXqvSZGRkwM3NDRMmTMCSJUuquxySyMWLF+Hh4YGlS5diypQp1V1OmfierFl48ivVWR988AEaN26MZs2aITs7G9HR0Vi3bh1mzZpVLaGkprty5Qr++OMPLF26FGZmZnjnnXequySq4/ierJkYTKjOUqlUWLp0Ka5cuQK9Xo+nnnoKK1as4IdXJa1btw4ffPAB3N3dsXHjRqMrJoiqA9+TNRMP5RAREZFs8HJhIiIikg0GEyIiIpINBhOiarR+/XooFArDj1KpRJMmTTBmzBij+8DExMRAoVAgIiKi2DpOnDiBMWPGwMPDA5aWlrC2tkaHDh2wZMkS/O9//zO0CwoKgkKhQJ8+fYqto+gW9suWLXuiY9YFQUFBZX5JIBGZhie/EslAeHg4vL29kZeXhyNHjmDRokWIjY3FyZMnYWVlVWq/tWvX4u2334aXlxemTp2KVq1aQafTISkpCatXr0Z8fDx27Nhh1Gf//v04dOhQsZsSVlRlxiQiqigGEyIZ8PHxMXxNeffu3VFQUIB//etfiIqKKnYX1SLx8fEICwtDSEgIoqKijL6aPCQkBJMnT8a+ffuM+rRo0QJ6vR7Tpk1DYmJiifcAKktlxiQiMgUP5RDJUJcuXQCUfTv6hQsXQqFQYM2aNSXeyM/CwgLPPfec0TKVSoUPP/wQycnJ2Lp1q8l1VWbMRyUlJeGVV16Bu7s7NBoN3N3dMXToUKNt/e2336BQKPDll18W6793714oFAqjew3t3LkTbdq0gVqtRrNmzfDxxx9j3rx5JgevIkV9jx8/jtDQUNja2sLOzg7Dhw/HzZs3K7VOIqoYBhMiGbpw4QKA/7+x4qMKCgpw6NAh+Pn5wdXV1aR1v/zyy/Dz88OsWbOg0+kq3O9xxnzYxYsX4eXlhZUrV2L//v1YvHgx0tPT0bFjR8Mdl9u2bYv27dsjPDy8WP/169fD0dER/fr1AwDs27cPoaGhsLe3x9atW7FkyRJs3rwZGzZsqHSNRZ5//nl4enoiIiIC8+bNQ1RUFHr37m3S60ZEpuGhHCIZKCgogF6vx/379xEbG4sFCxbAxsam1L0Pt27dQm5uLjw8PEweS6FQYPHixejZsye++OILjB8/vkL9HmfMhw0ZMgRDhgwxPC4oKMCAAQPQqFEjbNq0CRMnTgQAjBkzBhMnTsT58+fRokULAMCdO3ewc+dOjB8/Hkrlg4+vOXPmwMXFBfv37zd8Y2+fPn1KvCOxqUJDQw1fYd6rVy80atQIr776Kr799ttSD7ER0ePhHhMiGejSpQtUKhVsbGwwYMAAODk5Ye/evWjUqJEk4wUHB6NXr1744IMPcO/ePUnGKE12djamT58OT09PKJVKKJVKWFtbIycnB2fPnjW0e/XVV6FWq7F+/XrDss2bNyM/Px9jxowBAOTk5CApKQmDBw82uo2AtbU1Bg4c+Ni1Pho+XnrpJSiVShw+fPix101EJWMwIZKBr7/+GomJiTh+/DiuXbuGEydOoGvXrqW2d3BwgFarRVpaWqXHXLx4MW7dumV0iXBZnsSYADBs2DB8+umnGDt2LPbv349jx44hMTERDRs2RF5enqFdgwYN8Nxzz+Hrr79GQUEBgAeHcTp16oTWrVsDeLAHRQhRYoB7EqHOycnJ6LFSqYS9vT1u37792OsmopIxmBDJQMuWLeHv74927drB2dm53Pbm5uYIDg5GcnIyrly5Uqkx27Vrh6FDh2LFihW4ceNGlYyZmZmJ6OhoTJs2De+//z6Cg4PRsWNH+Pr6lvj9J0Xf53Lw4EGcOXMGiYmJhr0lAFC/fn0oFIoS679+/XqlaixrHXq9Hrdv34a9vf1jr5uISsZgQlRDzZgxA0IIvPHGG/jrr7+KPa/T6bB79+4y17FgwQL89ddfmD9/fpWMqVAoIIQodkXPunXrDHtFHtarVy+4uLggPDwc4eHhsLS0xNChQw3PW1lZwd/fH1FRUUb1FN0t+nFt3LjR6PG3334LvV7PL1QjkhBPfiWqoQICArBq1Sq8/fbb8PPzQ1hYGFq3bg2dTofjx49jzZo18PHxKfNcCw8PD4SFheHjjz+ukjFtbW3x7LPPYunSpXBwcIC7uztiY2Px5Zdfol69esXam5ubY+TIkVixYgVsbW0RGhoKOzs7ozYffPAB+vfvj969e+Odd95BQUEBli5dCmtr62J7YYKDgxEbGwu9Xl+h7Y2MjIRSqURISAhOnz6N2bNno23btnjppZfK7Ofp6Qng/6+uIqKK4x4TohrsjTfeQFJSEvz8/LB48WL06tULgwcPxubNmzFs2DCsWbOm3HXMmjULtra2VTbmpk2b0L17d0ybNg2hoaFISkrCwYMHiwWOImPGjEF+fj5u3rxpdBinSJ8+fbB9+3bcvn0bL7/8MiZNmoTnn38egwYNKhZ2CgoKStwzU5rIyEikpKQgNDQUc+bMwcCBA3HgwAGjE21LotfrKxx+iMiYQgghqrsIIqInSafToV27dnBxccGBAwdM7j9v3jzMnz8fN2/ehIODgwQVElFpeCiHiGq8119/HSEhIXB2dsb169exevVqnD17tsKHqIhIPhhMiKjGu3fvHqZMmYKbN29CpVKhQ4cO2LNnD3r27FndpRGRiXgoh4iIiGSDJ78SERGRbDCYEBERkWwwmBAREZFsVPnJr4WFhbh27RpsbGygUCiqengiIiKqBkII3Lt3D40bN4aZWen7Rao8mFy7dg2urq5VPSwRERHJwOXLl9GkSZNSn6/yYGJjYwPgQWGmfNukHOl0Ohw4cAC9evWCSqWq7nLob5wX+eLcyBPnRZ5q27xkZWXB1dXVkANKU+XBpOjwja2tba0IJlqtFra2trXiTVNbcF7ki3MjT5wXeaqt81LeaRw8+ZWIiIhkg8GEiIiIZIPBhIiIiGTD5GBy9epVDB8+HPb29tBqtWjXrh2Sk5OlqI2IiIjqGJNOfr1z5w66du2K7t27Y+/evXB0dERqairq1asnUXlERERUl5gUTBYvXgxXV1eEh4cblrm7uz/pmoiIiKiOMimY7Nq1C71798aLL76I2NhYuLi44O2338Ybb7xRap/8/Hzk5+cbHmdlZQF4cBmUTqerZNnyUFR/Td+O2obzIl+cG3nivMhTbZuXim6HQgghKrpSS0tLAMCkSZPw4osv4tixY3j33XfxxRdfYOTIkSX2mTdvHubPn19s+aZNm6DVais6NBEREdVgubm5GDZsGDIzM8v8HjOTgomFhQX8/f1x9OhRw7KJEyciMTER8fHxJfYpaY+Jq6srbt26VSu+YO3gwYMICQmpVV9+U9NxXuSLcyNPnBd5qm3zkpWVBQcHh3KDiUmHcpydndGqVSujZS1btsT27dtL7aNWq6FWq4stV6lUteKFBmrXttQmnBf54tzIE+dFnmrLvFR0G0y6XLhr1644d+6c0bLz58/Dzc3NlNUQERERlcikYPLee+8hISEBCxcuxIULF7Bp0yasWbMG48aNk6o+IiIiqkNMOpTTsWNH7NixAzNmzMAHH3wADw8PrFy5Eq+++qpU9REREclS2q0c5OTrK9w+Ly8XaRfOV7h9QUEBfj2ZilxlHMzNzU2qzcOzBTSail9gYqVWwsPByqQxpGLy3YUHDBiAAQMGSFELERFRjZB2Kwc9Vu6CQnmvwn3+unUZt6OXSVjV/7MfMAUWDq4Vbi/0Njj07nOyCCcmBxMiIqK6LidfD1W9n6Fu+EOF+1h5APU7ekpY1cOiTGqdfzMYOfn9pCnFRAwmRERElaC72xnL+w9Dc0drSdav1+vx048/oeszXaFUSvfnOjUjGxM3pkq2flMxmBAREVWC0NvCw9YLreztJFm/TqdDmjINLRu0lPRy4cL7mRD6m5Kt31Qm312YiIiISCoMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkG7yJHxERkYnydAUAgFNXMyUbIycvH0k3Aac/78BKo5ZsnAsZ2ZKtuzIYTIiIiEyU+vcf8/cjT0o8khL/vZAo8RgPWKnlEQnkUQUREVEN0qu1EwCguaM1NCpzScY4l56JyREnsXyIL7yc7SQZo4iVWgkPBytJx6goBhMiIiITNbCywCudmko6hl6vBwA0b2gFHxdpg4mc8ORXIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg3eXZiIiKgK5ObmIiUlpcLtz6XfRf71Czh7SoPC2/VMGsvb2xtardbECuWBwYSIiKgKpKSkwM/Pz+R+wzaYPlZycjI6dOhgekcZYDAhIiKqAt7e3khOTq5w++y8fHx3OB79uwfAWqM2eayaisGEiIioCmi1WpP2Yuh0Oty5lYGATv5QqVQSViYvPPmViIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkgzfxIyKqRXJzc5GSkmJSn+y8fBw9mYr6DkmVuoutVqs1qQ9RWRhMiIhqkZSUFPj5+VWq75JK9ElOTjbpjrlE5WEwISKqRby9vZGcnGxSn3PpdzFp20mseNEXXs71TB6P6EliMCEiqkW0Wq3JezDM/rwNdVweWvq0RTs3e4kqI6oYnvxKREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREsmFSMJk3bx4UCoXRj5OTk1S1ERERUR2jNLVD69at8f333xsem5ubP9GCiIiIqO4yOZgolUruJSEiIiJJmBxMfv/9dzRu3BhqtRqdO3fGwoUL0axZs1Lb5+fnIz8/3/A4KysLAKDT6aDT6SpRsnwU1V/Tt6O24bzIF+dGnvR6veG/nBv5qG2/LxXdDoUQQlR0pXv37kVubi5atGiBGzduYMGCBUhJScHp06dhb29fYp958+Zh/vz5xZZv2rQJWq22okMTEZFELmcDy04qMcVXD1fr6q6Gaqvc3FwMGzYMmZmZsLW1LbWdScHkUTk5OWjevDmmTZuGSZMmldimpD0mrq6uuHXrVpmF1QQ6nQ4HDx5ESEgIVCpVdZdDf+O8yBfnxnQXb+cgJ79A0jHOX8/EtB1nseT5lmjhZCfpWFZqc7jbW0k6Rm1R235fsrKy4ODgUG4wMflQzsOsrKzg6+uL33//vdQ2arUaarW62HKVSlUrXmigdm1LbcJ5kS/OTcWk3cpByMqfqmy8aTvOVsk4h6cEwcOB4aSiasvvS0W34bGCSX5+Ps6ePYtu3bo9zmqIiKgEOfkPzv1Y+XI7eDpKd4wlJy8f0THxGBAUACtN8X9IPikXMrLx7tZfDdtFVBKTgsmUKVMwcOBANG3aFBkZGViwYAGysrIwatQoqeojIqrzPB2t4eMi3SEWnU6H6w2BDm71a8W/zKlmMymYXLlyBUOHDsWtW7fQsGFDdOnSBQkJCXBzc5OqPiIiIqpDTAomW7ZskaoOIiIiIt4rh4iIiOSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiKpEQUEBYmNjceTIEcTGxqKgQNqvWSeimonBhIgkFxkZCU9PT4SEhGDFihUICQmBp6cnIiMjq7s0IpIZBhMiklRkZCSGDBkCX19fxMXFYfPmzYiLi4Ovry+GDBnCcEJERhhMiEgyBQUFmDx5MgYMGICoqCh07twZGo0GnTt3RlRUFAYMGIApU6bwsA4RGTCYEJFk4uLicPHiRcycORNmZsYfN2ZmZpgxYwbS0tIQFxdXTRUSkdwwmBCRZNLT0wEAPj4+JT5ftLyoHRGRSffKISIyhbOzMwDg1KlT6NKlS7HnT506ZdSOilMos5CWdQ5mltaSjaHX63FNfw1n/3cWSqV0fxbSsrKhUGZJtn6qHRhMiEgy3bp1g7u7OxYuXIioqCij5woLC7Fo0SJ4eHigW7du1VNgDaCq9zNmHltYJWN9vu9zycdQ1QsG0E/ycajmYjAhIsmYm5tj+fLlGDJkCAYPHoypU6ciLy8PCQkJWLp0KaKjoxEREQFzc/PqLlW2dHc7Y3n/YWjuKO0ek59+/Aldn+kq6R6T1IxsTNyYKtn6qXZgMCEiSYWGhiIiIgKTJ0/Gs88+a1ju4eGBiIgIhIaGVmN18if0tvCw9UIrezvJxtDpdEhTpqFlg5ZQqVSSjVN4PxNCf1Oy9VPtwGBCRJILDQ3FgAED8J///AeHDh1Cjx49MGHCBFhYWFR3aUQkM7wqh4gkFxkZCS8vL0yZMgV79uzBlClT4OXlxS9XI6JiGEyISFJF3/x648YNo+U3btzgN78SUTEMJkQkmYKCAoSFhUEIgeDgYKOvpA8ODoYQAmFhYfzmVyIyYDAhIsnExMQgIyMDzzzzDHbu3Gn0lfQ7d+5E165dkZGRgZiYmOoulYhkgie/EpFkigLH/PnzYWZmZrRnxMzMDPPmzUNISAhiYmIQHBxcTVXKV57uwet16mqmpOPk5OUj6Sbg9OcdWGnUko1zISNbsnVT7cFgQkQkU6l//yF/P/JkFYymxH8vJFbBOICVmn96qHR8dxCRZIKCgrBgwQLMnTsX3bp1Q2xsLI4cOQIrKysEBgZi/vz5hnZUXK/WTgCA5o7W0Kik+xK6c+mZmBxxEsuH+MLLWbrvSwEehBIPBytJx6CajcGEiCQTFBSEhg0b4scff4SdnR3y8vIAACtWrIBGo0FeXh4cHR0ZTErRwMoCr3RqKvk4er0eANC8oRV8XKQNJkTl4cmvRCQZc3NzjB49GgCQn59v9Nxff/0FABg1ahS/kp6IDBhMiEgyBQUF2LZtG/z9/dGkSROj55o0aQJ/f39ERETwcmEiMmAwISLJxMXF4eLFi/jPf/6DP/74AwcPHsSkSZNw8OBBpKam4pNPPkFaWhri4uKqu1QikgmeY0JEkklPTwcA+Pj4wNzcHIGBgcjJyUFgYCDMzc3h4+Nj1I6IiHtMiEgyzs7OAIBTp06V+HzR8qJ2REQMJkQkmW7dusHd3R0LFy5EYWGh0XOFhYVYtGgRPDw80K1bt2qqkIjkhsGEiCRjbm6O5cuXIzo6GoMHD0ZCQgLy8vKQkJCAwYMHIzo6GsuWLeNVOURkwHNMiEhSoaGhiIiIwOTJk/Hss88alnt4eCAiIgKhoaHVWB0RyQ2DCRFJLjQ0FIMGDcLhw4exd+9e9O3bF927d+eeEiIqhsGEiKpESVflEBE9iueYEBERkWxwjwkRVUpubi5SUlJM6pOdl4+jJ1NR3yEJ1hp1hft5e3tDq9WaWiIR1UAMJkRUKSkpKfDz86tU3yUmtk9OTkaHDh0qNVZdU5nAeC79LvKvX8DZUxoU3q5nUl+GRnrSGEyIqFK8vb2RnJxsUp9z6XcxadtJrHjRF17O9UwaiyrmcQLjsA2m92FopCeNwYSIKkWr1Zr8B8nsz9tQx+WhpU9btHOzl6iyuq0ygTE7Lx/fHY5H/+4BJh1iKxqP6EliMCEiqkUqExh1Oh3u3MpAQCd/qFQqiSojqhhelUNERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESy8VjBZNGiRVAoFHj33XefUDlERERUl1U6mCQmJmLNmjVo06bNk6yHiIiI6rBKBZPs7Gy8+uqrWLt2LerXr/+kayIiIqI6SlmZTuPGjUP//v3Rs2dPLFiwoMy2+fn5yM/PNzzOysoCAOh0Ouh0usoMLxtF9df07ahtOC/ypdfrDf/l/MgHf2fkqbbNS0W3w+RgsmXLFvzyyy9ITEysUPtFixZh/vz5xZYfOHAAWq3W1OFl6eDBg9VdApWA8yI/l7MBQImEhARcPVXd1dCj+DsjT7VlXnJzcyvUzqRgcvnyZbzzzjs4cOAALC0tK9RnxowZmDRpkuFxVlYWXF1d0atXL9ja2poyvOzodDocPHgQISEhUKlU1V0O/Y3zIl+/XfofcDIJXbp0QdumDaq7HPobf2fkqbbNS9ERk/KYFEySk5ORkZEBPz8/w7KCggIcOXIEn376KfLz82Fubm7UR61WQ61WF1uXSqWqFS80ULu2pTbhvMiPUqk0/JdzIz/8nZGn2jIvFd0Gk4JJcHAwTp48abRszJgx8Pb2xvTp04uFEiIiIiJTmBRMbGxs4OPjY7TMysoK9vb2xZYTUc2SdisHOfl6ScdIvZlj+G/R3hMpWKmV8HCwkmz9RCQd6T4ZiKjGSLuVg+7LYqpsvMkRJ8tv9JgOTwliOCGqgR47mMTExDyBMoioOhXtKVn5cjt4OlpLN05ePqJj4jEgKABWmuLnnj0JFzKy8e7WXyXf+0NE0uAeEyIy8HS0ho+LnWTr1+l0uN4Q6OBWv1aczEdETx5v4kdERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssF75RARAEChzEJa1jmYWUp3Ez+9Xo9r+ms4+7+zUCql+fhJy8qGQpklybqJSHoMJiRrubm5SElJMalPdl4+jp5MRX2HJFibeAdbb29vaLVak/rUFqp6P2PmsYVVMtbn+z6XdP2qesEA+kk6BhFJg8GEZC0lJQV+fn6V6rukEn2Sk5PRoUOHSo1X0+nudsby/sPQ3FHaPSY//fgTuj7TVbI9JqkZ2Zi4MVWSdROR9BhMSNa8vb2RnJxsUp9z6XcxadtJrHjRF17O9Uwer64Selt42Hqhlb2dZGPodDqkKdPQskFLqFQqScYovJ8Job8pybqJSHoMJiRrWq3W5D0YZn/ehjouDy192qKdm71ElRERkRR4VQ4RERHJBoMJERERyQaDCREREckGgwkRERHJBoMJERERyQaDCREREckGgwkRERHJBr/HhIiQpysAAJy6minpODl5+Ui6CTj9eQdWJt4uoKIuZGRLsl4iqhoMJkSE1L//mL8febIKRlPivxcSJR/FSs2PN6KaiL+5RIRerZ0AAM0draFRmUs2zrn0TEyOOInlQ3zh5SzdV99bqZXwcLCSbP1EJB0GEyJCAysLvNKpqeTj6PV6AEDzhlbwcZEumBBRzcWTX4mIiEg2GEyIiIhINhhMiIiISDYYTIiIiEg2GEyIiIhINhhMiIiISDYYTIiIiEg2GEwqqaCgALGxsThy5AhiY2NRUFBQ3SURERHVeAwmlRAZGQlPT0+EhIRgxYoVCAkJgaenJyIjI6u7NCIiohqNwcREkZGRGDJkCHx9fREXF4fNmzcjLi4Ovr6+GDJkCMMJERHRY2AwMUFBQQEmT56MAQMGICoqCp07d4ZGo0Hnzp0RFRWFAQMGYMqUKTysQ0REVEkMJiaIi4vDxYsXMXPmTJiZGb90ZmZmmDFjBtLS0hAXF1dNFRIREdVsDCYmSE9PBwD4+PiU+HzR8qJ2REREZBoGExM4OzsDAE6dOlXi80XLi9oRERGRaRhMTNCtWze4u7tj4cKFKCwsNHqusLAQixYtgoeHB7p161ZNFRIREdVsDCYmMDc3x/LlyxEdHY3BgwcjISEBeXl5SEhIwODBgxEdHY1ly5bB3Ny8ukslIiKqkZTVXUBNExoaioiICEyePBnPPvusYbmHhwciIiIQGhpajdURERHVbAwmlRAaGopBgwbh8OHD2Lt3L/r27Yvu3btzTwkREdFjYjCpJHNzcwQGBiInJweBgYEMJURERE8AzzEhIiIi2WAwISIiItlgMCEiIiLZYDAhIiIi2WAwISIiItlgMCEiIiLZYDAhIiIi2WAwISIiItngF6xRlUu7lYOcfL1k60+9mWP4r1Ip7VvcSq2Eh4OVpGMQEdUlDCZUpdJu5aD7spgqGWtyxMkqGefwlKA6GU5yc3ORkpJiUp9z6XeRf/0Czp7SoPB2vQr38/b2hlarNbFCIqqJGEyoShXtKVn5cjt4OlpLM0ZePqJj4jEgKABWGrUkYwDAhYxsvLv1V0n3/shZSkoK/Pz8KtV32AbT2icnJ6NDhw6VGouIahYGE6oWno7W8HGxk2TdOp0O1xsCHdzqQ6VSSTIGPdiLkZycbFKf7Lx8fHc4Hv27B8DahNDo7e1tanlEVEMxmBBRpWi1WpP3Yuh0Oty5lYGATv4MjURUIl6VQ0RERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESyYVIwWbVqFdq0aQNbW1vY2toiICAAe/fulao2IiIiqmNMCiZNmjTBv//9byQlJSEpKQk9evTAoEGDcPr0aanqIyIiojrEpO8xGThwoNHjDz/8EKtWrUJCQgJat279RAsjIiKiuqfSX7BWUFCAbdu2IScnBwEBAaW2y8/PR35+vuFxVlYWgAdftKTT6So7vCwU1V/Tt6Mq6fV6w3+let2qal6qYltqG/7OyBPnRZ5q27xUdDsUQghhyopPnjyJgIAA3L9/H9bW1ti0aRP69etXavt58+Zh/vz5xZZv2rSJN+Wqgy5nA8tOKjHFVw9XaW6VU2Vq07YQEUktNzcXw4YNQ2ZmJmxtbUttZ3Iw+euvv3Dp0iXcvXsX27dvx7p16xAbG4tWrVqV2L6kPSaurq64detWmYXVBDqdDgcPHkRISAi/XruCTl/LwuBVCYgK64LWjaWZ/6qal6rYltqGvzPyxHmRp9o2L1lZWXBwcCg3mJh8KMfCwgKenp4AAH9/fyQmJuLjjz/GF198UWJ7tVoNtbr4zbpUKlWteKGB2rUtUlMqlVAos3A5NxUWWdLsZtDr9bimv4YL9y5AqZTudlCXc7OhUGZBqVRy/k3E3xl54rzIU22Zl4puw2N/agshjPaIEJVHVe9nzDy2UPJxPt/3ueRjqOoFAyj9UCYREZnGpGAyc+ZM9O3bF66urrh37x62bNmCmJgY7Nu3T6r6qBbS3e2M5f2HobmjdHtMfvrxJ3R9pquke0xSM7IxcWOqZOsnIqqLTPrUvnHjBkaMGIH09HTY2dmhTZs22LdvH0JCQqSqj2ohobeFh60XWtnbSbJ+nU6HNGUaWjZoKenuz8L7mRD6m5Ktn4ioLjIpmHz55ZdS1UFERETEe+UQERGRfDCYEBERkWwwmBAREZFsMJgQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbEh3hzOiEuTpCgAAp65mSjZGTl4+km4CTn/egZVGLdk4FzKyJVs3EVFdxWBCVSr17z/m70eelHgkJf57IVHiMR6wUvPXiIjoSeEnKlWpXq2dAADNHa2hUZlLMsa59ExMjjiJ5UN84eUszR2Mi1iplfBwsJJ0DCKiuoTBhKpUAysLvNKpqaRj6PV6AEDzhlbwcZE2mBAR0ZPFk1+JiIhINhhMKqmgoACxsbE4cuQIYmNjUVBQUN0lERER1XgMJpUQGRkJDw8PhISEYMWKFQgJCYGHhwciIyOruzQiIqIajcHERJGRkXjhhRdw+fJlo+WXL1/GCy+8wHBCRET0GBhMTFBQUIDhw4eX2Wb48OE8rENERFRJDCYmOHDgAPLy8gAAFhYWmDZtGlatWoVp06bBwsICAJCXl4cDBw5UZ5lEREQ1FoOJCZYsWQIAMDc3x71797BgwQI4OztjwYIFuHfvHszNzY3aERERkWkYTExw6tQpAMDAgQMNe0iKWFhYoF+/fkbtiIiIyDT8grW/5ebmIiUlpcw2ZmYPclx8fDySkpKQm6/D0ZOpqO+QBK1ahWPHjhna/fLLL2Wuy9vbG1qt9skUT0REVEswmPwtJSUFfn5+FWp748YNdOzY0fD40QM3GRkZ5a4rOTkZHTp0MLVMIiKiWo3B5G/e3t5ITk4us01eXh6eeeaZctf1448/QqPRlDseERERGWMw+ZtWq63QHoxBgwZh586dZT7ftWvXJ1kaERFRncGTX00UFRWFQYMGlfjcoEGDEBUVVbUFERER1SIMJpUQFRWF3NxcvDTiNVi6t8dLI15Dbm4uQwkREdFj4qGcStJoNJjxryX4eVUCZoR1KfecEiIiIiof95gQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbDCYEBERkWzwC9ZI1nJzc5GSkmJSn3Ppd5F//QLOntKg8HY9k/p6e3tDq9Wa1IeIiJ4cBhOStZSUFPj5+VWq77ANpvdJTk6u0M0ciYhIGgwmJGve3t5ITk42qU92Xj6+OxyP/t0DYK1RmzweERFVHwYTkjWtVmvyHgydToc7tzIQ0MkfKpVKosqIiEgKPPmViIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZKPWfiV92q0c5OTrJR0j9WaO4b9KpXQvpZVaCQ8HK8nWT0REJBe1Mpik3cpB92UxVTbe5IiTko9xeEoQwwkREdV6tTKYFO0pWflyO3g6Wks3Tl4+omPiMSAoAFYm3sW2oi5kZOPdrb9KvveHiIhIDmplMCni6WgNHxc7ydav0+lwvSHQwa0+72JLRET0BPDkVyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINk76SftGiRYiMjERKSgo0Gg2efvppLF68GF5eXlLVV2kKZRbSss7BzFK6e+Xo9Xpc01/D2f+dlezuwmlZ2VAosyRZNxERkdyY9Nc0NjYW48aNQ8eOHaHX6/HPf/4TvXr1wpkzZ2BlJa8736rq/YyZxxZWyVif7/tc0vWr6gUD6CfpGERERHJgUjDZt2+f0ePw8HA4OjoiOTkZzz777BMt7HHp7nbG8v7D0FzCuwvr9Xr89ONP6PpMV8n2mKRmZGPixlRJ1k1ERCQ3j/XXNDMzEwDQoEGDUtvk5+cjPz/f8Dgr68FhCZ1OB51O9zjDl0qv10PobeGqbY6nbG0lGQN4sA1pyjR42nhKdnfhv7KzIPQ3odfrJXu9apui14mvl/xwbuSJ8yJPtW1eKrodCiGEqMwAQggMGjQId+7cQVxcXKnt5s2bh/nz5xdbvmnTJmi12soMXa7L2cCyk0pM8dXDVbodJlWiNm0LERHVXbm5uRg2bBgyMzNhW8ZOg0oHk3HjxuG7777Djz/+iCZNmpTarqQ9Jq6urrh161aZhT2O09eyMHhVAqLCuqB1Y2n3mBw8eBAhISGS7TGpqm2pTapiXqhyODfyxHmRp9o2L1lZWXBwcCg3mFTqUM6ECROwa9cuHDlypMxQAgBqtRpqtbrYcpVKJdkLrRMKAEDKjRzJzv0AgJy8fCTdBJyuZcNKU3wbn4SL/7sPAFAqlbXijVmVpHyP0ePh3MgT50Weasu8VHQbTPqrLYTAhAkTsGPHDsTExMDDw6NSxUktNSMbAPB+5MkqGE2J/15IlHwUK7V0AYuIiEguTPprN27cOGzatAk7d+6EjY0Nrl+/DgCws7ODRqORpMDK6NXaCQDQ3NEaGpW5ZOOcS8/E5IiTWD7EF17OdpKNY6VWwsNBXpdjExERScGkYLJq1SoAQFBQkNHy8PBwjB49+knV9NgaWFnglU5NJR9Hr9cDAJo3tIKPi3TBhIiIqK4w+VAOERERkVR4rxwiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINBhMiIiKSDQYTIiIikg0GEyIiIpINZXUXIBe5ublISUkxqc+59LvIv34BZ09pUHi7nkl9vb29odVqTepDRERU2zGY/C0lJQV+fn6V6jtsg+l9kpOT0aFDh0qNR0REVFsxmPzN29sbycnJJvXJzsvHd4fj0b97AKw1apPHIyIiImMMJn/TarUm78HQ6XS4cysDAZ38oVKpJKqMiIio7uDJr0RERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbDCZEREQkGwwmREREJBsMJkRERCQbVX53YSEEACArK6uqh37idDodcnNzkZWVxbsLywjnRb44N/LEeZGn2jYvRX/3i3JAaao8mNy7dw8A4OrqWtVDExERUTW7d+8e7OzsSn1eIcqLLk9YYWEhrl27BhsbGygUiqoc+onLysqCq6srLl++DFtb2+ouh/7GeZEvzo08cV7kqbbNixAC9+7dQ+PGjWFmVvqZJFW+x8TMzAxNmjSp6mElZWtrWyveNLUN50W+ODfyxHmRp9o0L2XtKSnCk1+JiIhINhhMiIiISDYYTB6DWq3G3LlzoVarq7sUegjnRb44N/LEeZGnujovVX7yKxEREVFpuMeEiIiIZIPBhIiIiGSDwYSIiIhkg8GEiIiIZIPBpBSLFi1Cx44dYWNjA0dHRwwePBjnzp0rt19sbCz8/PxgaWmJZs2aYfXq1VVQbd2xatUqtGnTxvCFQwEBAdi7d2+ZfTgnVW/RokVQKBR49913y2zHuZHevHnzoFAojH6cnJzK7MN5qRpXr17F8OHDYW9vD61Wi3bt2iE5ObnMPnVhbhhMShEbG4tx48YhISEBBw8ehF6vR69evZCTk1Nqn7S0NPTr1w/dunXD8ePHMXPmTEycOBHbt2+vwsprtyZNmuDf//43kpKSkJSUhB49emDQoEE4ffp0ie05J1UvMTERa9asQZs2bcpsx7mpOq1bt0Z6errh5+TJk6W25bxUjTt37qBr165QqVTYu3cvzpw5g+XLl6NevXql9qkzcyOoQjIyMgQAERsbW2qbadOmCW9vb6Nl//jHP0SXLl2kLq9Oq1+/vli3bl2Jz3FOqta9e/fEU089JQ4ePCgCAwPFO++8U2pbzk3VmDt3rmjbtm2F23Neqsb06dPFM888Y1KfujI33GNSQZmZmQCABg0alNomPj4evXr1MlrWu3dvJCUlQafTSVpfXVRQUIAtW7YgJycHAQEBJbbhnFStcePGoX///ujZs2e5bTk3Vef3339H48aN4eHhgVdeeQV//PFHqW05L1Vj165d8Pf3x4svvghHR0e0b98ea9euLbNPXZkbBpMKEEJg0qRJeOaZZ+Dj41Nqu+vXr6NRo0ZGyxo1agS9Xo9bt25JXWadcfLkSVhbW0OtVuOtt97Cjh070KpVqxLbck6qzpYtW/DLL79g0aJFFWrPuakanTt3xtdff439+/dj7dq1uH79Op5++mncvn27xPacl6rxxx9/YNWqVXjqqaewf/9+vPXWW5g4cSK+/vrrUvvUlbmp8rsL10Tjx4/HiRMn8OOPP5bbVqFQGD0Wf3+x7qPLqfK8vLzw66+/4u7du9i+fTtGjRqF2NjYUsMJ50R6ly9fxjvvvIMDBw7A0tKywv04N9Lr27ev4f99fX0REBCA5s2bY8OGDZg0aVKJfTgv0issLIS/vz8WLlwIAGjfvj1Onz6NVatWYeTIkaX2qwtzwz0m5ZgwYQJ27dqFw4cPo0mTJmW2dXJywvXr142WZWRkQKlUwt7eXsoy6xQLCwt4enrC398fixYtQtu2bfHxxx+X2JZzUjWSk5ORkZEBPz8/KJVKKJVKxMbG4pNPPoFSqURBQUGxPpyb6mFlZQVfX1/8/vvvJT7Peakazs7Oxf4x1bJlS1y6dKnUPnVlbrjHpBRCCEyYMAE7duxATEwMPDw8yu0TEBCA3bt3Gy07cOAA/P39oVKppCq1zhNCID8/v8TnOCdVIzg4uNiVHmPGjIG3tzemT58Oc3PzYn04N9UjPz8fZ8+eRbdu3Up8nvNSNbp27VrsKyjOnz8PNze3UvvUmbmpvvNu5S0sLEzY2dmJmJgYkZ6ebvjJzc01tHn//ffFiBEjDI//+OMPodVqxXvvvSfOnDkjvvzyS6FSqURERER1bEKtNGPGDHHkyBGRlpYmTpw4IWbOnCnMzMzEgQMHhBCcEzl59Koczk31mDx5soiJiRF//PGHSEhIEAMGDBA2Njbi4sWLQgjOS3U5duyYUCqV4sMPPxS///672Lhxo9BqteKbb74xtKmrc8NgUgoAJf6Eh4cb2owaNUoEBgYa9YuJiRHt27cXFhYWwt3dXaxatapqC6/lXnvtNeHm5iYsLCxEw4YNRXBwsCGUCME5kZNHgwnnpnq8/PLLwtnZWahUKtG4cWMRGhoqTp8+bXie81J9du/eLXx8fIRarRbe3t5izZo1Rs/X1blRCPH3mTNERERE1YwnvxIREZFsMJgQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbDCYEBERkWwwmBAREZFsMJgQERGRbDCYENET4+7ujpUrVxoeKxQKREVFVVs9pbl48SIUCgV+/fXXCvcJCgrCu+++W2abR7efiEzHm/gR1QAXL16Eh4cHjh8/jnbt2lV3ORWWnp6O+vXrV3cZT0RkZGTtulEakUwxmBCRZJycnKq7hCemQYMG1V0CUZ3AQzlEVUQIgSVLlqBZs2bQaDRo27YtIiIiDM/fuXMHr776Kho2bAiNRoOnnnoK4eHhAAAPDw8AQPv27aFQKBAUFFThcd3d3bFgwQKMHDkS1tbWcHNzw86dO3Hz5k0MGjQI1tbW8PX1RVJSklG/o0eP4tlnn4VGo4GrqysmTpyInJwcw/MZGRkYOHAgNBoNPDw8sHHjxmJjP3ooZ/r06WjRogW0Wi2aNWuG2bNnQ6fTGZ6fN28e2rVrh//+979wd3eHnZ0dXnnlFdy7d6/C21s07qpVq9C3b19Dfdu2bSuzT2xsLDp16gS1Wg1nZ2e8//770Ov1hucfPZRT3va/9tprGDBggNEyvV4PJycnfPXVVyZtD1FdwmBCVEVmzZqF8PBwrFq1CqdPn8Z7772H4cOHIzY2FgAwe/ZsnDlzBnv37sXZs2exatUqODg4AACOHTsGAPj++++Rnp6OyMhIk8b+6KOP0LVrVxw/fhz9+/fHiBEjMHLkSAwfPhy//PILPD09MXLkSBTd0/PkyZPo3bs3QkNDceLECWzduhU//vgjxo8fb1jn6NGjcfHiRRw6dAgRERH4/PPPkZGRUWYdNjY2WL9+Pc6cOYOPP/4Ya9euxUcffWTUJjU1FVFRUYiOjkZ0dDRiY2Px73//26TtBR68ni+88AJ+++03DB8+HEOHDsXZs2dLbHv16lX069cPHTt2xG+//YZVq1bhyy+/xIIFC0pdf3nbP3bsWOzbtw/p6emGZXv27EF2djZeeuklk7eHqM6o3psbE9UN2dnZwtLSUhw9etRo+euvvy6GDh0qhBBi4MCBYsyYMSX2T0tLEwDE8ePHTR7bzc1NDB8+3PA4PT1dABCzZ882LIuPjxcARHp6uhBCiBEjRog333zTaD1xcXHCzMxM5OXliXPnzgkAIiEhwfD82bNnBQDx0UcfGZYBEDt27Ci1tiVLlgg/Pz/D47lz5wqtViuysrIMy6ZOnSo6d+5s0jYDEG+99ZbRss6dO4uwsDAhRPHXc+bMmcLLy0sUFhYa2n/22WfC2tpaFBQUCCGECAwMFO+8844QQlR4+1u1aiUWL15seDx48GAxevRok7aFqK7hOSZEVeDMmTO4f/8+QkJCjJb/9ddfaN++PQAgLCwML7zwAn755Rf06tULgwcPxtNPP/1Exm/Tpo3h/xs1agQA8PX1LbYsIyMDTk5OSE5OxoULF4wOTwghUFhYiLS0NJw/fx5KpRL+/v6G5729vVGvXr0y64iIiMDKlStx4cIFZGdnQ6/Xw9bW1qiNu7s7bGxsDI+dnZ3L3RNTkoCAgGKPS7sK5+zZswgICIBCoTAs69q1K7Kzs3HlyhU0bdq0WPuKbP/YsWOxZs0aTJs2DRkZGfjuu+/www8/mLwtRHUJgwlRFSgsLAQAfPfdd3BxcTF6Tq1WAwD69u2LP//8E9999x2+//57BAcHY9y4cVi2bNljj//w1SRFf3xLWlZUZ2FhIf7xj39g4sSJxdbVtGlTnDt3zqhfRSQkJOCVV17B/Pnz0bt3b9jZ2WHLli1Yvnx5qbUWjVFU1+MqrV4hRLHnxN+HtUrqU9ZzDxs5ciTef/99xMfHIz4+Hu7u7ujWrVtlSieqMxhMiKpAq1atoFarcenSJQQGBpbarmHDhhg9ejRGjx6Nbt26YerUqVi2bBksLCwAAAUFBVVSb4cOHXD69Gl4enqW+HzLli2h1+uRlJSETp06AQDOnTuHu3fvlrrOn376CW5ubvjnP/9pWPbnn38+0boflpCQgJEjRxo9Lto79ahWrVph+/btRgHl6NGjsLGxKRYkgYpvv729PQYPHozw8HDEx8djzJgxT2jriGovnvxKVAVsbGwwZcoUvPfee9iwYQNSU1Nx/PhxfPbZZ9iwYQMAYM6cOdi5cycuXLiA06dPIzo6Gi1btgQAODo6QqPRYN++fbhx4wYyMzMBADt27IC3t/cTr3f69OmIj4/HuHHj8Ouvv+L333/Hrl27MGHCBACAl5cX+vTpgzfeeAM///wzkpOTMXbsWGg0mlLX6enpiUuXLmHLli1ITU3FJ598gh07dphc26efforg4OBy223btg1fffUVzp8/j7lz5+LYsWNGJ+8+7O2338bly5cxYcIEpKSkYOfOnZg7dy4mTZoEM7PiH5OmbP/YsWOxYcMGnD17FqNGjTJ5e4nqGgYToiryr3/9C3PmzMGiRYvQsmVL9O7dG7t37zZcCmxhYYEZM2agTZs2ePbZZ2Fubo4tW7YAAJRKJT755BN88cUXaNy4MQYNGgQAyMzMNBxWeZLatGmD2NhY/P777+jWrRvat2+P2bNnw9nZ2dAmPDwcrq6uCAwMRGhoKN588004OjqWus5Bgwbhvffew/jx49GuXTscPXoUs2fPNrm2W7duITU1tdx28+fPx5YtW9CmTRts2LABGzduRKtWrUps6+Ligj179uDYsWNo27Yt3nrrLbz++uuYNWtWqeuv6Pb37NkTzs7O6N27Nxo3blzxDSWqoxSi6GApEVEtoVAosGPHDgwePLi6S0Fubi4aN26Mr776CqGhodVdDpHs8RwTIiIJFBYW4vr161i+fDns7Ozw3HPPVXdJRDUCgwkRkQQuXboEDw8PNGnSBOvXr4dSyY9boorgoRwiIiKSDZ78SkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESywWBCREREssFgQkRERLLBYEJERESy8X9f84g8PMXvhQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# PICNIC average ploidy vs estimated copy-neutral \n",
    "tmp = df_ploidies.boxplot(column=\"PICNIC avg. pl.\", by = \"est. median. ploidy\" )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d2fdc84-4b91-4b13-b4c7-32aaeb1308bf",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Convert gene-level integer CN into log2R-like format in order to make it compatible with TCGA and CCLE\n",
    "\n",
    "1) Copy-neutral state was defined from average ploiy, as median of integer CN values in non-disrupted genes.\n",
    "\n",
    "2) Compute log2(CN/neutral-CN) for min and max CN; keep the value with most extreme estimate\n",
    "\n",
    "3) Replace estimates below thresholds with zeroes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9ce3ceb1-1a71-4ee7-a7fd-b8aaba851d3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3.0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "estimated_CN = est_ploidies[\"median_pl\"].to_dict()\n",
    "estimated_CN[1287381]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2cd2fa84-c594-499c-b28a-e451b17fed5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1287381</th>\n",
       "      <th>924100</th>\n",
       "      <th>910924</th>\n",
       "      <th>687561</th>\n",
       "      <th>1287706</th>\n",
       "      <th>687452</th>\n",
       "      <th>906798</th>\n",
       "      <th>906797</th>\n",
       "      <th>906800</th>\n",
       "      <th>910922</th>\n",
       "      <th>...</th>\n",
       "      <th>909785</th>\n",
       "      <th>909904</th>\n",
       "      <th>909905</th>\n",
       "      <th>687592</th>\n",
       "      <th>1303911</th>\n",
       "      <th>946358</th>\n",
       "      <th>909907</th>\n",
       "      <th>1298146</th>\n",
       "      <th>908452</th>\n",
       "      <th>908450</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OR4G11P</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-34P13.7</th>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               1287381  924100   910924    687561   1287706  687452   906798   \\\n",
       "OR4G11P       0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "OR4F5         0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "RP11-34P13.7  0.736966      0.0      0.0  0.321928      0.0      0.0      0.0   \n",
       "\n",
       "              906797   906800   910922   ...  909785    909904   909905   \\\n",
       "OR4G11P           0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "OR4F5             0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "RP11-34P13.7      0.0      0.0     -1.0  ...      0.0  0.415037      0.0   \n",
       "\n",
       "              687592    1303911   946358   909907    1298146   908452   \\\n",
       "OR4G11P           0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "OR4F5             0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "RP11-34P13.7      0.0 -0.584963  0.321928     -1.0 -0.584963  0.584963   \n",
       "\n",
       "              908450   \n",
       "OR4G11P           0.0  \n",
       "OR4F5             0.0  \n",
       "RP11-34P13.7      0.0  \n",
       "\n",
       "[3 rows x 996 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc = gdsc.apply(lambda x : CN2log2R(x,estimated_CN[x.name] ))\n",
    "# drop genes without any determined value\n",
    "gdsc = gdsc.dropna(axis=0,how=\"all\")\n",
    "# fill with zeroes the remaining ones\n",
    "gdsc.fillna(0,inplace=True)\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1ab61add-8438-464f-8963-f41cd7c7879c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_781/188133694.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  gdsc = gdsc.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))\n"
     ]
    }
   ],
   "source": [
    "gdsc = gdsc.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cd93f386-1501-4368-b36b-3c3858e29b1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Symbol rows are not empty.\n",
      "Ok: All Symbol are mapped to GeneID\n",
      "12 Symbol mapped to multiple GeneID\n",
      "Ok: All GeneID are unique\n",
      "140871 Symbol can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "NCBI = pd.read_csv(\"/Volumes/Expansion/Thesis Work/Supplementary Files/Homo_sapiens.gene_info/Homo_sapiens.gene_info.txt\",sep = \"\\t\")\n",
    "NCBI = NCBI[[\"#tax_id\",\"GeneID\",\"Symbol\",\"Synonyms\",\"type_of_gene\"]]\n",
    "NCBI = NCBI.loc[NCBI[\"#tax_id\"] == 9606]\n",
    "NCBI = NCBI.loc[NCBI[\"type_of_gene\"] != \"unknown\"]\n",
    "ncbi_symbols = parse_mapping_table(NCBI, \"Symbol\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "daf2ac34-849f-4551-9154-051c333e09bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Synonyms rows are not empty.\n",
      "Ok: All Synonyms are mapped to GeneID\n",
      "3376 Synonyms mapped to multiple GeneID\n",
      "53208 different Synonyms mapped to the same GeneID\n",
      "11014 Synonyms can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "ncbi_synonyms = expand(NCBI[[\"Synonyms\",\"GeneID\"]],column=\"Synonyms\",sep=\"|\") \n",
    "ncbi_synonyms = parse_mapping_table(ncbi_synonyms, \"Synonyms\",\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dd4906b6-a83a-4089-ba01-e3560dcf16bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 24508 \n",
      "\tdirectly via main_mapper 21819 \n",
      "\tvia alternative mapper 858 \n",
      "\tvia one of multiple synonyms in alternative mapper 1831 \n",
      "\tLOC 0 \n",
      "Unmapped: 21624 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 105 \n",
      "\tLOC not found in Entrez 0 \n",
      "\tNot found at all: 21519\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 181\n",
      "Warning: query IDs not mapped to any target IDs excluded: 21624\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " [143872, 414212, 414213, 284697, 341019, 51236, 414761, 442416, 728113, 26165, 6218, 221262, 100132948, 768096, 6241, 642658, 401508, 100289124, 728695, 125050, 387707, 84631, 4253, 643749, 54438, 3255, 100034743, 2749, 653505, 26824, 23254, 728798, 100462820, 119016, 26871, 286464, 51463, 642826, 653067, 100132115, 9502, 10529, 23334, 440107, 256815, 645425, 399687, 647507, 728917, 84316, 200030, 729438, 100302179, 106478953, 246126, 1903, 100033392, 92017, 374, 445815, 79741, 399761, 647060, 284565, 6551, 161176, 643479, 644509, 100418977, 29099, 9645, 10160, 653234, 644019, 158135, 114112, 79817, 23499, 463, 677844, 26583, 23523, 84458, 1517, 574445, 26095, 653308]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1287381</th>\n",
       "      <th>924100</th>\n",
       "      <th>910924</th>\n",
       "      <th>687561</th>\n",
       "      <th>1287706</th>\n",
       "      <th>687452</th>\n",
       "      <th>906798</th>\n",
       "      <th>906797</th>\n",
       "      <th>906800</th>\n",
       "      <th>910922</th>\n",
       "      <th>...</th>\n",
       "      <th>909785</th>\n",
       "      <th>909904</th>\n",
       "      <th>909905</th>\n",
       "      <th>687592</th>\n",
       "      <th>1303911</th>\n",
       "      <th>946358</th>\n",
       "      <th>909907</th>\n",
       "      <th>1298146</th>\n",
       "      <th>908452</th>\n",
       "      <th>908450</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.736966</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.736966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    1287381   924100   910924    687561    1287706   687452    906798   \\\n",
       "1  0.000000  0.000000      0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "2  0.000000  0.584963      0.0 -0.415037  0.321928  0.000000  0.584963   \n",
       "9 -0.584963  0.584963      0.0 -0.415037 -0.415037 -0.584963  0.000000   \n",
       "\n",
       "    906797   906800   910922   ...  909785    909904    909905    687592   \\\n",
       "1  0.584963      0.0      0.0  ...      0.0 -0.584963  0.584963  0.415037   \n",
       "2  0.584963      0.0      0.0  ...      0.0 -0.584963  0.000000  0.000000   \n",
       "9  0.000000      0.0      0.0  ...      0.0 -0.584963  0.000000  0.415037   \n",
       "\n",
       "    1303911   946358   909907    1298146   908452    908450   \n",
       "1  0.000000  0.000000      0.0  0.000000  0.584963  0.415037  \n",
       "2  0.736966  0.321928      0.0 -0.584963  0.000000  0.736966  \n",
       "9  0.000000  0.000000     -1.0 -0.584963  0.000000  0.000000  \n",
       "\n",
       "[3 rows x 996 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc,query2target,not_mapped  = apply_mappers(gdsc, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")\n",
    "gdsc.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "73218f7d-3ea3-45ff-97b3-ac3d3cdc6973",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87 duplicated IDs in 181 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 25\n",
      "Merged  156 duplicated rows into 75 rows\n"
     ]
    }
   ],
   "source": [
    "gdsc = handle_dups(gdsc,corr_thr = 0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "01ddcc07-31ee-481e-b9eb-16b1face2bca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>683665</th>\n",
       "      <th>683667</th>\n",
       "      <th>684052</th>\n",
       "      <th>684055</th>\n",
       "      <th>684057</th>\n",
       "      <th>684059</th>\n",
       "      <th>684062</th>\n",
       "      <th>684072</th>\n",
       "      <th>684681</th>\n",
       "      <th>687448</th>\n",
       "      <th>...</th>\n",
       "      <th>1659818</th>\n",
       "      <th>1659819</th>\n",
       "      <th>1659823</th>\n",
       "      <th>1659928</th>\n",
       "      <th>1659929</th>\n",
       "      <th>1660034</th>\n",
       "      <th>1660035</th>\n",
       "      <th>1660036</th>\n",
       "      <th>1674021</th>\n",
       "      <th>1789883</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.321928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>-0.415037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.584963</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>-1.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.415037</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.415037</td>\n",
       "      <td>0.584963</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 996 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         683665    683667   684052    684055    684057   684059    684062   \\\n",
       "gene_id                                                                      \n",
       "1            0.0  0.000000      0.0  0.000000 -0.415037      0.0 -0.415037   \n",
       "2            0.0  0.000000      0.0  0.584963  0.000000      0.0  0.000000   \n",
       "9            0.0  0.321928      0.0  0.584963  0.584963      0.0  0.321928   \n",
       "10           0.0  0.321928      0.0  0.584963  0.584963      0.0  0.321928   \n",
       "12           0.0 -1.000000      0.0  0.000000 -1.000000      0.0 -0.415037   \n",
       "\n",
       "          684072    684681   687448   ...  1659818  1659819   1659823  \\\n",
       "gene_id                               ...                               \n",
       "1        0.000000  0.415037      0.0  ...      0.0      0.0  0.000000   \n",
       "2        0.584963  0.000000      0.0  ...      0.0      0.0  0.000000   \n",
       "9        0.000000 -0.584963      0.0  ...      0.0      0.0 -0.415037   \n",
       "10       0.000000 -0.584963      0.0  ...      0.0      0.0 -0.415037   \n",
       "12       0.000000  0.000000      0.0  ...     -1.0      0.0  0.321928   \n",
       "\n",
       "         1659928   1659929  1660034   1660035   1660036   1674021   1789883  \n",
       "gene_id                                                                      \n",
       "1            0.0 -0.415037      0.0 -0.584963  0.000000  0.000000  0.321928  \n",
       "2            0.0  0.000000      0.0  0.000000  0.000000  0.584963 -0.415037  \n",
       "9            0.0  0.000000     -1.0 -1.584963  0.000000 -1.000000 -1.000000  \n",
       "10           0.0  0.000000     -1.0 -1.584963  0.000000 -1.000000 -1.000000  \n",
       "12           0.0 -0.415037      0.0  0.415037  0.584963  0.000000  0.000000  \n",
       "\n",
       "[5 rows x 996 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gdsc.index.name = \"gene_id\"\n",
    "gdsc = gdsc.T.sort_index().T\n",
    "gdsc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d93c1162-f61c-4ce4-b929-72da6f6e8269",
   "metadata": {},
   "outputs": [],
   "source": [
    "gdsc.to_csv(preprocessed_dir+\"/\"+\"GDSC\"+\".Segment_Mean.CNA.tsv\",\n",
    "                 sep = \"\\t\",header=True,index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65bf64c4-bc10-4e60-b6d8-36cae57307f8",
   "metadata": {
    "tags": []
   },
   "source": [
    "# PDX \n",
    "\n",
    "For PDX dataset only gene-level estimated copy-number (non-integer) reported. \n",
    "From ploidy distributions, calculated as average over all genes we concluded that CN estimates were called under assumption that copy-neutral state of each xenograft corresponds CN = 2.\n",
    "\n",
    "\n",
    "For gene ID conversion we used the same approach as for RNA-seq."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "2db009b4-fc5f-452e-b4b1-924ae9121425",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(23852, 375)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.58</td>\n",
       "      <td>1.60</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.00</td>\n",
       "      <td>3.94</td>\n",
       "      <td>2.04</td>\n",
       "      <td>11.39</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.01</td>\n",
       "      <td>...</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.14</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.99</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.43</td>\n",
       "      <td>2.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>2.58</td>\n",
       "      <td>1.60</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.00</td>\n",
       "      <td>3.94</td>\n",
       "      <td>2.04</td>\n",
       "      <td>11.39</td>\n",
       "      <td>2.17</td>\n",
       "      <td>2.01</td>\n",
       "      <td>...</td>\n",
       "      <td>2.08</td>\n",
       "      <td>2.10</td>\n",
       "      <td>2.14</td>\n",
       "      <td>2.95</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.99</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.43</td>\n",
       "      <td>2.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>2.87</td>\n",
       "      <td>2.97</td>\n",
       "      <td>2.01</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.10</td>\n",
       "      <td>1.58</td>\n",
       "      <td>2.01</td>\n",
       "      <td>1.64</td>\n",
       "      <td>1.89</td>\n",
       "      <td>1.99</td>\n",
       "      <td>...</td>\n",
       "      <td>2.04</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.58</td>\n",
       "      <td>2.08</td>\n",
       "      <td>1.95</td>\n",
       "      <td>1.92</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1.28</td>\n",
       "      <td>1.33</td>\n",
       "      <td>2.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>5.74</td>\n",
       "      <td>1.64</td>\n",
       "      <td>2.06</td>\n",
       "      <td>2.01</td>\n",
       "      <td>2.07</td>\n",
       "      <td>1.74</td>\n",
       "      <td>2.06</td>\n",
       "      <td>1.59</td>\n",
       "      <td>1.40</td>\n",
       "      <td>2.53</td>\n",
       "      <td>...</td>\n",
       "      <td>2.03</td>\n",
       "      <td>2.07</td>\n",
       "      <td>2.25</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.01</td>\n",
       "      <td>2.00</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.85</td>\n",
       "      <td>1.93</td>\n",
       "      <td>1.45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X-1004  X-1008  X-1027  X-1095  X-1119  X-1156  X-1167  X-1169  \\\n",
       "Sample                                                                     \n",
       "A1BG        2.58    1.60    2.17    2.08    2.00    3.94    2.04   11.39   \n",
       "A1BG-AS1    2.58    1.60    2.17    2.08    2.00    3.94    2.04   11.39   \n",
       "A1CF        2.87    2.97    2.01    2.06    2.10    1.58    2.01    1.64   \n",
       "A2LD1       5.74    1.64    2.06    2.01    2.07    1.74    2.06    1.59   \n",
       "\n",
       "          X-1172  X-1173  ...  X-5694  X-5696  X-5713  X-5717  X-5727  X-5739  \\\n",
       "Sample                    ...                                                   \n",
       "A1BG        2.17    2.01  ...    2.08    2.10    2.14    2.95    2.06    2.07   \n",
       "A1BG-AS1    2.17    2.01  ...    2.08    2.10    2.14    2.95    2.06    2.07   \n",
       "A1CF        1.89    1.99  ...    2.04    0.97    1.58    2.08    1.95    1.92   \n",
       "A2LD1       1.40    2.53  ...    2.03    2.07    2.25    2.00    1.01    2.00   \n",
       "\n",
       "          X-5808  X-5959  X-5975  X-6047  \n",
       "Sample                                    \n",
       "A1BG        1.99    2.07    1.43    2.03  \n",
       "A1BG-AS1    1.99    2.07    1.43    2.03  \n",
       "A1CF        1.54    1.28    1.33    2.10  \n",
       "A2LD1       1.08    1.85    1.93    1.45  \n",
       "\n",
       "[4 rows x 375 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PDX_xls = \"/Volumes/Expansion/Thesis Work/Datasets/PDX/CNA/nm.3954-S2.xlsx\"\n",
    "pdx = pd.read_excel(PDX_xls,\"copy number\")\n",
    "pdx.set_index(\"Sample\",drop=True,inplace=True)\n",
    "focal  = pdx.T[\"FocalCNScore\"]\n",
    "pdx.drop([\"ArmLevelCNScore\",\"FocalCNScore\"],inplace = True)\n",
    "print(pdx.shape)\n",
    "pdx.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e55e64f3-d8fa-46fd-95f1-3687097bc98a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Strings containing duplicated gene IDs: 544\n",
      "268 duplicated IDs in 544 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 134\n",
      "Merged  410 duplicated rows into 205 rows\n"
     ]
    }
   ],
   "source": [
    "pdx.index.name = \"gene_id\"\n",
    "ids = pdx.index\n",
    "ids = list(set(ids[ids.duplicated()]))\n",
    "print(\"Strings containing duplicated gene IDs:\",pdx.loc[ids,:].shape[0])\n",
    "pdx = handle_dups(pdx,corr_thr = 0.75)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "60c0da35-bad9-4f43-884b-8698ef323acd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'CN Averaged over all')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAHFCAYAAAAHcXhbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA8+UlEQVR4nO3deZyN9f//8edhmBlmsc/STGNkz04xUkafLBMilCLZP7IkSdmSUTFRiU+FFkbKoAUpW5KxhEJE+KJs48M02WbGYCzz/v3Rz/l0jDFzmHHONR732+3cbq739b6u63Wdi86z9/U+17EZY4wAAAAsqoCrCwAAALgZhBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkgl82cOVM2m83+8vDwUEhIiLp3767//ve/9n7x8fGy2WyKj4+/4WMcPHgw276RkZGKjIx0+hhZsdlsio6OzrX9RUdHO7xfhQsXVnh4uJ577jmdPn3a3u/q99XLy0uBgYFq0qSJYmJilJSU5LBfY4weeughlSxZUomJiZmO+8wzz6hw4cLaunVrrp2LO8nt6wS4Mw9XFwDkV7GxsapcubLOnTunNWvWKCYmRqtXr9aOHTtUtGjRm9p3y5YttWHDBgUFBeVStTm3YcMGhYSE5Pp+ly1bJn9/f6WmpmrJkiWaPHmyfv75Z61fv142m83e78r7evHiRSUlJWndunUaP3683nrrLc2bN08PPfSQpL8/zGfMmKHq1aurd+/e+uabb+z7WL58uT744ANFR0erdu3auX4uAG4twgyQR6pVq6Z69epJkpo0aaLLly/rtdde08KFC9W5c+eb2nfp0qVVunTp3CjTaQ0aNMiT/datW1elSpWSJDVt2lQnTpzQp59+qvXr1+u+++6z9/vn+ypJ7du31/PPP69GjRqpXbt22rdvnwICAiRJd955pyZOnKhevXopNjZW3bt316lTp9SzZ0/VqVNHI0eOzJNzAXBrcZsJuEWuhIBDhw5dt9+iRYsUERGhIkWKyNfXV02bNtWGDRsc+lzrNpMxRhMmTFBYWJi8vLxUp04dLV261GG7M2fOqFixYurTp0+m4x48eFAFCxbUm2++ed36rr59caWWVatWqW/fvipVqpRKliypdu3a6ejRo9fd1/Xk9P2S/g4tb7/9tlJTU/XBBx84rOvZs6eioqL0/PPPKyEhQQMGDNDx48c1a9YseXhk//9zX3zxherXry9/f38VKVJE5cqVU48ePezrz58/rxdeeEG1atWSv7+/SpQooYiICH399deZ9mWz2TRgwADFxsaqUqVK8vb2Vr169bRx40YZY/Tmm28qPDxcPj4+evDBB/X77787bB8ZGalq1app7dq1atCggby9vXXHHXdo1KhRunz5crbnkpiYqD59+igkJMR+O2/MmDG6dOmSQ7+pU6eqZs2a8vHxka+vrypXrqwRI0Zku3/AVQgzwC1y5YPpeiMqcXFxatOmjfz8/DRnzhxNnz5dp06dUmRkpNatW3fd/Y8ZM0ZDhw5V06ZNtXDhQvXt21e9e/fWnj177H18fHzUo0cPzZ49W8nJyQ7bT5kyRYULF3b4oHZGr169VKhQIcXFxWnChAmKj4/XU089dUP7knL2fv3Tww8/rIIFC2rNmjWZ1n388ccqUKCAmjRpori4OI0ZM0Z33313tvvcsGGDOnbsqHLlymnu3LlavHixXnnlFYcP//T0dJ08eVJDhgzRwoULNWfOHPso0axZszLt89tvv9XHH3+sN954Q3PmzFFqaqpatmypF154QT/++KPee+89ffjhh9q1a5fat28vY4zD9omJiXriiSfUuXNnff311+rQoYNef/11Pffcc9c9l8TERN17771avny5XnnlFS1dulQ9e/ZUTEyMevfube83d+5c9evXT40bN9aCBQu0cOFCPf/880pLS8v2/QJcxgDIVbGxsUaS2bhxo7l48aJJTU013377rSldurTx9fU1iYmJxhhjVq1aZSSZVatWGWOMuXz5sgkODjbVq1c3ly9ftu8vNTXVlClTxjRs2DDTMQ4cOGCMMebUqVPGy8vLPProow61/Pjjj0aSady4sb3tjz/+MAUKFDDvvPOOve3cuXOmZMmSpnv37tmenyQzevToTLX069fPod+ECROMJHPs2LHr7m/06NFGkklMTDQXL140p06dMp999pnx9vY2oaGh5ty5cw7H2bRpU5b7CggIMFWqVLnmunHjxhlJpmbNmubSpUvZnqcxxrz11ltGkjl9+nSO+htjzKVLl8zFixdNz549Te3atR3WSTKBgYHmzJkz9raFCxcaSaZWrVomIyPD3j5p0iQjyWzfvt3e1rhxYyPJfP311w777d27tylQoIA5dOiQw7H+eZ369OljfHx8HPr88xx37txpjDFmwIABplixYjk+X8AdMDID5JEGDRqoUKFC8vX1VatWrRQYGKilS5fa53Ncbc+ePTp69Ki6dOmiAgX+90/Tx8dH7du318aNG3X27NlrbrthwwadP38+01ychg0bKiwszKGtXLlyatWqlaZMmWL/v/64uDidOHFCAwYMuOHzfeSRRxyWa9SoISlnt4kkKTAwUIUKFVLx4sX11FNPqU6dOlq2bJm8vLxyXIO5ahTjirS0NPvozL59+3L0LTBJuueeeyRJjz/+uD7//HOHb6P90xdffKH77rtPPj4+8vDwUKFChTR9+nTt3r07U98mTZo4TACvUqWKJCkqKsphovOV9qvfP19f30zvdadOnZSRkXHNUakrvv32WzVp0kTBwcG6dOmS/RUVFSVJWr16tSTp3nvv1enTp/Xkk0/q66+/1vHjx7PcJ+AuCDNAHpk1a5Y2bdqkrVu36ujRo9q+fbvDRNarnThxQpKu+Q2l4OBgZWRk6NSpU9fdNjAwMNO6a7U999xz2rdvn1asWCFJev/99xUREaE6depkf2JZKFmypMOyp6enJOncuXM52v7777/Xpk2btG3bNh0/flzr1q1T1apVc3z8tLQ0nThxQsHBwZnWvfjiizp8+LAWL16sokWLqkePHlkGn3964IEHtHDhQl26dElPP/20QkJCVK1aNc2ZM8feZ/78+Xr88cd1xx136LPPPtOGDRu0adMm9ejRQ+fPn8+0zxIlSjgsFy5c+LrtV+/jWmH4yjW+8vfgWv7880998803KlSokMPryu22K6GlS5cumjFjhg4dOqT27durTJkyql+/vv3vCuCO+DYTkEeqVKni8K2b7FwJA8eOHcu07ujRoypQoICKFy9+3W2v9TyVxMRElS1b1qHtwQcfVLVq1fTee+/Jx8dHv/zyiz777LMc15oXatasaf82041YvHixLl++nOmZOitXrtS0adM0atQotWjRQu+//74ef/xxvfvuuxo4cGC2+23Tpo3atGmj9PR0bdy4UTExMerUqZPKli2riIgIffbZZwoPD9e8efMcRlbS09Nv+Fyu588//8zUduW6Xx0o/6lUqVKqUaOGxo4de831/wyB3bt3V/fu3ZWWlqY1a9Zo9OjRatWqlfbu3ZtppA9wB4zMAG6iUqVKuuOOOxQXF+cwapCWlqavvvrK/g2na2nQoIG8vLw0e/Zsh/b169dneZtn4MCBWrx4sYYPH66AgAA99thjuXcyt9jhw4c1ZMgQ+fv7O3xTKyUlRT169FDNmjX18ssvS5Iee+wxdejQQcOHD9cff/yR42N4enqqcePGGj9+vCTZH7Z35UF//wwyiYmJ1/w2U25ITU3VokWLHNri4uJUoEABPfDAA1lu16pVK/3222+66667VK9evUyva41oFS1aVFFRURo5cqQuXLignTt35vr5ALmBkRnATRQoUEATJkxQ586d1apVK/Xp00fp6el68803dfr0ab3xxhtZblu8eHENGTJEr7/+unr16qXHHntMCQkJio6OvuZtJkl66qmnNHz4cK1Zs0Yvv/yy/baGu/vtt9/s8z2SkpK0du1axcbGqmDBglqwYIHDt5+ef/55JSYm2m+vXDFlyhTdfffd6tGjh/1JzNfyyiuv6MiRI/rXv/6lkJAQnT59WpMnT1ahQoXUuHFjSX+HhPnz56tfv37q0KGDEhIS9NprrykoKEj79u3L9fMvWbKk+vbtq8OHD6tixYpasmSJPvroI/Xt21d33nlnltu9+uqrWrFihRo2bKiBAweqUqVKOn/+vA4ePKglS5Zo2rRpCgkJUe/eveXt7a377rtPQUFBSkxMVExMjPz9/e1ziAB3Q5gB3EinTp1UtGhRxcTEqGPHjipYsKAaNGigVatWqWHDhtfd9tVXX1XRokU1ZcoUffrpp6pcubKmTZumt95665r9vb291bp1a3322Wd65pln8uJ08kT37t0l/T2npFixYqpSpYqGDh2qXr16OQSZpUuXasaMGRo7dqx9MvIVpUuXztHtpvr162vz5s0aOnSo/vrrLxUrVkz16tXTDz/8YJ9r0r17dyUlJWnatGmaMWOGypUrp2HDhunIkSMaM2ZMrp9/YGCg3n//fQ0ZMkQ7duxQiRIlNGLEiGyPFRQUpM2bN+u1117Tm2++qSNHjsjX11fh4eFq0aKF/Rbm/fffr5kzZ+rzzz/XqVOnVKpUKTVq1EizZs1y2YMagezYTE5mwQHIdy5cuKCyZcuqUaNG+vzzz11dDnIgMjJSx48f12+//ebqUgC3wsgMcJv566+/tGfPHsXGxurPP//UsGHDXF0SANwUwgxwm1m8eLG6d++uoKAgTZky5aa+jg0A7oDbTAAAwNL4ajYAALA0wgwAALA0wgwAALC0fD8BOCMjQ0ePHpWvr2+WD8YCAADuxRij1NRUBQcHO/z47rXk+zBz9OhRhYaGuroMAABwAxISEhQSEnLdPvk+zPj6+kr6+83w8/NzcTUAACAnUlJSFBoaav8cv558H2au3Fry8/MjzAAAYDE5mSLCBGAAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBpLg0zU6dOVY0aNew/NRAREaGlS5fa13fr1k02m83h1aBBAxdWDAAA3I1Lf5spJCREb7zxhsqXLy9J+uSTT9SmTRtt3bpVd999tySpRYsWio2NtW9TuHBhl9QKAADck0vDTOvWrR2Wx44dq6lTp2rjxo32MOPp6anAwEBXlAcAACzAbebMXL58WXPnzlVaWpoiIiLs7fHx8SpTpowqVqyo3r17KykpyYVVAgAAd+PSkRlJ2rFjhyIiInT+/Hn5+PhowYIFqlq1qiQpKipKjz32mMLCwnTgwAGNGjVKDz74oLZs2SJPT89r7i89PV3p6en25ZSUlFtyHgAAwDVsxhjjygIuXLigw4cP6/Tp0/rqq6/08ccfa/Xq1fZA80/Hjh1TWFiY5s6dq3bt2l1zf9HR0RozZkym9uTkZPn5+eV6/UBeKDts8Q1ve/CNlrlYCQC4RkpKivz9/XP0+e3y20yFCxdW+fLlVa9ePcXExKhmzZqaPHnyNfsGBQUpLCxM+/bty3J/w4cPV3Jysv2VkJCQV6UDAAA34PLbTFczxjjcJvqnEydOKCEhQUFBQVlu7+npmeUtKAAAkP+4NMyMGDFCUVFRCg0NVWpqqubOnav4+HgtW7ZMZ86cUXR0tNq3b6+goCAdPHhQI0aMUKlSpfToo4+6smwAAOBGXBpm/vzzT3Xp0kXHjh2Tv7+/atSooWXLlqlp06Y6d+6cduzYoVmzZun06dMKCgpSkyZNNG/ePPn6+rqybAAA4EZcGmamT5+e5Tpvb28tX778FlYDAACsyOUTgAEAAG4GYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFiah6sLAPKrssMWu7oEALgtMDIDAAAsjTADAAAsjTADAAAsjTADAAAsjTADAAAszaVhZurUqapRo4b8/Pzk5+eniIgILV261L7eGKPo6GgFBwfL29tbkZGR2rlzpwsrBgAA7salYSYkJERvvPGGNm/erM2bN+vBBx9UmzZt7IFlwoQJmjhxot577z1t2rRJgYGBatq0qVJTU11ZNgAAcCMuDTOtW7fWww8/rIoVK6pixYoaO3asfHx8tHHjRhljNGnSJI0cOVLt2rVTtWrV9Mknn+js2bOKi4tzZdkAAMCNuM2cmcuXL2vu3LlKS0tTRESEDhw4oMTERDVr1szex9PTU40bN9b69euz3E96erpSUlIcXgAAIP9y+ROAd+zYoYiICJ0/f14+Pj5asGCBqlatag8sAQEBDv0DAgJ06NChLPcXExOjMWPG5GnNgDu7mScPH3yjZS5WAgC3hstHZipVqqRt27Zp48aN6tu3r7p27apdu3bZ19tsNof+xphMbf80fPhwJScn218JCQl5VjsAAHA9l4/MFC5cWOXLl5ck1atXT5s2bdLkyZM1dOhQSVJiYqKCgoLs/ZOSkjKN1vyTp6enPD0987ZoAADgNlw+MnM1Y4zS09MVHh6uwMBArVixwr7uwoULWr16tRo2bOjCCgEAgDtx6cjMiBEjFBUVpdDQUKWmpmru3LmKj4/XsmXLZLPZNGjQII0bN04VKlRQhQoVNG7cOBUpUkSdOnVyZdkAAMCNuDTM/Pnnn+rSpYuOHTsmf39/1ahRQ8uWLVPTpk0lSS+99JLOnTunfv366dSpU6pfv76+++47+fr6urJsAADgRmzGGOPqIvJSSkqK/P39lZycLD8/P1eXg9vIzXyryFX4NhMAd+HM57fbzZkBAABwBmEGAABYGmEGAABYmsufMwO4MyvOe7kZPD0YgBUxMgMAACyNMAMAACyNMAMAACyNMAMAACyNMAMAACyNMAMAACyNMAMAACyNMAMAACyNMAMAACyNJwDDEm63J/ECAHKOkRkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBpHq4uAED+UHbY4hve9uAbLXOxEgC3G0ZmAACApRFmAACApRFmAACApRFmAACApRFmAACApbk0zMTExOiee+6Rr6+vypQpo7Zt22rPnj0Ofbp16yabzebwatCggYsqBgAA7salYWb16tXq37+/Nm7cqBUrVujSpUtq1qyZ0tLSHPq1aNFCx44ds7+WLFnioooBAIC7celzZpYtW+awHBsbqzJlymjLli164IEH7O2enp4KDAy81eUBAAALcKs5M8nJyZKkEiVKOLTHx8erTJkyqlixonr37q2kpKQs95Genq6UlBSHFwAAyL/cJswYYzR48GA1atRI1apVs7dHRUVp9uzZ+uGHH/T2229r06ZNevDBB5Wenn7N/cTExMjf39/+Cg0NvVWnAAAAXMBmjDGuLkKS+vfvr8WLF2vdunUKCQnJst+xY8cUFhamuXPnql27dpnWp6enOwSdlJQUhYaGKjk5WX5+fnlSO/LezTwqH+6PnzMAcLWUlBT5+/vn6PPbLX6b6dlnn9WiRYu0Zs2a6wYZSQoKClJYWJj27dt3zfWenp7y9PTMizIBAIAbcmmYMcbo2Wef1YIFCxQfH6/w8PBstzlx4oQSEhIUFBR0CyoEAADuzqVzZvr376/PPvtMcXFx8vX1VWJiohITE3Xu3DlJ0pkzZzRkyBBt2LBBBw8eVHx8vFq3bq1SpUrp0UcfdWXpAADATbh0ZGbq1KmSpMjISIf22NhYdevWTQULFtSOHTs0a9YsnT59WkFBQWrSpInmzZsnX19fF1QMAADcjctvM12Pt7e3li9ffouqAQAAVuQ2X80GAAC4EYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaR6uLgAAyg5bfMPbHnyjZS5WAsCKGJkBAACWRpgBAACWlith5vTp07mxGwAAAKc5HWbGjx+vefPm2Zcff/xxlSxZUnfccYd+/fXXXC0OAAAgO06HmQ8++EChoaGSpBUrVmjFihVaunSpoqKi9OKLL+Z6gQAAANfj9LeZjh07Zg8z3377rR5//HE1a9ZMZcuWVf369XO9QAAAgOtxemSmePHiSkhIkCQtW7ZMDz30kCTJGKPLly/nbnUAAADZcHpkpl27durUqZMqVKigEydOKCoqSpK0bds2lS9fPtcLBAAAuB6nw8w777yjsmXLKiEhQRMmTJCPj4+kv28/9evXL9cLBAAAuB6nw0yhQoU0ZMiQTO2DBg3KjXoAAACcckPPmfn000/VqFEjBQcH69ChQ5KkSZMm6euvv87V4gAAALLjdJiZOnWqBg8erKioKJ0+fdo+6bdYsWKaNGlSbtcHAABwXU6HmXfffVcfffSRRo4cqYIFC9rb69Wrpx07duRqcQAAANlxOswcOHBAtWvXztTu6emptLS0XCkKAAAgp5wOM+Hh4dq2bVum9qVLl6pq1aq5URMAAECOOR1mXnzxRfXv31/z5s2TMUY///yzxo4dqxEjRjj9cwYxMTG655575OvrqzJlyqht27bas2ePQx9jjKKjoxUcHCxvb29FRkZq586dzpYNAADyKafDTPfu3TV69Gi99NJLOnv2rDp16qRp06Zp8uTJeuKJJ5za1+rVq9W/f39t3LhRK1as0KVLl9SsWTOH21UTJkzQxIkT9d5772nTpk0KDAxU06ZNlZqa6mzpAAAgH7IZY8yNbnz8+HFlZGSoTJkyuVLMX3/9pTJlymj16tV64IEHZIxRcHCwBg0apKFDh0qS0tPTFRAQoPHjx6tPnz7Z7jMlJUX+/v5KTk6Wn59frtSJW6/ssMWuLgFu6uAbLV1dAoA84Mzn9w09Z+aKUqVK5VqQkaTk5GRJUokSJST9Pdk4MTFRzZo1s/fx9PRU48aNtX79+lw7LgAAsK4cPQG4du3astlsOdrhL7/8ckOFGGM0ePBgNWrUSNWqVZMkJSYmSpICAgIc+gYEBNgf1ne19PR0paen25dTUlJuqB4AAGANOQozbdu2zeMypAEDBmj79u1at25dpnVXByljTJbhKiYmRmPGjMmTGgEAgPvJUZgZPXp0nhbx7LPPatGiRVqzZo1CQkLs7YGBgZL+HqEJCgqytyclJWUarbli+PDhGjx4sH05JSVFoaGheVQ5AABwNad/aPKKzZs3a/fu3bLZbKpSpYrq1q3r9D6MMXr22We1YMECxcfHKzw83GF9eHi4AgMDtWLFCvuD+i5cuKDVq1dr/Pjx19ynp6enPD09nT8hAABgSU6HmSNHjujJJ5/Ujz/+qGLFikmSTp8+rYYNG2rOnDlOjYL0799fcXFx+vrrr+Xr62ufI+Pv7y9vb2/ZbDYNGjRI48aNU4UKFVShQgWNGzdORYoUUadOnZwtHQAA5ENOf5upR48eunjxonbv3q2TJ0/q5MmT2r17t4wx6tmzp1P7mjp1qpKTkxUZGamgoCD7a968efY+L730kgYNGqR+/fqpXr16+u9//6vvvvtOvr6+zpYOAADyIaefM+Pt7a3169dn+n2mX375Rffdd5/OnTuXqwXeLJ4zkz/wnBlkhefMAPlTnj5n5s4779TFixcztV+6dEl33HGHs7sDAAC4KU6HmQkTJujZZ5/V5s2bdWVQZ/PmzXruuef01ltv5XqBAAAA1+P0babixYvr7NmzunTpkjw8/p4/fOXPRYsWdeh78uTJ3Kv0BnGbKX/gNhOywm0mIH9y5vPb6W8zTZo06UbrAgAAyHVOh5muXbvmRR0AAAA35IYfmpeUlKSkpCRlZGQ4tNeoUeOmiwIAAMgpp8PMli1b1LVrV/uzZf7JZrPp8uXLuVYcAABAdpwOM927d1fFihU1ffp0BQQE5PjXtAEAAPKC02HmwIEDmj9/vsqXL58X9QAAADjF6efM/Otf/9Kvv/6aF7UAAAA4zemRmY8//lhdu3bVb7/9pmrVqqlQoUIO6x955JFcKw4AACA7ToeZ9evXa926dVq6dGmmdUwABgAAt5rTt5kGDhyoLl266NixY8rIyHB4EWQAAMCt5nSYOXHihJ5//nkFBATkRT0AAABOcTrMtGvXTqtWrcqLWgAAAJzm9JyZihUravjw4Vq3bp2qV6+eaQLwwIEDc604AACA7Dj9q9nh4eFZ78xm0/79+2+6qNzEr2bnD/xqNrLCr2YD+VOe/mr2gQMHbrgwAACA3Ob0nBkAAAB3ckO/mn3kyBEtWrRIhw8f1oULFxzWTZw4MVcKAwAAyAmnw8zKlSv1yCOPKDw8XHv27FG1atV08OBBGWNUp06dvKgRAAAgS07fZho+fLheeOEF/fbbb/Ly8tJXX32lhIQENW7cWI899lhe1AgAAJAlp8PM7t271bVrV0mSh4eHzp07Jx8fH7366qsaP358rhcIAABwPU6HmaJFiyo9PV2SFBwcrD/++MO+7vjx47lXGQAAQA44PWemQYMG+vHHH1W1alW1bNlSL7zwgnbs2KH58+erQYMGeVEjAABAlpwOMxMnTtSZM2ckSdHR0Tpz5ozmzZun8uXL65133sn1AgEAAK7H6TBTrlw5+5+LFCmiKVOm5GpBAAAAznB6zkxCQoKOHDliX/755581aNAgffjhh7laGAAAQE44HWY6depk/9XsxMREPfTQQ/r55581YsQIvfrqq7leIAAAwPU4HWZ+++033XvvvZKkzz//XNWrV9f69esVFxenmTNn5nZ9AAAA1+V0mLl48aI8PT0lSd9//70eeeQRSVLlypV17Nix3K0OAAAgG06HmbvvvlvTpk3T2rVrtWLFCrVo0UKSdPToUZUsWTLXCwQAALgep8PM+PHj9cEHHygyMlJPPvmkatasKUlatGiR/fYTAADAreL0V7MjIyN1/PhxpaSkqHjx4vb2f//73ypSpEiuFgcAAJAdp8OMJBUsWNAhyEhS2bJlc6MeAAAApzh9mwkAAMCdEGYAAIClEWYAAIClEWYAAICl3dAE4JUrV2rlypVKSkpSRkaGw7oZM2bkSmEAAAA54fTIzJgxY9SsWTOtXLlSx48f16lTpxxezlizZo1at26t4OBg2Ww2LVy40GF9t27dZLPZHF4NGjRwtmQAAJCPOT0yM23aNM2cOVNdunS56YOnpaWpZs2a6t69u9q3b3/NPi1atFBsbKx9uXDhwjd9XAAAkH84HWYuXLighg0b5srBo6KiFBUVdd0+np6eCgwMzJXjAQCA/Mfp20y9evVSXFxcXtRyTfHx8SpTpowqVqyo3r17Kykp6br909PTlZKS4vACAAD5l9MjM+fPn9eHH36o77//XjVq1FChQoUc1k+cODHXiouKitJjjz2msLAwHThwQKNGjdKDDz6oLVu22H+5+2oxMTEaM2ZMrtUAAADcm9NhZvv27apVq5Yk6bfffnNYZ7PZcqWoKzp27Gj/c7Vq1VSvXj2FhYVp8eLFateu3TW3GT58uAYPHmxfTklJUWhoaK7WBQAA3IfTYWbVqlV5UUeOBAUFKSwsTPv27cuyj6enZ5ajNgAAIP+x1EPzTpw4oYSEBAUFBbm6FAAA4CZu6KF5ueXMmTP6/fff7csHDhzQtm3bVKJECZUoUULR0dFq3769goKCdPDgQY0YMUKlSpXSo48+6sKqAQCAO3FpmNm8ebOaNGliX74y16Vr166aOnWqduzYoVmzZun06dMKCgpSkyZNNG/ePPn6+rqqZAAA4GZcGmYiIyNljMly/fLly29hNQAAwIosNWcGAADgaoQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaYQZAABgaR6uLgC3Xtlhi29424NvtMzFSgAAuHmMzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEsjzAAAAEvjCcAALM2KT7S2Ys2AO2NkBgAAWBphBgAAWBphBgAAWBphBgAAWBphBgAAWJpLw8yaNWvUunVrBQcHy2azaeHChQ7rjTGKjo5WcHCwvL29FRkZqZ07d7qmWAAA4JZcGmbS0tJUs2ZNvffee9dcP2HCBE2cOFHvvfeeNm3apMDAQDVt2lSpqam3uFIAAOCuXPqcmaioKEVFRV1znTFGkyZN0siRI9WuXTtJ0ieffKKAgADFxcWpT58+t7JUAADgptx2zsyBAweUmJioZs2a2ds8PT3VuHFjrV+/Psvt0tPTlZKS4vACAAD5l9s+ATgxMVGSFBAQ4NAeEBCgQ4cOZbldTEyMxowZk6e1uYObeYKoFY8LAEBW3HZk5gqbzeawbIzJ1PZPw4cPV3Jysv2VkJCQ1yUCAAAXctuRmcDAQEl/j9AEBQXZ25OSkjKN1vyTp6enPD0987w+AADgHtx2ZCY8PFyBgYFasWKFve3ChQtavXq1GjZs6MLKAACAO3HpyMyZM2f0+++/25cPHDigbdu2qUSJErrzzjs1aNAgjRs3ThUqVFCFChU0btw4FSlSRJ06dXJh1QAAwJ24NMxs3rxZTZo0sS8PHjxYktS1a1fNnDlTL730ks6dO6d+/frp1KlTql+/vr777jv5+vq6qmQAAOBmXBpmIiMjZYzJcr3NZlN0dLSio6NvXVEAAMBS3HbODAAAQE4QZgAAgKURZgAAgKW57XNmACCv3cwTrQ++0TIXKwFwMxiZAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlubh6gIAwIrKDltsueMefKOl5Y4L5AQjMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNIIMwAAwNLcOsxER0fLZrM5vAIDA11dFgAAcCMeri4gO3fffbe+//57+3LBggVdWA0AAHA3bh9mPDw8GI0BAABZcuvbTJK0b98+BQcHKzw8XE888YT2799/3f7p6elKSUlxeAEAgPzLrcNM/fr1NWvWLC1fvlwfffSREhMT1bBhQ504cSLLbWJiYuTv729/hYaG3sKKAQDArebWYSYqKkrt27dX9erV9dBDD2nx4sWSpE8++STLbYYPH67k5GT7KyEh4VaVCwAAXMDt58z8U9GiRVW9enXt27cvyz6enp7y9PS8hVUBAABXcuuRmaulp6dr9+7dCgoKcnUpAADATbh1mBkyZIhWr16tAwcO6KefflKHDh2UkpKirl27uro0AADgJtz6NtORI0f05JNP6vjx4ypdurQaNGigjRs3KiwszNWlAQAAN+HWYWbu3LmuLgEAALg5t77NBAAAkB3CDAAAsDTCDAAAsDS3njMDALC+ssMW3/C2B99omYuVIL9iZAYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaP2fgQjfziG8AcBb/zck5foLBWhiZAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlmYzxhhXF5GXUlJS5O/vr+TkZPn5+eX6/nmiJgAgt/D04P9x5vObkRkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBphBkAAGBpHq4uAAAA3DxXPZHeHZ5azMgMAACwNMIMAACwNMIMAACwNMIMAACwNEuEmSlTpig8PFxeXl6qW7eu1q5d6+qSAACAm3D7MDNv3jwNGjRII0eO1NatW3X//fcrKipKhw8fdnVpAADADbh9mJk4caJ69uypXr16qUqVKpo0aZJCQ0M1depUV5cGAADcgFuHmQsXLmjLli1q1qyZQ3uzZs20fv16F1UFAADciVs/NO/48eO6fPmyAgICHNoDAgKUmJh4zW3S09OVnp5uX05OTpYkpaSk5EmNGeln82S/AIDbz818Vrnq8yivPl+v7NcYk21ftw4zV9hsNodlY0ymtitiYmI0ZsyYTO2hoaF5UhsAALnFf5KrK3BeXtecmpoqf3//6/Zx6zBTqlQpFSxYMNMoTFJSUqbRmiuGDx+uwYMH25czMjJ08uRJlSxZMssAlN+kpKQoNDRUCQkJ8vPzc3U5ty2ug/vgWrgProV7sMJ1MMYoNTVVwcHB2fZ16zBTuHBh1a1bVytWrNCjjz5qb1+xYoXatGlzzW08PT3l6enp0FasWLG8LNNt+fn5ue1f0tsJ18F9cC3cB9fCPbj7dchuROYKtw4zkjR48GB16dJF9erVU0REhD788EMdPnxYzzzzjKtLAwAAbsDtw0zHjh114sQJvfrqqzp27JiqVaumJUuWKCwszNWlAQAAN+D2YUaS+vXrp379+rm6DMvw9PTU6NGjM91uw63FdXAfXAv3wbVwD/ntOthMTr7zBAAA4Kbc+qF5AAAA2SHMAAAASyPMAAAASyPMAAAASyPMWMyaNWvUunVrBQcHy2azaeHChdluM3v2bNWsWVNFihRRUFCQunfvrhMnTuR9sflYTEyM7rnnHvn6+qpMmTJq27at9uzZk+12q1evVt26deXl5aVy5cpp2rRpt6Da/O1GrsX8+fPVtGlTlS5dWn5+foqIiNDy5ctvUcX5143+u7jixx9/lIeHh2rVqpV3Rd4GbvQ6pKena+TIkQoLC5Onp6fuuusuzZgx4xZUfPMIMxaTlpammjVr6r333stR/3Xr1unpp59Wz549tXPnTn3xxRfatGmTevXqlceV5m+rV69W//79tXHjRq1YsUKXLl1Ss2bNlJaWluU2Bw4c0MMPP6z7779fW7du1YgRIzRw4EB99dVXt7Dy/OdGrsWaNWvUtGlTLVmyRFu2bFGTJk3UunVrbd269RZWnv/cyLW4Ijk5WU8//bT+9a9/3YJK87cbvQ6PP/64Vq5cqenTp2vPnj2aM2eOKleufIuqvkkGliXJLFiw4Lp93nzzTVOuXDmHtv/85z8mJCQkDyu7/SQlJRlJZvXq1Vn2eemll0zlypUd2vr06WMaNGiQ1+XdVnJyLa6latWqZsyYMXlU1e3JmWvRsWNH8/LLL5vRo0ebmjVr5n1xt5GcXIelS5caf39/c+LEiVtYWe5hZCafa9iwoY4cOaIlS5bIGKM///xTX375pVq2bOnq0vKV5ORkSVKJEiWy7LNhwwY1a9bMoa158+bavHmzLl68mKf13U5yci2ulpGRodTUVKe2QfZyei1iY2P1xx9/aPTo0beirNtOTq7DokWLVK9ePU2YMEF33HGHKlasqCFDhujcuXO3qsybYoknAOPGNWzYULNnz1bHjh11/vx5Xbp0SY888ojeffddV5eWbxhjNHjwYDVq1EjVqlXLsl9iYmKmX3sPCAjQpUuXdPz4cQUFBeV1qfleTq/F1d5++22lpaXp8ccfz8Pqbi85vRb79u3TsGHDtHbtWnl48JGU23J6Hfbv369169bJy8tLCxYs0PHjx9WvXz+dPHnSEvNmGJnJ53bt2qWBAwfqlVde0ZYtW7Rs2TIdOHCAH+rMRQMGDND27ds1Z86cbPvabDaHZfP/H8B9dTtujDPX4oo5c+YoOjpa8+bNU5kyZfKwuttLTq7F5cuX1alTJ40ZM0YVK1a8hdXdPnL6byIjI0M2m02zZ8/Wvffeq4cfflgTJ07UzJkzrTE648p7XLg5ysGcmaeeesp06NDBoW3t2rVGkjl69GgeVnd7GDBggAkJCTH79+/Ptu/9999vBg4c6NA2f/584+HhYS5cuJBXJd42nLkWV8ydO9d4e3ubb7/9Ng8ru/3k9FqcOnXKSDIFCxa0v2w2m71t5cqVt6ji/MmZfxNPP/20ueuuuxzadu3aZSSZvXv35lWJuYYxvXzu7NmzmYZuCxYsKOl/owJwnjFGzz77rBYsWKD4+HiFh4dnu01ERIS++eYbh7bvvvtO9erVU6FChfKq1HzvRq6F9PeITI8ePTRnzhzmkOUSZ6+Fn5+fduzY4dA2ZcoU/fDDD/ryyy9zfC3h6Eb+Tdx333364osvdObMGfn4+EiS9u7dqwIFCigkJCSvS755Lo1ScFpqaqrZunWr2bp1q5FkJk6caLZu3WoOHTpkjDFm2LBhpkuXLvb+sbGxxsPDw0yZMsX88ccfZt26daZevXrm3nvvddUp5At9+/Y1/v7+Jj4+3hw7dsz+Onv2rL3P1ddi//79pkiRIub55583u3btMtOnTzeFChUyX375pStOId+4kWsRFxdnPDw8zPvvv++wzenTp11xCvnGjVyLq/Ftppt3I9chNTXVhISEmA4dOpidO3ea1atXmwoVKphevXq54hScRpixmFWrVhlJmV5du3Y1xhjTtWtX07hxY4dt/vOf/5iqVasab29vExQUZDp37myOHDly64vPR651DSSZ2NhYe59rXYv4+HhTu3ZtU7hwYVO2bFkzderUW1t4PnQj16Jx48bX/XeEG3Oj/y7+iTBz8270Ouzevds89NBDxtvb24SEhJjBgwc7BCB3ZjOGew0AAMC6+DYTAACwNMIMAACwNMIMAACwNMIMAACwNMIMAACwNMIMAACwNMIMAACwNMIMgNvSwYMHZbPZtG3bNleXkmvKli2rSZMm2ZdtNpsWLlzosnqAW4UwA1hAYmKinn32WZUrV06enp4KDQ1V69attXLlSnufsmXLymazaePGjQ7bDho0SJGRkTk6TrNmzVSwYMFM+wAAd0aYAdzcwYMHVbduXf3www+aMGGCduzYoWXLlqlJkybq37+/Q18vLy8NHTr0ho5z+PBhbdiwQQMGDND06dNzo/TrunDhQp4fw4p4XwDnEWYAN9evXz/ZbDb9/PPP6tChgypWrKi7775bgwcPzjSC0qdPH23cuFFLlixx+jixsbFq1aqV+vbtq3nz5iktLU2StHz5cnl5een06dMO/QcOHKjGjRvbl9evX68HHnhA3t7eCg0N1cCBA+37kP4eOXr99dfVrVs3+fv7q3fv3pKkoUOHqmLFiipSpIjKlSunUaNG6eLFiw7Hev3111WmTBn5+vqqV69eGjZsmGrVqpWp/ipVqsjLy0uVK1fWlClTHNb//PPPql27try8vFSvXj1t3bo12/fk1KlTevrpp1W8eHEVKVJEUVFR2rdvnyQpOTlZ3t7eWrZsmcM28+fPV9GiRXXmzBlJ0n//+1917NhRxYsXV8mSJdWmTRsdPHjQ3r9bt25q27atYmJiFBwcrIoVK16zlj/++ENt2rRRQECAfHx8dM899+j777/P9hyA2wFhBnBjJ0+e1LJly9S/f38VLVo00/pixYo5LJctW1bPPPOMhg8froyMjBwfxxij2NhYPfXUU6pcubIqVqyozz//XJL00EMPqVixYvrqq6/s/S9fvqzPP/9cnTt3liTt2LFDzZs3V7t27bR9+3bNmzdP69at04ABAxyO8+abb6patWrasmWLRo0aJUny9fXVzJkztWvXLk2ePFkfffSR3nnnHfs2s2fP1tixYzV+/Hht2bJFd955p6ZOneqw348++kgjR47U2LFjtXv3bo0bN06jRo3SJ598IklKS0tTq1atVKlSJW3ZskXR0dEaMmRItu9Lt27dtHnzZi1atEgbNmyQMUYPP/ywLl68KH9/f7Vs2VKzZ8922CYuLk5t2rSRj4+Pzp49qyZNmsjHx0dr1qzRunXr5OPjoxYtWjiMwKxcuVK7d+/WihUr9O23316zljNnzujhhx/W999/r61bt6p58+Zq3bq1Dh8+nO15APmea3/nEsD1/PTTT0aSmT9/frZ9w8LCzDvvvGOSkpKMr6+vmTVrljHGmOeee+66v1JsjDHfffedKV26tLl48aIxxph33nnH3Hffffb1AwcONA8++KB9efny5aZw4cLm5MmTxhhjunTpYv7973877HPt2rWmQIEC5ty5c/b62rZtm+15TJgwwdStW9e+XL9+fdO/f3+HPvfdd5/DLyuHhoaauLg4hz6vvfaaiYiIMMYY88EHH5gSJUqYtLQ0+/qpU6caSWbr1q3XrGPv3r1Gkvnxxx/tbcePHzfe3t7m888/N8YYM3/+fOPj42Pfb3JysvHy8jKLFy82xhgzffp0U6lSJZORkWHfR3p6uvH29jbLly83xvz968UBAQEmPT092/fmalWrVjXvvvuuffnK34ErJJkFCxY4vV/AahiZAdyY+f8/am+z2XK8TenSpTVkyBC98sorOZ5/MX36dHXs2FEeHh6SpCeffFI//fST9uzZI0nq3Lmz4uPjdfToUUl/j5Y8/PDDKl68uCRpy5Ytmjlzpnx8fOyv5s2bKyMjQwcOHLAfp169epmO/eWXX6pRo0YKDAyUj4+PRo0a5TDasGfPHt17770O2/xz+a+//lJCQoJ69uzpcPzXX39df/zxhyRp9+7dqlmzpooUKWLfLiIi4rrvye7du+Xh4aH69evb20qWLKlKlSpp9+7dkqSWLVvKw8NDixYtkiR99dVX8vX1VbNmzezvy++//y5fX197XSVKlND58+fttUlS9erVVbhw4evWk5aWppdeeklVq1ZVsWLF5OPjo//7v/9jZAaQ5OHqAgBkrUKFCrLZbNq9e7fatm2b4+0GDx6sKVOmZJo3ci0nT57UwoULdfHiRYfbN5cvX9aMGTM0fvx43Xvvvbrrrrs0d+5c9e3bVwsWLFBsbKy9b0ZGhvr06aOBAwdm2v+dd95p//PVt8o2btyoJ554QmPGjFHz5s3l7++vuXPn6u2333bod3WYuxLyrhxb+vtW0z+DhyQVLFgwU/+cymobY4y9nsKFC6tDhw6Ki4vTE088obi4OIdQmJGRobp162a6FSX9HTqvuNYtxKu9+OKLWr58ud566y2VL19e3t7e6tChAxOGARFmALdWokQJNW/eXO+//74GDhyY6UPv9OnTmebNSLKPcERHR6t169bXPcbs2bMVEhKS6XkkK1euVExMjMaOHSsPDw916tTJ3rdAgQJq2bKlvW+dOnW0c+dOlS9f3qnz+/HHHxUWFqaRI0fa2w4dOuTQp1KlSvr555/VpUsXe9vmzZvtfw4ICNAdd9yh/fv32+fwXK1q1ar69NNPde7cOXl7e0tStl8/r1q1qi5duqSffvpJDRs2lCSdOHFCe/fuVZUqVez9OnfurGbNmmnnzp1atWqVXnvtNfu6OnXqaN68eSpTpoz8/Pyyezuua+3aterWrZseffRRSX/PofnnRGLgtubSm1wAsrV//34TGBhoqlatar788kuzd+9es2vXLjN58mRTuXJle7+r50tcuHDB3HXXXcbLy+u6c2Zq1qxphg4dmqk9JSXFeHp6moULFxpj/jeHpEaNGqZnz54OfX/99Vfj7e1t+vXrZ7Zu3Wr27t1rvv76azNgwIAs6zPGmIULFxoPDw8zZ84c8/vvv5vJkyebEiVKGH9/f3ufzz77zHh7e5uZM2eavXv3mtdee834+fmZWrVq2ft89NFHxtvb20yaNMns2bPHbN++3cyYMcO8/fbbxhhjUlNTTalSpcyTTz5pdu7caRYvXmzKly9/3TkzxhjTpk0bU7VqVbN27Vqzbds206JFC1O+fHlz4cIFe5+MjAwTEhJiatasae666y6H7dPS0kyFChVMZGSkWbNmjdm/f7+Jj483AwcONAkJCcaYv+fMtGnTJssarmjbtq2pVauW2bp1q9m2bZtp3bq18fX1Nc8991yW77GYM4PbBHNmADcXHh6uX375RU2aNNELL7ygatWqqWnTplq5cmWmb/X8U6FChfTaa6/p/PnzWfbZsmWLfv31V7Vv3z7TuitzP648c6ZChQq65557tH379kwjIDVq1NDq1au1b98+3X///apdu7ZGjRqloKCg655bmzZt9Pzzz2vAgAGqVauW1q9fb/+W0xWdO3fW8OHDNWTIENWpU0cHDhxQt27d5OXlZe/Tq1cvffzxx5o5c6aqV6+uxo0ba+bMmQoPD5f090jVN998o127dql27doaOXKkxo8ff93apL+/7l23bl21atVKERERMsZoyZIlKlSokL2PzWbTk08+qV9//TXT+1KkSBGtWbNGd955p9q1a6cqVaqoR48eOnfunNMjNe+8846KFy+uhg0bqnXr1mrevLnq1Knj1D6A/MpmzA3cTAYAF2ratKkCAwP16aefuroUAG6AOTMA3NrZs2c1bdo0NW/eXAULFtScOXP0/fffa8WKFa4uDYCbYGQGgFs7d+6cWrdurV9++UXp6emqVKmSXn75ZbVr187VpQFwE4QZAABgaUwABgAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlkaYAQAAlvb/ALn/nJJpIJmlAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "average_ploidies = pdx.apply(np.mean)\n",
    "p = plt.hist(average_ploidies,bins=30)\n",
    "plt.title(\"Ploidy in PDX samples\")\n",
    "plt.ylabel(\"n samples\")\n",
    "plt.xlabel(\"CN Averaged over all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6af5fd0d-a1a7-4578-8143-5300b35980b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_781/3035632997.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  pdx = pdx.applymap(lambda x: np.log2(x/2))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.021480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.117695</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>0.049631</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.021480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.521051</td>\n",
       "      <td>0.570463</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.007196</td>\n",
       "      <td>-0.286304</td>\n",
       "      <td>-0.081614</td>\n",
       "      <td>-0.007232</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>-1.043943</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>-0.036526</td>\n",
       "      <td>-0.058894</td>\n",
       "      <td>-0.377070</td>\n",
       "      <td>-0.643856</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.070389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            X-1004    X-1008    X-1027    X-1095    X-1119    X-1156  \\\n",
       "gene_id                                                                \n",
       "A1BG      0.367371 -0.321928  0.117695  0.056584  0.000000  0.978196   \n",
       "A1BG-AS1  0.367371 -0.321928  0.117695  0.056584  0.000000  0.978196   \n",
       "A1CF      0.521051  0.570463  0.007196  0.042644  0.070389 -0.340075   \n",
       "\n",
       "            X-1167    X-1169    X-1172    X-1173  ...    X-5694    X-5696  \\\n",
       "gene_id                                           ...                       \n",
       "A1BG      0.028569  2.509696  0.117695  0.007196  ...  0.056584  0.070389   \n",
       "A1BG-AS1  0.028569  2.509696  0.117695  0.007196  ...  0.056584  0.070389   \n",
       "A1CF      0.007196 -0.286304 -0.081614 -0.007232  ...  0.028569 -1.043943   \n",
       "\n",
       "            X-5713    X-5717    X-5727    X-5739    X-5808    X-5959  \\\n",
       "gene_id                                                                \n",
       "A1BG      0.097611  0.560715  0.042644  0.049631 -0.007232  0.049631   \n",
       "A1BG-AS1  0.097611  0.560715  0.042644  0.049631 -0.007232  0.049631   \n",
       "A1CF     -0.340075  0.056584 -0.036526 -0.058894 -0.377070 -0.643856   \n",
       "\n",
       "            X-5975    X-6047  \n",
       "gene_id                       \n",
       "A1BG     -0.483985  0.021480  \n",
       "A1BG-AS1 -0.483985  0.021480  \n",
       "A1CF     -0.588574  0.070389  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx = pdx.applymap(lambda x: np.log2(x/2))\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6bf257f3-3551-437c-bd01-6a6722ca9fd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/34/h8l5tm2s0f71rm7c8sx59rk00000gn/T/ipykernel_781/268329890.py:1: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.\n",
      "  pdx = pdx.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.521051</td>\n",
       "      <td>0.570463</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.286304</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.043943</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.37707</td>\n",
       "      <td>-0.643856</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            X-1004    X-1008  X-1027  X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                  \n",
       "A1BG      0.367371 -0.321928     0.0     0.0     0.0  0.978196     0.0   \n",
       "A1BG-AS1  0.367371 -0.321928     0.0     0.0     0.0  0.978196     0.0   \n",
       "A1CF      0.521051  0.570463     0.0     0.0     0.0 -0.340075     0.0   \n",
       "\n",
       "            X-1169  X-1172  X-1173  ...  X-5694    X-5696    X-5713    X-5717  \\\n",
       "gene_id                             ...                                         \n",
       "A1BG      2.509696     0.0     0.0  ...     0.0  0.000000  0.000000  0.560715   \n",
       "A1BG-AS1  2.509696     0.0     0.0  ...     0.0  0.000000  0.000000  0.560715   \n",
       "A1CF     -0.286304     0.0     0.0  ...     0.0 -1.043943 -0.340075  0.000000   \n",
       "\n",
       "          X-5727  X-5739   X-5808    X-5959    X-5975  X-6047  \n",
       "gene_id                                                        \n",
       "A1BG         0.0     0.0  0.00000  0.000000 -0.483985     0.0  \n",
       "A1BG-AS1     0.0     0.0  0.00000  0.000000 -0.483985     0.0  \n",
       "A1CF         0.0     0.0 -0.37707 -0.643856 -0.588574     0.0  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx = pdx.applymap(lambda x :  clean_logR(x, pos_seg_mean_thr, neg_seg_mean_thr))\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b4725a10-2268-448a-b1ff-36e740cd8c13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 23262 \n",
      "\tdirectly via main_mapper 20493 \n",
      "\tvia alternative mapper 572 \n",
      "\tvia one of multiple synonyms in alternative mapper 1336 \n",
      "\tLOC 861 \n",
      "Unmapped: 251 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 77 \n",
      "\tLOC not found in Entrez 37 \n",
      "\tNot found at all: 137\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 105\n",
      "Warning: query IDs not mapped to any target IDs excluded: 251\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IDs mapped to multiple target IDs are kept:\n",
      " [143872, 286464, 51463, 642826, 653067, 399761, 647060, 284565, 84631, 161176, 284697, 341019, 83869, 4253, 9502, 83871, 10529, 5414, 51754, 440107, 9645, 645425, 728113, 26165, 114112, 399687, 728695, 6218, 23499, 100132948, 100134869, 84316, 200030, 729438, 6241, 642658, 100302179, 401508, 286183, 84968, 119016, 84458, 26220, 574445, 1903, 26095, 445815, 3192, 387707, 79741]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.367371</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978196</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.509696</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.483985</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.761285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.700440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.327687</td>\n",
       "      <td>-0.494109</td>\n",
       "      <td>-0.535332</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.761285</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.700440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.327687</td>\n",
       "      <td>-0.494109</td>\n",
       "      <td>-0.535332</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           X-1004    X-1008  X-1027    X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                   \n",
       "1        0.367371 -0.321928     0.0  0.000000     0.0  0.978196     0.0   \n",
       "2        0.761285  0.000000     0.0  0.500802     0.0  0.700440     0.0   \n",
       "3        0.761285  0.000000     0.0  0.500802     0.0  0.700440     0.0   \n",
       "\n",
       "           X-1169  X-1172  X-1173  ...  X-5694  X-5696  X-5713    X-5717  \\\n",
       "gene_id                            ...                                     \n",
       "1        2.509696     0.0     0.0  ...     0.0     0.0     0.0  0.560715   \n",
       "2        0.000000     0.0     0.0  ...     0.0     0.0     0.0  0.739848   \n",
       "3        0.000000     0.0     0.0  ...     0.0     0.0     0.0  0.739848   \n",
       "\n",
       "         X-5727    X-5739    X-5808    X-5959    X-5975  X-6047  \n",
       "gene_id                                                          \n",
       "1           0.0  0.000000  0.000000  0.000000 -0.483985     0.0  \n",
       "2           0.0  0.739848  0.327687 -0.494109 -0.535332     0.0  \n",
       "3           0.0  0.739848  0.327687 -0.494109 -0.535332     0.0  \n",
       "\n",
       "[3 rows x 375 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pdx,query2target,not_mapped = apply_mappers(pdx, ncbi_symbols, ncbi_synonyms, verbose = True,handle_duplicates = \"keep\")\n",
    "pdx.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ee210596-9af5-4b6d-a210-c7a4362b822e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5694</th>\n",
       "      <th>X-5696</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143872</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.350497</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.349235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.514573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143872</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.560715</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.350497</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.349235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.321928</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.514573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.550901</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.524915</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.902389</td>\n",
       "      <td>-1.321928</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>0.718088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.01450</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.494109</td>\n",
       "      <td>-1.494109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>1.350497</td>\n",
       "      <td>0.632268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.448901</td>\n",
       "      <td>-0.268817</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.480265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>-0.330973</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.021480</td>\n",
       "      <td>0.599318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387707</th>\n",
       "      <td>0.389567</td>\n",
       "      <td>-0.666576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>-0.473931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.386468</td>\n",
       "      <td>-0.610433</td>\n",
       "      <td>-0.577767</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>387707</th>\n",
       "      <td>0.389567</td>\n",
       "      <td>-0.666576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.340075</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.014500</td>\n",
       "      <td>-0.473931</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.386468</td>\n",
       "      <td>-0.610433</td>\n",
       "      <td>-0.577767</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79741</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.589763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.242977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.798366</td>\n",
       "      <td>-0.985645</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.70044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.459432</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79741</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.589763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.242977</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.798366</td>\n",
       "      <td>-0.985645</td>\n",
       "      <td>-0.358454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.70044</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.367732</td>\n",
       "      <td>-0.588574</td>\n",
       "      <td>0.459432</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>105 rows × 375 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           X-1004    X-1008  X-1027    X-1095  X-1119    X-1156  X-1167  \\\n",
       "gene_id                                                                   \n",
       "143872   0.000000  0.560715     0.0  0.000000     0.0 -0.330973     0.0   \n",
       "143872   0.000000  0.560715     0.0  0.000000     0.0 -0.330973     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "286464   0.000000  0.378512     0.0  0.550901     0.0 -0.524915     0.0   \n",
       "...           ...       ...     ...       ...     ...       ...     ...   \n",
       "3192     1.350497  0.632268     0.0  0.000000     0.0  0.000000     0.0   \n",
       "387707   0.389567 -0.666576     0.0  0.000000     0.0 -0.340075     0.0   \n",
       "387707   0.389567 -0.666576     0.0  0.000000     0.0 -0.340075     0.0   \n",
       "79741    0.000000  0.589763     0.0  0.000000     0.0 -0.242977     0.0   \n",
       "79741    0.000000  0.589763     0.0  0.000000     0.0 -0.242977     0.0   \n",
       "\n",
       "           X-1169    X-1172    X-1173  ...    X-5694    X-5696    X-5713  \\\n",
       "gene_id                                ...                                 \n",
       "143872  -0.367732  0.350497  0.000000  ...  0.000000 -0.349235  0.000000   \n",
       "143872  -0.367732  0.350497  0.000000  ...  0.000000 -0.349235  0.000000   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "286464  -0.902389 -1.321928  0.367371  ...  0.000000 -1.014500  0.718088   \n",
       "...           ...       ...       ...  ...       ...       ...       ...   \n",
       "3192     0.448901 -0.268817  0.000000  ...  0.480265  0.000000  0.000000   \n",
       "387707  -0.367732  0.000000  0.000000  ...  0.000000 -1.014500 -0.473931   \n",
       "387707  -0.367732  0.000000  0.000000  ...  0.000000 -1.014500 -0.473931   \n",
       "79741    0.000000  0.000000  0.000000  ... -0.798366 -0.985645  0.000000   \n",
       "79741    0.000000  0.000000  0.000000  ... -0.798366 -0.985645 -0.358454   \n",
       "\n",
       "           X-5717   X-5727    X-5739    X-5808    X-5959    X-5975    X-6047  \n",
       "gene_id                                                                       \n",
       "143872  -0.321928 -1.01450 -0.588574  0.000000  0.000000  0.000000 -0.514573  \n",
       "143872  -0.321928 -1.01450 -0.588574  0.000000  0.000000  0.000000 -0.514573  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "286464   0.000000 -1.01450  0.000000  0.000000  0.000000 -1.494109 -1.494109  \n",
       "...           ...      ...       ...       ...       ...       ...       ...  \n",
       "3192     0.000000  0.00000 -0.330973  0.000000  0.000000  1.021480  0.599318  \n",
       "387707   0.000000  0.00000  0.000000 -0.386468 -0.610433 -0.577767  0.000000  \n",
       "387707   0.000000  0.00000  0.000000 -0.386468 -0.610433 -0.577767  0.000000  \n",
       "79741    0.000000  0.70044  0.000000 -0.367732 -0.588574  0.459432  0.000000  \n",
       "79741    0.000000  0.70044  0.000000 -0.367732 -0.588574  0.459432  0.000000  \n",
       "\n",
       "[105 rows x 375 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dups = list(set(pdx[pdx.index.duplicated(keep=False)].index.values))\n",
    "pdx.loc[dups,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "b9c81a94-e790-4c23-9705-f142a6932f8c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 duplicated IDs in 105 rows found.\n",
      "duplicate rows removed due to low correlation of duplicated profiles 4\n",
      "Merged  101 duplicated rows into 48 rows\n"
     ]
    }
   ],
   "source": [
    "# most of these dupliates correspond to genes merged in the current assembly, e.g. gene - gene-AS\n",
    "pdx = handle_dups(pdx,corr_thr = 0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "838ef404-97a6-46b8-8e0b-1f381a83b992",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdx = pdx.T.sort_index().T\n",
    "pdx.to_csv(preprocessed_dir+\"/\"+\"PDX\"+\".Segment_Mean.CNA.tsv\",\n",
    "                 sep = \"\\t\",header=True,index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
