{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "11776fbc-1395-462f-9d68-5a90c212fc5e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Files\n",
      "/Volumes/Expansion/Thesis Work/Codes\n"
     ]
    }
   ],
   "source": [
    "# change directory to use \"mapper\" package used in MOLI for preprocessing steps\n",
    "\n",
    "import os\n",
    "\n",
    "print(os.getcwd())\n",
    "os.chdir(\"/Volumes/Expansion/Thesis Work/Codes/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe184bcf-2062-40e3-a585-f863941ce185",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "\n",
    "from __future__ import print_function\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import zscore\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import xlrd\n",
    "from mapper import expand, parse_mapping_table, apply_mappers\n",
    "%matplotlib inline\n",
    "import warnings\n",
    "\n",
    "# suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2ebeac51-e4a0-4274-9a8a-a158f3efa614",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define directories\n",
    "\n",
    "gene_id = \"ENTREZID\"\n",
    "raw_data_dir = \"/Volumes/Expansion/Thesis Work/Datasets/PDX/Expression/\"\n",
    "preprocessing_data_dir = \"/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/\"\n",
    "root_dir = \"/Volumes/Expansion/Thesis Work/Supplementary Files/\"\n",
    "tcga_tmp_dir = \"/Volumes/Expansion/Thesis Work/Datasets/TCGA/Expression/expression__2016_01_28\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f438434f-9ee2-4cb8-9a6a-c5ffaf04360c",
   "metadata": {},
   "source": [
    "# PDX \n",
    "\n",
    "- Gene symbols are converted to Gene IDs \n",
    "- gene expression profiles (FPKM) are converted to log2(TPM+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d2548978-52fc-4372-8620-d1a9dc25158e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22665, 399)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5974</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6030</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>2.75</td>\n",
       "      <td>8.97</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.08</td>\n",
       "      <td>17.52</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>21.24</td>\n",
       "      <td>0.46</td>\n",
       "      <td>18.70</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.30</td>\n",
       "      <td>14.41</td>\n",
       "      <td>5.09</td>\n",
       "      <td>6.35</td>\n",
       "      <td>0.09</td>\n",
       "      <td>3.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>2.48</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>15.21</td>\n",
       "      <td>0.09</td>\n",
       "      <td>3.42</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.30</td>\n",
       "      <td>8.36</td>\n",
       "      <td>3.71</td>\n",
       "      <td>4.52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.30</td>\n",
       "      <td>2.83</td>\n",
       "      <td>2.87</td>\n",
       "      <td>0.72</td>\n",
       "      <td>3.41</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.84</td>\n",
       "      <td>2.96</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>4.63</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>4.87</td>\n",
       "      <td>0.81</td>\n",
       "      <td>6.45</td>\n",
       "      <td>4.94</td>\n",
       "      <td>11.07</td>\n",
       "      <td>0.87</td>\n",
       "      <td>3.28</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.61</td>\n",
       "      <td>7.10</td>\n",
       "      <td>...</td>\n",
       "      <td>1.33</td>\n",
       "      <td>2.51</td>\n",
       "      <td>2.67</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.44</td>\n",
       "      <td>2.96</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.99</td>\n",
       "      <td>2.57</td>\n",
       "      <td>1.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.01</td>\n",
       "      <td>71.17</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>58.16</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>94.12</td>\n",
       "      <td>0.02</td>\n",
       "      <td>...</td>\n",
       "      <td>463.32</td>\n",
       "      <td>0.04</td>\n",
       "      <td>33.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>254.97</td>\n",
       "      <td>41.52</td>\n",
       "      <td>2.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X-1004  X-1008  X-1027  X-1095  X-1119  X-1156  X-1167  X-1169  \\\n",
       "Sample                                                                     \n",
       "A1BG        2.75    8.97    0.14    0.13    0.08   17.52    0.18    0.41   \n",
       "A1BG-AS1    2.48    3.25    0.00    0.00    0.00    6.52    0.00    0.00   \n",
       "A1CF        0.02    0.03    1.30    2.83    2.87    0.72    3.41    0.01   \n",
       "A2LD1       4.87    0.81    6.45    4.94   11.07    0.87    3.28    0.32   \n",
       "A2M         0.01   71.17    0.00    3.69    0.00   58.16    0.00    0.02   \n",
       "\n",
       "          X-1172  X-1173  ...  X-5713  X-5717  X-5727  X-5739  X-5808  X-5959  \\\n",
       "Sample                    ...                                                   \n",
       "A1BG        0.08    0.00  ...   21.24    0.46   18.70    0.00    8.30   14.41   \n",
       "A1BG-AS1    0.00    0.00  ...   15.21    0.09    3.42    0.00    8.30    8.36   \n",
       "A1CF        1.84    2.96  ...    0.01    0.45    0.00    0.01    0.00    0.00   \n",
       "A2LD1       0.61    7.10  ...    1.33    2.51    2.67    3.13    0.44    2.96   \n",
       "A2M        94.12    0.02  ...  463.32    0.04   33.35    0.00  254.97   41.52   \n",
       "\n",
       "          X-5974  X-5975  X-6030  X-6047  \n",
       "Sample                                    \n",
       "A1BG        5.09    6.35    0.09    3.22  \n",
       "A1BG-AS1    3.71    4.52    0.00    3.24  \n",
       "A1CF        0.02    0.03    4.63    0.02  \n",
       "A2LD1       0.00    1.99    2.57    1.50  \n",
       "A2M         2.32    0.00    0.00    0.16  \n",
       "\n",
       "[5 rows x 399 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX expression dataset (PDX)\n",
    "\n",
    "exprs  = pd.read_excel(raw_data_dir+\"nm.3954-S2.xlsx\",\"RNAseq_fpkm\")\n",
    "exprs.set_index(\"Sample\", inplace = True, drop = True)\n",
    "print(exprs.shape)\n",
    "exprs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1dba0322-6c74-42cf-bed7-2f6d9281d46e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(67445, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Tumor Type</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Volume (mm3)</th>\n",
       "      <th>body weight (g)</th>\n",
       "      <th>Days Post T0</th>\n",
       "      <th>% TVol Difference</th>\n",
       "      <th>% BW Difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>X-007</td>\n",
       "      <td>GC</td>\n",
       "      <td>BGJ398</td>\n",
       "      <td>202.3</td>\n",
       "      <td>21.5</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>X-007</td>\n",
       "      <td>GC</td>\n",
       "      <td>BGJ398</td>\n",
       "      <td>590.3</td>\n",
       "      <td>23.0</td>\n",
       "      <td>4</td>\n",
       "      <td>191.8</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>X-007</td>\n",
       "      <td>GC</td>\n",
       "      <td>BGJ398</td>\n",
       "      <td>796.3</td>\n",
       "      <td>22.7</td>\n",
       "      <td>7</td>\n",
       "      <td>293.6</td>\n",
       "      <td>5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>X-007</td>\n",
       "      <td>GC</td>\n",
       "      <td>BGJ398</td>\n",
       "      <td>1004.5</td>\n",
       "      <td>23.4</td>\n",
       "      <td>11</td>\n",
       "      <td>396.5</td>\n",
       "      <td>8.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>X-007</td>\n",
       "      <td>GC</td>\n",
       "      <td>BKM120</td>\n",
       "      <td>288.8</td>\n",
       "      <td>20.4</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Model Tumor Type Treatment  Volume (mm3)  body weight (g)  Days Post T0  \\\n",
       "0  X-007         GC    BGJ398         202.3             21.5             0   \n",
       "1  X-007         GC    BGJ398         590.3             23.0             4   \n",
       "2  X-007         GC    BGJ398         796.3             22.7             7   \n",
       "3  X-007         GC    BGJ398        1004.5             23.4            11   \n",
       "4  X-007         GC    BKM120         288.8             20.4             0   \n",
       "\n",
       "   % TVol Difference  % BW Difference  \n",
       "0                0.0              0.0  \n",
       "1              191.8              7.0  \n",
       "2              293.6              5.6  \n",
       "3              396.5              8.8  \n",
       "4                0.0              0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PCT raw dataset\n",
    "\n",
    "pct  = pd.read_excel(raw_data_dir+\"nm.3954-S2.xlsx\", \"PCT raw data\")\n",
    "print(pct.shape)\n",
    "pct.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a122868-0311-469f-aefe-3d72be41aca5",
   "metadata": {},
   "source": [
    "# Mapping of gene symbols to EntrezID using current gene_info file provided by NCBI:\n",
    "\n",
    "\n",
    " * Download the unzip the file  \n",
    "\\# wget ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz\n",
    "\\# gunzip Homo_sapiens.gene_info.gz;\n",
    " * Specify *hgnc_file* variable in this notebook\n",
    " * Mapping strategy\n",
    " \n",
    "1). Unknown genes and genes belonging to organisms other than H.sapiens were excluded.\n",
    "\n",
    "2). First each query symbol was tried to match with any of current \"Symbol\" directly. If the query symbol mapped to one of current symbol but has no Gene ID, the query symbol was marked as not mapped. \n",
    "\n",
    "3). If the query symbol matches none of current symbols, we tried to match it with one of \"Synonyms\". Genes matched no synonym as well as matched ambiguous synonyms correponding more than one Gene ID were condiered not mapped. \n",
    "\n",
    "4). At this point, many of not recognized symbols had LOCXXXXXXXXX format. \n",
    "According to the documntation provided by NCBI: \"Symbols beginning with LOC. When a published symbol is not available, and orthologs have not yet been determined, Gene will provide a symbol that is constructed as 'LOC' + the GeneID. This is not retained when a replacement symbol has been identified, although queries by the LOC term are still supported. In other words, a record with the symbol LOC12345 is equivalent to GeneID = 12345. So if the symbol changes, the record can still be retrieved on the web using LOC12345 as a query, or from any file using GeneID = 12345\" e.g. :\n",
    "\n",
    "    - LOC100093631 -> 100093631 \n",
    "    - LOC100129726 -> 100129726\n",
    "    - etc. \n",
    "\n",
    "Therefore all genes started with LOC were converted to Gene IDs removing \"LOC\" from query term. If resulting Gene ID matched none of current Gene IDs the symbol was considered not mapped. \n",
    "\n",
    "4). Several pairs of query gene symbols matched current symbol and synonym of the same Entrez gene ID, e.g. AGAP8 and AGAP4, ANXA8L1 and ANXA8L2, etc.\n",
    "Expressions of these genes were summarized because such genes were merged to a single gene in newer assembly versions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "510cac0f-7242-4788-8b24-9a9df7b6ec7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(48763, 9) 48763\n",
      "Index(['Approved symbol', 'Approved name', 'Status', 'Previous symbols',\n",
      "       'Alias symbols', 'Chromosome', 'Accession numbers', 'RefSeq IDs',\n",
      "       'NCBI Gene ID'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Approved symbol</th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous symbols</th>\n",
       "      <th>Alias symbols</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Accession numbers</th>\n",
       "      <th>RefSeq IDs</th>\n",
       "      <th>NCBI Gene ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HGNC:5</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NM_130786</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:37133</th>\n",
       "      <td>A1BG-AS1</td>\n",
       "      <td>A1BG antisense RNA 1</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NCRNA00181, A1BGAS, A1BG-AS</td>\n",
       "      <td>FLJ23569</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>BC040926</td>\n",
       "      <td>NR_015380</td>\n",
       "      <td>503538.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:24086</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACF, ASP, ACF64, ACF65, APOBEC1CF</td>\n",
       "      <td>10q11.23</td>\n",
       "      <td>AF271790</td>\n",
       "      <td>NM_014576</td>\n",
       "      <td>29974.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:6</th>\n",
       "      <td>A1S9T</td>\n",
       "      <td>symbol withdrawn, see [HGNC:12469](/data/gene-...</td>\n",
       "      <td>Symbol Withdrawn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HGNC:7</th>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FWP007, S863-7, CPAMD5</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>BX647329, X68728, M11313</td>\n",
       "      <td>NM_000014</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Approved symbol                                      Approved name  \\\n",
       "HGNC ID                                                                         \n",
       "HGNC:5                A1BG                             alpha-1-B glycoprotein   \n",
       "HGNC:37133        A1BG-AS1                               A1BG antisense RNA 1   \n",
       "HGNC:24086            A1CF                     APOBEC1 complementation factor   \n",
       "HGNC:6               A1S9T  symbol withdrawn, see [HGNC:12469](/data/gene-...   \n",
       "HGNC:7                 A2M                              alpha-2-macroglobulin   \n",
       "\n",
       "                      Status             Previous symbols  \\\n",
       "HGNC ID                                                     \n",
       "HGNC:5              Approved                          NaN   \n",
       "HGNC:37133          Approved  NCRNA00181, A1BGAS, A1BG-AS   \n",
       "HGNC:24086          Approved                          NaN   \n",
       "HGNC:6      Symbol Withdrawn                          NaN   \n",
       "HGNC:7              Approved                          NaN   \n",
       "\n",
       "                                Alias symbols Chromosome  \\\n",
       "HGNC ID                                                    \n",
       "HGNC:5                                    NaN   19q13.43   \n",
       "HGNC:37133                           FLJ23569   19q13.43   \n",
       "HGNC:24086  ACF, ASP, ACF64, ACF65, APOBEC1CF   10q11.23   \n",
       "HGNC:6                                    NaN        NaN   \n",
       "HGNC:7                 FWP007, S863-7, CPAMD5   12p13.31   \n",
       "\n",
       "                   Accession numbers RefSeq IDs  NCBI Gene ID  \n",
       "HGNC ID                                                        \n",
       "HGNC:5                           NaN  NM_130786           1.0  \n",
       "HGNC:37133                  BC040926  NR_015380      503538.0  \n",
       "HGNC:24086                  AF271790  NM_014576       29974.0  \n",
       "HGNC:6                           NaN        NaN           NaN  \n",
       "HGNC:7      BX647329, X68728, M11313  NM_000014           2.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read HGNC file\n",
    "\n",
    "hgnc_file = root_dir + \"HGNC/HGNC.txt\"\n",
    "\n",
    "hgnc = pd.read_csv(hgnc_file, sep = \"\\t\", index_col = 0)\n",
    "print(hgnc.shape, len(set(hgnc.index.values)))\n",
    "print(hgnc.columns)\n",
    "hgnc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "84b99fc9-e2b4-4def-8dfb-cbe86f5aafea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2212 rows with both Previous symbols and NCBI Gene ID empty\n",
      "Ok: no duplicated pairs detected\n",
      "29026 rows with empty Previous symbols were excluded\n",
      "196 Previous symbols ids mapped to no NCBI Gene ID\n",
      "85 Previous symbols mapped to multiple NCBI Gene ID\n",
      "5229 different Previous symbols mapped to the same NCBI Gene ID\n",
      "9783 Previous symbols can be mapped directly to NCBI Gene ID\n"
     ]
    }
   ],
   "source": [
    "# filter status with approved gene symbols\n",
    "\n",
    "approved = hgnc.loc[hgnc[\"Status\"] == \"Approved\",:]\n",
    "hgnc_prev = expand(approved[[\"Previous symbols\",\"NCBI Gene ID\"]], column = \"Previous symbols\", sep = \", \")\n",
    "hgnc_prev = parse_mapping_table(hgnc_prev, \"Previous symbols\", \"NCBI Gene ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "33a4f7fe-313d-4477-a79e-06ff67861558",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1899 rows with both Alias symbols and NCBI Gene ID empty\n",
      "Ok: no duplicated pairs detected\n",
      "19505 rows with empty Alias symbols were excluded\n",
      "689 Alias symbols ids mapped to no NCBI Gene ID\n",
      "1033 Alias symbols mapped to multiple NCBI Gene ID\n",
      "29209 different Alias symbols mapped to the same NCBI Gene ID\n",
      "11274 Alias symbols can be mapped directly to NCBI Gene ID\n"
     ]
    }
   ],
   "source": [
    "# find gene symbols with aliases\n",
    "\n",
    "hgnc_syn = expand(approved[[\"Alias symbols\",\"NCBI Gene ID\"]], column = \"Alias symbols\", sep = \", \")\n",
    "hgnc_syn = parse_mapping_table(hgnc_syn, \"Alias symbols\", \"NCBI Gene ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "414a951d-9890-4b38-8ed1-731426324389",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Symbol rows are not empty.\n",
      "Ok: All Symbol are mapped to GeneID\n",
      "12 Symbol mapped to multiple GeneID\n",
      "Ok: All GeneID are unique\n",
      "140871 Symbol can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "# map gene symbols to gene id\n",
    "\n",
    "NCBI = pd.read_csv(root_dir + \"Homo_sapiens.gene_info/Homo_sapiens.gene_info.txt\", sep = \"\\t\")\n",
    "NCBI.columns\n",
    "NCBI = NCBI[[\"#tax_id\",\"GeneID\",\"Symbol\",\"Synonyms\",\"type_of_gene\"]]\n",
    "NCBI = NCBI.loc[NCBI[\"#tax_id\"] == 9606]\n",
    "NCBI = NCBI.loc[NCBI[\"type_of_gene\"] != \"unknown\"]\n",
    "ncbi_symbols = parse_mapping_table(NCBI, \"Symbol\", \"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f387662-cdb9-43b9-9e44-97f01b94ef52",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ok: no empty rows detected\n",
      "Ok: no duplicated pairs detected\n",
      "Ok: All Synonyms rows are not empty.\n",
      "Ok: All Synonyms are mapped to GeneID\n",
      "3376 Synonyms mapped to multiple GeneID\n",
      "53208 different Synonyms mapped to the same GeneID\n",
      "11014 Synonyms can be mapped directly to GeneID\n"
     ]
    }
   ],
   "source": [
    "# map synonyms to gene id\n",
    "\n",
    "ncbi_synonyms = expand(NCBI[[\"Synonyms\",\"GeneID\"]], column = \"Synonyms\", sep = \"|\")\n",
    "ncbi_synonyms = parse_mapping_table(ncbi_synonyms, \"Synonyms\", \"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9a87b90-cebf-4822-9855-69e33402f4a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped: 22416 \n",
      "\tdirectly via main_mapper 19626 \n",
      "\tvia alternative mapper 548 \n",
      "\tvia one of multiple synonyms in alternative mapper 1426 \n",
      "\tLOC 816 \n",
      "Unmapped: 249 \n",
      "\trecognized symbols without Entrez ID 0 \n",
      "\tmultiple query_ids map to the same target_id 0 \n",
      "\tquery_ids map to multiple target_ids in the main mapper 0 \n",
      "\tquery_ids map to multiple target_ids in the alternative mapper 80 \n",
      "\tLOC not found in Entrez 34 \n",
      "\tNot found at all: 135\n",
      "Warning: query IDs mapping to duplicated target IDs in mapping table: 75\n",
      "Warning: query IDs not mapped to any target IDs excluded: 249\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5974</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6030</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.75</td>\n",
       "      <td>8.97</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.08</td>\n",
       "      <td>17.52</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.41</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>21.24</td>\n",
       "      <td>0.46</td>\n",
       "      <td>18.70</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8.30</td>\n",
       "      <td>14.41</td>\n",
       "      <td>5.09</td>\n",
       "      <td>6.35</td>\n",
       "      <td>0.09</td>\n",
       "      <td>3.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.01</td>\n",
       "      <td>71.17</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.69</td>\n",
       "      <td>0.00</td>\n",
       "      <td>58.16</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>94.12</td>\n",
       "      <td>0.02</td>\n",
       "      <td>...</td>\n",
       "      <td>463.32</td>\n",
       "      <td>0.04</td>\n",
       "      <td>33.35</td>\n",
       "      <td>0.00</td>\n",
       "      <td>254.97</td>\n",
       "      <td>41.52</td>\n",
       "      <td>2.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1.47</td>\n",
       "      <td>0.74</td>\n",
       "      <td>5.64</td>\n",
       "      <td>1.98</td>\n",
       "      <td>9.98</td>\n",
       "      <td>3.71</td>\n",
       "      <td>12.41</td>\n",
       "      <td>2.57</td>\n",
       "      <td>1.38</td>\n",
       "      <td>4.77</td>\n",
       "      <td>...</td>\n",
       "      <td>1.19</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0.93</td>\n",
       "      <td>1.27</td>\n",
       "      <td>1.99</td>\n",
       "      <td>3.34</td>\n",
       "      <td>0.27</td>\n",
       "      <td>1.64</td>\n",
       "      <td>11.12</td>\n",
       "      <td>1.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.99</td>\n",
       "      <td>3.06</td>\n",
       "      <td>9.65</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.11</td>\n",
       "      <td>4.48</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.92</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    X-1004  X-1008  X-1027  X-1095  X-1119  X-1156  X-1167  X-1169  X-1172  \\\n",
       "1     2.75    8.97    0.14    0.13    0.08   17.52    0.18    0.41    0.08   \n",
       "2     0.01   71.17    0.00    3.69    0.00   58.16    0.00    0.02   94.12   \n",
       "3     0.00    0.01    0.00    0.00    0.00    0.10    0.00    0.00    0.02   \n",
       "9     1.47    0.74    5.64    1.98    9.98    3.71   12.41    2.57    1.38   \n",
       "10    0.00    0.00    0.99    3.06    9.65    0.00    2.31    0.00    0.11   \n",
       "\n",
       "    X-1173  ...  X-5713  X-5717  X-5727  X-5739  X-5808  X-5959  X-5974  \\\n",
       "1     0.00  ...   21.24    0.46   18.70    0.00    8.30   14.41    5.09   \n",
       "2     0.02  ...  463.32    0.04   33.35    0.00  254.97   41.52    2.32   \n",
       "3     0.00  ...    0.03    0.10    0.01    0.02    0.01    0.00    0.00   \n",
       "9     4.77  ...    1.19    1.93    0.93    1.27    1.99    3.34    0.27   \n",
       "10    4.48  ...    0.02    0.00    0.00    0.00    0.02    0.00    0.03   \n",
       "\n",
       "    X-5975  X-6030  X-6047  \n",
       "1     6.35    0.09    3.22  \n",
       "2     0.00    0.00    0.16  \n",
       "3     0.01    0.00    0.02  \n",
       "9     1.64   11.12    1.60  \n",
       "10    0.16    1.92    0.04  \n",
       "\n",
       "[5 rows x 399 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show mappend and unmapped gene ids\n",
    "\n",
    "exprss = apply_mappers(exprs, ncbi_symbols, ncbi_synonyms, verbose = True, handle_duplicates = \"sum\")\n",
    "df = pd.DataFrame.from_records(exprss[0])\n",
    "df.set_index(exprss[0].index.values, inplace= True)\n",
    "exprs = df\n",
    "exprs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7cb7d15-7ecb-4987-b66a-2f26dabee430",
   "metadata": {},
   "source": [
    "### FPKM to TPM conversion\n",
    "Let $X_i$ is a number of fragments mapped to a transcript, $N$ is a total number of fragments sequensed (and assigned to any transcript) and $\\widetilde{l_i}$  is an effective length of a transcript (i.e. how many fragments with average length $\\mu_{frag}$ can generate a transcript of length $l_i$: $\\widetilde{l_i} = l_i - \\mu_{frag}+1$)\n",
    "\n",
    "*FPKM* - fragments per kilobase of exon (i.e. effective length) per million reads mapped\n",
    "\n",
    "$FPKM_i = \\frac{X_i}{\\frac{$\\widetilde{l_i}}{10^3}*\\frac{N}{10^6}} = \\frac{X_i}{l_iN}*10^9 $\n",
    "\n",
    "*TPM* - transcripts per million of transcripts.\n",
    "\n",
    "In turn, $\\frac{X_i}{\\widetilde{l_i}}$ is estimated number of transcripts \n",
    "\n",
    "$TPM_i = \\frac{\\frac{X_i}{\\widetilde{l_i}}}{\\sum_j{\\frac{X_j}{\\widetilde{l_j}}}}*10^6 $\n",
    "\n",
    "\n",
    "### how to convert FPKM to TPM\n",
    "Divide both numerator and denominator by $N$ and mutiply by $10^9$:\n",
    "\n",
    "$TPM_i = \\frac{\\frac{X_i}{N\\widetilde{l_i}}*10^9}{\\sum_j{\\frac{X_j}{N\\widetilde{l_j}}*10^9}}*10^6 = \\frac{FPKM_i}{\\sum_j{FPKM_j}} * 10^6$\n",
    "\n",
    "Sources:\n",
    " - What the FPKM? https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/\n",
    " - https://www.biostars.org/p/160989/\n",
    " - Lior Pachter https://arxiv.org/pdf/1104.3889.pdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7fa1d138-5dbc-4194-a049-cdf8669f8855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "X-1004    387513.24\n",
       "X-1008    488175.21\n",
       "X-1027    722266.30\n",
       "X-1095    448322.63\n",
       "X-1119    608236.65\n",
       "dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##  FPKM convert to log2(TPM+1)\n",
    "\n",
    "sum_fpkm = exprs.apply(sum, axis = 0)\n",
    "sum_fpkm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed80e41a-8f9e-4bc3-96c7-16fa80c8e8b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5974</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6030</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8.096532</td>\n",
       "      <td>19.374550</td>\n",
       "      <td>1.193834</td>\n",
       "      <td>1.289970</td>\n",
       "      <td>1.131528</td>\n",
       "      <td>33.732665</td>\n",
       "      <td>1.364690</td>\n",
       "      <td>1.780637</td>\n",
       "      <td>1.165838</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>44.914047</td>\n",
       "      <td>1.791006</td>\n",
       "      <td>35.027009</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>17.015718</td>\n",
       "      <td>28.605711</td>\n",
       "      <td>11.796351</td>\n",
       "      <td>12.305588</td>\n",
       "      <td>1.170204</td>\n",
       "      <td>7.070138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.025806</td>\n",
       "      <td>146.787821</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>9.230680</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>109.660492</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.038080</td>\n",
       "      <td>196.108029</td>\n",
       "      <td>1.030471</td>\n",
       "      <td>...</td>\n",
       "      <td>958.921680</td>\n",
       "      <td>1.068783</td>\n",
       "      <td>61.684532</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>492.991273</td>\n",
       "      <td>80.541231</td>\n",
       "      <td>5.920930</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.301622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.020484</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.186830</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.041459</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.062025</td>\n",
       "      <td>1.171958</td>\n",
       "      <td>1.018196</td>\n",
       "      <td>1.041289</td>\n",
       "      <td>1.019296</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.017804</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.037703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4.793419</td>\n",
       "      <td>2.515849</td>\n",
       "      <td>8.808754</td>\n",
       "      <td>5.416462</td>\n",
       "      <td>17.408087</td>\n",
       "      <td>7.931403</td>\n",
       "      <td>26.143367</td>\n",
       "      <td>5.893263</td>\n",
       "      <td>3.860700</td>\n",
       "      <td>8.267369</td>\n",
       "      <td>...</td>\n",
       "      <td>3.460344</td>\n",
       "      <td>4.318785</td>\n",
       "      <td>2.692252</td>\n",
       "      <td>3.621875</td>\n",
       "      <td>4.839913</td>\n",
       "      <td>7.398548</td>\n",
       "      <td>1.572694</td>\n",
       "      <td>3.919868</td>\n",
       "      <td>22.029635</td>\n",
       "      <td>4.016218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.370686</td>\n",
       "      <td>7.825442</td>\n",
       "      <td>16.865535</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.680192</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.228027</td>\n",
       "      <td>7.825538</td>\n",
       "      <td>...</td>\n",
       "      <td>1.041350</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.038592</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.063633</td>\n",
       "      <td>1.284865</td>\n",
       "      <td>4.631016</td>\n",
       "      <td>1.075405</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      X-1004      X-1008    X-1027    X-1095     X-1119      X-1156  \\\n",
       "1   8.096532   19.374550  1.193834  1.289970   1.131528   33.732665   \n",
       "2   1.025806  146.787821  1.000000  9.230680   1.000000  109.660492   \n",
       "3   1.000000    1.020484  1.000000  1.000000   1.000000    1.186830   \n",
       "9   4.793419    2.515849  8.808754  5.416462  17.408087    7.931403   \n",
       "10  1.000000    1.000000  2.370686  7.825442  16.865535    1.000000   \n",
       "\n",
       "       X-1167    X-1169      X-1172    X-1173  ...      X-5713    X-5717  \\\n",
       "1    1.364690  1.780637    1.165838  1.000000  ...   44.914047  1.791006   \n",
       "2    1.000000  1.038080  196.108029  1.030471  ...  958.921680  1.068783   \n",
       "3    1.000000  1.000000    1.041459  1.000000  ...    1.062025  1.171958   \n",
       "9   26.143367  5.893263    3.860700  8.267369  ...    3.460344  4.318785   \n",
       "10   5.680192  1.000000    1.228027  7.825538  ...    1.041350  1.000000   \n",
       "\n",
       "       X-5727    X-5739      X-5808     X-5959     X-5974     X-5975  \\\n",
       "1   35.027009  1.000000   17.015718  28.605711  11.796351  12.305588   \n",
       "2   61.684532  1.000000  492.991273  80.541231   5.920930   1.000000   \n",
       "3    1.018196  1.041289    1.019296   1.000000   1.000000   1.017804   \n",
       "9    2.692252  3.621875    4.839913   7.398548   1.572694   3.919868   \n",
       "10   1.000000  1.000000    1.038592   1.000000   1.063633   1.284865   \n",
       "\n",
       "       X-6030    X-6047  \n",
       "1    1.170204  7.070138  \n",
       "2    1.000000  1.301622  \n",
       "3    1.000000  1.037703  \n",
       "9   22.029635  4.016218  \n",
       "10   4.631016  1.075405  \n",
       "\n",
       "[5 rows x 399 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert FPKM to TPM\n",
    "\n",
    "tpm = exprs / sum_fpkm * 1000000 + 1\n",
    "tpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "859b7c7b-c0bd-4772-9734-3d211dbc9741",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(22377, 399)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X-1004</th>\n",
       "      <th>X-1008</th>\n",
       "      <th>X-1027</th>\n",
       "      <th>X-1095</th>\n",
       "      <th>X-1119</th>\n",
       "      <th>X-1156</th>\n",
       "      <th>X-1167</th>\n",
       "      <th>X-1169</th>\n",
       "      <th>X-1172</th>\n",
       "      <th>X-1173</th>\n",
       "      <th>...</th>\n",
       "      <th>X-5713</th>\n",
       "      <th>X-5717</th>\n",
       "      <th>X-5727</th>\n",
       "      <th>X-5739</th>\n",
       "      <th>X-5808</th>\n",
       "      <th>X-5959</th>\n",
       "      <th>X-5974</th>\n",
       "      <th>X-5975</th>\n",
       "      <th>X-6030</th>\n",
       "      <th>X-6047</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.017304</td>\n",
       "      <td>4.276091</td>\n",
       "      <td>0.255603</td>\n",
       "      <td>0.367337</td>\n",
       "      <td>0.178272</td>\n",
       "      <td>5.076074</td>\n",
       "      <td>0.448574</td>\n",
       "      <td>0.832394</td>\n",
       "      <td>0.221367</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>5.489095</td>\n",
       "      <td>0.840770</td>\n",
       "      <td>5.130396</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.088796</td>\n",
       "      <td>4.838231</td>\n",
       "      <td>3.560269</td>\n",
       "      <td>3.621242</td>\n",
       "      <td>0.226760</td>\n",
       "      <td>2.821738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.036757</td>\n",
       "      <td>7.197588</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.206437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.776900</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053917</td>\n",
       "      <td>7.615505</td>\n",
       "      <td>0.043304</td>\n",
       "      <td>...</td>\n",
       "      <td>9.905269</td>\n",
       "      <td>0.095969</td>\n",
       "      <td>5.946837</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.945418</td>\n",
       "      <td>6.331656</td>\n",
       "      <td>2.565824</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.380310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.029254</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.247114</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.058607</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.086818</td>\n",
       "      <td>0.228921</td>\n",
       "      <td>0.026016</td>\n",
       "      <td>0.058371</td>\n",
       "      <td>0.027573</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025460</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.053393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.261055</td>\n",
       "      <td>1.331045</td>\n",
       "      <td>3.138938</td>\n",
       "      <td>2.437351</td>\n",
       "      <td>4.121686</td>\n",
       "      <td>2.987576</td>\n",
       "      <td>4.708373</td>\n",
       "      <td>2.559067</td>\n",
       "      <td>1.948862</td>\n",
       "      <td>3.047428</td>\n",
       "      <td>...</td>\n",
       "      <td>1.790916</td>\n",
       "      <td>2.110625</td>\n",
       "      <td>1.428814</td>\n",
       "      <td>1.856737</td>\n",
       "      <td>2.274981</td>\n",
       "      <td>2.887242</td>\n",
       "      <td>0.653238</td>\n",
       "      <td>1.970805</td>\n",
       "      <td>4.461374</td>\n",
       "      <td>2.005837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.245304</td>\n",
       "      <td>2.968172</td>\n",
       "      <td>4.076006</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.505940</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.296342</td>\n",
       "      <td>2.968190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058455</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054629</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.089000</td>\n",
       "      <td>0.361617</td>\n",
       "      <td>2.211329</td>\n",
       "      <td>0.104881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 399 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      X-1004    X-1008    X-1027    X-1095    X-1119    X-1156    X-1167  \\\n",
       "1   3.017304  4.276091  0.255603  0.367337  0.178272  5.076074  0.448574   \n",
       "2   0.036757  7.197588  0.000000  3.206437  0.000000  6.776900  0.000000   \n",
       "3   0.000000  0.029254  0.000000  0.000000  0.000000  0.247114  0.000000   \n",
       "9   2.261055  1.331045  3.138938  2.437351  4.121686  2.987576  4.708373   \n",
       "10  0.000000  0.000000  1.245304  2.968172  4.076006  0.000000  2.505940   \n",
       "\n",
       "      X-1169    X-1172    X-1173  ...    X-5713    X-5717    X-5727    X-5739  \\\n",
       "1   0.832394  0.221367  0.000000  ...  5.489095  0.840770  5.130396  0.000000   \n",
       "2   0.053917  7.615505  0.043304  ...  9.905269  0.095969  5.946837  0.000000   \n",
       "3   0.000000  0.058607  0.000000  ...  0.086818  0.228921  0.026016  0.058371   \n",
       "9   2.559067  1.948862  3.047428  ...  1.790916  2.110625  1.428814  1.856737   \n",
       "10  0.000000  0.296342  2.968190  ...  0.058455  0.000000  0.000000  0.000000   \n",
       "\n",
       "      X-5808    X-5959    X-5974    X-5975    X-6030    X-6047  \n",
       "1   4.088796  4.838231  3.560269  3.621242  0.226760  2.821738  \n",
       "2   8.945418  6.331656  2.565824  0.000000  0.000000  0.380310  \n",
       "3   0.027573  0.000000  0.000000  0.025460  0.000000  0.053393  \n",
       "9   2.274981  2.887242  0.653238  1.970805  4.461374  2.005837  \n",
       "10  0.054629  0.000000  0.089000  0.361617  2.211329  0.104881  \n",
       "\n",
       "[5 rows x 399 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take logarithm and save the file\n",
    "\n",
    "tpm = tpm.applymap(np.log2)\n",
    "tpm.to_csv(preprocessing_data_dir + \"exprs/PDX.FPKM2TPMplus1log2.Expr.tsv\",sep=\"\\t\")\n",
    "print(tpm.shape)\n",
    "tpm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b5daea0-2e78-40da-a78f-9b691346ecb2",
   "metadata": {},
   "source": [
    "# TCGA \n",
    "\n",
    "* from http://gdac.broadinstitute.org/runs/stddata__2016_01_18/data/ download RSEM files, \"scaled estimate\" per gene. \n",
    "\n",
    "* RSEM Scaled estimate is an aboundance of a transcript divided by sum of aboundance over all the transcripts.  Therefore $TPM_i=ScaledEstimate_i*10^6$\n",
    "\n",
    "* Resulted TPM were log2-transformed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c3c5033b-bde4-460a-9078-1f86fcb70fdd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THYM (20531, 122)\n",
      "THCA (20531, 568)\n",
      "MESO (20531, 87)\n",
      "UCS (20531, 57)\n",
      "UVM (20531, 80)\n",
      "PRAD (20531, 550)\n",
      "STES (20531, 646)\n",
      "LUAD (20531, 576)\n",
      "LUSC (20531, 552)\n",
      "SKCM (20531, 473)\n",
      "PAAD (20531, 183)\n",
      "OV (20531, 307)\n",
      "STAD (20531, 450)\n",
      "LIHC (20531, 423)\n",
      "UCEC (20531, 381)\n",
      "SARC (20531, 265)\n",
      "PCPG (20531, 187)\n",
      "READ (20531, 72)\n",
      "TGCT (20531, 156)\n",
      "ACC (20531, 79)\n",
      "CHOL (20531, 45)\n",
      "COAD (20531, 191)\n",
      "CESC (20531, 309)\n",
      "DLBC (20531, 48)\n",
      "BLCA (20531, 427)\n",
      "COADREAD (20531, 263)\n",
      "ESCA (20531, 196)\n",
      "GBM (20531, 171)\n",
      "KICH (20531, 91)\n",
      "HNSC (20531, 566)\n",
      "BRCA (20531, 1212)\n",
      "GBMLGG (20531, 701)\n",
      "LAML (20531, 173)\n",
      "KIRP (20531, 323)\n",
      "KIPAN (20531, 1020)\n",
      "KIRC (20531, 606)\n",
      "LGG (20531, 530)\n",
      "LIHC (20531, 423)\n",
      "MESO (20531, 87)\n",
      "PAAD (20531, 183)\n",
      "OV (20531, 307)\n",
      "LUSC (20531, 552)\n",
      "LUAD (20531, 576)\n",
      "PCPG (20531, 187)\n",
      "READ (20531, 72)\n",
      "SARC (20531, 265)\n",
      "PRAD (20531, 550)\n",
      "STAD (20531, 450)\n",
      "SKCM (20531, 473)\n",
      "TGCT (20531, 156)\n",
      "THYM (20531, 122)\n",
      "UCS (20531, 57)\n",
      "STES (20531, 646)\n",
      "UCEC (20531, 381)\n",
      "THCA (20531, 568)\n",
      "UVM (20531, 80)\n"
     ]
    }
   ],
   "source": [
    "# save TCGA expression datasets as to cohorts\n",
    "\n",
    "f_ext = \".rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes__data.data.txt\"\n",
    "for fpath in os.listdir(tcga_tmp_dir):\n",
    "    if fpath.startswith(\"gdac.broadinstitute.org\") and not fpath.endswith(\".tar.gz\"):\n",
    "        fpath = fpath.replace(\".tar\",\"\")\n",
    "        cohort = fpath.split(\"_\")[1].replace(\".Merge\",\"\")\n",
    "        #print(fpath, cohort)\n",
    "        fname = cohort + f_ext\n",
    "        try:\n",
    "            exprs = pd.read_csv(tcga_tmp_dir+\"/\"+fpath+\"/\"+fname,sep=\"\\t\",index_col=0, low_memory=False)\n",
    "            # drop \"gene_id\" and keep only \"scaled_estimate\" columns\n",
    "            exprs = exprs.loc[:,exprs.T.loc[exprs.T[\"gene_id\"] == \"scaled_estimate\", :].index]\n",
    "            exprs = exprs.loc[exprs.index != \"gene_id\",]\n",
    "            exprs.rename(index = lambda x: int(x.split(\"|\")[1]),\n",
    "                         columns = lambda x: x.replace(\".1\",\"\"), inplace = True)\n",
    "            exprs.index.name = \"ENTREZID\"\n",
    "            # convert scaled_extimates to log2(TPM+1)\n",
    "            exprs = exprs.applymap(lambda x: np.log2(float(x) * 1000000 + 1))\n",
    "            exprs = exprs.sort_index()\n",
    "            exprs.to_csv(preprocessing_data_dir + \"exprs/\" +\"TCGA-\" + cohort + \"_exprs.RSEMscaled_est2TPMplus1log2.tsv\",\n",
    "                         sep = \"\\t\")\n",
    "            print(cohort, exprs.shape)\n",
    "            \n",
    "        except:\n",
    "            print(cohort, \"No expression data\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
