{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import libraries\n",
    "\n",
    "import os\n",
    "import torch \n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import math\n",
    "import mygene\n",
    "import sklearn.preprocessing as sk\n",
    "import seaborn as sns\n",
    "from sklearn import metrics\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import classification_report\n",
    "from torch.utils.data.sampler import WeightedRandomSampler\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import average_precision_score, precision_recall_curve\n",
    "import random\n",
    "import warnings\n",
    "from random import randint\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define directories\n",
    "\n",
    "cell_line_dir = \"/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC/\"\n",
    "models_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_Mutation_CNA/Models/GDSC_PDX_TCGA_Second_Strategy/\"\n",
    "finetuned_dir = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_Mutation_CNA/Models/GDSC_PDX_TCGA_Second_Strategy/Second_Strategy_Finetuned_Models/\"\n",
    "DEGs_dir = \"/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/DNA_REPi/\"\n",
    "dataset_dir = \"/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/\"\n",
    "pdx_data_dir = \"/Volumes/Expansion/Thesis Work/Datasets/PDX/Expression/\"\n",
    "save_results_to = \"/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_Mutation_CNA/Predictions/GDSC_PDX_TCGA_Second_Strategy/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set random seeds\n",
    "\n",
    "torch.manual_seed(42)\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Supplementary Files/GDSC\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(cell_line_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>Whole Exome Sequencing (WES)</th>\n",
       "      <th>Copy Number Alterations (CNA)</th>\n",
       "      <th>Gene Expression</th>\n",
       "      <th>Methylation</th>\n",
       "      <th>Drug\\nResponse</th>\n",
       "      <th>GDSC\\nTissue descriptor 1</th>\n",
       "      <th>GDSC\\nTissue\\ndescriptor 2</th>\n",
       "      <th>Cancer Type\\n(matching TCGA label)</th>\n",
       "      <th>Microsatellite \\ninstability Status (MSI)</th>\n",
       "      <th>Screen Medium</th>\n",
       "      <th>Growth Properties</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COSMIC identifier</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1298218</th>\n",
       "      <td>KP-2</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>pancreas</td>\n",
       "      <td>PAAD</td>\n",
       "      <td></td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330932</th>\n",
       "      <td>KO52</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>leukemia</td>\n",
       "      <td>acute_myeloid_leukaemia</td>\n",
       "      <td>LAML</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331030</th>\n",
       "      <td>SC-1</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>lymphoma</td>\n",
       "      <td>B_cell_lymphoma</td>\n",
       "      <td>DLBC</td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>R</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1503373</th>\n",
       "      <td>U-CH2</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>bone</td>\n",
       "      <td>bone_other</td>\n",
       "      <td></td>\n",
       "      <td>MSS/MSI-L</td>\n",
       "      <td>D/F12</td>\n",
       "      <td>Adherent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>TOTAL:</td>\n",
       "      <td>1001</td>\n",
       "      <td>996</td>\n",
       "      <td>968</td>\n",
       "      <td>957</td>\n",
       "      <td>990</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Sample Name Whole Exome Sequencing (WES)  \\\n",
       "COSMIC identifier                                            \n",
       "1298218                  KP-2                            Y   \n",
       "1330932                  KO52                            Y   \n",
       "1331030                  SC-1                            Y   \n",
       "1503373                 U-CH2                            Y   \n",
       "                       TOTAL:                         1001   \n",
       "\n",
       "                  Copy Number Alterations (CNA) Gene Expression Methylation  \\\n",
       "COSMIC identifier                                                             \n",
       "1298218                                       N               Y           N   \n",
       "1330932                                       Y               N           N   \n",
       "1331030                                       Y               N           N   \n",
       "1503373                                       Y               N           N   \n",
       "                                            996             968         957   \n",
       "\n",
       "                  Drug\\nResponse GDSC\\nTissue descriptor 1  \\\n",
       "COSMIC identifier                                            \n",
       "1298218                        Y                  pancreas   \n",
       "1330932                        N                  leukemia   \n",
       "1331030                        N                  lymphoma   \n",
       "1503373                        N                      bone   \n",
       "                             990                             \n",
       "\n",
       "                  GDSC\\nTissue\\ndescriptor 2  \\\n",
       "COSMIC identifier                              \n",
       "1298218                             pancreas   \n",
       "1330932              acute_myeloid_leukaemia   \n",
       "1331030                      B_cell_lymphoma   \n",
       "1503373                           bone_other   \n",
       "                                               \n",
       "\n",
       "                  Cancer Type\\n(matching TCGA label)  \\\n",
       "COSMIC identifier                                      \n",
       "1298218                                         PAAD   \n",
       "1330932                                         LAML   \n",
       "1331030                                         DLBC   \n",
       "1503373                                                \n",
       "                                                       \n",
       "\n",
       "                  Microsatellite \\ninstability Status (MSI) Screen Medium  \\\n",
       "COSMIC identifier                                                           \n",
       "1298218                                                             D/F12   \n",
       "1330932                                           MSS/MSI-L         D/F12   \n",
       "1331030                                           MSS/MSI-L             R   \n",
       "1503373                                           MSS/MSI-L         D/F12   \n",
       "                                                                            \n",
       "\n",
       "                  Growth Properties  \n",
       "COSMIC identifier                    \n",
       "1298218                    Adherent  \n",
       "1330932                          NA  \n",
       "1331030                          NA  \n",
       "1503373                    Adherent  \n",
       "                                     "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read cell line details table in GDSC\n",
    "\n",
    "GDSC_cell_line_details = pd.read_excel(\"GDSC_Cell_Lines_Details.xlsx\", keep_default_na = False)\n",
    "GDSC_cell_line_details.set_index(\"COSMIC identifier\", inplace = True)\n",
    "GDSC_cell_line_details.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC expression dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/exprs/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>107984199</th>\n",
       "      <th>107984208</th>\n",
       "      <th>107984325</th>\n",
       "      <th>107984411</th>\n",
       "      <th>107984648</th>\n",
       "      <th>107984889</th>\n",
       "      <th>107984989</th>\n",
       "      <th>107986313</th>\n",
       "      <th>107986782</th>\n",
       "      <th>107986810</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>-0.4711563767951986</td>\n",
       "      <td>-0.1857522957766964</td>\n",
       "      <td>0.9702409548662462</td>\n",
       "      <td>-0.4081281805960821</td>\n",
       "      <td>-0.4161511169999803</td>\n",
       "      <td>-0.4384159723467665</td>\n",
       "      <td>-1.154692246005183</td>\n",
       "      <td>0.5262132484107975</td>\n",
       "      <td>-0.7950924576033422</td>\n",
       "      <td>-0.6337442716864214</td>\n",
       "      <td>...</td>\n",
       "      <td>1.283534690590172</td>\n",
       "      <td>-0.3367265464586691</td>\n",
       "      <td>-0.0039762988723447</td>\n",
       "      <td>0.7378729872604209</td>\n",
       "      <td>0.9757614264575688</td>\n",
       "      <td>0.1494417647095801</td>\n",
       "      <td>2.062294063911948</td>\n",
       "      <td>-0.8140636761719959</td>\n",
       "      <td>0.2870333828833719</td>\n",
       "      <td>0.9524265085537148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684055</th>\n",
       "      <td>1.3565140948247052</td>\n",
       "      <td>-0.277542411913307</td>\n",
       "      <td>0.0887087882196536</td>\n",
       "      <td>-0.2754829982302413</td>\n",
       "      <td>-0.5153856346587746</td>\n",
       "      <td>-0.6248712076433911</td>\n",
       "      <td>0.1243408990866293</td>\n",
       "      <td>-1.2699667684674136</td>\n",
       "      <td>2.0788360886382717</td>\n",
       "      <td>2.5197991997459197</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.5571028630154293</td>\n",
       "      <td>-1.724090733574364</td>\n",
       "      <td>-0.3749080365068901</td>\n",
       "      <td>-0.0016055431935905</td>\n",
       "      <td>-0.3788535663547365</td>\n",
       "      <td>-1.3502092798315848</td>\n",
       "      <td>0.6949790577550328</td>\n",
       "      <td>0.4333032198982747</td>\n",
       "      <td>-0.1280986270308099</td>\n",
       "      <td>-1.629249958712223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684057</th>\n",
       "      <td>0.6510004960254864</td>\n",
       "      <td>1.6578876382433665</td>\n",
       "      <td>-0.4977106390881842</td>\n",
       "      <td>-0.1212172092933644</td>\n",
       "      <td>-0.398455281697677</td>\n",
       "      <td>-0.2804901106672752</td>\n",
       "      <td>0.6455898116420014</td>\n",
       "      <td>0.215833091774836</td>\n",
       "      <td>-0.1723595788875009</td>\n",
       "      <td>-0.2170870770552223</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.024863133395904</td>\n",
       "      <td>0.4073439017665206</td>\n",
       "      <td>-0.4440356638831038</td>\n",
       "      <td>0.8630347078663075</td>\n",
       "      <td>1.4486171428489227</td>\n",
       "      <td>-1.4777214830771732</td>\n",
       "      <td>0.5220400480464272</td>\n",
       "      <td>0.6441211958947753</td>\n",
       "      <td>0.9057518606131394</td>\n",
       "      <td>0.3018783753479829</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18597 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID            1                    2                    9          \\\n",
       "683665    -0.4711563767951986  -0.1857522957766964   0.9702409548662462   \n",
       "684055     1.3565140948247052   -0.277542411913307   0.0887087882196536   \n",
       "684057     0.6510004960254864   1.6578876382433665  -0.4977106390881842   \n",
       "\n",
       "ENTREZID            10                   12                   13         \\\n",
       "683665    -0.4081281805960821  -0.4161511169999803  -0.4384159723467665   \n",
       "684055    -0.2754829982302413  -0.5153856346587746  -0.6248712076433911   \n",
       "684057    -0.1212172092933644   -0.398455281697677  -0.2804901106672752   \n",
       "\n",
       "ENTREZID           14                   15                   16         \\\n",
       "683665    -1.154692246005183   0.5262132484107975  -0.7950924576033422   \n",
       "684055    0.1243408990866293  -1.2699667684674136   2.0788360886382717   \n",
       "684057    0.6455898116420014    0.215833091774836  -0.1723595788875009   \n",
       "\n",
       "ENTREZID            18         ...            107984199            107984208  \\\n",
       "683665    -0.6337442716864214  ...    1.283534690590172  -0.3367265464586691   \n",
       "684055     2.5197991997459197  ...  -0.5571028630154293   -1.724090733574364   \n",
       "684057    -0.2170870770552223  ...   -0.024863133395904   0.4073439017665206   \n",
       "\n",
       "ENTREZID            107984325            107984411            107984648  \\\n",
       "683665    -0.0039762988723447   0.7378729872604209   0.9757614264575688   \n",
       "684055    -0.3749080365068901  -0.0016055431935905  -0.3788535663547365   \n",
       "684057    -0.4440356638831038   0.8630347078663075   1.4486171428489227   \n",
       "\n",
       "ENTREZID            107984889           107984989            107986313  \\\n",
       "683665     0.1494417647095801   2.062294063911948  -0.8140636761719959   \n",
       "684055    -1.3502092798315848  0.6949790577550328   0.4333032198982747   \n",
       "684057    -1.4777214830771732  0.5220400480464272   0.6441211958947753   \n",
       "\n",
       "ENTREZID            107986782           107986810  \n",
       "683665     0.2870333828833719  0.9524265085537148  \n",
       "684055    -0.1280986270308099  -1.629249958712223  \n",
       "684057     0.9057518606131394  0.3018783753479829  \n",
       "\n",
       "[3 rows x 18597 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC expression dataset (DNA_REPi)\n",
    "\n",
    "GDSCE = pd.read_csv(\"GDSC_exprs.z.DNA_REPi.tsv\", \n",
    "                    sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "GDSCE = pd.DataFrame.transpose(GDSCE)\n",
    "GDSCE.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/mutations\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC mutation dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/mutations/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>143872</th>\n",
       "      <th>728577</th>\n",
       "      <th>2</th>\n",
       "      <th>2050</th>\n",
       "      <th>10243</th>\n",
       "      <th>79365</th>\n",
       "      <th>256006</th>\n",
       "      <th>4609</th>\n",
       "      <th>118788</th>\n",
       "      <th>113675</th>\n",
       "      <th>...</th>\n",
       "      <th>26517</th>\n",
       "      <th>28972</th>\n",
       "      <th>53917</th>\n",
       "      <th>283507</th>\n",
       "      <th>514</th>\n",
       "      <th>339210</th>\n",
       "      <th>64981</th>\n",
       "      <th>7335</th>\n",
       "      <th>5739</th>\n",
       "      <th>83442</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684055</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684057</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684059</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684062</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 18377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        143872  728577  2       2050    10243   79365   256006  4609    \\\n",
       "683665     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684055     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684057     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684059     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684062     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        118788  113675  ...  26517   28972   53917   283507  514     339210  \\\n",
       "683665     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684055     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684057     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684059     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "684062     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "        64981   7335    5739    83442   \n",
       "683665     0.0     0.0     0.0     0.0  \n",
       "684055     0.0     0.0     0.0     0.0  \n",
       "684057     0.0     0.0     0.0     0.0  \n",
       "684059     0.0     0.0     0.0     0.0  \n",
       "684062     0.0     0.0     0.0     0.0  \n",
       "\n",
       "[5 rows x 18377 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC mutation dataset (DNA_REPi)\n",
    "\n",
    "GDSCM = pd.read_csv(\"GDSC_mutations.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\", index_col=0, decimal=\".\")\n",
    "GDSCM.drop_duplicates(keep='last')\n",
    "GDSCM = pd.DataFrame.transpose(GDSCM)\n",
    "GDSCM = GDSCM.loc[:, ~GDSCM.columns.duplicated()]\n",
    "\n",
    "GDSCM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC CNA dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/CNA/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read GDSC CNA dataset (DNA_REPi)\n",
    "\n",
    "GDSCC = pd.read_csv(\"GDSC_CNA.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\", index_col=0, decimal=\".\")\n",
    "GDSCC.drop_duplicates(keep='last')\n",
    "GDSCC = pd.DataFrame.transpose(GDSCC)\n",
    "GDSCC = GDSCC.loc[:, ~GDSCC.columns.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response\n"
     ]
    }
   ],
   "source": [
    "# change directory to read GDSC response dataset (DNA_REPi)\n",
    "\n",
    "os.chdir(dataset_dir + \"/response/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>5.12075503641523</td>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-3.03711635474651</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-0.712119156515401</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-4.40897202584936</td>\n",
       "      <td>Gemcitabine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-1.59752443691324</td>\n",
       "      <td>Mitomycin C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             response             logIC50           drug  exprs  CNA  \\\n",
       "sample_name                                                            \n",
       "683665              0    5.12075503641523  Pyrimethamine      1    1   \n",
       "683665              0   -3.03711635474651    Doxorubicin      1    1   \n",
       "683665              0  -0.712119156515401      Etoposide      1    1   \n",
       "683665              0   -4.40897202584936    Gemcitabine      1    1   \n",
       "683665              0   -1.59752443691324    Mitomycin C      1    1   \n",
       "\n",
       "             mutations  \n",
       "sample_name             \n",
       "683665               1  \n",
       "683665               1  \n",
       "683665               1  \n",
       "683665               1  \n",
       "683665               1  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read GDSC response dataset (DNA_REPi)\n",
    "\n",
    "GDSCR = pd.read_csv(\"GDSC_response.DNA_REPi.tsv\",\n",
    "                    sep=\"\\t\",\n",
    "                    index_col=0,\n",
    "                    decimal=\",\")\n",
    "GDSCR.dropna(inplace=True)\n",
    "GDSCR.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "GDSCR[\"response\"] = GDSCR.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "GDSCR = GDSCR.loc[GDSCR[\"drug\"] != \"5-fluorouracil\", ]\n",
    "\n",
    "GDSCR.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "drug\n",
       "Etoposide         775\n",
       "SN-38             771\n",
       "Mitomycin C       767\n",
       "Doxorubicin       766\n",
       "5-Fluorouracil    764\n",
       "Gemcitabine       761\n",
       "Bleomycin         759\n",
       "Temozolomide      752\n",
       "Cisplatin         738\n",
       "Methotrexate      738\n",
       "Camptothecin      737\n",
       "Cytarabine        735\n",
       "Pyrimethamine     372\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# print count of cell lines for each drug (DNA_REPi)\n",
    "\n",
    "GDSCR[\"drug\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/exprs_homogenized\n"
     ]
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA expression datasets homogenized with GDSC expression dataset (5-Fluorouracil)\n",
    "\n",
    "os.chdir(dataset_dir + \"exprs_homogenized/\")\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100507206</th>\n",
       "      <th>100507254</th>\n",
       "      <th>100507436</th>\n",
       "      <th>100507472</th>\n",
       "      <th>100526773</th>\n",
       "      <th>100527978</th>\n",
       "      <th>100532746</th>\n",
       "      <th>100820829</th>\n",
       "      <th>102724473</th>\n",
       "      <th>105375355</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>4.11649564280422</td>\n",
       "      <td>3.59209496158439</td>\n",
       "      <td>6.28559101273304</td>\n",
       "      <td>2.67926549063284</td>\n",
       "      <td>3.04621480878604</td>\n",
       "      <td>2.69044790660834</td>\n",
       "      <td>8.01420464403606</td>\n",
       "      <td>2.73140674359613</td>\n",
       "      <td>8.71768731731273</td>\n",
       "      <td>2.89488898093329</td>\n",
       "      <td>...</td>\n",
       "      <td>2.71033163066003</td>\n",
       "      <td>3.2543218456306</td>\n",
       "      <td>5.67560198356418</td>\n",
       "      <td>3.07143466694584</td>\n",
       "      <td>3.0597754133993</td>\n",
       "      <td>2.63635721806084</td>\n",
       "      <td>2.61292331163008</td>\n",
       "      <td>3.18623983789074</td>\n",
       "      <td>0</td>\n",
       "      <td>2.4522722382887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>4.00318976265845</td>\n",
       "      <td>3.59209496158439</td>\n",
       "      <td>7.55375516379688</td>\n",
       "      <td>4.03669966422185</td>\n",
       "      <td>3.41557516120483</td>\n",
       "      <td>5.94712898968845</td>\n",
       "      <td>8.14610109582064</td>\n",
       "      <td>2.54373640699005</td>\n",
       "      <td>7.8874359290969</td>\n",
       "      <td>3.43325748497387</td>\n",
       "      <td>...</td>\n",
       "      <td>2.80929358976047</td>\n",
       "      <td>3.3020335598854</td>\n",
       "      <td>8.94930072777643</td>\n",
       "      <td>3.07143466694584</td>\n",
       "      <td>3.0597754133993</td>\n",
       "      <td>2.63635721806084</td>\n",
       "      <td>2.61292331163008</td>\n",
       "      <td>4.60199637750889</td>\n",
       "      <td>0</td>\n",
       "      <td>2.89512503666773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1173</th>\n",
       "      <td>3.74198340066366</td>\n",
       "      <td>3.67493736139833</td>\n",
       "      <td>6.16750446793135</td>\n",
       "      <td>3.50545765081401</td>\n",
       "      <td>3.04295501121306</td>\n",
       "      <td>2.98779164904873</td>\n",
       "      <td>8.37697489020915</td>\n",
       "      <td>2.64942372799293</td>\n",
       "      <td>8.2967442081863</td>\n",
       "      <td>4.13640544224621</td>\n",
       "      <td>...</td>\n",
       "      <td>2.71033163066003</td>\n",
       "      <td>3.2543218456306</td>\n",
       "      <td>4.67270724878415</td>\n",
       "      <td>3.07143466694584</td>\n",
       "      <td>3.0597754133993</td>\n",
       "      <td>2.63635721806084</td>\n",
       "      <td>2.61292331163008</td>\n",
       "      <td>4.99706823903714</td>\n",
       "      <td>0</td>\n",
       "      <td>3.34491841640924</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18189 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID         1                 2                 9          \\\n",
       "X-1027    4.11649564280422  3.59209496158439  6.28559101273304   \n",
       "X-1119    4.00318976265845  3.59209496158439  7.55375516379688   \n",
       "X-1173    3.74198340066366  3.67493736139833  6.16750446793135   \n",
       "\n",
       "ENTREZID         10                12                13         \\\n",
       "X-1027    2.67926549063284  3.04621480878604  2.69044790660834   \n",
       "X-1119    4.03669966422185  3.41557516120483  5.94712898968845   \n",
       "X-1173    3.50545765081401  3.04295501121306  2.98779164904873   \n",
       "\n",
       "ENTREZID         14                15                16         \\\n",
       "X-1027    8.01420464403606  2.73140674359613  8.71768731731273   \n",
       "X-1119    8.14610109582064  2.54373640699005   7.8874359290969   \n",
       "X-1173    8.37697489020915  2.64942372799293   8.2967442081863   \n",
       "\n",
       "ENTREZID         18         ...         100507206        100507254  \\\n",
       "X-1027    2.89488898093329  ...  2.71033163066003  3.2543218456306   \n",
       "X-1119    3.43325748497387  ...  2.80929358976047  3.3020335598854   \n",
       "X-1173    4.13640544224621  ...  2.71033163066003  3.2543218456306   \n",
       "\n",
       "ENTREZID         100507436         100507472        100526773  \\\n",
       "X-1027    5.67560198356418  3.07143466694584  3.0597754133993   \n",
       "X-1119    8.94930072777643  3.07143466694584  3.0597754133993   \n",
       "X-1173    4.67270724878415  3.07143466694584  3.0597754133993   \n",
       "\n",
       "ENTREZID         100527978         100532746         100820829 102724473  \\\n",
       "X-1027    2.63635721806084  2.61292331163008  3.18623983789074         0   \n",
       "X-1119    2.63635721806084  2.61292331163008  4.60199637750889         0   \n",
       "X-1173    2.63635721806084  2.61292331163008  4.99706823903714         0   \n",
       "\n",
       "ENTREZID         105375355  \n",
       "X-1027     2.4522722382887  \n",
       "X-1119    2.89512503666773  \n",
       "X-1173    3.34491841640924  \n",
       "\n",
       "[3 rows x 18189 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX expression dataset homogenized with GDSC expression dataset (5-Fluorouracil)\n",
    "\n",
    "PDXEfluorouracil = pd.read_csv(\"PDX_exprs.5-Fluorouracil.eb_with.GDSC_exprs.5-Fluorouracil.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "PDXEfluorouracil = pd.DataFrame.transpose(PDXEfluorouracil)\n",
    "PDXEfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>100188893</th>\n",
       "      <th>100190940</th>\n",
       "      <th>100190949</th>\n",
       "      <th>100191040</th>\n",
       "      <th>100192386</th>\n",
       "      <th>100268168</th>\n",
       "      <th>100271715</th>\n",
       "      <th>100287718</th>\n",
       "      <th>100288778</th>\n",
       "      <th>100289635</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IC</th>\n",
       "      <td>5.72964960061116</td>\n",
       "      <td>4.41876362025309</td>\n",
       "      <td>4.86023990366878</td>\n",
       "      <td>2.47277876318537</td>\n",
       "      <td>4.81396331677419</td>\n",
       "      <td>2.6727941627861</td>\n",
       "      <td>8.72268461889477</td>\n",
       "      <td>2.5342747028707</td>\n",
       "      <td>8.30356599972612</td>\n",
       "      <td>2.66680810277045</td>\n",
       "      <td>...</td>\n",
       "      <td>9.61327758421809</td>\n",
       "      <td>2.69169602849574</td>\n",
       "      <td>2.74156299885291</td>\n",
       "      <td>3.14358416474115</td>\n",
       "      <td>3.63806017988108</td>\n",
       "      <td>2.83936659460498</td>\n",
       "      <td>2.84468471240513</td>\n",
       "      <td>2.9097524321588</td>\n",
       "      <td>4.74904572228446</td>\n",
       "      <td>4.63830114019886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IX</th>\n",
       "      <td>5.16820185705619</td>\n",
       "      <td>5.0656506669595</td>\n",
       "      <td>5.05015666335564</td>\n",
       "      <td>2.49748103264488</td>\n",
       "      <td>6.24486748738657</td>\n",
       "      <td>3.89184369807008</td>\n",
       "      <td>8.07246325142995</td>\n",
       "      <td>2.65471072107481</td>\n",
       "      <td>8.69196063245573</td>\n",
       "      <td>3.97461007293206</td>\n",
       "      <td>...</td>\n",
       "      <td>9.28039145658571</td>\n",
       "      <td>2.68523221938676</td>\n",
       "      <td>2.70770230320119</td>\n",
       "      <td>3.1831168236566</td>\n",
       "      <td>3.17601340812919</td>\n",
       "      <td>2.66266337139924</td>\n",
       "      <td>3.05740520413346</td>\n",
       "      <td>2.9097524321588</td>\n",
       "      <td>4.66674620630048</td>\n",
       "      <td>4.95684240309785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AG-3593</th>\n",
       "      <td>4.09489404413517</td>\n",
       "      <td>3.56187701617794</td>\n",
       "      <td>5.74195136365074</td>\n",
       "      <td>3.52505514866919</td>\n",
       "      <td>0.688407410705727</td>\n",
       "      <td>2.04131612004455</td>\n",
       "      <td>8.50563761093224</td>\n",
       "      <td>2.83121226059543</td>\n",
       "      <td>8.98277420895832</td>\n",
       "      <td>4.7862922506412</td>\n",
       "      <td>...</td>\n",
       "      <td>9.44820752386781</td>\n",
       "      <td>3.3077275660044</td>\n",
       "      <td>3.04450667890368</td>\n",
       "      <td>3.53272687811095</td>\n",
       "      <td>2.90679846697165</td>\n",
       "      <td>2.45941479590968</td>\n",
       "      <td>2.65824917267036</td>\n",
       "      <td>2.9097524321588</td>\n",
       "      <td>5.80963066021836</td>\n",
       "      <td>4.86261821976041</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18197 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID             1                 2                 9          \\\n",
       "TCGA-3A-A9IC  5.72964960061116  4.41876362025309  4.86023990366878   \n",
       "TCGA-3A-A9IX  5.16820185705619   5.0656506669595  5.05015666335564   \n",
       "TCGA-AG-3593  4.09489404413517  3.56187701617794  5.74195136365074   \n",
       "\n",
       "ENTREZID             10                 12                13         \\\n",
       "TCGA-3A-A9IC  2.47277876318537   4.81396331677419   2.6727941627861   \n",
       "TCGA-3A-A9IX  2.49748103264488   6.24486748738657  3.89184369807008   \n",
       "TCGA-AG-3593  3.52505514866919  0.688407410705727  2.04131612004455   \n",
       "\n",
       "ENTREZID             14                15                16         \\\n",
       "TCGA-3A-A9IC  8.72268461889477   2.5342747028707  8.30356599972612   \n",
       "TCGA-3A-A9IX  8.07246325142995  2.65471072107481  8.69196063245573   \n",
       "TCGA-AG-3593  8.50563761093224  2.83121226059543  8.98277420895832   \n",
       "\n",
       "ENTREZID             18         ...         100188893         100190940  \\\n",
       "TCGA-3A-A9IC  2.66680810277045  ...  9.61327758421809  2.69169602849574   \n",
       "TCGA-3A-A9IX  3.97461007293206  ...  9.28039145658571  2.68523221938676   \n",
       "TCGA-AG-3593   4.7862922506412  ...  9.44820752386781   3.3077275660044   \n",
       "\n",
       "ENTREZID             100190949         100191040         100192386  \\\n",
       "TCGA-3A-A9IC  2.74156299885291  3.14358416474115  3.63806017988108   \n",
       "TCGA-3A-A9IX  2.70770230320119   3.1831168236566  3.17601340812919   \n",
       "TCGA-AG-3593  3.04450667890368  3.53272687811095  2.90679846697165   \n",
       "\n",
       "ENTREZID             100268168         100271715        100287718  \\\n",
       "TCGA-3A-A9IC  2.83936659460498  2.84468471240513  2.9097524321588   \n",
       "TCGA-3A-A9IX  2.66266337139924  3.05740520413346  2.9097524321588   \n",
       "TCGA-AG-3593  2.45941479590968  2.65824917267036  2.9097524321588   \n",
       "\n",
       "ENTREZID             100288778         100289635  \n",
       "TCGA-3A-A9IC  4.74904572228446  4.63830114019886  \n",
       "TCGA-3A-A9IX  4.66674620630048  4.95684240309785  \n",
       "TCGA-AG-3593  5.80963066021836  4.86261821976041  \n",
       "\n",
       "[3 rows x 18197 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA expression dataset homogenized with GDSC expression dataset (5-Fluorouracil)\n",
    "\n",
    "TCGAEfluorouracil = pd.read_csv(\"TCGA_exprs.Fluorouracil.eb_with.GDSC_exprs.5-Fluorouracil.tsv\",\n",
    "                   sep = \"\\t\", index_col=0, decimal = \",\")\n",
    "TCGAEfluorouracil = pd.DataFrame.transpose(TCGAEfluorouracil)\n",
    "TCGAEfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/mutations'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX mutation dataset (5-Fluorouracil)\n",
    "\n",
    "os.chdir(dataset_dir + \"mutations\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>3075.0</th>\n",
       "      <th>171017.0</th>\n",
       "      <th>2058.0</th>\n",
       "      <th>79890.0</th>\n",
       "      <th>7701.0</th>\n",
       "      <th>3104.0</th>\n",
       "      <th>4642.0</th>\n",
       "      <th>9778.0</th>\n",
       "      <th>100132406.0</th>\n",
       "      <th>23607.0</th>\n",
       "      <th>...</th>\n",
       "      <th>253738.0</th>\n",
       "      <th>154807.0</th>\n",
       "      <th>79140.0</th>\n",
       "      <th>6307.0</th>\n",
       "      <th>6352.0</th>\n",
       "      <th>56005.0</th>\n",
       "      <th>115650.0</th>\n",
       "      <th>64946.0</th>\n",
       "      <th>5777.0</th>\n",
       "      <th>253639.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1173</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 15010 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID 3075.0      171017.0    2058.0      79890.0     7701.0       \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1119           0.0       -0.05         0.0       -0.05         0.0   \n",
       "X-1173           0.0         0.0         0.0       -0.05         0.0   \n",
       "\n",
       "ENTREZID 3104.0      4642.0      9778.0      100132406.0 23607.0      ...  \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0  ...   \n",
       "X-1119           0.0         0.0         0.0       -0.05         0.0  ...   \n",
       "X-1173           0.0         0.0         0.0       -0.05         0.0  ...   \n",
       "\n",
       "ENTREZID 253738.0    154807.0    79140.0     6307.0      6352.0       \\\n",
       "X-1027           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1119           0.0         0.0         0.0         0.0         0.0   \n",
       "X-1173           0.0         0.0         0.0         0.0         0.0   \n",
       "\n",
       "ENTREZID 56005.0     115650.0    64946.0     5777.0      253639.0     \n",
       "X-1027           0.0         0.0         0.0         0.0         0.0  \n",
       "X-1119           0.0         0.0         0.0         0.0         0.0  \n",
       "X-1173           0.0         0.0         0.0         0.0         0.0  \n",
       "\n",
       "[3 rows x 15010 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX mutation dataset (5-Fluorouracil)\n",
    "\n",
    "PDXMfluorouracil = pd.read_csv(\"PDX_mutations.5-Fluorouracil.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "PDXMfluorouracil.drop_duplicates(keep='last')\n",
    "PDXMfluorouracil = pd.DataFrame.transpose(PDXMfluorouracil)\n",
    "PDXMfluorouracil = PDXMfluorouracil.loc[:, ~PDXMfluorouracil.columns.duplicated()]\n",
    "\n",
    "PDXMfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>9729</th>\n",
       "      <th>2050</th>\n",
       "      <th>515</th>\n",
       "      <th>140803</th>\n",
       "      <th>55814</th>\n",
       "      <th>51207</th>\n",
       "      <th>27143</th>\n",
       "      <th>152586</th>\n",
       "      <th>283659</th>\n",
       "      <th>51725</th>\n",
       "      <th>...</th>\n",
       "      <th>170393</th>\n",
       "      <th>401940</th>\n",
       "      <th>3249</th>\n",
       "      <th>254958</th>\n",
       "      <th>388199</th>\n",
       "      <th>55049</th>\n",
       "      <th>162699</th>\n",
       "      <th>114299</th>\n",
       "      <th>55969</th>\n",
       "      <th>147841</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IX</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AG-3593</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18540 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             9729   2050   515    140803 55814  51207  27143  152586 283659  \\\n",
       "TCGA-3A-A9IC    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-3A-A9IX    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-AG-3593    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "             51725   ... 170393 401940 3249   254958 388199 55049  162699  \\\n",
       "TCGA-3A-A9IC    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-3A-A9IX    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "TCGA-AG-3593    0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0   \n",
       "\n",
       "             114299 55969  147841  \n",
       "TCGA-3A-A9IC    0.0    0.0    0.0  \n",
       "TCGA-3A-A9IX    0.0    0.0    0.0  \n",
       "TCGA-AG-3593    0.0    0.0    0.0  \n",
       "\n",
       "[3 rows x 18540 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA mutation dataset (5-Fluorouracil)\n",
    "\n",
    "TCGAMfluorouracil = pd.read_csv(\"TCGA_mutations.Fluorouracil.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "TCGAMfluorouracil.drop_duplicates(keep='last')\n",
    "TCGAMfluorouracil = pd.DataFrame.transpose(TCGAMfluorouracil)\n",
    "TCGAMfluorouracil = TCGAMfluorouracil.loc[:, ~TCGAMfluorouracil.columns.duplicated()]\n",
    "\n",
    "TCGAMfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/CNA'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX CNA dataset (5-Fluorouracil)\n",
    "\n",
    "os.chdir(dataset_dir + \"CNA\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ENTREZID</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>...</th>\n",
       "      <th>101340252</th>\n",
       "      <th>102723547</th>\n",
       "      <th>102724473</th>\n",
       "      <th>103091865</th>\n",
       "      <th>105375355</th>\n",
       "      <th>107126285</th>\n",
       "      <th>109623460</th>\n",
       "      <th>109731405</th>\n",
       "      <th>118126072</th>\n",
       "      <th>128385369</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-1.1844245711374275</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.7311832415721999</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4276061727818994</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1173</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3391373849195853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3504972470841331</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3391373849195853</td>\n",
       "      <td>0.3785116232537298</td>\n",
       "      <td>0.3673710656485294</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8196681834964555</td>\n",
       "      <td>0.3617683594191533</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3219280948873623</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 23205 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ENTREZID 1         2         3         9         10         \\\n",
       "X-1027         0.0       0.0       0.0       0.0       0.0   \n",
       "X-1119         0.0       0.0       0.0       0.0       0.0   \n",
       "X-1173         0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "ENTREZID           12        13                  14        15         \\\n",
       "X-1027                   0.0       0.0                 0.0       0.0   \n",
       "X-1119                   0.0       0.0                 0.0       0.0   \n",
       "X-1173    0.3391373849195853       0.0  0.3504972470841331       0.0   \n",
       "\n",
       "ENTREZID 16         ...           101340252           102723547  \\\n",
       "X-1027         0.0  ...                 0.0                 0.0   \n",
       "X-1119         0.0  ...                 0.0                 0.0   \n",
       "X-1173         0.0  ...  0.3391373849195853  0.3785116232537298   \n",
       "\n",
       "ENTREZID           102724473 103091865           105375355  \\\n",
       "X-1027                   0.0       0.0                 0.0   \n",
       "X-1119                   0.0       0.0  0.7311832415721999   \n",
       "X-1173    0.3673710656485294       0.0  0.8196681834964555   \n",
       "\n",
       "ENTREZID            107126285 109623460 109731405           118126072  \\\n",
       "X-1027    -1.1844245711374275       0.0       0.0                 0.0   \n",
       "X-1119                    0.0       0.0       0.0  0.4276061727818994   \n",
       "X-1173     0.3617683594191533       0.0       0.0  0.3219280948873623   \n",
       "\n",
       "ENTREZID 128385369  \n",
       "X-1027         0.0  \n",
       "X-1119         0.0  \n",
       "X-1173         0.0  \n",
       "\n",
       "[3 rows x 23205 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX CNA dataset (5-Fluorouracil)\n",
    "\n",
    "PDXCfluorouracil = pd.read_csv(\"PDX_CNA.5-Fluorouracil.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "PDXCfluorouracil.drop_duplicates(keep='last')\n",
    "PDXCfluorouracil = pd.DataFrame.transpose(PDXCfluorouracil)\n",
    "PDXCfluorouracil = PDXCfluorouracil.loc[:, ~PDXCfluorouracil.columns.duplicated()]\n",
    "\n",
    "PDXCfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>24</th>\n",
       "      <th>27</th>\n",
       "      <th>29</th>\n",
       "      <th>32</th>\n",
       "      <th>34</th>\n",
       "      <th>35</th>\n",
       "      <th>36</th>\n",
       "      <th>...</th>\n",
       "      <th>728851</th>\n",
       "      <th>100130391</th>\n",
       "      <th>100290877</th>\n",
       "      <th>100310869</th>\n",
       "      <th>100420569</th>\n",
       "      <th>100421288</th>\n",
       "      <th>100422928</th>\n",
       "      <th>100506036</th>\n",
       "      <th>100506076</th>\n",
       "      <th>100506123</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IC</th>\n",
       "      <td>-0.2644</td>\n",
       "      <td>-0.2644</td>\n",
       "      <td>-0.2644</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3214</td>\n",
       "      <td>-0.2518</td>\n",
       "      <td>-0.2497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.2497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3A-A9IX</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AG-3593</th>\n",
       "      <td>-0.7299</td>\n",
       "      <td>-0.7299</td>\n",
       "      <td>-0.7299</td>\n",
       "      <td>-0.3147</td>\n",
       "      <td>0.4133</td>\n",
       "      <td>-0.7013</td>\n",
       "      <td>-0.294</td>\n",
       "      <td>-0.3147</td>\n",
       "      <td>-0.3051</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 31922 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             9         10        11        24        27        29         \\\n",
       "TCGA-3A-A9IC   -0.2644   -0.2644   -0.2644       0.0    0.3214   -0.2518   \n",
       "TCGA-3A-A9IX       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "TCGA-AG-3593   -0.7299   -0.7299   -0.7299   -0.3147    0.4133   -0.7013   \n",
       "\n",
       "             32        34        35        36         ... 728851    100130391  \\\n",
       "TCGA-3A-A9IC   -0.2497       0.0   -0.2497       0.0  ...       0.0       0.0   \n",
       "TCGA-3A-A9IX       0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "TCGA-AG-3593    -0.294   -0.3147   -0.3051       0.0  ...       0.0       0.0   \n",
       "\n",
       "             100290877 100310869 100420569 100421288 100422928 100506036  \\\n",
       "TCGA-3A-A9IC       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "TCGA-3A-A9IX       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "TCGA-AG-3593       0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "             100506076 100506123  \n",
       "TCGA-3A-A9IC       0.0       0.0  \n",
       "TCGA-3A-A9IX       0.0       0.0  \n",
       "TCGA-AG-3593       0.0       0.0  \n",
       "\n",
       "[3 rows x 31922 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA CNA dataset (5-Fluorouracil)\n",
    "\n",
    "TCGACfluorouracil = pd.read_csv(\"TCGA_CNA.Fluorouracil.tsv\",\n",
    "                        sep=\"\\t\", index_col=0, decimal=\",\")\n",
    "TCGACfluorouracil.drop_duplicates(keep='last')\n",
    "TCGACfluorouracil = pd.DataFrame.transpose(TCGACfluorouracil)\n",
    "TCGACfluorouracil = TCGACfluorouracil.loc[:, ~TCGACfluorouracil.columns.duplicated()]\n",
    "\n",
    "TCGACfluorouracil.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variance threshold for GDSC expression dataset (DNA_REPi)\n",
    "\n",
    "selector = VarianceThreshold(0.05)\n",
    "selector.fit_transform(GDSCE)\n",
    "GDSCE = GDSCE[GDSCE.columns[selector.get_support(indices=True)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fill NA values and binarize GDSC CNA and mutation datasets (DNA_REPi)\n",
    "\n",
    "GDSCM = GDSCM.fillna(0)\n",
    "GDSCM[GDSCM != 0.0] = 1\n",
    "GDSCC = GDSCC.fillna(0)\n",
    "GDSCC[GDSCC != 0.0] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC, PDX & TCGA # of common genes: 12578\n"
     ]
    }
   ],
   "source": [
    "# select shared genes between GDSC, PDX, and TCGA expression, mutation and CNA datasets \n",
    "\n",
    "ls = GDSCE.columns.intersection(GDSCM.columns)\n",
    "ls = ls.intersection(GDSCC.columns)\n",
    "ls = ls.intersection(PDXEfluorouracil.columns)\n",
    "ls = ls.intersection(PDXMfluorouracil.columns)\n",
    "ls = ls.intersection(PDXCfluorouracil.columns)\n",
    "ls = ls.intersection(TCGAEfluorouracil.columns)\n",
    "ls = ls.intersection(TCGAMfluorouracil.columns)\n",
    "ls = ls.intersection(TCGACfluorouracil.columns)\n",
    "\n",
    "ls = pd.unique(ls)\n",
    "\n",
    "print(f\"GDSC, PDX & TCGA # of common genes: {ls.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common cell lines: 801\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between GDSC expression, mutation, CNA, and response datasets (DNA_REPi)\n",
    "\n",
    "ls2 = GDSCE.index.intersection(GDSCM.index)\n",
    "ls2 = ls2.intersection(GDSCC.index)\n",
    "ls2 = ls2.intersection(GDSCR.index)\n",
    "\n",
    "print(f\"GDSC # of common cell lines: {ls2.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (801, 12578)\n",
      "GDSC # of common samples and genes (Mutation): (801, 12578)\n",
      "GDSC # of common samples and genes (CNA): (801, 12578)\n",
      "GDSC # of samples (Response): 9435\n"
     ]
    }
   ],
   "source": [
    "# subset shared genes and samples in GDSC expression, mutation, CNA, and response datasets\n",
    "\n",
    "GDSCE = GDSCE.loc[ls2, ls]\n",
    "GDSCM = GDSCM.loc[ls2, ls]\n",
    "GDSCC = GDSCC.loc[ls2, ls]\n",
    "GDSCR = GDSCR.loc[ls2, :]\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCE.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (Mutation): {GDSCM.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCC.shape}\")\n",
    "print(f\"GDSC # of samples (Response): {GDSCR.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples (5-Fluorouracil): 23\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between PDX expression, mutation, and CNA datasets (5-Fluorouracil) \n",
    "\n",
    "ls3 = PDXEfluorouracil.index.intersection(PDXMfluorouracil.index)\n",
    "ls3 = ls3.intersection(PDXCfluorouracil.index)\n",
    "\n",
    "print(f\"PDX # of common samples (5-Fluorouracil): {ls3.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA # of common samples (5-Fluorouracil): 33\n"
     ]
    }
   ],
   "source": [
    "# select shared samples between TCGA expression, mutation, and CNA datasets (5-Fluorouracil) \n",
    "\n",
    "ls4 = TCGAEfluorouracil.index.intersection(TCGAMfluorouracil.index)\n",
    "ls4 = ls4.intersection(TCGACfluorouracil.index)\n",
    "\n",
    "print(f\"TCGA # of common samples (5-Fluorouracil): {ls4.shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDX # of common samples and genes for Expression (5-Fluorouracil): (23, 12578)\n",
      "PDX # of common samples and genes for Mutation (5-Fluorouracil): (23, 12578)\n",
      "PDX # of common samples and genes for CNA (5-Fluorouracil): (23, 12578)\n",
      "\n",
      "TCGA # of common samples and genes for Expression (5-Fluorouracil): (33, 12578)\n",
      "TCGA # of common samples and genes for Mutation (5-Fluorouracil): (33, 12578)\n",
      "TCGA # of common samples and genes for CNA (5-Fluorouracil): (33, 12578)\n"
     ]
    }
   ],
   "source": [
    "# select shared genes and samples between PDX and TCGA expression, mutation, and CNA datasets (5-Fluorouracil)\n",
    "\n",
    "PDXEfluorouracil = PDXEfluorouracil.loc[ls3, ls]\n",
    "PDXMfluorouracil = PDXMfluorouracil.loc[ls3, ls]\n",
    "PDXCfluorouracil = PDXCfluorouracil.loc[ls3, ls]\n",
    "\n",
    "TCGAEfluorouracil = TCGAEfluorouracil.loc[ls4, ls]\n",
    "TCGAMfluorouracil = TCGAMfluorouracil.loc[ls4, ls]\n",
    "TCGACfluorouracil = TCGACfluorouracil.loc[ls4, ls]\n",
    "\n",
    "print(f\"PDX # of common samples and genes for Expression (5-Fluorouracil): {PDXEfluorouracil.shape}\")\n",
    "print(f\"PDX # of common samples and genes for Mutation (5-Fluorouracil): {PDXMfluorouracil.shape}\")\n",
    "print(f\"PDX # of common samples and genes for CNA (5-Fluorouracil): {PDXCfluorouracil.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for Expression (5-Fluorouracil): {TCGAEfluorouracil.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for Mutation (5-Fluorouracil): {TCGAMfluorouracil.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for CNA (5-Fluorouracil): {TCGACfluorouracil.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/GDSC_DEGs_inhibitors/DNA_REPi\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(DEGs_dir)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 858 DEGs (p-adjusted < 0.05 and |logFC| > 1) among DNA replication inhibitors\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLFN11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCKAP1L</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDC4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KRT18</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S100A14</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Frequency\n",
       "0      SLFN11          9\n",
       "1     NCKAP1L          8\n",
       "2        SDC4          7\n",
       "3       KRT18          7\n",
       "4     S100A14          7"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read DNA_REPi_Differentially_Expressed_Genes common among DNA replication inhibitors\n",
    "\n",
    "DEGs_filtered_data = pd.read_excel(\"DNA_REPi_Differentially_Expressed_Genes (EnsemblID).xlsx\",\n",
    "                                    sheet_name = \"Common DEGs\")\n",
    "print(f\"There are {DEGs_filtered_data.shape[0]} DEGs (p-adjusted < 0.05 and |logFC| > 1) among DNA replication inhibitors\\n\")\n",
    "\n",
    "DEGs_filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>ELF1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>546</th>\n",
       "      <td>PROSER2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>547</th>\n",
       "      <td>CLIP4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>THAP10</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>SHTN1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene Symbol  Frequency\n",
       "545        ELF1          1\n",
       "546     PROSER2          1\n",
       "547       CLIP4          1\n",
       "548      THAP10          1\n",
       "549       SHTN1          1"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter = DEGs_filtered_data[\"Frequency\"] == 1\n",
    "DEGs_freq_one =  DEGs_filtered_data[filter]\n",
    "DEGs_freq_one.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 24 DEGs unique to 5-FU, so they were removed from the DEGs list\n"
     ]
    }
   ],
   "source": [
    "DEGs_fluorouracil_data = pd.read_excel(\"DNA_REPi_Differentially_Expressed_Genes (EnsemblID).xlsx\",\n",
    "                                              sheet_name = \"5-Fluorouracil\")\n",
    "\n",
    "filter = DEGs_fluorouracil_data[\"Gene.Symbol\"].isin(DEGs_freq_one[\"Gene Symbol\"])\n",
    "only_fluorouracil_degs = DEGs_fluorouracil_data.loc[filter, \"Gene.Symbol\"]\n",
    "\n",
    "print(f'There are {only_fluorouracil_degs.shape[0]} DEGs unique to 5-FU, so they were removed from the DEGs list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLFN11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NCKAP1L</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDC4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KRT18</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S100A14</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Gene Symbol  Frequency\n",
       "0      SLFN11          9\n",
       "1     NCKAP1L          8\n",
       "2        SDC4          7\n",
       "3       KRT18          7\n",
       "4     S100A14          7"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filter = DEGs_filtered_data[\"Gene Symbol\"].isin(only_fluorouracil_degs)\n",
    "DEGs_filtered_data = DEGs_filtered_data[~filter]\n",
    "\n",
    "DEGs_filtered_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1 input query terms found dup hits:\t[('CLECL1P', 2)]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "834"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get gene symbol from mygene database\n",
    "\n",
    "mg = mygene.MyGeneInfo()\n",
    "DEGs_entrez_id = mg.querymany(DEGs_filtered_data[\"Gene Symbol\"], \n",
    "                              species=\"human\", \n",
    "                              scopes = \"symbol\",\n",
    "                              field = \"entrezgene\",\n",
    "                              as_dataframe=True)[\"entrezgene\"]\n",
    "\n",
    "DEGs_entrez_id.dropna(inplace=True)\n",
    "\n",
    "DEGs_entrez_id.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign GDSC datasets to new variables\n",
    "\n",
    "exprs_z = GDSCE\n",
    "mut = GDSCM\n",
    "cna = GDSCC\n",
    "responses = GDSCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>response</th>\n",
       "      <th>logIC50</th>\n",
       "      <th>drug</th>\n",
       "      <th>exprs</th>\n",
       "      <th>CNA</th>\n",
       "      <th>mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>5.12075503641523</td>\n",
       "      <td>Pyrimethamine</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-3.03711635474651</td>\n",
       "      <td>Doxorubicin</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683665</th>\n",
       "      <td>0</td>\n",
       "      <td>-0.712119156515401</td>\n",
       "      <td>Etoposide</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        response             logIC50           drug  exprs  CNA  mutations\n",
       "683665         0    5.12075503641523  Pyrimethamine      1    1          1\n",
       "683665         0   -3.03711635474651    Doxorubicin      1    1          1\n",
       "683665         0  -0.712119156515401      Etoposide      1    1          1"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# GDSC response dataset\n",
    "\n",
    "responses.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'5-Fluorouracil',\n",
       " 'Bleomycin',\n",
       " 'Camptothecin',\n",
       " 'Cisplatin',\n",
       " 'Cytarabine',\n",
       " 'Doxorubicin',\n",
       " 'Etoposide',\n",
       " 'Gemcitabine',\n",
       " 'Methotrexate',\n",
       " 'Mitomycin C',\n",
       " 'Pyrimethamine',\n",
       " 'SN-38',\n",
       " 'Temozolomide'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list drug names in GDSC response dataset\n",
    "\n",
    "drugs = set(responses[\"drug\"].values)\n",
    "drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert Entrez gene IDs to integer\n",
    "\n",
    "GDSCE.index = GDSCE.index.astype(str)\n",
    "GDSCM.index = GDSCM.index.astype(str)\n",
    "GDSCC.index = GDSCC.index.astype(str)\n",
    "responses.index = responses.index.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge GDSC cell line details with response dataset\n",
    "\n",
    "dat = pd.merge(GDSC_cell_line_details,\n",
    "         responses,\n",
    "         left_index = True, \n",
    "         right_index=True,\n",
    "         how = \"inner\")\n",
    "\n",
    "dat[\"GDSC\\nTissue descriptor 1\"].value_counts()\n",
    "#dat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get tissue descriptors\n",
    "\n",
    "dat[\"GDSC\\nTissue\\ndescriptor 2\"].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: count, dtype: int64)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# list count of cell lines as to tissues treated with fluorouracil\n",
    "\n",
    "filter = (dat[\"drug\"] == \"5-Fluorouracil\")\n",
    "dat.loc[filter,\"GDSC\\nTissue descriptor 1\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['683665', '683665', '683665', '683665', '683665', '683665', '683665',\n",
       "       '683665', '683665', '683665',\n",
       "       ...\n",
       "       '1659818', '1659818', '1659818', '1659818', '1659819', '1659819',\n",
       "       '1659819', '1659819', '1659819', '687800'],\n",
       "      dtype='object', length=9435)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert Entrez gene IDs to string\n",
    "\n",
    "responses.index = responses.index.astype(str)\n",
    "responses.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC # of common samples and genes (Expression): (9435, 631)\n",
      "GDSC # of common samples and genes (Mutation): (9435, 631)\n",
      "GDSC # of common samples and genes (CNA): (9435, 631)\n",
      "GDSC # of common samples (Response): (9435, 6)\n",
      "\n",
      "PDX # of common samples and genes for 5-Fluorouracil (Expression): (23, 631)\n",
      "PDX # of common samples and genes for 5-Fluorouracil (Mutation): (23, 631)\n",
      "PDX # of common samples and genes for 5-Fluorouracil (CNA): (23, 631)\n",
      "\n",
      "TCGA # of common samples and genes for 5-Fluorouracil (Expression): (33, 631)\n",
      "TCGA # of common samples and genes for 5-Fluorouracil (Mutation): (33, 631)\n",
      "TCGA # of common samples and genes for 5-Fluorouracil (CNA): (33, 631)\n"
     ]
    }
   ],
   "source": [
    "# filter GDSC expression, mutation, and CNA datasets (DNA_REPi) as to drugs\n",
    "# subset selected DEGs\n",
    "\n",
    "expression_zscores = []\n",
    "CNA = []\n",
    "mutations = []\n",
    "for drug in drugs:\n",
    "    samples = responses.loc[responses[\"drug\"] == drug, :].index.values\n",
    "    e_z = exprs_z.loc[samples, :]\n",
    "    c = cna.loc[samples, :]\n",
    "    m = mut.loc[samples, :]\n",
    "\n",
    "    expression_zscores.append(e_z)\n",
    "    CNA.append(c)\n",
    "    mutations.append(m)\n",
    "\n",
    "GDSCEv2 = pd.concat(expression_zscores, axis=0)\n",
    "GDSCMv2 = pd.concat(mutations, axis=0)\n",
    "GDSCCv2 = pd.concat(CNA, axis=0)\n",
    "GDSCRv2 = responses\n",
    "\n",
    "ls5 = list(set(GDSCE.columns).intersection(set(DEGs_entrez_id.astype(int))))\n",
    "ls6 = GDSCEv2.index.intersection(GDSCMv2.index)\n",
    "ls6 = ls6.intersection(GDSCCv2.index)\n",
    "\n",
    "GDSCEv2 = GDSCEv2.loc[ls6, ls5]\n",
    "GDSCMv2 = GDSCMv2.loc[ls6, ls5]\n",
    "GDSCCv2 = GDSCCv2.loc[ls6, ls5]\n",
    "GDSCRv2 = GDSCRv2.loc[ls6, :]\n",
    "\n",
    "PDXEfluorouracil = PDXEfluorouracil.loc[:,ls5]\n",
    "PDXMfluorouracil = PDXMfluorouracil.loc[:,ls5]\n",
    "PDXCfluorouracil = PDXCfluorouracil.loc[:,ls5]\n",
    "\n",
    "TCGAEfluorouracil = TCGAEfluorouracil.loc[:,ls5]\n",
    "TCGAMfluorouracil = TCGAMfluorouracil.loc[:,ls5]\n",
    "TCGACfluorouracil = TCGACfluorouracil.loc[:,ls5]\n",
    "\n",
    "responses.index = responses.index.values + \"_\" + responses[\"drug\"].values\n",
    "\n",
    "print(f\"GDSC # of common samples and genes (Expression): {GDSCEv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (Mutation): {GDSCMv2.shape}\")\n",
    "print(f\"GDSC # of common samples and genes (CNA): {GDSCCv2.shape}\")\n",
    "print(f\"GDSC # of common samples (Response): {GDSCRv2.shape}\\n\")\n",
    "\n",
    "print(f\"PDX # of common samples and genes for 5-Fluorouracil (Expression): {PDXEfluorouracil.shape}\")\n",
    "print(f\"PDX # of common samples and genes for 5-Fluorouracil (Mutation): {PDXMfluorouracil.shape}\")\n",
    "print(f\"PDX # of common samples and genes for 5-Fluorouracil (CNA): {PDXCfluorouracil.shape}\\n\")\n",
    "\n",
    "print(f\"TCGA # of common samples and genes for 5-Fluorouracil (Expression): {TCGAEfluorouracil.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for 5-Fluorouracil (Mutation): {TCGAMfluorouracil.shape}\")\n",
    "print(f\"TCGA # of common samples and genes for 5-Fluorouracil (CNA): {TCGACfluorouracil.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 0 ... 0 1 0]\n"
     ]
    }
   ],
   "source": [
    "# GDSC response values\n",
    "\n",
    "Y = GDSCRv2['response'].values\n",
    "print(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/Expansion/Thesis Work/Results/preprocessed_results2/response'"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# change directory to read PDX and TCGA response datasets (5-Fluorouracil)\n",
    "\n",
    "os.chdir(dataset_dir + \"response\")\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 23 samples with response data in the PDX (5-Fluorouracil)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0])"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX response dataset (5-Fluorouracil)\n",
    "\n",
    "PDXRfluorouracil = pd.read_csv(\"PDX_response.5-Fluorouracil.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "PDXRfluorouracil.dropna(inplace=True)\n",
    "PDXRfluorouracil.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "PDXRfluorouracil[\"response\"] = PDXRfluorouracil.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytsfluorouracil_pdx = PDXRfluorouracil[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytsfluorouracil_pdx.shape[0]} samples with response data in the PDX (5-Fluorouracil)\\n\")\n",
    "Ytsfluorouracil_pdx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tumor Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X-0933</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1027</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1055</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1119</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X-1167</th>\n",
       "      <td>CRC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Tumor Type\n",
       "Model            \n",
       "X-0933        CRC\n",
       "X-1027        CRC\n",
       "X-1055        CRC\n",
       "X-1119        CRC\n",
       "X-1167        CRC"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read PDX dataset to select cohorts (5-Fluorouracil)\n",
    "\n",
    "pdx_cohort = pd.read_excel(pdx_data_dir + \"/nm.3954-S2.xlsx\",\n",
    "                              \"PCT raw data\")\n",
    "pdx_cohort = pdx_cohort.loc[pdx_cohort[\"Treatment\"] == \"5FU\",]\n",
    "pdx_cohort = pdx_cohort.loc[:,[\"Model\",\"Tumor Type\"]]\n",
    "pdx_cohort = pdx_cohort.drop_duplicates()\n",
    "pdx_cohort.set_index(\"Model\", inplace = True) \n",
    "pdx_cohort.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Tumor Type\n",
       "CRC    23\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merged PDX cohort table with PDX response dataset (5-Fluorouracil)\n",
    "\n",
    "dat_pdx = pd.merge(pdx_cohort,\n",
    "                   PDXRfluorouracil,\n",
    "                   left_index = True,\n",
    "                   right_index = True)\n",
    "dat_pdx[\"Tumor Type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['X-1027', 'X-1119', 'X-1173', 'X-1270', 'X-1329', 'X-2145', 'X-2182',\n",
       "       'X-2483', 'X-2538', 'X-2573', 'X-2659', 'X-2846', 'X-2861', 'X-3093',\n",
       "       'X-3205', 'X-3267', 'X-3671', 'X-3792', 'X-4087', 'X-5405', 'X-5438',\n",
       "       'X-5495', 'X-5578'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter CRC cohort samples (5-Fluorouracil)\n",
    "\n",
    "filter = (dat_pdx[\"Tumor Type\"] == \"CRC\")\n",
    "pdx_crc_samples = dat_pdx.loc[filter,].index\n",
    "pdx_crc_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0])"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of CRC cohort samples (5-Fluorouracil)\n",
    "\n",
    "Ytsfluorouracil_pdx_crc = PDXRfluorouracil.loc[pdx_crc_samples,\"response\"].values\n",
    "Ytsfluorouracil_pdx_crc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    22\n",
       "1     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistant CRC samples (5-Fluorouracil)\n",
    "\n",
    "PDXRfluorouracil.loc[pdx_crc_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 33 samples with response data in the TCGA (5-Fluorouracil)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0,\n",
       "       1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1])"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read TCGA response dataset (5-Fluorouracil)\n",
    "\n",
    "TCGARfluorouracil = pd.read_csv(\"TCGA_response.Fluorouracil.tsv\", \n",
    "                      sep = \"\\t\", \n",
    "                      index_col=0, \n",
    "                      decimal = \",\")\n",
    "TCGARfluorouracil.dropna(inplace=True)\n",
    "TCGARfluorouracil.rename(mapper=str, axis='index', inplace=True)\n",
    "d = {\"R\": 0, \"S\": 1}\n",
    "TCGARfluorouracil[\"response\"] = TCGARfluorouracil.loc[:, \"response\"].apply(lambda x: d[x])\n",
    "Ytsfluorouracil_tcga = TCGARfluorouracil[\"response\"].values    \n",
    "\n",
    "print(f\"There are {Ytsfluorouracil_tcga.shape[0]} samples with response data in the TCGA (5-Fluorouracil)\\n\")\n",
    "Ytsfluorouracil_tcga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cohort\n",
       "STAD    18\n",
       "READ     7\n",
       "PAAD     6\n",
       "ESCA     2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TCGA cohort table with TCGA response dataset (5-Fluorouracil)\n",
    "\n",
    "dat_tcga = TCGARfluorouracil[\"cohort\"]\n",
    "\n",
    "dat_tcga.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCGA-CG-4444', 'TCGA-D7-8579', 'TCGA-D7-A748', 'TCGA-F1-A448',\n",
       "       'TCGA-FP-7829', 'TCGA-FP-8209', 'TCGA-HU-8243', 'TCGA-KB-A6F7',\n",
       "       'TCGA-VQ-A8DT', 'TCGA-VQ-A8DU', 'TCGA-VQ-A8DZ', 'TCGA-VQ-A8E2',\n",
       "       'TCGA-VQ-A8P3', 'TCGA-VQ-A91Q', 'TCGA-VQ-AA68', 'TCGA-VQ-AA6B',\n",
       "       'TCGA-VQ-AA6F', 'TCGA-VQ-AA6G'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter STAD cohort samples (5-Fluorouracil)\n",
    "\n",
    "filter = (TCGARfluorouracil[\"cohort\"] == \"STAD\")\n",
    "tcga_stad_samples = dat_tcga.loc[filter,].index\n",
    "tcga_stad_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0])"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of STAD cohort samples (5-Fluorouracil)\n",
    "\n",
    "Ytsfluorouracil_tcga_stad = TCGARfluorouracil.loc[tcga_stad_samples,\"response\"].values\n",
    "Ytsfluorouracil_tcga_stad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "1    13\n",
       "0     5\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistance STAD samples (5-Fluorouracil)\n",
    "\n",
    "TCGARfluorouracil.loc[tcga_stad_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCGA-3A-A9IC', 'TCGA-3A-A9IX', 'TCGA-HZ-7920', 'TCGA-HZ-A49I',\n",
       "       'TCGA-IB-7888', 'TCGA-OE-A75W'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter PAAD cohort samples (5-Fluorouracil)\n",
    "\n",
    "filter = (TCGARfluorouracil[\"cohort\"] == \"PAAD\")\n",
    "tcga_paad_samples = dat_tcga.loc[filter,].index\n",
    "tcga_paad_samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 1, 0, 0, 0])"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter responses of PAAD cohort samples (5-Fluorouracil)\n",
    "\n",
    "Ytsfluorouracil_tcga_paad = TCGARfluorouracil.loc[tcga_paad_samples,\"response\"].values\n",
    "Ytsfluorouracil_tcga_paad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "response\n",
       "0    4\n",
       "1    2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count of sensitive and resistance PAAD samples (5-Fluorouracil)\n",
    "\n",
    "TCGARfluorouracil.loc[tcga_paad_samples,\"response\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# maximum iteration\n",
    "\n",
    "max_iter = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use count of nodes and dropout rates found in trained pan-drug dataset\n",
    "\n",
    "hdm1 = 128\n",
    "hdm2 = 1024\n",
    "hdm3 = 256\n",
    "rate1 = 0.5\n",
    "rate2 = 0.7\n",
    "rate3 = 0.4\n",
    "rate4 = 0.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GDSC - Training AUC (DNA_REPi): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi): 0.34\n",
      "PDX - Test AUC (5-Fluorouracil): 0.14\n",
      "PDX - Test AUCPR (5-Fluorouracil): 0.05\n",
      "PDX - Test AUC (5-Fluorouracil, CRC): 0.14\n",
      "PDX - Test AUCPR (5-Fluorouracil, CRC): 0.05\n",
      "TCGA - Test AUC (5-Fluorouracil): 0.53\n",
      "TCGA - Test AUCPR (5-Fluorouracil): 0.74\n",
      "TCGA - Test AUC (5-Fluorouracil, STAD): 0.75\n",
      "TCGA - Test AUCPR (5-Fluorouracil, STAD): 0.89\n",
      "TCGA - Test AUC (5-Fluorouracil, PAAD): 0.25\n",
      "TCGA - Test AUCPR (5-Fluorouracil, PAAD): 0.32\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.34\n",
      "PDX - Test AUC (5-Fluorouracil, Finetuned): 0.18\n",
      "PDX - Test AUCPR (5-Fluorouracil, Finetuned): 0.05\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.34\n",
      "PDX - Test AUC (5-Fluorouracil, CRC, Finetuned): 0.18\n",
      "PDX - Test AUCPR (5-Fluorouracil, CRC, Finetuned): 0.05\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.33\n",
      "TCGA - Test AUC (5-Fluorouracil, Finetuned): 0.57\n",
      "TCGA - Test AUCPR (5-Fluorouracil, Finetuned): 0.78\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "TCGA - Test AUC (5-Fluorouracil, STAD, Finetuned): 0.74\n",
      "TCGA - Test AUCPR (5-Fluorouracil, STAD, Finetuned): 0.9\n",
      "\n",
      "GDSC - Training AUC (DNA_REPi, Finetuned): 0.81\n",
      "GDSC - Training AUCPR (DNA_REPi, Finetuned): 0.32\n",
      "TCGA - Test AUC (5-Fluorouracil, PAAD, Finetuned): 0.5\n",
      "TCGA - Test AUCPR (5-Fluorouracil, PAAD, Finetuned): 0.45\n",
      "\n",
      "\n",
      "Models for maximum finetuning (5-Fluorouracil, PDX)\n",
      "Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "\n",
      "Models for maximum finetuning (5-Fluorouracil, PDX, CRC)\n",
      "Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_12.pt\n",
      "\n",
      "Models for maximum finetuning (5-Fluorouracil, TCGA)\n",
      "Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_16.pt\n",
      "Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_16.pt\n",
      "CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_16.pt\n",
      "Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_16.pt\n",
      "\n",
      "Models for maximum finetuning (5-Fluorouracil, TCGA, STAD)\n",
      "Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_35.pt\n",
      "Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_35.pt\n",
      "CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_35.pt\n",
      "Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_35.pt\n",
      "\n",
      "Models for maximum finetuning (5-Fluorouracil, TCGA, PAAD)\n",
      "Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_50.pt\n",
      "Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_50.pt\n",
      "CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_50.pt\n",
      "Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_50.pt\n"
     ]
    }
   ],
   "source": [
    "# load pre-trained modules and make predictions\n",
    "\n",
    "scalerGDSC = sk.StandardScaler()\n",
    "scalerGDSC.fit(GDSCEv2.values)\n",
    "X_trainE = scalerGDSC.transform(GDSCEv2.values)\n",
    "X_testEfluorouracil_pdx = scalerGDSC.transform(PDXEfluorouracil.values) \n",
    "X_testEfluorouracil_pdx_crc = scalerGDSC.transform(PDXEfluorouracil.loc[pdx_crc_samples,].values) \n",
    "\n",
    "X_testEfluorouracil_tcga = scalerGDSC.transform(TCGAEfluorouracil.values)    \n",
    "X_testEfluorouracil_tcga_stad = scalerGDSC.transform(TCGAEfluorouracil.loc[tcga_stad_samples,].values)    \n",
    "X_testEfluorouracil_tcga_paad = scalerGDSC.transform(TCGAEfluorouracil.loc[tcga_paad_samples,].values)    \n",
    "\n",
    "X_trainM = np.nan_to_num(GDSCMv2.values)\n",
    "X_testMfluorouracil_pdx = np.nan_to_num(PDXMfluorouracil.values.astype(np.float32))\n",
    "X_testMfluorouracil_pdx_crc = np.nan_to_num(PDXMfluorouracil.loc[pdx_crc_samples,].values.astype(np.float32))\n",
    "X_testMfluorouracil_tcga = np.nan_to_num(TCGAMfluorouracil.values.astype(np.float32))\n",
    "X_testMfluorouracil_tcga_stad = np.nan_to_num(TCGAMfluorouracil.loc[tcga_stad_samples,].values)\n",
    "X_testMfluorouracil_tcga_paad = np.nan_to_num(TCGAMfluorouracil.loc[tcga_paad_samples,].values.astype(np.float32))\n",
    "\n",
    "X_trainC = np.nan_to_num(GDSCCv2.values)\n",
    "X_testCfluorouracil_pdx = np.nan_to_num(PDXCfluorouracil.values.astype(np.float32))\n",
    "X_testCfluorouracil_pdx_crc = np.nan_to_num(PDXCfluorouracil.loc[pdx_crc_samples,].values.astype(np.float32))\n",
    "X_testCfluorouracil_tcga = np.nan_to_num(TCGACfluorouracil.values)\n",
    "X_testCfluorouracil_tcga_stad = np.nan_to_num(TCGACfluorouracil.loc[tcga_stad_samples,].values)\n",
    "X_testCfluorouracil_tcga_paad = np.nan_to_num(TCGACfluorouracil.loc[tcga_paad_samples,].values)\n",
    "\n",
    "TX_testEfluorouracil_pdx = torch.FloatTensor(X_testEfluorouracil_pdx)\n",
    "TX_testEfluorouracil_pdx_crc = torch.FloatTensor(X_testEfluorouracil_pdx_crc)\n",
    "\n",
    "TX_testMfluorouracil_pdx = torch.FloatTensor(X_testMfluorouracil_pdx.astype(np.float32))\n",
    "TX_testMfluorouracil_pdx_crc = torch.FloatTensor(X_testMfluorouracil_pdx_crc.astype(np.float32))\n",
    "\n",
    "TX_testCfluorouracil_pdx = torch.FloatTensor(X_testCfluorouracil_pdx.astype(np.float32))\n",
    "TX_testCfluorouracil_pdx_crc = torch.FloatTensor(X_testCfluorouracil_pdx_crc.astype(np.float32))\n",
    "\n",
    "ty_testEfluorouracil_pdx = torch.FloatTensor(Ytsfluorouracil_pdx.astype(int))\n",
    "ty_testEfluorouracil_pdx_crc = torch.FloatTensor(Ytsfluorouracil_pdx_crc.astype(int))\n",
    "\n",
    "TX_testEfluorouracil_tcga = torch.FloatTensor(X_testEfluorouracil_tcga)\n",
    "TX_testEfluorouracil_tcga_stad = torch.FloatTensor(X_testEfluorouracil_tcga_stad)\n",
    "TX_testEfluorouracil_tcga_paad = torch.FloatTensor(X_testEfluorouracil_tcga_paad)\n",
    "\n",
    "TX_testMfluorouracil_tcga = torch.FloatTensor(X_testMfluorouracil_tcga.astype(np.float32))\n",
    "TX_testMfluorouracil_tcga_stad = torch.FloatTensor(X_testMfluorouracil_tcga_stad.astype(np.float32))\n",
    "TX_testMfluorouracil_tcga_paad = torch.FloatTensor(X_testMfluorouracil_tcga_paad.astype(np.float32))\n",
    "\n",
    "TX_testCfluorouracil_tcga = torch.FloatTensor(X_testCfluorouracil_tcga.astype(np.float32))\n",
    "TX_testCfluorouracil_tcga_stad = torch.FloatTensor(X_testCfluorouracil_tcga_stad.astype(np.float32))\n",
    "TX_testCfluorouracil_tcga_paad = torch.FloatTensor(X_testCfluorouracil_tcga_paad.astype(np.float32))\n",
    "\n",
    "ty_testEfluorouracil_tcga = torch.FloatTensor(Ytsfluorouracil_tcga.astype(int))\n",
    "ty_testEfluorouracil_tcga_stad = torch.FloatTensor(Ytsfluorouracil_tcga_stad.astype(int))\n",
    "ty_testEfluorouracil_tcga_paad = torch.FloatTensor(Ytsfluorouracil_tcga_paad.astype(int))\n",
    "\n",
    "n_sampE, IE_dim = X_trainE.shape\n",
    "n_sampM, IM_dim = X_trainM.shape\n",
    "n_sampC, IC_dim = X_trainC.shape\n",
    "\n",
    "h_dim1 = hdm1\n",
    "h_dim2 = hdm2\n",
    "h_dim3 = hdm3        \n",
    "Z_in = h_dim1 + h_dim2 + h_dim3\n",
    "\n",
    "class AEE(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEE, self).__init__()\n",
    "        self.EnE = torch.nn.Sequential(\n",
    "            nn.Linear(IE_dim, h_dim1),\n",
    "            nn.BatchNorm1d(h_dim1),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate1))\n",
    "    def forward(self, x):\n",
    "        output = self.EnE(x)\n",
    "        return output  \n",
    "\n",
    "\n",
    "class AEM(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEM, self).__init__()\n",
    "        self.EnM = torch.nn.Sequential(\n",
    "            nn.Linear(IM_dim, h_dim2),\n",
    "            nn.BatchNorm1d(h_dim2),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate2))\n",
    "    def forward(self, x):\n",
    "        output = self.EnM(x)\n",
    "        return output       \n",
    "\n",
    "class AEC(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(AEC, self).__init__()\n",
    "        self.EnC = torch.nn.Sequential(\n",
    "            nn.Linear(IC_dim, h_dim3),\n",
    "            nn.BatchNorm1d(h_dim3),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(rate3))\n",
    "    def forward(self, x):\n",
    "        output = self.EnC(x)\n",
    "        return output \n",
    "        \n",
    "class Classifier(nn.Module):\n",
    "    def __init__(self):\n",
    "        super(Classifier, self).__init__()\n",
    "        self.FC = torch.nn.Sequential(\n",
    "            nn.Linear(Z_in, 1),\n",
    "            nn.Dropout(rate4),\n",
    "            nn.Sigmoid())\n",
    "    def forward(self, x):\n",
    "        return self.FC(x)\n",
    "\n",
    "torch.cuda.manual_seed_all(42)\n",
    "\n",
    "AutoencoderE = torch.load(models_dir + \"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy.pt\")\n",
    "AutoencoderM = torch.load(models_dir + \"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy.pt\")\n",
    "AutoencoderC = torch.load(models_dir + \"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy.pt\")\n",
    "\n",
    "Clas = torch.load(models_dir + \"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy.pt\")\n",
    "\n",
    "AutoencoderE.eval()\n",
    "AutoencoderM.eval()\n",
    "AutoencoderC.eval()\n",
    "Clas.eval()\n",
    "\n",
    "ZEX = AutoencoderE(torch.FloatTensor(X_trainE))\n",
    "ZMX = AutoencoderM(torch.FloatTensor(X_trainM))\n",
    "ZCX = AutoencoderC(torch.FloatTensor(X_trainC))\n",
    "ZTX = torch.cat((ZEX, ZMX, ZCX), 1)\n",
    "ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "PredX = Clas(ZTX)\n",
    "AUCT = roc_auc_score(Y, PredX.detach().numpy())\n",
    "AUCTaucpr = average_precision_score(Y, PredX.detach().numpy())\n",
    "\n",
    "ZETfluorouracil_pdx = AutoencoderE(TX_testEfluorouracil_pdx)\n",
    "ZMTfluorouracil_pdx = AutoencoderM(TX_testMfluorouracil_pdx)\n",
    "ZCTfluorouracil_pdx = AutoencoderC(TX_testCfluorouracil_pdx)\n",
    "ZTTfluorouracil_pdx = torch.cat((ZETfluorouracil_pdx, ZMTfluorouracil_pdx, ZCTfluorouracil_pdx), 1)\n",
    "ZTTfluorouracil_pdx = F.normalize(ZTTfluorouracil_pdx, p=2, dim=0)\n",
    "PredTfluorouracil_pdx = Clas(ZTTfluorouracil_pdx)\n",
    "AUCTfluorouracil_pdx = roc_auc_score(Ytsfluorouracil_pdx.astype(int), PredTfluorouracil_pdx.detach().numpy())\n",
    "AUCTfluorouracil_pdx_aucpr = average_precision_score(Ytsfluorouracil_pdx.astype(int), PredTfluorouracil_pdx.detach().numpy())\n",
    "\n",
    "ZETfluorouracil_pdx_crc = AutoencoderE(TX_testEfluorouracil_pdx_crc)\n",
    "ZMTfluorouracil_pdx_crc = AutoencoderM(TX_testMfluorouracil_pdx_crc)\n",
    "ZCTfluorouracil_pdx_crc = AutoencoderC(TX_testCfluorouracil_pdx_crc)\n",
    "ZTTfluorouracil_pdx_crc = torch.cat((ZETfluorouracil_pdx_crc, ZMTfluorouracil_pdx_crc, ZCTfluorouracil_pdx_crc), 1)\n",
    "ZTTfluorouracil_pdx_crc = F.normalize(ZTTfluorouracil_pdx_crc, p=2, dim=0)\n",
    "PredTfluorouracil_pdx_crc = Clas(ZTTfluorouracil_pdx_crc)\n",
    "AUCTfluorouracil_pdx_crc = roc_auc_score(Ytsfluorouracil_pdx_crc.astype(int), PredTfluorouracil_pdx_crc.detach().numpy())\n",
    "AUCTfluorouracil_pdx_crc_aucpr = average_precision_score(Ytsfluorouracil_pdx_crc.astype(int), PredTfluorouracil_pdx_crc.detach().numpy())\n",
    "\n",
    "ZETfluorouracil_tcga = AutoencoderE(TX_testEfluorouracil_tcga)\n",
    "ZMTfluorouracil_tcga = AutoencoderM(TX_testMfluorouracil_tcga)\n",
    "ZCTfluorouracil_tcga = AutoencoderC(TX_testCfluorouracil_tcga)\n",
    "ZTTfluorouracil_tcga = torch.cat((ZETfluorouracil_tcga, ZMTfluorouracil_tcga, ZCTfluorouracil_tcga), 1)\n",
    "ZTTfluorouracil_tcga = F.normalize(ZTTfluorouracil_tcga, p=2, dim=0)\n",
    "PredTfluorouracil_tcga = Clas(ZTTfluorouracil_tcga)\n",
    "AUCTfluorouracil_tcga = roc_auc_score(Ytsfluorouracil_tcga.astype(int), PredTfluorouracil_tcga.detach().numpy())\n",
    "AUCTfluorouracil_tcga_aucpr = average_precision_score(Ytsfluorouracil_tcga.astype(int), PredTfluorouracil_tcga.detach().numpy())\n",
    "\n",
    "ZETfluorouracil_tcga_stad = AutoencoderE(TX_testEfluorouracil_tcga_stad)\n",
    "ZMTfluorouracil_tcga_stad = AutoencoderM(TX_testMfluorouracil_tcga_stad)\n",
    "ZCTfluorouracil_tcga_stad = AutoencoderC(TX_testCfluorouracil_tcga_stad)\n",
    "ZTTfluorouracil_tcga_stad = torch.cat((ZETfluorouracil_tcga_stad, ZMTfluorouracil_tcga_stad, ZCTfluorouracil_tcga_stad), 1)\n",
    "ZTTfluorouracil_tcga_stad = F.normalize(ZTTfluorouracil_tcga_stad, p=2, dim=0)\n",
    "PredTfluorouracil_tcga_stad = Clas(ZTTfluorouracil_tcga_stad)\n",
    "AUCTfluorouracil_tcga_stad = roc_auc_score(Ytsfluorouracil_tcga_stad.astype(int), PredTfluorouracil_tcga_stad.detach().numpy())\n",
    "AUCTfluorouracil_tcga_stad_aucpr = average_precision_score(Ytsfluorouracil_tcga_stad.astype(int), PredTfluorouracil_tcga_stad.detach().numpy())\n",
    "\n",
    "ZETfluorouracil_tcga_paad = AutoencoderE(TX_testEfluorouracil_tcga_paad)\n",
    "ZMTfluorouracil_tcga_paad = AutoencoderM(TX_testMfluorouracil_tcga_paad)\n",
    "ZCTfluorouracil_tcga_paad = AutoencoderC(TX_testCfluorouracil_tcga_paad)\n",
    "ZTTfluorouracil_tcga_paad = torch.cat((ZETfluorouracil_tcga_paad, ZMTfluorouracil_tcga_paad, ZCTfluorouracil_tcga_paad), 1)\n",
    "ZTTfluorouracil_tcga_paad = F.normalize(ZTTfluorouracil_tcga_paad, p=2, dim=0)\n",
    "PredTfluorouracil_tcga_paad = Clas(ZTTfluorouracil_tcga_paad)\n",
    "AUCTfluorouracil_tcga_paad = roc_auc_score(Ytsfluorouracil_tcga_paad.astype(int), PredTfluorouracil_tcga_paad.detach().numpy())\n",
    "AUCTfluorouracil_tcga_paad_aucpr = average_precision_score(Ytsfluorouracil_tcga_paad.astype(int), PredTfluorouracil_tcga_paad.detach().numpy())\n",
    "\n",
    "AUCT_finetuned = dict()\n",
    "AUCTaucpr_finetuned = dict()\n",
    "AUCTfluorouracil_pdx_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_finetuned = dict()\n",
    "AUCTfluorouracil_pdx_crc_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_stad_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_paad_finetuned = dict()\n",
    "AUCTfluorouracil_pdx_aucpr_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_aucpr_finetuned = dict()\n",
    "AUCTfluorouracil_pdx_crc_aucpr_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_stad_aucpr_finetuned = dict()\n",
    "AUCTfluorouracil_tcga_paad_aucpr_finetuned = dict()\n",
    "for iter in range(max_iter):\n",
    "   iter += 1  \n",
    "\n",
    "   AutoencoderE_finetuned = torch.load(finetuned_dir + f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{iter}.pt\")\n",
    "   AutoencoderM_finetuned = torch.load(finetuned_dir + f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{iter}.pt\")\n",
    "   AutoencoderC_finetuned = torch.load(finetuned_dir + f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{iter}.pt\")\n",
    "   Clas_finetuned = torch.load(finetuned_dir + f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{iter}.pt\")\n",
    "   \n",
    "   AutoencoderE_finetuned.eval()\n",
    "   AutoencoderM_finetuned.eval()\n",
    "   AutoencoderC_finetuned.eval()\n",
    "   Clas_finetuned.eval()\n",
    "      \n",
    "   ZEX = AutoencoderE_finetuned(torch.FloatTensor(X_trainE))\n",
    "   ZMX = AutoencoderM_finetuned(torch.FloatTensor(X_trainM))\n",
    "   ZCX = AutoencoderC_finetuned(torch.FloatTensor(X_trainC))\n",
    "   ZTX = torch.cat((ZEX, ZMX, ZCX), 1)\n",
    "   ZTX = F.normalize(ZTX, p=2, dim=0)\n",
    "   PredX = Clas_finetuned(ZTX)\n",
    "   AUCT_finetuned[iter] = roc_auc_score(Y, PredX.detach().numpy())\n",
    "   AUCTaucpr_finetuned[iter] = average_precision_score(Y, PredX.detach().numpy())\n",
    "   \n",
    "   ZETfluorouracil_pdx = AutoencoderE_finetuned(TX_testEfluorouracil_pdx)\n",
    "   ZMTfluorouracil_pdx = AutoencoderM_finetuned(TX_testMfluorouracil_pdx)\n",
    "   ZCTfluorouracil_pdx = AutoencoderC_finetuned(TX_testCfluorouracil_pdx)\n",
    "   ZTTfluorouracil_pdx = torch.cat((ZETfluorouracil_pdx, ZMTfluorouracil_pdx, ZCTfluorouracil_pdx), 1)\n",
    "   ZTTfluorouracil_pdx = F.normalize(ZTTfluorouracil_pdx, p=2, dim=0)\n",
    "   PredTfluorouracil_pdx = Clas_finetuned(ZTTfluorouracil_pdx)\n",
    "   AUCTfluorouracil_pdx_finetuned[iter] = roc_auc_score(Ytsfluorouracil_pdx.astype(int), PredTfluorouracil_pdx.detach().numpy())\n",
    "   AUCTfluorouracil_pdx_aucpr_finetuned[iter] = average_precision_score(Ytsfluorouracil_pdx.astype(int), PredTfluorouracil_pdx.detach().numpy())\n",
    "\n",
    "   ZETfluorouracil_pdx_crc = AutoencoderE_finetuned(TX_testEfluorouracil_pdx_crc)\n",
    "   ZMTfluorouracil_pdx_crc = AutoencoderM_finetuned(TX_testMfluorouracil_pdx_crc)\n",
    "   ZCTfluorouracil_pdx_crc = AutoencoderC_finetuned(TX_testCfluorouracil_pdx_crc)\n",
    "   ZTTfluorouracil_pdx_crc = torch.cat((ZETfluorouracil_pdx_crc, ZMTfluorouracil_pdx_crc, ZCTfluorouracil_pdx_crc), 1)\n",
    "   ZTTfluorouracil_pdx_crc = F.normalize(ZTTfluorouracil_pdx_crc, p=2, dim=0)\n",
    "   PredTfluorouracil_pdx_crc = Clas_finetuned(ZTTfluorouracil_pdx_crc)\n",
    "   AUCTfluorouracil_pdx_crc_finetuned[iter] = roc_auc_score(Ytsfluorouracil_pdx_crc.astype(int), PredTfluorouracil_pdx_crc.detach().numpy())\n",
    "   AUCTfluorouracil_pdx_crc_aucpr_finetuned[iter] = average_precision_score(Ytsfluorouracil_pdx_crc.astype(int), PredTfluorouracil_pdx_crc.detach().numpy())\n",
    "\n",
    "   ZETfluorouracil_tcga = AutoencoderE_finetuned(TX_testEfluorouracil_tcga)\n",
    "   ZMTfluorouracil_tcga = AutoencoderM_finetuned(TX_testMfluorouracil_tcga)\n",
    "   ZCTfluorouracil_tcga = AutoencoderC_finetuned(TX_testCfluorouracil_tcga)\n",
    "   ZTTfluorouracil_tcga = torch.cat((ZETfluorouracil_tcga, ZMTfluorouracil_tcga, ZCTfluorouracil_tcga), 1)\n",
    "   ZTTfluorouracil_tcga = F.normalize(ZTTfluorouracil_tcga, p=2, dim=0)\n",
    "   PredTfluorouracil_tcga = Clas_finetuned(ZTTfluorouracil_tcga)\n",
    "   AUCTfluorouracil_tcga_finetuned[iter] = roc_auc_score(Ytsfluorouracil_tcga.astype(int), PredTfluorouracil_tcga.detach().numpy())\n",
    "   AUCTfluorouracil_tcga_aucpr_finetuned[iter] = average_precision_score(Ytsfluorouracil_tcga.astype(int), PredTfluorouracil_tcga.detach().numpy())\n",
    "\n",
    "   ZETfluorouracil_tcga_stad = AutoencoderE_finetuned(TX_testEfluorouracil_tcga_stad)\n",
    "   ZMTfluorouracil_tcga_stad = AutoencoderM_finetuned(TX_testMfluorouracil_tcga_stad)\n",
    "   ZCTfluorouracil_tcga_stad = AutoencoderC_finetuned(TX_testCfluorouracil_tcga_stad)\n",
    "   ZTTfluorouracil_tcga_stad = torch.cat((ZETfluorouracil_tcga_stad, ZMTfluorouracil_tcga_stad, ZCTfluorouracil_tcga_stad), 1)\n",
    "   ZTTfluorouracil_tcga_stad = F.normalize(ZTTfluorouracil_tcga_stad, p=2, dim=0)\n",
    "   PredTfluorouracil_tcga_stad = Clas_finetuned(ZTTfluorouracil_tcga_stad)\n",
    "   AUCTfluorouracil_tcga_stad_finetuned[iter] = roc_auc_score(Ytsfluorouracil_tcga_stad.astype(int), PredTfluorouracil_tcga_stad.detach().numpy())\n",
    "   AUCTfluorouracil_tcga_stad_aucpr_finetuned[iter] = average_precision_score(Ytsfluorouracil_tcga_stad.astype(int), PredTfluorouracil_tcga_stad.detach().numpy())\n",
    "\n",
    "   ZETfluorouracil_tcga_paad = AutoencoderE_finetuned(TX_testEfluorouracil_tcga_paad)\n",
    "   ZMTfluorouracil_tcga_paad = AutoencoderM_finetuned(TX_testMfluorouracil_tcga_paad)\n",
    "   ZCTfluorouracil_tcga_paad = AutoencoderC_finetuned(TX_testCfluorouracil_tcga_paad)\n",
    "   ZTTfluorouracil_tcga_paad = torch.cat((ZETfluorouracil_tcga_paad, ZMTfluorouracil_tcga_paad, ZCTfluorouracil_tcga_paad), 1)\n",
    "   ZTTfluorouracil_tcga_paad = F.normalize(ZTTfluorouracil_tcga_paad, p=2, dim=0)\n",
    "   PredTfluorouracil_tcga_paad = Clas_finetuned(ZTTfluorouracil_tcga_paad)\n",
    "   AUCTfluorouracil_tcga_paad_finetuned[iter] = roc_auc_score(Ytsfluorouracil_tcga_paad.astype(int), PredTfluorouracil_tcga_paad.detach().numpy())\n",
    "   AUCTfluorouracil_tcga_paad_aucpr_finetuned[iter] = average_precision_score(Ytsfluorouracil_tcga_paad.astype(int), PredTfluorouracil_tcga_paad.detach().numpy())\n",
    "\n",
    "max_key_pdx = max(AUCTfluorouracil_pdx_aucpr_finetuned, key=AUCTfluorouracil_pdx_aucpr_finetuned.get)\n",
    "max_key_pdx_crc = max(AUCTfluorouracil_pdx_crc_aucpr_finetuned, key=AUCTfluorouracil_pdx_crc_aucpr_finetuned.get)\n",
    "max_key_tcga = max(AUCTfluorouracil_tcga_aucpr_finetuned, key=AUCTfluorouracil_tcga_aucpr_finetuned.get)\n",
    "max_key_tcga_stad = max(AUCTfluorouracil_tcga_stad_aucpr_finetuned, key=AUCTfluorouracil_tcga_stad_aucpr_finetuned.get)\n",
    "max_key_tcga_paad = max(AUCTfluorouracil_tcga_paad_aucpr_finetuned, key=AUCTfluorouracil_tcga_paad_aucpr_finetuned.get)\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi): {round(AUCT,2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi): {round(AUCTaucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (5-Fluorouracil): {round(AUCTfluorouracil_pdx,2)}\")\n",
    "print(f\"PDX - Test AUCPR (5-Fluorouracil): {round(AUCTfluorouracil_pdx_aucpr,2)}\")\n",
    "print(f\"PDX - Test AUC (5-Fluorouracil, CRC): {round(AUCTfluorouracil_pdx_crc,2)}\")\n",
    "print(f\"PDX - Test AUCPR (5-Fluorouracil, CRC): {round(AUCTfluorouracil_pdx_crc_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil): {round(AUCTfluorouracil_tcga,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil): {round(AUCTfluorouracil_tcga_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil, STAD): {round(AUCTfluorouracil_tcga_stad,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil, STAD): {round(AUCTfluorouracil_tcga_stad_aucpr,2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil, PAAD): {round(AUCTfluorouracil_tcga_paad,2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil, PAAD): {round(AUCTfluorouracil_tcga_paad_aucpr,2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUC (5-Fluorouracil, Finetuned): {round(AUCTfluorouracil_pdx_finetuned[max_key_pdx],2)}\")\n",
    "print(f\"PDX - Test AUCPR (5-Fluorouracil, Finetuned): {round(AUCTfluorouracil_pdx_aucpr_finetuned[max_key_pdx],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"PDX - Test AUC (5-Fluorouracil, CRC, Finetuned): {round(AUCTfluorouracil_pdx_crc_finetuned[max_key_pdx_crc],2)}\")\n",
    "print(f\"PDX - Test AUCPR (5-Fluorouracil, CRC, Finetuned): {round(AUCTfluorouracil_pdx_crc_aucpr_finetuned[max_key_pdx_crc],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil, Finetuned): {round(AUCTfluorouracil_tcga_finetuned[max_key_tcga],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil, Finetuned): {round(AUCTfluorouracil_tcga_aucpr_finetuned[max_key_tcga],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_tcga_stad],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga_stad],2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil, STAD, Finetuned): {round(AUCTfluorouracil_tcga_stad_finetuned[max_key_tcga_stad],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil, STAD, Finetuned): {round(AUCTfluorouracil_tcga_stad_aucpr_finetuned[max_key_tcga_stad],2)}\\n\")\n",
    "\n",
    "print(f\"GDSC - Training AUC (DNA_REPi, Finetuned): {round(AUCT_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"GDSC - Training AUCPR (DNA_REPi, Finetuned): {round(AUCTaucpr_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"TCGA - Test AUC (5-Fluorouracil, PAAD, Finetuned): {round(AUCTfluorouracil_tcga_paad_finetuned[max_key_tcga_paad],2)}\")\n",
    "print(f\"TCGA - Test AUCPR (5-Fluorouracil, PAAD, Finetuned): {round(AUCTfluorouracil_tcga_paad_aucpr_finetuned[max_key_tcga_paad],2)}\\n\\n\")\n",
    "\n",
    "print(\"Models for maximum finetuning (5-Fluorouracil, PDX)\")\n",
    "print(f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "print(f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (5-Fluorouracil, PDX, CRC)\")\n",
    "print(f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "print(f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_pdx_crc}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (5-Fluorouracil, TCGA)\")\n",
    "print(f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "print(f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (5-Fluorouracil, TCGA, STAD)\")\n",
    "print(f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_stad}.pt\")\n",
    "print(f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_stad}.pt\")\n",
    "print(f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_stad}.pt\")\n",
    "print(f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_stad}.pt\")\n",
    "\n",
    "print(\"\\nModels for maximum finetuning (5-Fluorouracil, TCGA, PAAD)\")\n",
    "print(f\"Exprs_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n",
    "print(f\"Mut_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n",
    "print(f\"CNA_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n",
    "print(f\"Class_Fluorouracil_GDSC_PDX_TCGA_DNA_REPi_Expression_Mutation_CNA_Second_Strategy_Finetuned_{max_key_tcga_paad}.pt\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEE(\n",
       "  (EnE): Sequential(\n",
       "    (0): Linear(in_features=631, out_features=128, bias=True)\n",
       "    (1): BatchNorm1d(128, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.5, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show expression layer parameters\n",
    "\n",
    "AutoencoderE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEM(\n",
       "  (EnM): Sequential(\n",
       "    (0): Linear(in_features=631, out_features=1024, bias=True)\n",
       "    (1): BatchNorm1d(1024, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.7, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show mutation layer parameters\n",
    "\n",
    "AutoencoderM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AEC(\n",
       "  (EnC): Sequential(\n",
       "    (0): Linear(in_features=631, out_features=256, bias=True)\n",
       "    (1): BatchNorm1d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)\n",
       "    (2): ReLU()\n",
       "    (3): Dropout(p=0.4, inplace=False)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show CNA layer parameters\n",
    "\n",
    "AutoencoderC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Classifier(\n",
       "  (FC): Sequential(\n",
       "    (0): Linear(in_features=1408, out_features=1, bias=True)\n",
       "    (1): Dropout(p=0.3, inplace=False)\n",
       "    (2): Sigmoid()\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# show classification layer parameters\n",
    "\n",
    "Clas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data</th>\n",
       "      <th>AUC (GDSC)</th>\n",
       "      <th>AUCPR (GDSC)</th>\n",
       "      <th>AUC (PDX)</th>\n",
       "      <th>AUCPR (PDX)</th>\n",
       "      <th>AUC (PDX-CRC)</th>\n",
       "      <th>AUCPR (PDX-CRC)</th>\n",
       "      <th>AUC (TCGA)</th>\n",
       "      <th>AUCPR (TCGA)</th>\n",
       "      <th>AUC (TCGA-STAD)</th>\n",
       "      <th>AUCPR (TCGA-STAD)</th>\n",
       "      <th>AUC (TCGA-PAAD)</th>\n",
       "      <th>AUCPR (TCGA-PAAD)</th>\n",
       "      <th>Sample Size (GDSC)</th>\n",
       "      <th>Feature Size (GDSC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Expr + Mut + CNA</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.89</td>\n",
       "      <td>0.25</td>\n",
       "      <td>0.32</td>\n",
       "      <td>9435</td>\n",
       "      <td>631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Expr + Mut + CNA (Fine-tuned)</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.57</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.45</td>\n",
       "      <td>9435</td>\n",
       "      <td>631</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Data AUC (GDSC) AUCPR (GDSC) AUC (PDX)  \\\n",
       "0               Expr + Mut + CNA       0.81         0.34      0.14   \n",
       "1  Expr + Mut + CNA (Fine-tuned)       0.81         0.34      0.18   \n",
       "\n",
       "  AUCPR (PDX) AUC (PDX-CRC) AUCPR (PDX-CRC) AUC (TCGA) AUCPR (TCGA)  \\\n",
       "0        0.05          0.14            0.05       0.53         0.74   \n",
       "1        0.05          0.18            0.05       0.57         0.78   \n",
       "\n",
       "  AUC (TCGA-STAD) AUCPR (TCGA-STAD) AUC (TCGA-PAAD) AUCPR (TCGA-PAAD)  \\\n",
       "0            0.75              0.89            0.25              0.32   \n",
       "1            0.74               0.9             0.5              0.45   \n",
       "\n",
       "  Sample Size (GDSC) Feature Size (GDSC)  \n",
       "0               9435                 631  \n",
       "1               9435                 631  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe with the prediction results \n",
    "\n",
    "emc  = {\"Data\": [\"Expr + Mut + CNA\", \"Expr + Mut + CNA (Fine-tuned)\"],\n",
    "       \"AUC (GDSC)\": [round(AUCT, 2), round(AUCT_finetuned[max_key_pdx], 2)],\n",
    "       \"AUCPR (GDSC)\": [round(AUCTaucpr,2), round(AUCTaucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX)\": [round(AUCTfluorouracil_pdx,2), round(AUCTfluorouracil_pdx_finetuned[max_key_pdx],2)],\n",
    "       \"AUCPR (PDX)\":  [round(AUCTfluorouracil_pdx_aucpr,2), round(AUCTfluorouracil_pdx_aucpr_finetuned[max_key_pdx],2)],\n",
    "       \"AUC (PDX-CRC)\":  [round(AUCTfluorouracil_pdx_crc,2), round(AUCTfluorouracil_pdx_crc_finetuned[max_key_pdx_crc],2)],\n",
    "       \"AUCPR (PDX-CRC)\": [round(AUCTfluorouracil_pdx_crc_aucpr,2), round(AUCTfluorouracil_pdx_crc_aucpr_finetuned[max_key_pdx_crc],2)],\n",
    "       \"AUC (TCGA)\": [round(AUCTfluorouracil_tcga,2), round(AUCTfluorouracil_tcga_finetuned[max_key_tcga],2)],\n",
    "       \"AUCPR (TCGA)\":  [round(AUCTfluorouracil_tcga_aucpr,2), round(AUCTfluorouracil_tcga_aucpr_finetuned[max_key_tcga],2)],\n",
    "       \"AUC (TCGA-STAD)\":  [round(AUCTfluorouracil_tcga_stad,2), round(AUCTfluorouracil_tcga_stad_finetuned[max_key_tcga_stad],2)],\n",
    "       \"AUCPR (TCGA-STAD)\": [round(AUCTfluorouracil_tcga_stad_aucpr,2), round(AUCTfluorouracil_tcga_stad_aucpr_finetuned[max_key_tcga_stad],2)],\n",
    "       \"AUC (TCGA-PAAD)\":  [round(AUCTfluorouracil_tcga_paad,2), round(AUCTfluorouracil_tcga_paad_finetuned[max_key_tcga_paad],2)],\n",
    "       \"AUCPR (TCGA-PAAD)\": [round(AUCTfluorouracil_tcga_paad_aucpr,2), round(AUCTfluorouracil_tcga_paad_aucpr_finetuned[max_key_tcga_paad],2)],\n",
    "       \"Sample Size (GDSC)\": [GDSCEv2.shape[0], GDSCEv2.shape[0]],\n",
    "       \"Feature Size (GDSC)\": [GDSCEv2.shape[1], GDSCEv2.shape[1]] \n",
    "}\n",
    "\n",
    "emc_dataframe = pd.DataFrame.from_dict(emc, orient='index').transpose() \n",
    "os.makedirs(save_results_to, exist_ok = True)\n",
    "emc_dataframe.to_csv(save_results_to + \"GDSC_PDX_TCGA_Expression_Mutation_CNA_DNA_REPi_Fluorouracil_Second_Strategy.tsv\",\n",
    "                     sep = \"\\t\",\n",
    "                     index = False)\n",
    "\n",
    "emc_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Expansion/Thesis Work/Results/Transfer Learning/Drugs with same pathways/DNA_REPi/Expression_Mutation_CNA/Predictions/GDSC_PDX_TCGA_Second_Strategy\n"
     ]
    }
   ],
   "source": [
    "# change directory\n",
    "\n",
    "os.chdir(save_results_to)\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save predictions\n",
    "\n",
    "file = open('GDSC_PDX_TCGA_DNA_REPi_Fluorouracil_Predictions_Second_Strategy (Expression, Mutation, and CNA).txt', 'w')\n",
    "file.write(f\"GDSC Training (EMC) AUC (DNA_REPi): {round(AUCT,2)}\\n\")\n",
    "file.write(f\"GDSC Training (EMC) AUCPR (DNA_REPi): {round(AUCTaucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EMC) AUC (5-Fluorouracil): {round(AUCTfluorouracil_pdx,2)}\\n\")\n",
    "file.write(f\"PDX Test (EMC) AUCPR (5-Fluorouracil): {round(AUCTfluorouracil_pdx_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"PDX Test (EMC) AUC (5-Fluorouracil, CRC): {round(AUCTfluorouracil_pdx_crc,2)}\\n\")\n",
    "file.write(f\"PDX Test (EMC) AUCPR (5-Fluorouracil, CRC): {round(AUCTfluorouracil_pdx_crc_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EMC) AUC (5-Fluorouracil): {round(AUCTfluorouracil_tcga,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EMC) AUCPR (5-Fluorouracil): {round(AUCTfluorouracil_tcga_aucpr,2)}\\n\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EMC) AUC (5-Fluorouracil, STAD): {round(AUCTfluorouracil_tcga_stad,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EMC) AUCPR (5-Fluorouracil, STAD): {round(AUCTfluorouracil_tcga_stad_aucpr,2)}\\n\")\n",
    "\n",
    "file.write(f\"TCGA Test (EMC) AUC (5-Fluorouracil, PAAD): {round(AUCTfluorouracil_tcga_paad,2)}\\n\")\n",
    "file.write(f\"TCGA Test (EMC) AUCPR (5-Fluorouracil, PAAD): {round(AUCTfluorouracil_tcga_paad_aucpr,2)}\\n\")\n",
    "\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
